<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.model.DS.datasource.DSTranscript.DSTranscript Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>model</b></li><li class="navelem"><b>DS</b></li><li class="navelem"><b>datasource</b></li><li class="navelem"><b>DSTranscript</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.model.DS.datasource.DSTranscript.DSTranscript Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Inherits BaseDS.</p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ab8211408c1881ce8628cfec44b44680a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html#ab8211408c1881ce8628cfec44b44680a">__eq__</a> (self, other)</td></tr>
<tr class="separator:ab8211408c1881ce8628cfec44b44680a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5865cbc43ebbf03caa29773b45367b82"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html#a5865cbc43ebbf03caa29773b45367b82">__hash__</a> (self)</td></tr>
<tr class="separator:a5865cbc43ebbf03caa29773b45367b82"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ffa6e668a754630fe391c2a9d1028d9"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html#a3ffa6e668a754630fe391c2a9d1028d9">update</a> (self, other, gene_id_conflicts=None)</td></tr>
<tr class="separator:a3ffa6e668a754630fe391c2a9d1028d9"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> </h1>
<p>Each Transcript contains:</p><ul>
<li>id: Integer - A unique transcript ID.</li>
<li>transcript_id: String - The unique Transcript ID.</li>
<li>data_source: String - The name of the source.</li>
<li>gene_id: String - The unique gene ID.</li>
<li>raw_strand: String - The strand of the transcript (in the original annotation version).</li>
<li>raw_start_pos: Integer - The genomics coordinates of the first nucleotide of the transcript (in the original annotation version).</li>
<li>raw_end_pos: Integer - The genomics coordinates of the last nucleotide of the transcript (in the original annotation version).</li>
<li>strand: String - The strand of the transcript.</li>
<li>start_pos: Integer - The genomics coordinates of the first nucleotide of the transcript (in the current annotation version).</li>
<li>end_pos: Integer - The genomics coordinates of the last nucleotide of the transcript (in the current annotation version).</li>
<li>raw_cds_start_pos: Integer - The genomics coordinates of the start codon of the CDS (in the original annotation version).</li>
<li>raw_cds_stop_pos: Integer - The genomics coordinates of the stop codon of the CDS (in the original annotation version).</li>
<li>cds_start_pos: Integer - The genomics coordinates of the start codon of the CDS (in the current annotation version).</li>
<li>cds_stop_pos: Integer - The genomics coordinates of the stop codon of the CDS (in the current annotation version).</li>
<li>rna_biotype: String - The RNA biotype. </li>
</ul>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="ab8211408c1881ce8628cfec44b44680a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab8211408c1881ce8628cfec44b44680a">&#9670;&nbsp;</a></span>__eq__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.model.DS.datasource.DSTranscript.DSTranscript.__eq__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>other</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>__eq__ </h2>
<p>Tests the equality between two instances of this class. Two instances are considered equals if their "primary key-like" attributes (i.e. the attributes with unique constraint) are all equals.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> - Another <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> object to compare to this object.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Boolean - Are this object and the other one equal? </dd></dl>
<div class="fragment"><div class="line"><a name="l00091"></a><span class="lineno">   91</span>&#160;    <span class="keyword">def </span>__eq__( self, other ):</div><div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160;        </div><div class="line"><a name="l00093"></a><span class="lineno">   93</span>&#160;        <span class="comment"># Check if other object is of the same class</span></div><div class="line"><a name="l00094"></a><span class="lineno">   94</span>&#160;        <span class="keywordflow">if</span> ( type( other ) != type( self ) ):</div><div class="line"><a name="l00095"></a><span class="lineno">   95</span>&#160;            <span class="keywordflow">return</span> <span class="keyword">False</span></div><div class="line"><a name="l00096"></a><span class="lineno">   96</span>&#160;        </div><div class="line"><a name="l00097"></a><span class="lineno">   97</span>&#160;        <span class="comment"># Check if the transcript ID and the source are the same</span></div><div class="line"><a name="l00098"></a><span class="lineno">   98</span>&#160;        <span class="keywordflow">elif</span> ( ( self.transcript_id == other.transcript_id )</div><div class="line"><a name="l00099"></a><span class="lineno">   99</span>&#160;               <span class="keywordflow">and</span> ( self.data_source == other.data_source ) ):</div><div class="line"><a name="l00100"></a><span class="lineno">  100</span>&#160;            <span class="keywordflow">return</span> <span class="keyword">True</span></div><div class="line"><a name="l00101"></a><span class="lineno">  101</span>&#160;        </div><div class="line"><a name="l00102"></a><span class="lineno">  102</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00103"></a><span class="lineno">  103</span>&#160;            <span class="keywordflow">return</span> <span class="keyword">False</span></div><div class="line"><a name="l00104"></a><span class="lineno">  104</span>&#160;</div><div class="line"><a name="l00105"></a><span class="lineno">  105</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="a5865cbc43ebbf03caa29773b45367b82"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5865cbc43ebbf03caa29773b45367b82">&#9670;&nbsp;</a></span>__hash__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.model.DS.datasource.DSTranscript.DSTranscript.__hash__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>__hash__ </h2>
<p>Returns the hash value of a <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> object. The hash value of an instance is computed using its "primary key-like" attributes (i.e. the attributes with unique constraint).</p>
<dl class="section return"><dt>Returns</dt><dd>the hash value of the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> object. </dd></dl>
<div class="fragment"><div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;    <span class="keyword">def </span>__hash__( self ):</div><div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;        </div><div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;        <span class="keywordflow">return</span> hash( ( self.transcript_id, self.data_source ) )</div><div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;</div><div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="a3ffa6e668a754630fe391c2a9d1028d9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3ffa6e668a754630fe391c2a9d1028d9">&#9670;&nbsp;</a></span>update()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.model.DS.datasource.DSTranscript.DSTranscript.update </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>other</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id_conflicts</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>update </h2>
<p>Try to combine two instances of this class when they are "equals" (i.e. share the same values for "primary key-like attributes") by getting the missing attributes. It tests the equality between non-primary-key attributes of two equal instances of this class, and raise an DenCellORFObjUpdateException if <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> compared have different values for their non-primary key attributes. If:</p><ul>
<li>The self object is missing some values the other object has, then it updates the current object.</li>
<li>The information provided by both objects are contradictory, it logs a warning.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> - Another <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> object "equal" to the current one. </td></tr>
    <tr><td class="paramname">gene_id_conflicts</td><td>List - The list of all gene IDs that have been found associated to the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1DS_1_1datasource_1_1DSTranscript_1_1DSTranscript.html">DSTranscript</a> entry. None by default.</td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">DenCellORFTrUpdateException</td><td>When different values are found for the same attributes of two objects assumed to be the same. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;    <span class="keyword">def </span>update( self, other, gene_id_conflicts=None ):</div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;        </div><div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;        <span class="comment"># For each non-primary-key attribute, check if it is provided,</span></div><div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160;        <span class="comment"># try to update the current object by completing missing information</span></div><div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;        <span class="comment"># using information provided by the other one and if there are attributes</span></div><div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;        <span class="comment"># which have different values, raise an excpetion</span></div><div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;        attributes_to_check = [ <span class="stringliteral">&#39;raw_strand&#39;</span>, <span class="stringliteral">&#39;raw_start_pos&#39;</span>, <span class="stringliteral">&#39;raw_end_pos&#39;</span>, </div><div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;                                <span class="stringliteral">&#39;raw_cds_start_pos&#39;</span>, <span class="stringliteral">&#39;raw_cds_stop_pos&#39;</span>, <span class="stringliteral">&#39;rna_biotype&#39;</span> ]</div><div class="line"><a name="l00149"></a><span class="lineno">  149</span>&#160;        </div><div class="line"><a name="l00150"></a><span class="lineno">  150</span>&#160;        <span class="comment"># List of the attributes for which the value should be set to </span></div><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;        <span class="comment"># &#39;AMBIGUOUS&#39; if they are found different</span></div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;        att_to_flag_if_conflict = [ <span class="stringliteral">&#39;rna_biotype&#39;</span> ]</div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;</div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;        <span class="comment"># Initialize a dictionary attributes for which there is a conflict</span></div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;        conflicting_att = {}</div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;        </div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;        <span class="comment"># Check the gene IDs. If the two gene IDs are different, create a new ID </span></div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;        <span class="comment"># by concatenating the names of the two genes.</span></div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;        self_gene = self.gene_id</div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;        other_gene = other.gene_id</div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;        </div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;        <span class="keywordflow">if</span> ( self_gene != other_gene ):</div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;            </div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;            init_gene_id_conflicts = gene_id_conflicts</div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;            </div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;            <span class="comment"># If a conflict has already been registered for this DSTranscript, </span></div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;            <span class="comment"># then get the list of genes associated to it.</span></div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;            <span class="comment"># Otherwise, register this first conflict.</span></div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;            <span class="keywordflow">if</span> gene_id_conflicts:</div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;                <span class="comment"># Get the name of the gene which does not start with the CONFLICT prefix</span></div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;                <span class="keywordflow">if</span> ( ( <span class="keywordflow">not</span> self_gene.startswith( Constants.PREFIX_CONFLICT_GENE_TRANSCRIPT ) )</div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;                     <span class="keywordflow">and</span> ( self_gene <span class="keywordflow">not</span> <span class="keywordflow">in</span> gene_id_conflicts ) ):</div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;                    gene_id_conflicts.append( self_gene )</div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;                    </div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;                <span class="keywordflow">if</span> ( ( <span class="keywordflow">not</span> other_gene.startswith( Constants.PREFIX_CONFLICT_GENE_TRANSCRIPT ) )</div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;                    <span class="keywordflow">and</span> ( other_gene <span class="keywordflow">not</span> <span class="keywordflow">in</span> gene_id_conflicts ) ):</div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;                    gene_id_conflicts.append( other_gene )</div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;                </div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;                <span class="comment"># Register this new conflict and order the gene IDs</span></div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;                gene_id_conflicts = [ self_gene, other_gene ]</div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;        </div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;            <span class="comment"># Order the IDs in order to always get the same new ID when the same gene IDs are conflicting</span></div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;            gene_id_conflicts.sort()</div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;        </div><div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;            <span class="comment"># Register the new list of conflict on gene IDs only if it has been updated</span></div><div class="line"><a name="l00187"></a><span class="lineno">  187</span>&#160;            <span class="keywordflow">if</span> ( gene_id_conflicts != init_gene_id_conflicts ):</div><div class="line"><a name="l00188"></a><span class="lineno">  188</span>&#160;                <span class="comment"># Create a new &quot;conflict&quot; id </span></div><div class="line"><a name="l00189"></a><span class="lineno">  189</span>&#160;                new_gene_id = Constants.PREFIX_CONFLICT_GENE_TRANSCRIPT + <span class="stringliteral">&#39;_&#39;</span>.join( gene_id_conflicts )            </div><div class="line"><a name="l00190"></a><span class="lineno">  190</span>&#160;                self.gene_id = new_gene_id</div><div class="line"><a name="l00191"></a><span class="lineno">  191</span>&#160;                conflicting_att[ <span class="stringliteral">&#39;gene_id&#39;</span> ] = ( self_gene, other_gene, self.gene_id )</div><div class="line"><a name="l00192"></a><span class="lineno">  192</span>&#160;        </div><div class="line"><a name="l00193"></a><span class="lineno">  193</span>&#160;        </div><div class="line"><a name="l00194"></a><span class="lineno">  194</span>&#160;        <span class="comment"># Check all the other non-primary-key attributes</span></div><div class="line"><a name="l00195"></a><span class="lineno">  195</span>&#160;        <span class="keywordflow">for</span> att <span class="keywordflow">in</span> attributes_to_check:</div><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;            </div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;            self_att = getattr( self, att )</div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;            other_att = getattr( other, att )</div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;            </div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;            <span class="comment"># If the attribute treated is the raw strand, the &#39;-&#39; value </span></div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;            <span class="comment"># should not be considered as missing value</span></div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;            <span class="keywordflow">if</span> ( att == <span class="stringliteral">&#39;raw_strand&#39;</span> ):</div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;                self_att_is_empty = GeneralUtil.is_empty( val = self_att, </div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;                                                          empty_val = Constants.EMPTY_VALUES_WO_DASH )</div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;                other_att_is_empty = GeneralUtil.is_empty( val = other_att, </div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;                                                           empty_val = Constants.EMPTY_VALUES_WO_DASH )</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;                self_att_is_empty = GeneralUtil.is_empty( self_att )</div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;                other_att_is_empty = GeneralUtil.is_empty( other_att )</div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;        </div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;            <span class="comment"># If the other attribute is provided    </span></div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;            <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> other_att_is_empty ):</div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;                </div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;                <span class="comment"># and if the self attribute is empty, fill it using the information of the other object</span></div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;                <span class="keywordflow">if</span> self_att_is_empty:</div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;                    setattr( self, att, other_att )</div><div class="line"><a name="l00217"></a><span class="lineno">  217</span>&#160;                </div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;                <span class="comment"># and if the self attribute is filled and the two values of the attribute are different, </span></div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;                <span class="comment"># reset the value to &#39;AMBIGUOUS&#39; and keep track of this attribute and its values</span></div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;                <span class="keywordflow">elif</span> ( ( <span class="keywordflow">not</span> self_att_is_empty ) <span class="keywordflow">and</span> ( other_att != self_att ) ):</div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;                    <span class="keywordflow">if</span> att <span class="keywordflow">in</span> att_to_flag_if_conflict:</div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;                        setattr( self, att, Constants.DENCELLORFOBJ_AMBIGUOUS_ATT )</div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;                        setattr( self, att, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;                    conflicting_att[ att ] = ( self_att, other_att, getattr(self, att) )</div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;        </div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;        <span class="comment"># If there are attributes for which there is a conflict, </span></div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;        <span class="comment"># raise a DenCellORFTrUpdateException</span></div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;        <span class="keywordflow">if</span> conflicting_att != {}:</div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;            except_message = []</div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;            <span class="keywordflow">for</span> (att, val) <span class="keywordflow">in</span> conflicting_att.items():</div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;                except_message.append( <span class="stringliteral">&#39;the attribute &#39;</span> + att + <span class="stringliteral">&#39; has been found with the values &quot;&#39;</span> + </div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                                       str( val[0] ) + <span class="stringliteral">&#39;&quot; and &quot;&#39;</span> + str( val[1] ) + <span class="stringliteral">&#39;&quot;,&#39;</span> +</div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                                       <span class="stringliteral">&#39; hence its value has been set to &quot;&#39;</span> + str( val[2] ) + <span class="stringliteral">&#39;&quot;&#39;</span> )</div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;            except_message = <span class="stringliteral">&#39;, &#39;</span>.join( except_message ) + <span class="stringliteral">&#39;.&#39;</span></div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;            </div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;            <span class="comment"># If a conflict has been found regarding the gene ID, raise the exception </span></div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;            <span class="comment"># with the value of the new gene ID to create and to associate to this transcript</span></div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;            <span class="keywordflow">if</span> ( <span class="stringliteral">&#39;gene_id&#39;</span> <span class="keywordflow">in</span> conflicting_att.keys() ):</div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;                <span class="keywordflow">raise</span> DenCellORFTrUpdateException( message = except_message, </div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;                                                   gene_id = self.gene_id, </div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;                                                   gene_id_conflicts = gene_id_conflicts )</div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;                <span class="keywordflow">raise</span> DenCellORFTrUpdateException( message = except_message )</div></div><!-- fragment -->
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/model/DS/datasource/DSTranscript.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
