<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.util.genetics.GeneticsUtil.GeneticsUtil Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>util</b></li><li class="navelem"><b>genetics</b></li><li class="navelem"><b>GeneticsUtil</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1util_1_1genetics_1_1GeneticsUtil_1_1GeneticsUtil.html">GeneticsUtil</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1util_1_1genetics_1_1GeneticsUtil_1_1GeneticsUtil-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.util.genetics.GeneticsUtil.GeneticsUtil Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Inherits object.</p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a62175eb1cbf3d0699119cac0e43dba1f"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1util_1_1genetics_1_1GeneticsUtil_1_1GeneticsUtil.html#a62175eb1cbf3d0699119cac0e43dba1f">translate_dna</a> (dna_seq, include_stop=False, stop_codon=GeneticsConstants.STOP_CODON, to_stop=False)</td></tr>
<tr class="separator:a62175eb1cbf3d0699119cac0e43dba1f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ade5f2a150762cedf44d6cdbd35a8a62e"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1util_1_1genetics_1_1GeneticsUtil_1_1GeneticsUtil.html#ade5f2a150762cedf44d6cdbd35a8a62e">find_sqce_consensus</a> (list_of_sequences, sqce_type=Constants.SEQUENCE_TYPE_DNA, threshold=Constants.DEFAULT_SQCE_CONSENSUS_AMBIG_THRESHOLD, fasta_end_name='')</td></tr>
<tr class="separator:ade5f2a150762cedf44d6cdbd35a8a62e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aafe58a29bf44854ed66f9c1df64c26ac"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1util_1_1genetics_1_1GeneticsUtil_1_1GeneticsUtil.html#aafe58a29bf44854ed66f9c1df64c26ac">find_codon_consensus</a> (sequences, threshold=Constants.DEFAULT_SQCE_CONSENSUS_AMBIG_THRESHOLD, ambiguous=Constants.SEQUENCE_AMBIGUOUS_DNA_BASE)</td></tr>
<tr class="separator:aafe58a29bf44854ed66f9c1df64c26ac"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1util_1_1genetics_1_1GeneticsUtil_1_1GeneticsUtil.html">GeneticsUtil</a> </h1>
<p>This class contains various static methods which aims to simplify the operation performed on genetic-related objects, such as strings containing sequences. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="aafe58a29bf44854ed66f9c1df64c26ac"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aafe58a29bf44854ed66f9c1df64c26ac">&#9670;&nbsp;</a></span>find_codon_consensus()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.util.genetics.GeneticsUtil.GeneticsUtil.find_codon_consensus </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>threshold</em> = <code>Constants.DEFAULT_SQCE_CONSENSUS_AMBIG_THRESHOLD</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ambiguous</em> = <code>Constants.SEQUENCE_AMBIGUOUS_DNA_BASE</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>find_codon_consensus </h2>
<p>This is a static method that allows to find a consensus for the sequence of a codon (i.e. for a sequence of exactly 3 nucleotides).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>List of strings - The list of sequences that needs to be use to find the consensus. </td></tr>
    <tr><td class="paramname">threshold</td><td>Float - The threshold value required to add a particular letter to the sequence, i.e. the proportion of sequences of the list that needs to contain the letter at a particular place to consider it as part of the consensus. E.g. if the threshold is 0.7, at least 70 % of the sequences needs to get the letter 'N' at a given place to include this letter at this place in the consensus; otherwise the ambiguous letter will be added at this place. </td></tr>
    <tr><td class="paramname">ambiguous</td><td>String - The letter to use for an ambiguous nucleotide.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>consensus - The consensus computed. </dd></dl>
<div class="fragment"><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;    <span class="keyword">def </span>find_codon_consensus( sequences, threshold=Constants.DEFAULT_SQCE_CONSENSUS_AMBIG_THRESHOLD, \</div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;                              ambiguous=Constants.SEQUENCE_AMBIGUOUS_DNA_BASE ):</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;                    </div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;        <span class="comment"># Remove from the list of sequence any sequence that </span></div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;        <span class="comment"># would not have exactly 3 nucleotides</span></div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;        list_of_seq = []</div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;        <span class="keywordflow">for</span> seq <span class="keywordflow">in</span> sequences:</div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;            <span class="keywordflow">if</span> ( len( seq ) == 3 ):</div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;                list_of_seq.append( seq )</div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;    </div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;        <span class="comment"># Get the number of sequences to use for the consensus</span></div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;        sequences_count = len( list_of_seq )</div><div class="line"><a name="l00217"></a><span class="lineno">  217</span>&#160;        </div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;        <span class="comment"># If there is not any sequence in the list, return None</span></div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;        <span class="keywordflow">if</span> ( sequences_count == 0 ):</div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;            consensus = <span class="keywordtype">None</span></div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;        </div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;        <span class="comment"># If there is only one sequence in the list, return it</span></div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;        <span class="keywordflow">elif</span> ( sequences_count == 1 ):</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;            consensus = list_of_seq[ 0 ]</div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;        </div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;        <span class="comment"># Otherwise, compute the consensus for the start codon sequence</span></div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;            nt_consensus = []</div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;            </div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;            <span class="comment"># First, compute the consensus for each nucleotide</span></div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;            <span class="keywordflow">for</span> nt_pos <span class="keywordflow">in</span> range( 3 ):</div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;                </div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                <span class="comment"># For each position, get the number of time a nucleotide is found</span></div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                nt_count_dict = dict( Counter( [ seq[ nt_pos ] <span class="keywordflow">for</span> seq <span class="keywordflow">in</span> list_of_seq ] ) )</div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;                 </div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;                <span class="comment"># Get the nucleotide(s) found the highest number of time</span></div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;                nt_max_count = max( nt_count_dict.values() )</div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;                    </div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;                <span class="comment"># Convert the count into a proportion</span></div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;                nt_max_prop = float( nt_max_count ) / float( sequences_count )</div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;                </div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;                <span class="comment"># If the proportion is (strictly) lower than the threshold defined,</span></div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;                <span class="comment"># then replace the nucleotide at this position by the ambiguous letter </span></div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;                <span class="keywordflow">if</span> ( nt_max_prop &lt; threshold ):</div><div class="line"><a name="l00245"></a><span class="lineno">  245</span>&#160;                    nt_consensus.append( ambiguous )</div><div class="line"><a name="l00246"></a><span class="lineno">  246</span>&#160;                    </div><div class="line"><a name="l00247"></a><span class="lineno">  247</span>&#160;                <span class="comment"># Otherwise, use the letter(s) at this positions to compute the consensus</span></div><div class="line"><a name="l00248"></a><span class="lineno">  248</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00249"></a><span class="lineno">  249</span>&#160;                    nt_list = [ nt <span class="keywordflow">for</span> ( nt, count ) <span class="keywordflow">in</span> nt_count_dict.items() <span class="keywordflow">if</span> ( count == nt_max_count ) ]</div><div class="line"><a name="l00250"></a><span class="lineno">  250</span>&#160;                    <span class="comment"># If there is one single nucleotide with this proportion, use it in the consensus</span></div><div class="line"><a name="l00251"></a><span class="lineno">  251</span>&#160;                    <span class="keywordflow">if</span> ( len( nt_list ) == 1 ):</div><div class="line"><a name="l00252"></a><span class="lineno">  252</span>&#160;                        nt_consensus.append( nt_list[ 0 ] )</div><div class="line"><a name="l00253"></a><span class="lineno">  253</span>&#160;                    <span class="comment"># Otherwise, use the ambiguous letter for this position</span></div><div class="line"><a name="l00254"></a><span class="lineno">  254</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00255"></a><span class="lineno">  255</span>&#160;                        nt_consensus.append( ambiguous )</div><div class="line"><a name="l00256"></a><span class="lineno">  256</span>&#160;            </div><div class="line"><a name="l00257"></a><span class="lineno">  257</span>&#160;            <span class="comment"># Create a string from the consensus for the three nucleotides</span></div><div class="line"><a name="l00258"></a><span class="lineno">  258</span>&#160;            consensus = <span class="stringliteral">&#39;&#39;</span>.join( nt_consensus )</div><div class="line"><a name="l00259"></a><span class="lineno">  259</span>&#160;            </div><div class="line"><a name="l00260"></a><span class="lineno">  260</span>&#160;        <span class="keywordflow">return</span> consensus</div></div><!-- fragment -->
</div>
</div>
<a id="ade5f2a150762cedf44d6cdbd35a8a62e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ade5f2a150762cedf44d6cdbd35a8a62e">&#9670;&nbsp;</a></span>find_sqce_consensus()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.util.genetics.GeneticsUtil.GeneticsUtil.find_sqce_consensus </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>list_of_sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>sqce_type</em> = <code>Constants.SEQUENCE_TYPE_DNA</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>threshold</em> = <code>Constants.DEFAULT_SQCE_CONSENSUS_AMBIG_THRESHOLD</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>fasta_end_name</em> = <code>''</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>find_sqce_consensus </h2>
<p>This is a static method that allows to find a consensus from several DNA or protein sequences. This sequence uses MUSCLE algorithm to compute the consensus.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">list_of_sequences</td><td>List of strings - The list of sequences that needs to be use to find the consensus. </td></tr>
    <tr><td class="paramname">sqce_type</td><td>String - The type of sequence to compare. Should equal 'DNA' or 'PROT'. 'DNA' by default. </td></tr>
    <tr><td class="paramname">threshold</td><td>Float - The threshold value required to add a particular letter to the sequence, i.e. the proportion of sequences of the list that needs to contain the letter at a particular place to consider it as part of the consensus. E.g. if the threshold is 0.7, at least 70 % of the sequences needs to get the letter 'X' at a given place to include this letter at this place in the consensus; otherwise the ambiguous letter will be added at this place. </td></tr>
    <tr><td class="paramname">fasta_end_name</td><td>String - The string to add at the end of the fasta file name (Optional).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>str( consensus ) - The consensus sequence computed.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">DenCellORFException</td><td>When the provided sequence type is not 'DNA' or 'PROT'. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;    <span class="keyword">def </span>find_sqce_consensus( list_of_sequences, sqce_type=Constants.SEQUENCE_TYPE_DNA, \</div><div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;                             threshold=Constants.DEFAULT_SQCE_CONSENSUS_AMBIG_THRESHOLD, \</div><div class="line"><a name="l00127"></a><span class="lineno">  127</span>&#160;                             fasta_end_name = &#39;&#39; ):</div><div class="line"><a name="l00128"></a><span class="lineno">  128</span>&#160;                </div><div class="line"><a name="l00129"></a><span class="lineno">  129</span>&#160;        <span class="keywordflow">if</span> ( sqce_type == Constants.SEQUENCE_TYPE_DNA ):</div><div class="line"><a name="l00130"></a><span class="lineno">  130</span>&#160;            alphabet = generic_dna</div><div class="line"><a name="l00131"></a><span class="lineno">  131</span>&#160;            ambiguous = Constants.SEQUENCE_AMBIGUOUS_DNA_BASE</div><div class="line"><a name="l00132"></a><span class="lineno">  132</span>&#160;            </div><div class="line"><a name="l00133"></a><span class="lineno">  133</span>&#160;        <span class="keywordflow">elif</span> ( sqce_type == Constants.SEQUENCE_TYPE_PROT ):</div><div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;            alphabet = generic_protein</div><div class="line"><a name="l00135"></a><span class="lineno">  135</span>&#160;            ambiguous = Constants.SEQUENCE_AMBIGUOUS_PROT_AA</div><div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;        </div><div class="line"><a name="l00137"></a><span class="lineno">  137</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;            <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;MergeStrategy.find_sqce_consensus(): The type of sequence provided&#39;</span> +</div><div class="line"><a name="l00139"></a><span class="lineno">  139</span>&#160;                                       <span class="stringliteral">&#39; has to be &#39;</span> + Constants.SEQUENCE_TYPE_DNA + <span class="stringliteral">&#39; or &#39;</span> +</div><div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;                                       Constants.SEQUENCE_TYPE_PROT + </div><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;                                       <span class="stringliteral">&#39; (provided type: &#39;</span> + str( sqce_type ) + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;        </div><div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;        <span class="comment"># Store the input sequences in a fasta file in order to run Muscle</span></div><div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160;        input_sequences = ( SeqRecord( Seq( s, alphabet) ) <span class="keywordflow">for</span> s <span class="keywordflow">in</span> list_of_sequences )</div><div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;        </div><div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> os.path.exists( DefaultTemporaryFolder.TEMPORARY_FOLDER ) ):</div><div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;            os.makedirs( DefaultTemporaryFolder.TEMPORARY_FOLDER )</div><div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;            </div><div class="line"><a name="l00149"></a><span class="lineno">  149</span>&#160;        input_sequences_file = os.path.join( DefaultTemporaryFolder.TEMPORARY_FOLDER, </div><div class="line"><a name="l00150"></a><span class="lineno">  150</span>&#160;                                             <span class="stringliteral">&#39;input_sequences&#39;</span> + fasta_end_name + <span class="stringliteral">&#39;.fasta&#39;</span> )</div><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;        SeqIO.write( input_sequences, input_sequences_file, <span class="stringliteral">&#39;fasta&#39;</span> )</div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;        </div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;        </div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;        <span class="comment"># Perform the multiple sequences alignment and </span></div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;        <span class="comment"># store the output in a fasta file</span></div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;        aligned_sequences_file = os.path.join( DefaultTemporaryFolder.TEMPORARY_FOLDER, </div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;                                               <span class="stringliteral">&#39;aligned_sequences&#39;</span> + fasta_end_name + <span class="stringliteral">&#39;.fasta&#39;</span> )</div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;        muscle_cline = MuscleCommandline( cmd = <span class="stringliteral">&#39;/bin/muscle&#39;</span>,</div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;                                          input = input_sequences_file,</div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;                                          out = aligned_sequences_file )</div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;        </div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;        ( stdout, stderr ) = muscle_cline()</div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;        </div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;        </div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;        <span class="comment"># Read the fasta file containing aligned sequences</span></div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;        align = AlignIO.read( aligned_sequences_file, <span class="stringliteral">&#39;fasta&#39;</span> )</div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;        </div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;        summary_align = AlignInfo.SummaryInfo( align )</div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;        </div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;        </div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;        <span class="comment"># Compute the consensus</span></div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;        consensus = summary_align.gap_consensus( threshold = threshold, </div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;                                                 ambiguous = ambiguous )</div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;        </div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;        </div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;        <span class="comment"># Remove the temporary fasta files</span></div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;        os.remove( input_sequences_file )</div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;        os.remove( aligned_sequences_file )</div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;        </div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;        <span class="keywordflow">return</span> str( consensus )</div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;</div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;</div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a62175eb1cbf3d0699119cac0e43dba1f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a62175eb1cbf3d0699119cac0e43dba1f">&#9670;&nbsp;</a></span>translate_dna()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.util.genetics.GeneticsUtil.GeneticsUtil.translate_dna </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dna_seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>include_stop</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>stop_codon</em> = <code>GeneticsConstants.STOP_CODON</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>to_stop</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>translate_dna </h2>
<p>This is a static method that allows to translate a DNA sequence into an amino acid sequence.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dna_seq</td><td>String - The DNA sequence to translate. </td></tr>
    <tr><td class="paramname">include_stop</td><td>Boolean - Should the (last) stop be included in the sequence returned? False by default. </td></tr>
    <tr><td class="paramname">stop_codon</td><td>String - The symbol that has to be used for stop codons. '*' (GeneticsConstants.STOP_CODON) by default. </td></tr>
    <tr><td class="paramname">to_stop</td><td>Boolean - Should the translation be halted as soon as the first stop codon is met? False by default.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>aa_seq: String - The amino acid sequence corresponding to the translation of the provided DNA sequence.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">TranslateDNAException</td><td>When the length of the DNA sequence to translate is not a multiple of three. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00053"></a><span class="lineno">   53</span>&#160;    <span class="keyword">def </span>translate_dna( dna_seq, include_stop=False, stop_codon=GeneticsConstants.STOP_CODON, to_stop=False ):</div><div class="line"><a name="l00054"></a><span class="lineno">   54</span>&#160;        </div><div class="line"><a name="l00055"></a><span class="lineno">   55</span>&#160;        <span class="comment"># Get the sequence in uppercase</span></div><div class="line"><a name="l00056"></a><span class="lineno">   56</span>&#160;        dna_seq = dna_seq.upper()</div><div class="line"><a name="l00057"></a><span class="lineno">   57</span>&#160;        </div><div class="line"><a name="l00058"></a><span class="lineno">   58</span>&#160;        <span class="comment"># Record the initial sequence to translate</span></div><div class="line"><a name="l00059"></a><span class="lineno">   59</span>&#160;        initial_dna_seq = str( dna_seq )</div><div class="line"><a name="l00060"></a><span class="lineno">   60</span>&#160;        </div><div class="line"><a name="l00061"></a><span class="lineno">   61</span>&#160;        <span class="comment"># Get the length of the sequence</span></div><div class="line"><a name="l00062"></a><span class="lineno">   62</span>&#160;        dna_seq_len = len( dna_seq )</div><div class="line"><a name="l00063"></a><span class="lineno">   63</span>&#160;        </div><div class="line"><a name="l00064"></a><span class="lineno">   64</span>&#160;        <span class="comment"># If the length is not a multiple of three, remove the last nucleotides </span></div><div class="line"><a name="l00065"></a><span class="lineno">   65</span>&#160;        <span class="comment"># prior to perform the translation</span></div><div class="line"><a name="l00066"></a><span class="lineno">   66</span>&#160;        remainder = dna_seq_len % 3</div><div class="line"><a name="l00067"></a><span class="lineno">   67</span>&#160;        </div><div class="line"><a name="l00068"></a><span class="lineno">   68</span>&#160;        <span class="keywordflow">if</span> ( remainder != 0 ):</div><div class="line"><a name="l00069"></a><span class="lineno">   69</span>&#160;            dna_seq = dna_seq[ : -remainder ]</div><div class="line"><a name="l00070"></a><span class="lineno">   70</span>&#160;            </div><div class="line"><a name="l00071"></a><span class="lineno">   71</span>&#160;        <span class="comment"># Translate the DNA sequence into protein using Biopython</span></div><div class="line"><a name="l00072"></a><span class="lineno">   72</span>&#160;        dna_sequence = Seq( dna_seq, generic_dna )</div><div class="line"><a name="l00073"></a><span class="lineno">   73</span>&#160;        aa_seq = str( dna_sequence.translate( table = 1, </div><div class="line"><a name="l00074"></a><span class="lineno">   74</span>&#160;                                              stop_symbol = stop_codon, </div><div class="line"><a name="l00075"></a><span class="lineno">   75</span>&#160;                                              to_stop = to_stop ) )</div><div class="line"><a name="l00076"></a><span class="lineno">   76</span>&#160;            </div><div class="line"><a name="l00077"></a><span class="lineno">   77</span>&#160;        <span class="comment"># Remove the last stop codon if necessary</span></div><div class="line"><a name="l00078"></a><span class="lineno">   78</span>&#160;        <span class="keywordflow">if</span> ( ( aa_seq[ -1 ] == stop_codon ) </div><div class="line"><a name="l00079"></a><span class="lineno">   79</span>&#160;             <span class="keywordflow">and</span> ( <span class="keywordflow">not</span> include_stop ) ):</div><div class="line"><a name="l00080"></a><span class="lineno">   80</span>&#160;            aa_seq = aa_seq[ :-1 ]</div><div class="line"><a name="l00081"></a><span class="lineno">   81</span>&#160;        </div><div class="line"><a name="l00082"></a><span class="lineno">   82</span>&#160;        <span class="comment"># If the length of the provided sequence was a multiple of three, </span></div><div class="line"><a name="l00083"></a><span class="lineno">   83</span>&#160;        <span class="comment"># then return the amino acid sequence.</span></div><div class="line"><a name="l00084"></a><span class="lineno">   84</span>&#160;        <span class="comment"># Otherwise, raise a TranslateDNAException.</span></div><div class="line"><a name="l00085"></a><span class="lineno">   85</span>&#160;        <span class="keywordflow">if</span> ( remainder == 0 ):</div><div class="line"><a name="l00086"></a><span class="lineno">   86</span>&#160;            <span class="keywordflow">return</span> aa_seq</div><div class="line"><a name="l00087"></a><span class="lineno">   87</span>&#160;        </div><div class="line"><a name="l00088"></a><span class="lineno">   88</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00089"></a><span class="lineno">   89</span>&#160;            <span class="keywordflow">raise</span> TranslateDNAException( message = ( <span class="stringliteral">&#39;The length of the DNA sequence provided is not&#39;</span> +</div><div class="line"><a name="l00090"></a><span class="lineno">   90</span>&#160;                                                     <span class="stringliteral">&#39; a multiple of three. Hence the sequence has been&#39;</span> +</div><div class="line"><a name="l00091"></a><span class="lineno">   91</span>&#160;                                                     <span class="stringliteral">&#39; translated ignoring the last &#39;</span> + str( remainder ) +</div><div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160;                                                     <span class="stringliteral">&#39; nucleotides.&#39;</span> ),</div><div class="line"><a name="l00093"></a><span class="lineno">   93</span>&#160;                                         dna_seq = initial_dna_seq,</div><div class="line"><a name="l00094"></a><span class="lineno">   94</span>&#160;                                         aa_seq = aa_seq,</div><div class="line"><a name="l00095"></a><span class="lineno">   95</span>&#160;                                         remainder = remainder )</div><div class="line"><a name="l00096"></a><span class="lineno">   96</span>&#160;        </div><div class="line"><a name="l00097"></a><span class="lineno">   97</span>&#160;    </div><div class="line"><a name="l00098"></a><span class="lineno">   98</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/util/genetics/GeneticsUtil.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
