<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>execution</b></li><li class="navelem"><b>insertion</b></li><li class="navelem"><b>ParserStrategy</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html">ParserStrategy</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<div class="dynheader">
Inheritance diagram for fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.png" usemap="#fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy_map" alt=""/>
  <map id="fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy_map" name="fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy_map">
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Johnstone2016_1_1Johnstone2016.html" title="Source: Johnstone et al., 2016." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Johnstone2016.Johnstone2016" shape="rect" coords="552,56,1094,80"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Mackowiak2015_1_1Mackowiak2015.html" title="Source: Mackowiak et al., 2015." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Mackowiak2015.Mackowiak2015" shape="rect" coords="552,112,1094,136"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1NCBIGeneList_1_1NCBIGeneList.html" title="Source: NCBI." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.NCBIGeneList.NCBIGeneList" shape="rect" coords="552,168,1094,192"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html" title="Source: Samandi et al., 2017." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017" shape="rect" coords="552,224,1094,248"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html" title="Source: sORFs.org." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org" shape="rect" coords="552,280,1094,304"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1Erhard2018_1_1Erhard2018.html" title="Source: Erhard et al., 2018." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.human.Erhard2018.Erhard2018" shape="rect" coords="552,336,1094,360"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1HGNCGeneList_1_1HGNCGeneList.html" title="Source: HGNC." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.human.HGNCGeneList.HGNCGeneList" shape="rect" coords="552,392,1094,416"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1Laumont2016_1_1Laumont2016.html" title="Source: Laumont et al., 2016." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.human.Laumont2016.Laumont2016" shape="rect" coords="552,448,1094,472"/>
  </map>
</div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:af3a41c9649531a71f9ab74bf6c5f709a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#af3a41c9649531a71f9ab74bf6c5f709a">__init__</a> (self, data_path)</td></tr>
<tr class="separator:af3a41c9649531a71f9ab74bf6c5f709a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8cdab4b401f878c517c4b737b9ecf778"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a8cdab4b401f878c517c4b737b9ecf778">execute</a> ()</td></tr>
<tr class="separator:a8cdab4b401f878c517c4b737b9ecf778"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11654f0533e3a163aa528c8f996c3456"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a11654f0533e3a163aa528c8f996c3456">import_file_content</a> (self, file_sep=',', dtype_val='str', encoding_val='utf-8')</td></tr>
<tr class="separator:a11654f0533e3a163aa528c8f996c3456"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb455bdb96149592f7251e5f224c7fa3"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#abb455bdb96149592f7251e5f224c7fa3">check_headers</a> (self)</td></tr>
<tr class="separator:abb455bdb96149592f7251e5f224c7fa3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0b04e0792f45b98d4111aa27c51ac93"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ab0b04e0792f45b98d4111aa27c51ac93">initialize_ids</a> (self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, orf_stop, orf_strand, index, log_symb_from_tr=True, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:ab0b04e0792f45b98d4111aa27c51ac93"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa01883a5db8dba8aabdb8af61d3b1f94"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#aa01883a5db8dba8aabdb8af61d3b1f94">initialize_transcript_id</a> (self, transcript_id, orf_id, index)</td></tr>
<tr class="separator:aa01883a5db8dba8aabdb8af61d3b1f94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a07cdd23100d43df77f4001cd81fba324"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a07cdd23100d43df77f4001cd81fba324">initialize_gene_symbol</a> (self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, orf_stop, orf_strand, log_symb_from_tr=True, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:a07cdd23100d43df77f4001cd81fba324"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d5bdcc0943ba6c1c71bb3c245bc81b3"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4d5bdcc0943ba6c1c71bb3c245bc81b3">symbol_from_transcript_id</a> (self, ensembl_db, orf_id, transcript_id, log_warning=True)</td></tr>
<tr class="separator:a4d5bdcc0943ba6c1c71bb3c245bc81b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac199b748e002f7651a7ad708197c0008"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ac199b748e002f7651a7ad708197c0008">symbol_from_orf_loc</a> (self, ensembl_db, orf_id, chr_name, orf_start, orf_stop, orf_strand, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:ac199b748e002f7651a7ad708197c0008"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4251bb5eb9243ff185efba305843d86b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4251bb5eb9243ff185efba305843d86b">get_gene_from_symbol</a> (self, symbol, chr_name, orf_id)</td></tr>
<tr class="separator:a4251bb5eb9243ff185efba305843d86b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a10fea64161666d9bb47ee4b8553b52e5"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a10fea64161666d9bb47ee4b8553b52e5">find_gene_from_symbol</a> (self, symbol, chr_name)</td></tr>
<tr class="separator:a10fea64161666d9bb47ee4b8553b52e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9d3195a3bb1fd975359b771ebee910ef"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a9d3195a3bb1fd975359b771ebee910ef">get_gene_from_conflict_id</a> (self, conflict_symbol, chr_name)</td></tr>
<tr class="separator:a9d3195a3bb1fd975359b771ebee910ef"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a602be6456b764340546d4ff9fd07d19f"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a602be6456b764340546d4ff9fd07d19f">is_empty</a> (val, empty_val=Constants.EMPTY_VALUES)</td></tr>
<tr class="separator:a602be6456b764340546d4ff9fd07d19f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a846efdd59ec331984a46aa80a46afcf2"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a846efdd59ec331984a46aa80a46afcf2">get_ensembl_db</a> (annotation_version)</td></tr>
<tr class="separator:a846efdd59ec331984a46aa80a46afcf2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a19c8d87c05bf4071bb6c7b541106a12b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a19c8d87c05bf4071bb6c7b541106a12b">initialize_chromosome_name</a> (chr_name)</td></tr>
<tr class="separator:a19c8d87c05bf4071bb6c7b541106a12b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a37ad31eeda67bd00e0aba6bfe491eecf"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a37ad31eeda67bd00e0aba6bfe491eecf">get_uniq_gene_on_chromosome</a> (chr, list_of_genes)</td></tr>
<tr class="separator:a37ad31eeda67bd00e0aba6bfe491eecf"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html">ParserStrategy</a> </h1>
<p>This class is an abstract class and metaclass for "ParserStrategy classes" that allow to parse the data from pre-defined data sources (see the "parserstrategy" module for more information about the parsers, one class has been defined for each data source). Hence, this class cannot be instantiated and only contains abstract methods. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="af3a41c9649531a71f9ab74bf6c5f709a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af3a41c9649531a71f9ab74bf6c5f709a">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>data_path</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>Constructor of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html">ParserStrategy</a> </h2>
<p>Instance variables:</p><ul>
<li>data_source: String - The name of the data source.</li>
<li>data_path: String - The path to the data source file.</li>
<li>file_content: Pandas data frame - The Pandas data frame containing the content of the source file.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">data_path</td><td>String - The path to the data source file. </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented in <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Johnstone2016_1_1Johnstone2016.html#ae753acfa7094737b213a6ccc0d8e7108">fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Johnstone2016.Johnstone2016</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html#a35fbfff77c3823848928613b72dde15d">fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Mackowiak2015_1_1Mackowiak2015.html#a36aef72acaa66d578e5625a45faa2d48">fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Mackowiak2015.Mackowiak2015</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1Erhard2018_1_1Erhard2018.html#a2f19b197d39f9f946c867e73c863931e">fr.tagc.uorf.core.execution.insertion.parserstrategy.human.Erhard2018.Erhard2018</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1Laumont2016_1_1Laumont2016.html#a1549a5796cc5fd770fd743e674c6af82">fr.tagc.uorf.core.execution.insertion.parserstrategy.human.Laumont2016.Laumont2016</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1mouse_1_1sORFs__org__Mouse_1_1sORFs__org__Mouse.html#a2641e96fe30067be4e62f2ffc6056fc0">fr.tagc.uorf.core.execution.insertion.parserstrategy.mouse.sORFs_org_Mouse.sORFs_org_Mouse</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1sORFs__org__Human_1_1sORFs__org__Human.html#a2dfe4412ba42ed1fff25abe9cfff4555">fr.tagc.uorf.core.execution.insertion.parserstrategy.human.sORFs_org_Human.sORFs_org_Human</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html#ab35b35829138cef8d8e8ea3a671bd2b0">fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017</a>, <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1NCBIGeneList_1_1NCBIGeneList.html#ac174453335305ed863122cfa7488ed72">fr.tagc.uorf.core.execution.insertion.parserstrategy.common.NCBIGeneList.NCBIGeneList</a>, and <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1HGNCGeneList_1_1HGNCGeneList.html#ab4809de068f6e31d3d76b3202b05590c">fr.tagc.uorf.core.execution.insertion.parserstrategy.human.HGNCGeneList.HGNCGeneList</a>.</p>
<div class="fragment"><div class="line"><a name="l00040"></a><span class="lineno">   40</span>&#160;    <span class="keyword">def </span>__init__( self, data_path ):</div><div class="line"><a name="l00041"></a><span class="lineno">   41</span>&#160;        </div><div class="line"><a name="l00042"></a><span class="lineno">   42</span>&#160;        self.data_source = str( self.__class__.__name__ )</div><div class="line"><a name="l00043"></a><span class="lineno">   43</span>&#160;        self.data_path = data_path</div><div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;        self.file_content = <span class="keywordtype">None</span></div><div class="line"><a name="l00045"></a><span class="lineno">   45</span>&#160;</div><div class="line"><a name="l00046"></a><span class="lineno">   46</span>&#160;</div><div class="line"><a name="l00047"></a><span class="lineno">   47</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="abb455bdb96149592f7251e5f224c7fa3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abb455bdb96149592f7251e5f224c7fa3">&#9670;&nbsp;</a></span>check_headers()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.check_headers </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>check_headers </h2>
<p>This method allows to check if a list of expected headers is included in the pandas data frame column names (containing the content of the file).</p>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">MissingHeaderException</td><td>When one of the expected headers is missing. </td></tr>
    <tr><td class="paramname">DuplicateHeaderException</td><td>When several headers seems to be identical in the original file. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;    <span class="keyword">def </span>check_headers( self ):</div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;                </div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;        <span class="comment"># Get the list of the column names and sort it</span></div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;        col_names = sorted( list( self.file_content ) )</div><div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;        </div><div class="line"><a name="l00187"></a><span class="lineno">  187</span>&#160;        <span class="comment"># Check if the list of expected headers is included in the column name list.</span></div><div class="line"><a name="l00188"></a><span class="lineno">  188</span>&#160;        <span class="comment"># Raise a MissingHeaderException if some of the expected headers are missing.</span></div><div class="line"><a name="l00189"></a><span class="lineno">  189</span>&#160;        header_included = set( self.HEADER_LIST ).issubset( col_names )</div><div class="line"><a name="l00190"></a><span class="lineno">  190</span>&#160;        </div><div class="line"><a name="l00191"></a><span class="lineno">  191</span>&#160;        <span class="keywordflow">if</span> ( header_included == <span class="keyword">False</span> ):</div><div class="line"><a name="l00192"></a><span class="lineno">  192</span>&#160;            <span class="comment"># Get the list of missing headers and raise a MissingHeaderException</span></div><div class="line"><a name="l00193"></a><span class="lineno">  193</span>&#160;            missing_headers = [ h <span class="keywordflow">for</span> h <span class="keywordflow">in</span> self.HEADER_LIST <span class="keywordflow">if</span> h <span class="keywordflow">not</span> <span class="keywordflow">in</span> col_names ]</div><div class="line"><a name="l00194"></a><span class="lineno">  194</span>&#160;            <span class="keywordflow">raise</span> MissingHeaderException( self.data_source + </div><div class="line"><a name="l00195"></a><span class="lineno">  195</span>&#160;                                          <span class="stringliteral">&#39;.check_headers(): The following headers are missing: &#39;</span> +</div><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;                                          <span class="stringliteral">&#39;, &#39;</span>.join( missing_headers ) + <span class="stringliteral">&#39; in the data file.&#39;</span> +</div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;                                          <span class="stringliteral">&#39; Please see the documentation of the parser for more information.&#39;</span> )</div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;        </div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;        </div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;        <span class="comment"># Check if there are several columns harboring the same name</span></div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;        uniq_list = []</div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;        duplicate_list = []</div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;        </div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;        <span class="keywordflow">for</span> n <span class="keywordflow">in</span> col_names:</div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;            <span class="comment"># NB: When a new header is found to have the same title than a previous column, </span></div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;            <span class="comment">#     Pandas uses its name and add a dot followed by a unique number.</span></div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;            nbis = n.split( <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;              </div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;            <span class="comment"># If the header name does not contain any dot, the header is clearly not a duplicate</span></div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;            <span class="keywordflow">if</span> ( len( nbis ) == 1 ):</div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;                uniq_list.append( n )</div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;                  </div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;            <span class="comment"># If the header contains dots in its name but does not finish with an integer, </span></div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;            <span class="comment"># then it cannot be a duplicate</span></div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;            <span class="keywordflow">elif</span> ( ( len( nbis ) &gt; 1 ) <span class="keywordflow">and</span> ( <span class="keywordflow">not</span> nbis[ -1 ].isdigit() ) ):</div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;                uniq_list.append( n )</div><div class="line"><a name="l00217"></a><span class="lineno">  217</span>&#160;                </div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;            <span class="comment"># If the header contains dots and finish with an integer, </span></div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;            <span class="comment"># then it looks like a duplicate</span></div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;            <span class="keywordflow">elif</span> ( ( len( nbis ) &gt; 1 ) <span class="keywordflow">and</span> ( nbis[ -1 ].isdigit() ) ):</div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;                </div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;                <span class="comment"># Then search if the header without its last integer exist in the uniq_list.</span></div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;                <span class="comment"># If not, then this is not a duplicate.</span></div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;                <span class="keywordflow">if</span> ( ( <span class="stringliteral">&#39;.&#39;</span>.join(nbis[ :-1 ] ) ) <span class="keywordflow">not</span> <span class="keywordflow">in</span> uniq_list ):</div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;                    uniq_list.append( n )</div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;                </div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;                <span class="comment"># Otherwise, if the header without its last integer exists in the uniq_list, </span></div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;                <span class="comment"># then it should probably be a duplicate</span></div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;                    <span class="comment"># If this is the first time this duplicate is detected, then add it to the list of </span></div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;                    <span class="comment"># potential duplicates</span></div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;                    <span class="keywordflow">if</span> ( <span class="stringliteral">&#39;.&#39;</span>.join( nbis[ :-1 ] ) <span class="keywordflow">not</span> <span class="keywordflow">in</span> duplicate_list ):</div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                        duplicate_list.append( <span class="stringliteral">&#39;.&#39;</span>.join( nbis[ :-1 ] ) )</div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                        </div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;        <span class="comment"># If there are several column names that seems to be duplcates, raise a DuplicateHeaderException</span></div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;        <span class="keywordflow">if</span> ( len( duplicate_list ) != 0 ):</div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;            <span class="keywordflow">raise</span> DuplicateHeaderException( self.data_source + </div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;                                            <span class="stringliteral">&#39;.check_headers(): The following column names: &#39;</span> + </div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;                                            <span class="stringliteral">&#39;, &#39;</span>.join( duplicate_list ) + </div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;                                            <span class="stringliteral">&#39; have been identified as potential duplicates in the header&#39;</span> +</div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;                                            <span class="stringliteral">&#39; of the file located at &#39;</span> + self.data_path +</div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;                                            <span class="stringliteral">&#39;. This parser will use the following columns &#39;</span> + </div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;                                            <span class="stringliteral">&#39;, &#39;</span>.join( self.HEADER_LIST ) + <span class="stringliteral">&#39; to process the file.&#39;</span> +</div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;                                            <span class="stringliteral">&#39; Please make sure the columns used to parse this file are the right ones.&#39;</span> )</div><div class="line"><a name="l00245"></a><span class="lineno">  245</span>&#160;</div><div class="line"><a name="l00246"></a><span class="lineno">  246</span>&#160;    </div><div class="line"><a name="l00247"></a><span class="lineno">  247</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a8cdab4b401f878c517c4b737b9ecf778"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8cdab4b401f878c517c4b737b9ecf778">&#9670;&nbsp;</a></span>execute()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.execute </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>execute </h2>
<p>This is an abstract method to execute the strategy.</p>
<p>For each parser inheriting from the current metaclass, the following main steps are performed when executing this method:</p><ul>
<li>The file content is imported.</li>
<li>The headers of the file are checked in order to make sure the file is provided in its expected format.</li>
<li>Several variables are assigned to some of the useful data contained in the DataManager main dictionary in order to access these data quickly and easily anywhere in the program. These include the dictionaries containing all the entries of the DSORF, DSTranscript, DSORFTranscriptAsso and UTDSTranscriptGeneConflict tables as well as the autoincrement variable containing the value to use as autoincrement (unique ID) when creating new entries.</li>
<li>The number of rows expected to be treated is saved and used to set up a counter allowing to follow the progression of the file whilst the algorithm is running.</li>
<li>Each line of the file is parsed and data are extracted and used to instantiate appropriate DenCellORF objects (i.e. to register new entries in the database).<ul>
<li>In order to perform the parsing, for each entry, the information provided are used to instantiate new objects. When a new object is created, it is compared to the existing one (in the dictionary or in 'tracking lists / dictionaries'). If an existing object is the same as the newly created (i.e. sharing the same primary key and/or attributes on which an unique constraint has been defined), the older one is used (and eventually updated using the new information). Otherwise, the new instance is added to the list of objects to insert in the database and the trace of this new object is kept (in the dictionaries / lists) for further use.</li>
<li>For an easier reading and when feasible the information that will be used several time (such as the chromosome name or the ORF ID) are saved in local variables prior to the creation of the objects.</li>
<li>Some additional information related to the creation of the DenCellORF objects:<ul>
<li>An 'autoincrement' variable is saved and increment of 1 at each new instantiation, allowing to define unique IDs when necessary.</li>
<li>For DSORF: The 'start_pos', 'stop_pos', 'sequence' and 'genomic_length' attributes are expected to be computed later during the LiftOver nad ComputeMissingInfo strategies.</li>
</ul>
</li>
<li>If the transcript ID is missing, an 'unkown' transcript is created.</li>
<li>If the gene symbol is missing, the program try to get it from the transcript ID or from the coordinates of the ORF (see documentation of the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ab0b04e0792f45b98d4111aa27c51ac93">initialize_ids()</a> methods). If it cannot be recovered, an 'unknown' gene is created.</li>
</ul>
</li>
</ul>
<dl class="section return"><dt>Returns</dt><dd>None </dd></dl>
<div class="fragment"><div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160;    <span class="keyword">def </span>execute():</div><div class="line"><a name="l00093"></a><span class="lineno">   93</span>&#160;        </div><div class="line"><a name="l00094"></a><span class="lineno">   94</span>&#160;        <span class="keywordflow">return</span> <span class="keywordtype">None</span></div><div class="line"><a name="l00095"></a><span class="lineno">   95</span>&#160;</div><div class="line"><a name="l00096"></a><span class="lineno">   96</span>&#160;    </div><div class="line"><a name="l00097"></a><span class="lineno">   97</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a10fea64161666d9bb47ee4b8553b52e5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a10fea64161666d9bb47ee4b8553b52e5">&#9670;&nbsp;</a></span>find_gene_from_symbol()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.find_gene_from_symbol </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>symbol</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr_name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>find_gene_from_symbol </h2>
<p>This method allows to get the Gene entry related to a provided symbol or alias and chromosome. It first get all the entries of the Gene table that are related to the provided alias and try to find if there is one of them which is located on the chromosome.</p>
<p>If there is one single gene on this chromosome, the method returns this gene. If there is one single gene associated with that symbol and that is missing its chromosome name, the method updates the chromosome name of the gene, log a GeneRefLogger warning and return the updated Gene. Otherwise, if no gene can be found using the provided alias, the method returns the exception.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">symbol</td><td>String - The symbol or alias provided in the source as being associated with this ORF. </td></tr>
    <tr><td class="paramname">chr_name</td><td>String - The chromosome name provided in the source for this ORF.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>gene: Gene, or RefGeneUpdateException, or RefConflictException, or RefGeneSearchException - Return the instance associated with the provided gene alias and chromosome pair. </dd>
<dd>
new_obj_to_insert: List - List of objects to insert if new objects have been created in the process. </dd></dl>
<div class="fragment"><div class="line"><a name="l00793"></a><span class="lineno">  793</span>&#160;    <span class="keyword">def </span>find_gene_from_symbol( self, symbol, chr_name ):</div><div class="line"><a name="l00794"></a><span class="lineno">  794</span>&#160;        </div><div class="line"><a name="l00795"></a><span class="lineno">  795</span>&#160;        <span class="comment"># Initialize the list of new objects that will have to be inserted</span></div><div class="line"><a name="l00796"></a><span class="lineno">  796</span>&#160;        new_obj_to_insert = []</div><div class="line"><a name="l00797"></a><span class="lineno">  797</span>&#160;        </div><div class="line"><a name="l00798"></a><span class="lineno">  798</span>&#160;        <span class="comment"># Get the necessary information from the DataManager</span></div><div class="line"><a name="l00799"></a><span class="lineno">  799</span>&#160;        all_genes = DataManager.get_instance().get_data( Constants.DM_ALL_GENES )</div><div class="line"><a name="l00800"></a><span class="lineno">  800</span>&#160;        all_genealiases = DataManager.get_instance().get_data( Constants.DM_ALL_GENEALIASES )</div><div class="line"><a name="l00801"></a><span class="lineno">  801</span>&#160;        all_genesfromaliases = DataManager.get_instance().get_data( Constants.DM_ALL_GENESFROMALIASES )</div><div class="line"><a name="l00802"></a><span class="lineno">  802</span>&#160;        all_gene_id_asso = DataManager.get_instance().get_data( Constants.DM_ALL_GENE_ID_ASSO )</div><div class="line"><a name="l00803"></a><span class="lineno">  803</span>&#160;        all_gene_symbol_asso = DataManager.get_instance().get_data( Constants.DM_ALL_GENE_SYMBOL_ASSO )</div><div class="line"><a name="l00804"></a><span class="lineno">  804</span>&#160;            </div><div class="line"><a name="l00805"></a><span class="lineno">  805</span>&#160;        <span class="comment"># Get all the gene IDs corresponding to this symbol using the UTGeneFromAlias table</span></div><div class="line"><a name="l00806"></a><span class="lineno">  806</span>&#160;        <span class="comment"># To do this, instantiate a new UTGeneFromAlias object and if this UTGeneFromAlias already</span></div><div class="line"><a name="l00807"></a><span class="lineno">  807</span>&#160;        <span class="comment"># exists (meaning there is at least a Gene and a GeneAlias entry for this symbol), get it.</span></div><div class="line"><a name="l00808"></a><span class="lineno">  808</span>&#160;        utgenefromalias = UTGeneFromAlias( alias = symbol )</div><div class="line"><a name="l00809"></a><span class="lineno">  809</span>&#160;        </div><div class="line"><a name="l00810"></a><span class="lineno">  810</span>&#160;        existing_utgenefromalias = all_genesfromaliases.get( utgenefromalias )</div><div class="line"><a name="l00811"></a><span class="lineno">  811</span>&#160;        </div><div class="line"><a name="l00812"></a><span class="lineno">  812</span>&#160;        <span class="keywordflow">if</span> existing_utgenefromalias:</div><div class="line"><a name="l00813"></a><span class="lineno">  813</span>&#160;            utgenefromalias = existing_utgenefromalias</div><div class="line"><a name="l00814"></a><span class="lineno">  814</span>&#160;                        </div><div class="line"><a name="l00815"></a><span class="lineno">  815</span>&#160;            <span class="comment"># Get the list of gene IDS from the string</span></div><div class="line"><a name="l00816"></a><span class="lineno">  816</span>&#160;            gene_ids = utgenefromalias.gene_ids.split( Constants.GENEFROMALIAS_SEPARATOR_IDS )</div><div class="line"><a name="l00817"></a><span class="lineno">  817</span>&#160;                </div><div class="line"><a name="l00818"></a><span class="lineno">  818</span>&#160;            <span class="comment"># Get the list of Gene objects corresponding to these gene IDs</span></div><div class="line"><a name="l00819"></a><span class="lineno">  819</span>&#160;            list_of_genes = []</div><div class="line"><a name="l00820"></a><span class="lineno">  820</span>&#160;            <span class="keywordflow">for</span> id <span class="keywordflow">in</span> gene_ids:</div><div class="line"><a name="l00821"></a><span class="lineno">  821</span>&#160;                list_of_genes += all_gene_id_asso.get( id )</div><div class="line"><a name="l00822"></a><span class="lineno">  822</span>&#160;                    </div><div class="line"><a name="l00823"></a><span class="lineno">  823</span>&#160;            <span class="comment"># If there is an unique gene located on the expected chromosome or one single gene</span></div><div class="line"><a name="l00824"></a><span class="lineno">  824</span>&#160;            <span class="comment"># without chromosome name, get it. If this is not possible, catch the exception </span></div><div class="line"><a name="l00825"></a><span class="lineno">  825</span>&#160;            <span class="comment"># returned.</span></div><div class="line"><a name="l00826"></a><span class="lineno">  826</span>&#160;            <span class="comment"># NB: See the documentation of the get_uniq_gene_on_chromosome() method for more</span></div><div class="line"><a name="l00827"></a><span class="lineno">  827</span>&#160;            <span class="comment">#     information about this step.</span></div><div class="line"><a name="l00828"></a><span class="lineno">  828</span>&#160;            gene = ParserStrategy.get_uniq_gene_on_chromosome( chr_name, list_of_genes )</div><div class="line"><a name="l00829"></a><span class="lineno">  829</span>&#160;            </div><div class="line"><a name="l00830"></a><span class="lineno">  830</span>&#160;            <span class="keywordflow">if</span> isinstance( gene, RefGeneUpdateException ):</div><div class="line"><a name="l00831"></a><span class="lineno">  831</span>&#160;                GeneRefLogger.get_instance().gene_update_warning( gene.get_message() +</div><div class="line"><a name="l00832"></a><span class="lineno">  832</span>&#160;                                                                  <span class="stringliteral">&#39; Information from &#39;</span> + </div><div class="line"><a name="l00833"></a><span class="lineno">  833</span>&#160;                                                                  self.data_source + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00834"></a><span class="lineno">  834</span>&#160;                <span class="comment"># Get the gene associated with the exception</span></div><div class="line"><a name="l00835"></a><span class="lineno">  835</span>&#160;                gene = gene.get_gene()</div><div class="line"><a name="l00836"></a><span class="lineno">  836</span>&#160;                </div><div class="line"><a name="l00837"></a><span class="lineno">  837</span>&#160;                <span class="comment"># Update the entries of the data manager </span></div><div class="line"><a name="l00838"></a><span class="lineno">  838</span>&#160;                <span class="comment"># registering this gene</span></div><div class="line"><a name="l00839"></a><span class="lineno">  839</span>&#160;                all_gene_id_asso[ gene.gene_id ] = [ gene ]</div><div class="line"><a name="l00840"></a><span class="lineno">  840</span>&#160;                all_genes[ gene ] = gene</div><div class="line"><a name="l00841"></a><span class="lineno">  841</span>&#160;                </div><div class="line"><a name="l00842"></a><span class="lineno">  842</span>&#160;                <span class="comment"># Add the entry to the list of objects to add to the session</span></div><div class="line"><a name="l00843"></a><span class="lineno">  843</span>&#160;                new_obj_to_insert.append( gene )</div><div class="line"><a name="l00844"></a><span class="lineno">  844</span>&#160;                </div><div class="line"><a name="l00845"></a><span class="lineno">  845</span>&#160;            <span class="comment"># If the object returned by the get_uniq_gene_on_chromosome() method is an other RefException, </span></div><div class="line"><a name="l00846"></a><span class="lineno">  846</span>&#160;            <span class="comment"># create a &quot;fake&quot; gene in order to keep the relational integrity of the database and store this</span></div><div class="line"><a name="l00847"></a><span class="lineno">  847</span>&#160;            <span class="comment"># Gene instance as an attribute of the exception.</span></div><div class="line"><a name="l00848"></a><span class="lineno">  848</span>&#160;            <span class="keywordflow">elif</span> ( isinstance( gene, RefConflictException ) </div><div class="line"><a name="l00849"></a><span class="lineno">  849</span>&#160;                   <span class="keywordflow">or</span> isinstance( gene, RefGeneSearchException ) ):</div><div class="line"><a name="l00850"></a><span class="lineno">  850</span>&#160;                </div><div class="line"><a name="l00851"></a><span class="lineno">  851</span>&#160;                <span class="comment"># Create a &quot;fake&quot; gene symbol</span></div><div class="line"><a name="l00852"></a><span class="lineno">  852</span>&#160;                fake_symbol = Constants.PREFIX_UNKNOWN_GENE + symbol + <span class="stringliteral">&#39;_chr&#39;</span> + str( chr_name )</div><div class="line"><a name="l00853"></a><span class="lineno">  853</span>&#160;                </div><div class="line"><a name="l00854"></a><span class="lineno">  854</span>&#160;                <span class="comment"># Create a new Gene entry</span></div><div class="line"><a name="l00855"></a><span class="lineno">  855</span>&#160;                fake_gene = Gene( gene_id = fake_symbol,</div><div class="line"><a name="l00856"></a><span class="lineno">  856</span>&#160;                                  chromosome = chr_name )</div><div class="line"><a name="l00857"></a><span class="lineno">  857</span>&#160;                </div><div class="line"><a name="l00858"></a><span class="lineno">  858</span>&#160;                <span class="comment"># Create a new GeneAlias entry</span></div><div class="line"><a name="l00859"></a><span class="lineno">  859</span>&#160;                genealias = GeneAlias( gene_id = fake_gene.gene_id,</div><div class="line"><a name="l00860"></a><span class="lineno">  860</span>&#160;                                       alias = fake_symbol,</div><div class="line"><a name="l00861"></a><span class="lineno">  861</span>&#160;                                       crossref = Constants.FAKE_CROSSREF )</div><div class="line"><a name="l00862"></a><span class="lineno">  862</span>&#160;                </div><div class="line"><a name="l00863"></a><span class="lineno">  863</span>&#160;                <span class="comment"># Keep record of the new gene in the UTGeneFromAlias table</span></div><div class="line"><a name="l00864"></a><span class="lineno">  864</span>&#160;                utgenefromalias = UTGeneFromAlias( alias = fake_gene.gene_id,</div><div class="line"><a name="l00865"></a><span class="lineno">  865</span>&#160;                                                   gene_ids = fake_gene.gene_id )</div><div class="line"><a name="l00866"></a><span class="lineno">  866</span>&#160;                </div><div class="line"><a name="l00867"></a><span class="lineno">  867</span>&#160;                <span class="comment"># Add the newly created Gene, GeneAlias and UTGeneFromAlias </span></div><div class="line"><a name="l00868"></a><span class="lineno">  868</span>&#160;                <span class="comment"># entries to the list of objects to insert</span></div><div class="line"><a name="l00869"></a><span class="lineno">  869</span>&#160;                new_obj_to_insert.append( fake_gene )              </div><div class="line"><a name="l00870"></a><span class="lineno">  870</span>&#160;                new_obj_to_insert.append( genealias )</div><div class="line"><a name="l00871"></a><span class="lineno">  871</span>&#160;                new_obj_to_insert.append( utgenefromalias )</div><div class="line"><a name="l00872"></a><span class="lineno">  872</span>&#160;                </div><div class="line"><a name="l00873"></a><span class="lineno">  873</span>&#160;                <span class="comment"># Keep record of these new objects in the DataManager</span></div><div class="line"><a name="l00874"></a><span class="lineno">  874</span>&#160;                all_genes[ fake_gene ] = fake_gene</div><div class="line"><a name="l00875"></a><span class="lineno">  875</span>&#160;                all_genealiases[ genealias ] = genealias</div><div class="line"><a name="l00876"></a><span class="lineno">  876</span>&#160;                all_genesfromaliases[ utgenefromalias ] = utgenefromalias</div><div class="line"><a name="l00877"></a><span class="lineno">  877</span>&#160;                all_gene_id_asso[ fake_gene.gene_id ] = [ fake_gene ]</div><div class="line"><a name="l00878"></a><span class="lineno">  878</span>&#160;                all_gene_symbol_asso[ ( symbol, chr_name ) ] = fake_gene</div><div class="line"><a name="l00879"></a><span class="lineno">  879</span>&#160;                </div><div class="line"><a name="l00880"></a><span class="lineno">  880</span>&#160;                <span class="comment"># Register this Gene instance as an attribute of the RefException</span></div><div class="line"><a name="l00881"></a><span class="lineno">  881</span>&#160;                gene.set_gene( fake_gene )</div><div class="line"><a name="l00882"></a><span class="lineno">  882</span>&#160;                    </div><div class="line"><a name="l00883"></a><span class="lineno">  883</span>&#160;                </div><div class="line"><a name="l00884"></a><span class="lineno">  884</span>&#160;        <span class="comment"># Otherwise, if there is no Gene nor GeneAlias entry for this symbol, </span></div><div class="line"><a name="l00885"></a><span class="lineno">  885</span>&#160;        <span class="comment"># create a new Gene and GeneAlias, keep record of them and log a GeneRef warning.  </span></div><div class="line"><a name="l00886"></a><span class="lineno">  886</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00887"></a><span class="lineno">  887</span>&#160;            <span class="comment"># Log the missing reference</span></div><div class="line"><a name="l00888"></a><span class="lineno">  888</span>&#160;            GeneRefLogger.get_instance().missing_ref_warning( <span class="stringliteral">&#39;No gene has been found in the database&#39;</span> +</div><div class="line"><a name="l00889"></a><span class="lineno">  889</span>&#160;                                                              <span class="stringliteral">&#39; for the symbol / alias: &quot;&#39;</span> + str( symbol ) + </div><div class="line"><a name="l00890"></a><span class="lineno">  890</span>&#160;                                                              <span class="stringliteral">&#39;&quot; (alias from &#39;</span> + self.data_source + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l00891"></a><span class="lineno">  891</span>&#160;                </div><div class="line"><a name="l00892"></a><span class="lineno">  892</span>&#160;            <span class="comment"># Create a new Gene entry</span></div><div class="line"><a name="l00893"></a><span class="lineno">  893</span>&#160;            gene = Gene( gene_id = symbol,</div><div class="line"><a name="l00894"></a><span class="lineno">  894</span>&#160;                         chromosome = chr_name )</div><div class="line"><a name="l00895"></a><span class="lineno">  895</span>&#160;                </div><div class="line"><a name="l00896"></a><span class="lineno">  896</span>&#160;            <span class="comment"># Create a new GeneAlias entry                                  </span></div><div class="line"><a name="l00897"></a><span class="lineno">  897</span>&#160;            genealias = GeneAlias( gene_id = gene.gene_id, </div><div class="line"><a name="l00898"></a><span class="lineno">  898</span>&#160;                                   alias = gene.gene_id,</div><div class="line"><a name="l00899"></a><span class="lineno">  899</span>&#160;                                   crossref = self.data_source )</div><div class="line"><a name="l00900"></a><span class="lineno">  900</span>&#160;            </div><div class="line"><a name="l00901"></a><span class="lineno">  901</span>&#160;            <span class="comment"># Keep record of the new gene in the UTGeneFromAlias table</span></div><div class="line"><a name="l00902"></a><span class="lineno">  902</span>&#160;            utgenefromalias = UTGeneFromAlias( alias = gene.gene_id,</div><div class="line"><a name="l00903"></a><span class="lineno">  903</span>&#160;                                               gene_ids = gene.gene_id )</div><div class="line"><a name="l00904"></a><span class="lineno">  904</span>&#160;                                    </div><div class="line"><a name="l00905"></a><span class="lineno">  905</span>&#160;            <span class="comment"># Add the newly created Gene, GeneAlias and UTGeneFromAlias instances </span></div><div class="line"><a name="l00906"></a><span class="lineno">  906</span>&#160;            <span class="comment"># to the list of objects to insert</span></div><div class="line"><a name="l00907"></a><span class="lineno">  907</span>&#160;            new_obj_to_insert.append( gene )</div><div class="line"><a name="l00908"></a><span class="lineno">  908</span>&#160;            new_obj_to_insert.append( genealias )</div><div class="line"><a name="l00909"></a><span class="lineno">  909</span>&#160;            new_obj_to_insert.append( utgenefromalias )</div><div class="line"><a name="l00910"></a><span class="lineno">  910</span>&#160;                        </div><div class="line"><a name="l00911"></a><span class="lineno">  911</span>&#160;            <span class="comment"># Keep record of these new objects in the DataManager</span></div><div class="line"><a name="l00912"></a><span class="lineno">  912</span>&#160;            all_genes[ gene ] = gene</div><div class="line"><a name="l00913"></a><span class="lineno">  913</span>&#160;            all_genealiases[ genealias ] = genealias</div><div class="line"><a name="l00914"></a><span class="lineno">  914</span>&#160;            all_genesfromaliases[ utgenefromalias ] = utgenefromalias</div><div class="line"><a name="l00915"></a><span class="lineno">  915</span>&#160;            all_gene_id_asso[ gene.gene_id ] = [ gene ]</div><div class="line"><a name="l00916"></a><span class="lineno">  916</span>&#160;            all_gene_symbol_asso[ ( symbol, chr_name ) ] = gene</div><div class="line"><a name="l00917"></a><span class="lineno">  917</span>&#160;                </div><div class="line"><a name="l00918"></a><span class="lineno">  918</span>&#160;            <span class="comment"># Log the creation of this new Gene entry</span></div><div class="line"><a name="l00919"></a><span class="lineno">  919</span>&#160;            GeneRefLogger.get_instance().new_entry_warning( <span class="stringliteral">&#39;As no gene has been found associated with symbol &quot;&#39;</span> + </div><div class="line"><a name="l00920"></a><span class="lineno">  920</span>&#160;                                                            symbol + <span class="stringliteral">&#39;&quot; in the database, a new Gene and&#39;</span> +</div><div class="line"><a name="l00921"></a><span class="lineno">  921</span>&#160;                                                            <span class="stringliteral">&#39; GeneAlias have been created on the chromosome &#39;</span> + </div><div class="line"><a name="l00922"></a><span class="lineno">  922</span>&#160;                                                            chr_name + <span class="stringliteral">&#39; using this symbol&#39;</span> +</div><div class="line"><a name="l00923"></a><span class="lineno">  923</span>&#160;                                                            <span class="stringliteral">&#39; (alias and chromosome from &#39;</span> + </div><div class="line"><a name="l00924"></a><span class="lineno">  924</span>&#160;                                                            self.data_source + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l00925"></a><span class="lineno">  925</span>&#160;        </div><div class="line"><a name="l00926"></a><span class="lineno">  926</span>&#160;        <span class="keywordflow">return</span> ( gene, new_obj_to_insert )</div><div class="line"><a name="l00927"></a><span class="lineno">  927</span>&#160;    </div><div class="line"><a name="l00928"></a><span class="lineno">  928</span>&#160;        </div><div class="line"><a name="l00929"></a><span class="lineno">  929</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="a846efdd59ec331984a46aa80a46afcf2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a846efdd59ec331984a46aa80a46afcf2">&#9670;&nbsp;</a></span>get_ensembl_db()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.get_ensembl_db </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>annotation_version</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>get_ensembl_db </h2>
<p>This is a static method allowing to download the appropriate Ensembl database which is used by the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ab0b04e0792f45b98d4111aa27c51ac93">initialize_ids()</a> method to find the gene symbol from the transcript ID or ORF coordinates when it is not provided in the source file.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">annotation_version</td><td>String or Integer - The Ensembl annotation version corresponding to the datasource.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>ensembl_db: EnsemblRelase object - The Ensembl database object. </dd></dl>
<div class="fragment"><div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;    <span class="keyword">def </span>get_ensembl_db( annotation_version ):</div><div class="line"><a name="l00135"></a><span class="lineno">  135</span>&#160;        </div><div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;        <span class="comment"># Get the name of the species from the DataManager</span></div><div class="line"><a name="l00137"></a><span class="lineno">  137</span>&#160;        sp = DataManager.get_instance().get_data( Constants.SPECIES_FULL )</div><div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;        </div><div class="line"><a name="l00139"></a><span class="lineno">  139</span>&#160;        <span class="keywordflow">return</span> EnsemblUtil.get_ensembl_db( sp, annotation_version )</div><div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;</div><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;    </div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a9d3195a3bb1fd975359b771ebee910ef"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9d3195a3bb1fd975359b771ebee910ef">&#9670;&nbsp;</a></span>get_gene_from_conflict_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.get_gene_from_conflict_id </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>conflict_symbol</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr_name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>get_gene_from_conflict_id </h2>
<p>This method allows to get the unique Gene entry associated with a particular "fake" gene ID that have been created due to a conflict found on the gene ID / symbol during the merging of two DSTranscript. In order to reduce the computation time, all pairs (fake gene ID, chromosome) that have already been created are saved in a dictionary of the DataManager main dictionary. The instances are saved in this dictionary with associate to each unique (symbol, chromosome) key the correspond Gene. If this association has never been met before, new entries are created in the Gene, GeneAlias and UTGeneFromAlias tables.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">conflict_symbol</td><td>String - The "conflict" symbol of the gene (without its chromosome name). </td></tr>
    <tr><td class="paramname">chr_name</td><td>String - The chromosome name on which the transcript is located.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>gene: Gene - Return the Gene object associated with the provided conflict symbol and chromosome. </dd>
<dd>
new_obj_to_insert: List - List of objects to insert if new objects have been created in the process. </dd></dl>
<div class="fragment"><div class="line"><a name="l01148"></a><span class="lineno"> 1148</span>&#160;    <span class="keyword">def </span>get_gene_from_conflict_id( self, conflict_symbol, chr_name ):</div><div class="line"><a name="l01149"></a><span class="lineno"> 1149</span>&#160;        </div><div class="line"><a name="l01150"></a><span class="lineno"> 1150</span>&#160;        <span class="comment"># If the chromosome is known, add it to the name of the &quot;conflict&quot; symbol</span></div><div class="line"><a name="l01151"></a><span class="lineno"> 1151</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> ParserStrategy.is_empty( chr_name ) ):</div><div class="line"><a name="l01152"></a><span class="lineno"> 1152</span>&#160;            conflict_symbol = conflict_symbol + <span class="stringliteral">&#39;_chr_&#39;</span> + chr_name</div><div class="line"><a name="l01153"></a><span class="lineno"> 1153</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l01154"></a><span class="lineno"> 1154</span>&#160;            conflict_symbol = conflict_symbol + <span class="stringliteral">&#39;_chr_UNKNOWN&#39;</span></div><div class="line"><a name="l01155"></a><span class="lineno"> 1155</span>&#160;        </div><div class="line"><a name="l01156"></a><span class="lineno"> 1156</span>&#160;        <span class="comment"># Initialize the list of new objects that will have to be inserted</span></div><div class="line"><a name="l01157"></a><span class="lineno"> 1157</span>&#160;        new_obj_to_insert = []</div><div class="line"><a name="l01158"></a><span class="lineno"> 1158</span>&#160;        </div><div class="line"><a name="l01159"></a><span class="lineno"> 1159</span>&#160;        <span class="comment"># Get the necessary information from the DataManager</span></div><div class="line"><a name="l01160"></a><span class="lineno"> 1160</span>&#160;        all_gene_symbol_asso = DataManager.get_instance().get_data( Constants.DM_ALL_GENE_SYMBOL_ASSO )</div><div class="line"><a name="l01161"></a><span class="lineno"> 1161</span>&#160;        </div><div class="line"><a name="l01162"></a><span class="lineno"> 1162</span>&#160;        all_genes = DataManager.get_instance().get_data( Constants.DM_ALL_GENES )</div><div class="line"><a name="l01163"></a><span class="lineno"> 1163</span>&#160;        all_genealiases = DataManager.get_instance().get_data( Constants.DM_ALL_GENEALIASES )</div><div class="line"><a name="l01164"></a><span class="lineno"> 1164</span>&#160;        all_genesfromaliases = DataManager.get_instance().get_data( Constants.DM_ALL_GENESFROMALIASES )</div><div class="line"><a name="l01165"></a><span class="lineno"> 1165</span>&#160;        all_gene_id_asso = DataManager.get_instance().get_data( Constants.DM_ALL_GENE_ID_ASSO )</div><div class="line"><a name="l01166"></a><span class="lineno"> 1166</span>&#160;        all_gene_symbol_asso = DataManager.get_instance().get_data( Constants.DM_ALL_GENE_SYMBOL_ASSO )</div><div class="line"><a name="l01167"></a><span class="lineno"> 1167</span>&#160;        </div><div class="line"><a name="l01168"></a><span class="lineno"> 1168</span>&#160;        <span class="comment"># If this &quot;conflict&quot; symbol has already been created get it</span></div><div class="line"><a name="l01169"></a><span class="lineno"> 1169</span>&#160;        existing_gene = all_gene_symbol_asso.get( ( conflict_symbol, chr_name ) )</div><div class="line"><a name="l01170"></a><span class="lineno"> 1170</span>&#160;        </div><div class="line"><a name="l01171"></a><span class="lineno"> 1171</span>&#160;        <span class="comment"># Otherwise, create a new entry in the Gene, GeneAlias and UTGeneFromAlias tables and log it</span></div><div class="line"><a name="l01172"></a><span class="lineno"> 1172</span>&#160;        <span class="keywordflow">if</span> <span class="keywordflow">not</span> existing_gene:</div><div class="line"><a name="l01173"></a><span class="lineno"> 1173</span>&#160;                </div><div class="line"><a name="l01174"></a><span class="lineno"> 1174</span>&#160;            <span class="comment"># Create a new Gene entry</span></div><div class="line"><a name="l01175"></a><span class="lineno"> 1175</span>&#160;            gene = Gene( gene_id = conflict_symbol,</div><div class="line"><a name="l01176"></a><span class="lineno"> 1176</span>&#160;                         chromosome = chr_name )</div><div class="line"><a name="l01177"></a><span class="lineno"> 1177</span>&#160;                </div><div class="line"><a name="l01178"></a><span class="lineno"> 1178</span>&#160;            <span class="comment"># Create a new GeneAlias entry</span></div><div class="line"><a name="l01179"></a><span class="lineno"> 1179</span>&#160;            genealias = GeneAlias( gene_id = gene.gene_id, </div><div class="line"><a name="l01180"></a><span class="lineno"> 1180</span>&#160;                                   alias = gene.gene_id,</div><div class="line"><a name="l01181"></a><span class="lineno"> 1181</span>&#160;                                   crossref = self.data_source )</div><div class="line"><a name="l01182"></a><span class="lineno"> 1182</span>&#160;            </div><div class="line"><a name="l01183"></a><span class="lineno"> 1183</span>&#160;            <span class="comment"># Keep record of the new gene in the UTGeneFromAlias table</span></div><div class="line"><a name="l01184"></a><span class="lineno"> 1184</span>&#160;            utgenefromalias = UTGeneFromAlias( alias = gene.gene_id,</div><div class="line"><a name="l01185"></a><span class="lineno"> 1185</span>&#160;                                               gene_ids = gene.gene_id )</div><div class="line"><a name="l01186"></a><span class="lineno"> 1186</span>&#160;            </div><div class="line"><a name="l01187"></a><span class="lineno"> 1187</span>&#160;            <span class="comment"># Add the newly created Gene, GeneAlias and UTGeneFromAlias entries to the list of objects to insert</span></div><div class="line"><a name="l01188"></a><span class="lineno"> 1188</span>&#160;            new_obj_to_insert.append( gene )</div><div class="line"><a name="l01189"></a><span class="lineno"> 1189</span>&#160;            new_obj_to_insert.append( genealias )</div><div class="line"><a name="l01190"></a><span class="lineno"> 1190</span>&#160;            new_obj_to_insert.append( utgenefromalias )</div><div class="line"><a name="l01191"></a><span class="lineno"> 1191</span>&#160;            </div><div class="line"><a name="l01192"></a><span class="lineno"> 1192</span>&#160;            <span class="comment"># Keep record of these new objects in the DataManager</span></div><div class="line"><a name="l01193"></a><span class="lineno"> 1193</span>&#160;            all_genes[ gene ] = gene</div><div class="line"><a name="l01194"></a><span class="lineno"> 1194</span>&#160;            all_genealiases[ genealias ] = genealias</div><div class="line"><a name="l01195"></a><span class="lineno"> 1195</span>&#160;            all_genesfromaliases[ utgenefromalias ] = utgenefromalias</div><div class="line"><a name="l01196"></a><span class="lineno"> 1196</span>&#160;            all_gene_id_asso[ gene.gene_id ] = [ gene ]</div><div class="line"><a name="l01197"></a><span class="lineno"> 1197</span>&#160;            all_gene_symbol_asso[ ( conflict_symbol, chr_name ) ] = gene</div><div class="line"><a name="l01198"></a><span class="lineno"> 1198</span>&#160;                </div><div class="line"><a name="l01199"></a><span class="lineno"> 1199</span>&#160;            <span class="comment"># Log the creation of this new gene</span></div><div class="line"><a name="l01200"></a><span class="lineno"> 1200</span>&#160;            GeneRefLogger.get_instance().new_entry_warning( <span class="stringliteral">&#39;A new Gene and GeneAlias entries have been&#39;</span> +</div><div class="line"><a name="l01201"></a><span class="lineno"> 1201</span>&#160;                                                            <span class="stringliteral">&#39; created on chromosome &#39;</span> + chr_name +</div><div class="line"><a name="l01202"></a><span class="lineno"> 1202</span>&#160;                                                            <span class="stringliteral">&#39; for the &quot;conflicting&quot; gene with gene ID &quot;&#39;</span> + </div><div class="line"><a name="l01203"></a><span class="lineno"> 1203</span>&#160;                                                            conflict_symbol + <span class="stringliteral">&#39; (conflict found in &#39;</span> + </div><div class="line"><a name="l01204"></a><span class="lineno"> 1204</span>&#160;                                                            self.data_source + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l01205"></a><span class="lineno"> 1205</span>&#160;            </div><div class="line"><a name="l01206"></a><span class="lineno"> 1206</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l01207"></a><span class="lineno"> 1207</span>&#160;            gene = existing_gene</div><div class="line"><a name="l01208"></a><span class="lineno"> 1208</span>&#160;                </div><div class="line"><a name="l01209"></a><span class="lineno"> 1209</span>&#160;        <span class="keywordflow">return</span> ( gene, new_obj_to_insert )</div><div class="line"><a name="l01210"></a><span class="lineno"> 1210</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="a4251bb5eb9243ff185efba305843d86b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4251bb5eb9243ff185efba305843d86b">&#9670;&nbsp;</a></span>get_gene_from_symbol()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.get_gene_from_symbol </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>symbol</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>get_gene_from_symbol </h2>
<p>This method allows to get the unique Gene entry associated with a (symbol, chromosome) pair, when this is possible. In order to minimize the computation time, all associations that have already been searched are saved in a dictionary of the DataManager main dictionary. The object are saved in this dictionary with value for the (symbol, chromosome) key that may be either an instance of the Gene class or an instance of an exception (when it was not possible to find an unique gene corresponding to the pair). In such cases, the message related to the exception is logged as a GeneRef warning. If this association has never been searched, the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a10fea64161666d9bb47ee4b8553b52e5">find_gene_from_symbol()</a> method is called in order to get the instance corresponding to the (symbol, chromosome) pair.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">symbol</td><td>String - The symbol or alias of the gene. </td></tr>
    <tr><td class="paramname">chr_name</td><td>String - The chromosome name of the gene. </td></tr>
    <tr><td class="paramname">orf_id</td><td>String - The ORF ID related to this gene (as provided in the source file).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>gene: Gene - Return the Gene instance associated with the provided alias and chromosome. </dd>
<dd>
new_obj_to_insert: List - List of objects to insert if new objects have been created in the process. </dd></dl>
<div class="fragment"><div class="line"><a name="l00724"></a><span class="lineno">  724</span>&#160;    <span class="keyword">def </span>get_gene_from_symbol( self, symbol, chr_name, orf_id ):</div><div class="line"><a name="l00725"></a><span class="lineno">  725</span>&#160;        </div><div class="line"><a name="l00726"></a><span class="lineno">  726</span>&#160;        <span class="comment"># Initialize the list of new objects that will have to be inserted</span></div><div class="line"><a name="l00727"></a><span class="lineno">  727</span>&#160;        new_obj_to_insert = []</div><div class="line"><a name="l00728"></a><span class="lineno">  728</span>&#160;        </div><div class="line"><a name="l00729"></a><span class="lineno">  729</span>&#160;        <span class="comment"># Get the dictionary which associate to each (symbol, chromosome) tuple, </span></div><div class="line"><a name="l00730"></a><span class="lineno">  730</span>&#160;        <span class="comment"># the Gene or the Exception previously instantiated</span></div><div class="line"><a name="l00731"></a><span class="lineno">  731</span>&#160;        all_gene_symbol_asso = DataManager.get_instance().get_data( Constants.DM_ALL_GENE_SYMBOL_ASSO )</div><div class="line"><a name="l00732"></a><span class="lineno">  732</span>&#160;        </div><div class="line"><a name="l00733"></a><span class="lineno">  733</span>&#160;        <span class="comment"># If the gene associated with this (symbol, chromosome) tuple has already been searched, get it.</span></div><div class="line"><a name="l00734"></a><span class="lineno">  734</span>&#160;        <span class="comment"># Otherwise, search for it and record the result (Gene or Exception) in the DataManager dictionary.</span></div><div class="line"><a name="l00735"></a><span class="lineno">  735</span>&#160;        gene = all_gene_symbol_asso.get( ( symbol, chr_name ) )</div><div class="line"><a name="l00736"></a><span class="lineno">  736</span>&#160;        </div><div class="line"><a name="l00737"></a><span class="lineno">  737</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> gene ):</div><div class="line"><a name="l00738"></a><span class="lineno">  738</span>&#160;            <span class="comment"># Get the instance related to this tuple and store it in the DataManager</span></div><div class="line"><a name="l00739"></a><span class="lineno">  739</span>&#160;            <span class="comment"># NB: See the documentation of the find_gene_from_symbol() method for more</span></div><div class="line"><a name="l00740"></a><span class="lineno">  740</span>&#160;            <span class="comment">#     information about this step.</span></div><div class="line"><a name="l00741"></a><span class="lineno">  741</span>&#160;            ( gene, new_obj_to_insert ) = self.find_gene_from_symbol( symbol, chr_name )</div><div class="line"><a name="l00742"></a><span class="lineno">  742</span>&#160;            all_gene_symbol_asso[ ( symbol, chr_name ) ] = gene</div><div class="line"><a name="l00743"></a><span class="lineno">  743</span>&#160;                </div><div class="line"><a name="l00744"></a><span class="lineno">  744</span>&#160;        <span class="comment"># If the object associated with the (symbol, chr_name) tuple in the dictionary is an exception,</span></div><div class="line"><a name="l00745"></a><span class="lineno">  745</span>&#160;        <span class="comment"># log the appropriate GeneRefLogger warning and return the &quot;fake&quot; gene</span></div><div class="line"><a name="l00746"></a><span class="lineno">  746</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> isinstance( gene, Gene ) ):</div><div class="line"><a name="l00747"></a><span class="lineno">  747</span>&#160;            </div><div class="line"><a name="l00748"></a><span class="lineno">  748</span>&#160;            <span class="keywordflow">if</span> isinstance( gene, RefConflictException ):</div><div class="line"><a name="l00749"></a><span class="lineno">  749</span>&#160;                GeneRefLogger.get_instance().conflict_info_warning( gene.get_message() + </div><div class="line"><a name="l00750"></a><span class="lineno">  750</span>&#160;                                                                    <span class="stringliteral">&#39; Hence the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00751"></a><span class="lineno">  751</span>&#160;                                                                    orf_id + <span class="stringliteral">&#39;&quot; (from &#39;</span> + self.data_source  + </div><div class="line"><a name="l00752"></a><span class="lineno">  752</span>&#160;                                                                    <span class="stringliteral">&#39;) will be associated with the &quot;fake&quot; gene &quot;&#39;</span> + </div><div class="line"><a name="l00753"></a><span class="lineno">  753</span>&#160;                                                                    str( gene.get_gene().gene_id ) + <span class="stringliteral">&#39;&quot;.&#39;</span> )</div><div class="line"><a name="l00754"></a><span class="lineno">  754</span>&#160;                </div><div class="line"><a name="l00755"></a><span class="lineno">  755</span>&#160;            <span class="keywordflow">elif</span> isinstance( gene, RefGeneSearchException ):</div><div class="line"><a name="l00756"></a><span class="lineno">  756</span>&#160;                GeneRefLogger.get_instance().gene_search_warning( gene.get_message() + </div><div class="line"><a name="l00757"></a><span class="lineno">  757</span>&#160;                                                                  <span class="stringliteral">&#39; Hence the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00758"></a><span class="lineno">  758</span>&#160;                                                                  orf_id + <span class="stringliteral">&#39;&quot; (from &#39;</span> + self.data_source  +</div><div class="line"><a name="l00759"></a><span class="lineno">  759</span>&#160;                                                                  <span class="stringliteral">&#39;) will be associated with the &quot;fake&quot; gene &quot;&#39;</span> + </div><div class="line"><a name="l00760"></a><span class="lineno">  760</span>&#160;                                                                  str( gene.get_gene().gene_id ) + <span class="stringliteral">&#39;&quot;.&#39;</span> )</div><div class="line"><a name="l00761"></a><span class="lineno">  761</span>&#160;            gene = gene.get_gene()</div><div class="line"><a name="l00762"></a><span class="lineno">  762</span>&#160;                </div><div class="line"><a name="l00763"></a><span class="lineno">  763</span>&#160;        <span class="keywordflow">return</span> ( gene, new_obj_to_insert )</div><div class="line"><a name="l00764"></a><span class="lineno">  764</span>&#160;        </div><div class="line"><a name="l00765"></a><span class="lineno">  765</span>&#160;        </div><div class="line"><a name="l00766"></a><span class="lineno">  766</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="a37ad31eeda67bd00e0aba6bfe491eecf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a37ad31eeda67bd00e0aba6bfe491eecf">&#9670;&nbsp;</a></span>get_uniq_gene_on_chromosome()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.get_uniq_gene_on_chromosome </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>list_of_genes</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>get_uniq_gene_on_chromosome </h2>
<p>This is a static method that allows to get the only gene located (or suspected to be located) on a particular chromosome in a provided list of genes .</p>
<p>Please find below the algorithm run by this function to return the gene (or to return an exception containing an unique gene): If there is one single gene in the list:</p><ul>
<li>and the expected chromosome name is known:<ul>
<li>and this single gene is missing its chromosome information, then update the gene and return it (using a RefGeneUpdateException).</li>
<li>and the chromosome name of this single gene is known:<ul>
<li>and the two chromosome names are the same, then return this gene.</li>
<li>and the two chromosomes names are not equal but are in fact the same (in a biological way: this happens when the chromosome name equals 'X' or 'Y' while the gene chromosome equals 'X|Y', or in the opposite situation, i.e. the chromosome name equals 'X|Y' whilst the gene chromosome is 'X' or 'Y'), then return this gene.</li>
<li>and the two chromosome names are different, then return a RefConflictException.</li>
</ul>
</li>
</ul>
</li>
<li>and the expected chromosome name is not provided, then return the gene.</li>
</ul>
<p>Else, if there are several genes in the list:</p><ul>
<li>and there are genes which are on the expected chromosome:<ul>
<li>and there is one single gene on the expected chromosome, then return this gene.</li>
<li>and there are several genes on the expected chromosome, then return a RefGeneSearchException.</li>
</ul>
</li>
<li>and there is no gene on the expected chromosome<ul>
<li>and the expected chromosome is 'X' or 'Y' while there are genes of the list on 'X|Y',<ul>
<li>and there is one single gene in that list, then return this gene.</li>
<li>and there are several genes on the expected chromosome, then return a RefGeneSearchException.</li>
</ul>
</li>
<li>and the expected chromosome name is provided and there are genes missing their chromosome information:<ul>
<li>and there is one single gene missing its chromosome information, then update the gene and return it (through a RefGeneUpdateException).</li>
<li>and there are several genes missing their chromosome information, then return a RefGeneSearchException.</li>
</ul>
</li>
<li>and the expected chromosome name is not provided OR there is no gene of the list missing its chromosome information, then return a RefGeneSearchException. NB: Please not that the exceptions (RefGeneUpdateException, RefConflictException and RefGeneSearchException) are returned by this method and NOT raised!</li>
</ul>
</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chr</td><td>String - The chromosome name of the expected location of the gene. </td></tr>
    <tr><td class="paramname">list_of_genes</td><td>List of Gene objects - The list of genes to search in.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Gene - The instance corresponding to the single gene located on the expected chromosome. </dd>
<dd>
RefGeneUpdateException: When one single gene missing its chromosome information is found, then this location is updated using the expected location. </dd>
<dd>
RefConflictException: When there is one single gene is in the list and its chromosome location does not match the expected one. </dd>
<dd>
RefGeneSearchException: When there are several genes on the expected chromosome. </dd>
<dd>
RefGeneSearchException: When there is no gene on the expected chromosome and. several missing their chromosome name. </dd>
<dd>
RefGeneSearchException: When there is no gene on the expected chromosome and no one missing its chromosome name. </dd></dl>
<div class="fragment"><div class="line"><a name="l00989"></a><span class="lineno">  989</span>&#160;    <span class="keyword">def </span>get_uniq_gene_on_chromosome( chr, list_of_genes ):</div><div class="line"><a name="l00990"></a><span class="lineno">  990</span>&#160;        </div><div class="line"><a name="l00991"></a><span class="lineno">  991</span>&#160;        <span class="comment"># If there is a single gene in the list, then get it first,</span></div><div class="line"><a name="l00992"></a><span class="lineno">  992</span>&#160;        <span class="keywordflow">if</span> ( len( list_of_genes ) == 1 ):</div><div class="line"><a name="l00993"></a><span class="lineno">  993</span>&#160;            gene = list_of_genes[ 0 ]</div><div class="line"><a name="l00994"></a><span class="lineno">  994</span>&#160;            </div><div class="line"><a name="l00995"></a><span class="lineno">  995</span>&#160;            <span class="comment"># If the expected chromosome of the gene is known,</span></div><div class="line"><a name="l00996"></a><span class="lineno">  996</span>&#160;            <span class="keywordflow">if</span> ( chr != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00997"></a><span class="lineno">  997</span>&#160;                <span class="comment"># and if this single gene is missing its chromosome information, then update it using the </span></div><div class="line"><a name="l00998"></a><span class="lineno">  998</span>&#160;                <span class="comment"># expected location and return a RefGeneUpdateException</span></div><div class="line"><a name="l00999"></a><span class="lineno">  999</span>&#160;                <span class="keywordflow">if</span> ( gene.chromosome == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l01000"></a><span class="lineno"> 1000</span>&#160;                    gene.chromosome = chr</div><div class="line"><a name="l01001"></a><span class="lineno"> 1001</span>&#160;                    <span class="keywordflow">return</span> RefGeneUpdateException( <span class="stringliteral">&#39;The gene with ID &quot;&#39;</span> + str( gene.gene_id ) + </div><div class="line"><a name="l01002"></a><span class="lineno"> 1002</span>&#160;                                                   <span class="stringliteral">&#39;&quot; was missing its chromosome information.&#39;</span> +</div><div class="line"><a name="l01003"></a><span class="lineno"> 1003</span>&#160;                                                   <span class="stringliteral">&#39; This gene has been associated with chromosome &#39;</span> + </div><div class="line"><a name="l01004"></a><span class="lineno"> 1004</span>&#160;                                                   chr + <span class="stringliteral">&#39;.&#39;</span>, </div><div class="line"><a name="l01005"></a><span class="lineno"> 1005</span>&#160;                                                   gene = gene )</div><div class="line"><a name="l01006"></a><span class="lineno"> 1006</span>&#160;                    </div><div class="line"><a name="l01007"></a><span class="lineno"> 1007</span>&#160;                <span class="comment"># Else, if this single gene has a chromosome information</span></div><div class="line"><a name="l01008"></a><span class="lineno"> 1008</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l01009"></a><span class="lineno"> 1009</span>&#160;                    <span class="comment"># and this chromosome is the same as the expected one, then return the gene</span></div><div class="line"><a name="l01010"></a><span class="lineno"> 1010</span>&#160;                    <span class="keywordflow">if</span> ( gene.chromosome == chr ):</div><div class="line"><a name="l01011"></a><span class="lineno"> 1011</span>&#160;                        <span class="keywordflow">return</span> gene</div><div class="line"><a name="l01012"></a><span class="lineno"> 1012</span>&#160;                    </div><div class="line"><a name="l01013"></a><span class="lineno"> 1013</span>&#160;                    <span class="comment"># and this chromosome does not equal the expected one but is in fact the same </span></div><div class="line"><a name="l01014"></a><span class="lineno"> 1014</span>&#160;                    <span class="comment"># (as it is equal to X or Y),</span></div><div class="line"><a name="l01015"></a><span class="lineno"> 1015</span>&#160;                    <span class="comment"># then return the gene</span></div><div class="line"><a name="l01016"></a><span class="lineno"> 1016</span>&#160;                    <span class="keywordflow">elif</span> ( ( ( chr == Constants.CHROMOSOME_X ) </div><div class="line"><a name="l01017"></a><span class="lineno"> 1017</span>&#160;                             <span class="keywordflow">or</span> ( chr == Constants.CHROMOSOME_Y ) </div><div class="line"><a name="l01018"></a><span class="lineno"> 1018</span>&#160;                             <span class="keywordflow">or</span> ( chr == Constants.CHROMOSOME_NAME_XY ) )</div><div class="line"><a name="l01019"></a><span class="lineno"> 1019</span>&#160;                            <span class="keywordflow">and</span> ( gene.chromosome == Constants.CHROMOSOME_NAME_XY ) ):</div><div class="line"><a name="l01020"></a><span class="lineno"> 1020</span>&#160;                        <span class="keywordflow">return</span> gene</div><div class="line"><a name="l01021"></a><span class="lineno"> 1021</span>&#160;                    </div><div class="line"><a name="l01022"></a><span class="lineno"> 1022</span>&#160;                    <span class="comment"># and this chromosome does not equal the expected one but is in fact the sexual </span></div><div class="line"><a name="l01023"></a><span class="lineno"> 1023</span>&#160;                    <span class="comment"># chromosome and the chromosome registered in the gene entry is X or Y</span></div><div class="line"><a name="l01024"></a><span class="lineno"> 1024</span>&#160;                    <span class="keywordflow">elif</span> ( ( chr == Constants.CHROMOSOME_NAME_XY )</div><div class="line"><a name="l01025"></a><span class="lineno"> 1025</span>&#160;                           <span class="keywordflow">and</span> ( ( gene.chromosome == Constants.CHROMOSOME_Y )</div><div class="line"><a name="l01026"></a><span class="lineno"> 1026</span>&#160;                                 <span class="keywordflow">or</span> ( gene.chromosome == Constants.CHROMOSOME_X ) ) ):</div><div class="line"><a name="l01027"></a><span class="lineno"> 1027</span>&#160;                        <span class="comment"># Update the gene to change X or Y by the sexual chromosome (X|Y)</span></div><div class="line"><a name="l01028"></a><span class="lineno"> 1028</span>&#160;                        exception_message = ( <span class="stringliteral">&#39;The gene with ID &quot;&#39;</span> + str( gene.gene_id ) + </div><div class="line"><a name="l01029"></a><span class="lineno"> 1029</span>&#160;                                              <span class="stringliteral">&#39;&quot; is located on the sexual chromosome (&#39;</span> +</div><div class="line"><a name="l01030"></a><span class="lineno"> 1030</span>&#160;                                              str( chr ) + <span class="stringliteral">&#39;) while it was previously registered on&#39;</span> +</div><div class="line"><a name="l01031"></a><span class="lineno"> 1031</span>&#160;                                              <span class="stringliteral">&#39; the chromosome &#39;</span> + str( gene.chromosome ) + </div><div class="line"><a name="l01032"></a><span class="lineno"> 1032</span>&#160;                                              <span class="stringliteral">&#39; in the database. Hence, the chromosome of this gene has&#39;</span> +</div><div class="line"><a name="l01033"></a><span class="lineno"> 1033</span>&#160;                                              <span class="stringliteral">&#39; been set to &#39;</span> + str( chr ) + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l01034"></a><span class="lineno"> 1034</span>&#160;                        gene.chromosome = Constants.CHROMOSOME_NAME_XY</div><div class="line"><a name="l01035"></a><span class="lineno"> 1035</span>&#160;                        <span class="keywordflow">return</span> RefGeneUpdateException( exception_message,</div><div class="line"><a name="l01036"></a><span class="lineno"> 1036</span>&#160;                                                       gene = gene )</div><div class="line"><a name="l01037"></a><span class="lineno"> 1037</span>&#160;                        </div><div class="line"><a name="l01038"></a><span class="lineno"> 1038</span>&#160;                    <span class="comment"># otherwise, if the two chromosomes are different, then return a RefConflictException</span></div><div class="line"><a name="l01039"></a><span class="lineno"> 1039</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l01040"></a><span class="lineno"> 1040</span>&#160;                        <span class="keywordflow">return</span> RefConflictException( <span class="stringliteral">&#39;The gene with ID &quot;&#39;</span> + str( gene.gene_id ) + </div><div class="line"><a name="l01041"></a><span class="lineno"> 1041</span>&#160;                                                     <span class="stringliteral">&#39;&quot; is located on chromosome &#39;</span> +</div><div class="line"><a name="l01042"></a><span class="lineno"> 1042</span>&#160;                                                     str( gene.chromosome ) + <span class="stringliteral">&#39; according to the Gene table&#39;</span> +</div><div class="line"><a name="l01043"></a><span class="lineno"> 1043</span>&#160;                                                     <span class="stringliteral">&#39; while it was expected to be on chromosome &#39;</span> + str( chr ) + </div><div class="line"><a name="l01044"></a><span class="lineno"> 1044</span>&#160;                                                     <span class="stringliteral">&#39; according to the data parsed.&#39;</span> )</div><div class="line"><a name="l01045"></a><span class="lineno"> 1045</span>&#160;            </div><div class="line"><a name="l01046"></a><span class="lineno"> 1046</span>&#160;            <span class="comment"># Otherwise, if the expected location of the gene is not known, return this gene</span></div><div class="line"><a name="l01047"></a><span class="lineno"> 1047</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l01048"></a><span class="lineno"> 1048</span>&#160;                <span class="keywordflow">return</span> gene</div><div class="line"><a name="l01049"></a><span class="lineno"> 1049</span>&#160;        </div><div class="line"><a name="l01050"></a><span class="lineno"> 1050</span>&#160;                </div><div class="line"><a name="l01051"></a><span class="lineno"> 1051</span>&#160;        <span class="comment"># Else, if there are several genes in the list, then first create a dictionary that </span></div><div class="line"><a name="l01052"></a><span class="lineno"> 1052</span>&#160;        <span class="comment"># associate with each chromosome names and the lists of genes</span></div><div class="line"><a name="l01053"></a><span class="lineno"> 1053</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l01054"></a><span class="lineno"> 1054</span>&#160;            genes_chr_asso = {}</div><div class="line"><a name="l01055"></a><span class="lineno"> 1055</span>&#160;            <span class="keywordflow">for</span> g <span class="keywordflow">in</span> list_of_genes:</div><div class="line"><a name="l01056"></a><span class="lineno"> 1056</span>&#160;                <span class="keywordflow">if</span> ( g.chromosome <span class="keywordflow">not</span> <span class="keywordflow">in</span> genes_chr_asso.keys() ):</div><div class="line"><a name="l01057"></a><span class="lineno"> 1057</span>&#160;                    genes_chr_asso[ g.chromosome ] = [ g ]</div><div class="line"><a name="l01058"></a><span class="lineno"> 1058</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l01059"></a><span class="lineno"> 1059</span>&#160;                    genes_chr_asso[ g.chromosome ].append( g )</div><div class="line"><a name="l01060"></a><span class="lineno"> 1060</span>&#160;                    </div><div class="line"><a name="l01061"></a><span class="lineno"> 1061</span>&#160;            <span class="comment"># If there are genes of the list located on the expected chromosome,</span></div><div class="line"><a name="l01062"></a><span class="lineno"> 1062</span>&#160;            <span class="keywordflow">if</span> ( chr <span class="keywordflow">in</span> genes_chr_asso.keys() ):</div><div class="line"><a name="l01063"></a><span class="lineno"> 1063</span>&#160;                <span class="comment"># If there is one single gene on the expected chromosome, return it</span></div><div class="line"><a name="l01064"></a><span class="lineno"> 1064</span>&#160;                <span class="keywordflow">if</span> ( len( genes_chr_asso[ chr ] ) == 1 ):</div><div class="line"><a name="l01065"></a><span class="lineno"> 1065</span>&#160;                    <span class="keywordflow">return</span> genes_chr_asso[ chr ][ 0 ]</div><div class="line"><a name="l01066"></a><span class="lineno"> 1066</span>&#160;                </div><div class="line"><a name="l01067"></a><span class="lineno"> 1067</span>&#160;                <span class="comment"># If there are several genes on the expected chromosome, then return a RefGeneSearchException</span></div><div class="line"><a name="l01068"></a><span class="lineno"> 1068</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l01069"></a><span class="lineno"> 1069</span>&#160;                    <span class="keywordflow">if</span> ( chr != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l01070"></a><span class="lineno"> 1070</span>&#160;                        <span class="keywordflow">return</span> RefGeneSearchException( <span class="stringliteral">&#39;Several genes of the list have been found on the&#39;</span> +</div><div class="line"><a name="l01071"></a><span class="lineno"> 1071</span>&#160;                                                       <span class="stringliteral">&#39; chromosome &#39;</span> + chr + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l01072"></a><span class="lineno"> 1072</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l01073"></a><span class="lineno"> 1073</span>&#160;                        <span class="keywordflow">return</span> RefGeneSearchException( <span class="stringliteral">&#39;The expected location of the gene was not provided and&#39;</span> +</div><div class="line"><a name="l01074"></a><span class="lineno"> 1074</span>&#160;                                                       <span class="stringliteral">&#39; several genes of the list have been found without any&#39;</span> +</div><div class="line"><a name="l01075"></a><span class="lineno"> 1075</span>&#160;                                                       <span class="stringliteral">&#39; chromosome information.&#39;</span> )</div><div class="line"><a name="l01076"></a><span class="lineno"> 1076</span>&#160;                    </div><div class="line"><a name="l01077"></a><span class="lineno"> 1077</span>&#160;            <span class="comment"># If there is no gene of the list on the expected chromosome</span></div><div class="line"><a name="l01078"></a><span class="lineno"> 1078</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l01079"></a><span class="lineno"> 1079</span>&#160;                <span class="comment"># and the expected chromosome is &#39;X&#39; or &#39;Y&#39; while are genes </span></div><div class="line"><a name="l01080"></a><span class="lineno"> 1080</span>&#160;                <span class="comment"># located on &#39;X|Y&#39; according to the list</span></div><div class="line"><a name="l01081"></a><span class="lineno"> 1081</span>&#160;                <span class="keywordflow">if</span> ( ( ( chr == Constants.CHROMOSOME_X ) <span class="keywordflow">or</span> ( chr == Constants.CHROMOSOME_Y ) ) </div><div class="line"><a name="l01082"></a><span class="lineno"> 1082</span>&#160;                     <span class="keywordflow">and</span> ( Constants.CHROMOSOME_NAME_XY <span class="keywordflow">in</span> genes_chr_asso.keys() ) ):</div><div class="line"><a name="l01083"></a><span class="lineno"> 1083</span>&#160;                    <span class="comment"># and there is one single gene on &#39;X|Y&#39;, then return it</span></div><div class="line"><a name="l01084"></a><span class="lineno"> 1084</span>&#160;                    <span class="keywordflow">if</span> ( len( genes_chr_asso[ Constants.CHROMOSOME_NAME_XY ] ) == 1 ):</div><div class="line"><a name="l01085"></a><span class="lineno"> 1085</span>&#160;                        <span class="keywordflow">return</span> genes_chr_asso[ Constants.CHROMOSOME_NAME_XY ]</div><div class="line"><a name="l01086"></a><span class="lineno"> 1086</span>&#160;                    </div><div class="line"><a name="l01087"></a><span class="lineno"> 1087</span>&#160;                    <span class="comment"># and there are several genes on &#39;X|Y&#39;, then return a RefGeneSearchException</span></div><div class="line"><a name="l01088"></a><span class="lineno"> 1088</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l01089"></a><span class="lineno"> 1089</span>&#160;                        <span class="keywordflow">return</span> RefGeneSearchException( <span class="stringliteral">&#39;Several genes of the list have been found on the&#39;</span> +</div><div class="line"><a name="l01090"></a><span class="lineno"> 1090</span>&#160;                                                       <span class="stringliteral">&#39; chromosome &#39;</span> + Constants.CHROMOSOME_NAME_XY + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l01091"></a><span class="lineno"> 1091</span>&#160;                </div><div class="line"><a name="l01092"></a><span class="lineno"> 1092</span>&#160;                <span class="comment"># and the expected chromosome is provided, and there are genes </span></div><div class="line"><a name="l01093"></a><span class="lineno"> 1093</span>&#160;                <span class="comment"># missing their chromosome information </span></div><div class="line"><a name="l01094"></a><span class="lineno"> 1094</span>&#160;                <span class="keywordflow">elif</span> ( ( chr != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( <span class="keywordtype">None</span> <span class="keywordflow">in</span> genes_chr_asso.keys() ) ):</div><div class="line"><a name="l01095"></a><span class="lineno"> 1095</span>&#160;                    <span class="comment"># and there is one single gene missing its chromosome information, </span></div><div class="line"><a name="l01096"></a><span class="lineno"> 1096</span>&#160;                    <span class="comment"># then update this gene and return a RefGeneUpdateException</span></div><div class="line"><a name="l01097"></a><span class="lineno"> 1097</span>&#160;                    <span class="keywordflow">if</span> len( genes_chr_asso[ <span class="keywordtype">None</span> ] ) == 1:</div><div class="line"><a name="l01098"></a><span class="lineno"> 1098</span>&#160;                        gene = genes_chr_asso[ <span class="keywordtype">None</span> ][ 0 ]</div><div class="line"><a name="l01099"></a><span class="lineno"> 1099</span>&#160;                        gene.chromosome = chr</div><div class="line"><a name="l01100"></a><span class="lineno"> 1100</span>&#160;                        <span class="keywordflow">return</span> RefGeneUpdateException( <span class="stringliteral">&#39;The gene with ID &quot;&#39;</span> + str( gene.gene_id ) + </div><div class="line"><a name="l01101"></a><span class="lineno"> 1101</span>&#160;                                                       <span class="stringliteral">&#39;&quot; was missing its chromosome information.&#39;</span> +</div><div class="line"><a name="l01102"></a><span class="lineno"> 1102</span>&#160;                                                       <span class="stringliteral">&#39; This gene has been associated with chromosome &#39;</span> + </div><div class="line"><a name="l01103"></a><span class="lineno"> 1103</span>&#160;                                                       chr + <span class="stringliteral">&#39;.&#39;</span>, </div><div class="line"><a name="l01104"></a><span class="lineno"> 1104</span>&#160;                                                       gene = gene )</div><div class="line"><a name="l01105"></a><span class="lineno"> 1105</span>&#160;                    </div><div class="line"><a name="l01106"></a><span class="lineno"> 1106</span>&#160;                    <span class="comment"># and there are several genes missing their chromosome information,</span></div><div class="line"><a name="l01107"></a><span class="lineno"> 1107</span>&#160;                    <span class="comment"># then return a RefGeneSearchException</span></div><div class="line"><a name="l01108"></a><span class="lineno"> 1108</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l01109"></a><span class="lineno"> 1109</span>&#160;                        <span class="keywordflow">return</span> RefGeneSearchException( <span class="stringliteral">&#39;No gene of the lis was located on chromosome &#39;</span> + </div><div class="line"><a name="l01110"></a><span class="lineno"> 1110</span>&#160;                                                       chr + <span class="stringliteral">&#39; and there were several genes of the list&#39;</span> +</div><div class="line"><a name="l01111"></a><span class="lineno"> 1111</span>&#160;                                                       <span class="stringliteral">&#39; missing their chromosome information.&#39;</span> )</div><div class="line"><a name="l01112"></a><span class="lineno"> 1112</span>&#160;                </div><div class="line"><a name="l01113"></a><span class="lineno"> 1113</span>&#160;                <span class="comment"># and the expected chromosome is not provided</span></div><div class="line"><a name="l01114"></a><span class="lineno"> 1114</span>&#160;                <span class="comment">#     OR there is no gene missing their chromosome information,</span></div><div class="line"><a name="l01115"></a><span class="lineno"> 1115</span>&#160;                <span class="comment">#     then return a RefGeneSearchException</span></div><div class="line"><a name="l01116"></a><span class="lineno"> 1116</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l01117"></a><span class="lineno"> 1117</span>&#160;                    <span class="keywordflow">if</span> ( chr == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l01118"></a><span class="lineno"> 1118</span>&#160;                        <span class="keywordflow">return</span> RefGeneSearchException( <span class="stringliteral">&#39;The expected chromosome name has not been provided&#39;</span> +</div><div class="line"><a name="l01119"></a><span class="lineno"> 1119</span>&#160;                                                       <span class="stringliteral">&#39; and there was no gene missing its chromosome information.&#39;</span> )</div><div class="line"><a name="l01120"></a><span class="lineno"> 1120</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l01121"></a><span class="lineno"> 1121</span>&#160;                        <span class="keywordflow">return</span> RefGeneSearchException( <span class="stringliteral">&#39;No gene of the list has been found on the expected&#39;</span> +</div><div class="line"><a name="l01122"></a><span class="lineno"> 1122</span>&#160;                                                       <span class="stringliteral">&#39; chromosome (&#39;</span> + chr + <span class="stringliteral">&#39;) and no gene of the list&#39;</span> +</div><div class="line"><a name="l01123"></a><span class="lineno"> 1123</span>&#160;                                                       <span class="stringliteral">&#39; was missing its chromosome information.&#39;</span> )</div><div class="line"><a name="l01124"></a><span class="lineno"> 1124</span>&#160;    </div><div class="line"><a name="l01125"></a><span class="lineno"> 1125</span>&#160;                        </div><div class="line"><a name="l01126"></a><span class="lineno"> 1126</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a11654f0533e3a163aa528c8f996c3456"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a11654f0533e3a163aa528c8f996c3456">&#9670;&nbsp;</a></span>import_file_content()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.import_file_content </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>file_sep</em> = <code>','</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dtype_val</em> = <code>'str'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>encoding_val</em> = <code>'utf-8'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>import_file_content </h2>
<p>This method allows to import the data from a file as a Pandas data frame.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file_sep</td><td>String - The delimiter to use (see pandas.read_csv() help for more information). </td></tr>
    <tr><td class="paramname">dtype_val</td><td>String - The data type for data or columns (see pandas.read_csv() help for more information). 'str' by default. </td></tr>
    <tr><td class="paramname">encoding_val</td><td>String - Encoding to use for UTF when reading (see pandas.read_csv() help more information). 'utf-8' by default. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;    <span class="keyword">def </span>import_file_content( self, file_sep = &#39;,&#39;, dtype_val = &#39;str&#39;, encoding_val = &#39;utf-8&#39; ):</div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;        </div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;        <span class="comment"># Get the content of the file</span></div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;        self.file_content = pd.read_csv( self.data_path, </div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;                                         sep = file_sep, </div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;                                         dtype = dtype_val, </div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;                                         encoding = encoding_val )</div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;        </div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;        <span class="comment"># Replace NA values by None</span></div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;        self.file_content = self.file_content.where( pd.notnull( self.file_content ), <span class="keywordtype">None</span> )</div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;        </div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;        </div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="a19c8d87c05bf4071bb6c7b541106a12b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a19c8d87c05bf4071bb6c7b541106a12b">&#9670;&nbsp;</a></span>initialize_chromosome_name()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.initialize_chromosome_name </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr_name</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>initialize_chromosome_name </h2>
<p>This method allows to set the chromosome name to None if it is missing, or to set it to the defined value if it is the mitochondrial chromosome, or to set it to the defined value if it is the sexual chromosome.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chr_name</td><td>String - The chromosome name (as provided in the source).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>chr_name: String / None - The chromosome name (set to None if missing). </dd></dl>
<div class="fragment"><div class="line"><a name="l00342"></a><span class="lineno">  342</span>&#160;    <span class="keyword">def </span>initialize_chromosome_name( chr_name ):</div><div class="line"><a name="l00343"></a><span class="lineno">  343</span>&#160;        </div><div class="line"><a name="l00344"></a><span class="lineno">  344</span>&#160;        chr_name = str( chr_name )</div><div class="line"><a name="l00345"></a><span class="lineno">  345</span>&#160;        </div><div class="line"><a name="l00346"></a><span class="lineno">  346</span>&#160;        <span class="comment"># Get the name of the sexual chromosome for the species in the database</span></div><div class="line"><a name="l00347"></a><span class="lineno">  347</span>&#160;        sp_short_name = DataManager.get_instance().get_data( Constants.SPECIES_SHORT )</div><div class="line"><a name="l00348"></a><span class="lineno">  348</span>&#160;        sex_chr = Constants.SEX_CHROMOSOME[ sp_short_name ]</div><div class="line"><a name="l00349"></a><span class="lineno">  349</span>&#160;        </div><div class="line"><a name="l00350"></a><span class="lineno">  350</span>&#160;        <span class="comment"># If the chromosome name is missing, set it to none</span></div><div class="line"><a name="l00351"></a><span class="lineno">  351</span>&#160;        <span class="keywordflow">if</span> ParserStrategy.is_empty( chr_name ):</div><div class="line"><a name="l00352"></a><span class="lineno">  352</span>&#160;            <span class="keywordflow">return</span> <span class="keywordtype">None</span></div><div class="line"><a name="l00353"></a><span class="lineno">  353</span>&#160;            </div><div class="line"><a name="l00354"></a><span class="lineno">  354</span>&#160;        <span class="comment"># If the chromosome name is mitochondrial, set it to the appropriate value</span></div><div class="line"><a name="l00355"></a><span class="lineno">  355</span>&#160;        <span class="keywordflow">elif</span> ( chr_name <span class="keywordflow">in</span> Constants.MITOCHONDRIAL_CHR_LIST ):</div><div class="line"><a name="l00356"></a><span class="lineno">  356</span>&#160;            <span class="keywordflow">return</span> Constants.MITOCHONDRIAL_CHR</div><div class="line"><a name="l00357"></a><span class="lineno">  357</span>&#160;        </div><div class="line"><a name="l00358"></a><span class="lineno">  358</span>&#160;        <span class="comment"># If the chromosome name is an alias of the sexual chromosome (e.g. 23 for H. sapiens), </span></div><div class="line"><a name="l00359"></a><span class="lineno">  359</span>&#160;        <span class="comment"># set it to &#39;X|Y&#39;</span></div><div class="line"><a name="l00360"></a><span class="lineno">  360</span>&#160;        <span class="keywordflow">elif</span> ( chr_name == sex_chr ):</div><div class="line"><a name="l00361"></a><span class="lineno">  361</span>&#160;            <span class="keywordflow">return</span> Constants.CHROMOSOME_NAME_XY</div><div class="line"><a name="l00362"></a><span class="lineno">  362</span>&#160;        </div><div class="line"><a name="l00363"></a><span class="lineno">  363</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00364"></a><span class="lineno">  364</span>&#160;            <span class="keywordflow">return</span> chr_name</div><div class="line"><a name="l00365"></a><span class="lineno">  365</span>&#160;</div><div class="line"><a name="l00366"></a><span class="lineno">  366</span>&#160;</div><div class="line"><a name="l00367"></a><span class="lineno">  367</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="a07cdd23100d43df77f4001cd81fba324"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a07cdd23100d43df77f4001cd81fba324">&#9670;&nbsp;</a></span>initialize_gene_symbol()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.initialize_gene_symbol </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ensembl_db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>symbol</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transcript_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_stop</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>log_symb_from_tr</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>log_single_symb_from_pos</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>initialize_gene_symbol </h2>
<p>This method allows to try to get the symbol if it is not provided, using the transcript ID (if this is an Ensembl ID) or the ORF coordinates or to generate a "fake" symbol if the actual cannot be recovered.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ensembl_db</td><td>EnsemblRelease - The EnsemblRelease object related to the appropriate genome annotation. </td></tr>
    <tr><td class="paramname">orf_id</td><td>String - The ORF ID (as provided in the source). </td></tr>
    <tr><td class="paramname">chr_name</td><td>String - The chromosome name (after use of the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a19c8d87c05bf4071bb6c7b541106a12b">initialize_chromosome_name()</a> method). </td></tr>
    <tr><td class="paramname">symbol</td><td>String - The symbol or gene ID (as provided in the source). </td></tr>
    <tr><td class="paramname">transcript_id</td><td>String - The transcript ID (after use of the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#aa01883a5db8dba8aabdb8af61d3b1f94">initialize_transcript_id()</a> method). </td></tr>
    <tr><td class="paramname">orf_start</td><td>String (or Integer) - The ORF start genomic coordinates. </td></tr>
    <tr><td class="paramname">orf_stop</td><td>String (or Integer) - The ORF stop genomic coordinates. </td></tr>
    <tr><td class="paramname">orf_strand</td><td>String - The ORF strand. </td></tr>
    <tr><td class="paramname">log_symb_from_tr</td><td>Boolean - Does the recovery of the gene symbol from the transcript ID has to be logged? True by default. NB: We advise to set this parameter to False only if the data source does not provide any information about the gene ID (see documentation of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4d5bdcc0943ba6c1c71bb3c245bc81b3">symbol_from_transcript_id()</a> method for more information). Setting this parameter to False allows to decrease the number of logs, thus increasing the computation time. </td></tr>
    <tr><td class="paramname">log_single_symb_from_pos</td><td>Boolean - Does the recovery of a unique gene symbol from the ORF coordinates has to be logged? True by default. NB: We advise to set this parameter to False only if the data source does not provide any information about the gene ID nor about the transcript ID (see documentation of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ac199b748e002f7651a7ad708197c0008">symbol_from_orf_loc()</a> method for more information). Note that if several genes are overlapping with the position or if there are no genes but lncRNAs overlapping with the position, then the information will always be logged. Setting this parameter to False allows to decrease the number of logs, thus decreasing the computation time.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>symbol: String - The symbol or gene ID (either an "actual" symbol, as provided in the source or found in the Ensembl database using the transcript ID or ORF coordinates, or as a "fake" ID generated if the actual gene ID is unknown and may not be recover). </dd></dl>
<div class="fragment"><div class="line"><a name="l00451"></a><span class="lineno">  451</span>&#160;    <span class="keyword">def </span>initialize_gene_symbol( self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, \</div><div class="line"><a name="l00452"></a><span class="lineno">  452</span>&#160;                                orf_stop, orf_strand, log_symb_from_tr=True, log_single_symb_from_pos=True ):</div><div class="line"><a name="l00453"></a><span class="lineno">  453</span>&#160;    </div><div class="line"><a name="l00454"></a><span class="lineno">  454</span>&#160;        <span class="comment"># If the symbol is provided, make sure it is a string      </span></div><div class="line"><a name="l00455"></a><span class="lineno">  455</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> ParserStrategy.is_empty( symbol ) ):</div><div class="line"><a name="l00456"></a><span class="lineno">  456</span>&#160;            symbol = str( symbol )</div><div class="line"><a name="l00457"></a><span class="lineno">  457</span>&#160;        </div><div class="line"><a name="l00458"></a><span class="lineno">  458</span>&#160;        <span class="comment"># Otherwise, if the gene symbol is missing try to get it using the ORF coordinates</span></div><div class="line"><a name="l00459"></a><span class="lineno">  459</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00460"></a><span class="lineno">  460</span>&#160;            <span class="comment"># Try to get the gene symbol using the transcript ID if it is provided</span></div><div class="line"><a name="l00461"></a><span class="lineno">  461</span>&#160;            symbol = self.symbol_from_transcript_id( ensembl_db = ensembl_db, </div><div class="line"><a name="l00462"></a><span class="lineno">  462</span>&#160;                                                     orf_id = orf_id,</div><div class="line"><a name="l00463"></a><span class="lineno">  463</span>&#160;                                                     transcript_id = transcript_id,</div><div class="line"><a name="l00464"></a><span class="lineno">  464</span>&#160;                                                     log_warning = log_symb_from_tr )</div><div class="line"><a name="l00465"></a><span class="lineno">  465</span>&#160;            </div><div class="line"><a name="l00466"></a><span class="lineno">  466</span>&#160;            <span class="comment"># If the symbol ID cannot be recovered using the transcript ID, try to get it using the coordinates</span></div><div class="line"><a name="l00467"></a><span class="lineno">  467</span>&#160;            <span class="keywordflow">if</span> ( symbol == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00468"></a><span class="lineno">  468</span>&#160;                symbol = self.symbol_from_orf_loc( ensembl_db = ensembl_db,</div><div class="line"><a name="l00469"></a><span class="lineno">  469</span>&#160;                                                   orf_id = orf_id, </div><div class="line"><a name="l00470"></a><span class="lineno">  470</span>&#160;                                                   chr_name = chr_name, </div><div class="line"><a name="l00471"></a><span class="lineno">  471</span>&#160;                                                   orf_start = orf_start, </div><div class="line"><a name="l00472"></a><span class="lineno">  472</span>&#160;                                                   orf_stop = orf_stop, </div><div class="line"><a name="l00473"></a><span class="lineno">  473</span>&#160;                                                   orf_strand = orf_strand,</div><div class="line"><a name="l00474"></a><span class="lineno">  474</span>&#160;                                                   log_single_symb_from_pos = log_single_symb_from_pos )</div><div class="line"><a name="l00475"></a><span class="lineno">  475</span>&#160;        </div><div class="line"><a name="l00476"></a><span class="lineno">  476</span>&#160;        <span class="keywordflow">return</span> symbol</div><div class="line"><a name="l00477"></a><span class="lineno">  477</span>&#160;    </div><div class="line"><a name="l00478"></a><span class="lineno">  478</span>&#160;</div><div class="line"><a name="l00479"></a><span class="lineno">  479</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="ab0b04e0792f45b98d4111aa27c51ac93"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab0b04e0792f45b98d4111aa27c51ac93">&#9670;&nbsp;</a></span>initialize_ids()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.initialize_ids </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ensembl_db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>symbol</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transcript_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_stop</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>log_symb_from_tr</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>log_single_symb_from_pos</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>initialize_ids </h2>
<p>This method allows to:</p><ul>
<li>Set the chromosome name to None if it is missing, or to set it to the defined value if it is the mitochondrial chromosome, or to set it to the defined value if it is the sexual chromosome,</li>
<li>Generate a "fake" transcript ID if its is not provided,</li>
<li>Try to get the symbol if it is not provided, using the transcript ID (if this is an Ensembl ID) or the ORF coordinates or to generate a "fake" symbol if the actual cannot be recovered.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ensembl_db</td><td>EnsemblRelease - The EnsemblRelease object related to the appropriate genome annotation. </td></tr>
    <tr><td class="paramname">orf_id</td><td>String - The original ORF ID (as provided in the source). </td></tr>
    <tr><td class="paramname">chr_name</td><td>String - The chromosome name (as provided in the source). </td></tr>
    <tr><td class="paramname">symbol</td><td>String - The symbol or gene ID (as provided in the source). </td></tr>
    <tr><td class="paramname">transcript_id</td><td>String - The transcript ID (as provided in the source). </td></tr>
    <tr><td class="paramname">orf_start</td><td>String (or Integer) - The ORF start genomic coordinates. </td></tr>
    <tr><td class="paramname">orf_stop</td><td>String (or Integer) - The ORF stop genomic coordinates. </td></tr>
    <tr><td class="paramname">orf_strand</td><td>String - The strand of the ORF. </td></tr>
    <tr><td class="paramname">index</td><td>Integer - The index of the row being parsed. </td></tr>
    <tr><td class="paramname">log_symb_from_tr</td><td>Boolean - Does the recovery of the gene symbol from the transcript ID has to be logged? True by default. NB: We advise to set this parameter to False only if the data source does not provide any information about the gene ID (see documentation of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4d5bdcc0943ba6c1c71bb3c245bc81b3">symbol_from_transcript_id()</a> method for more information). Setting this parameter to False allows to decrease the number of logs, thus decreasing the computation time. </td></tr>
    <tr><td class="paramname">log_single_symb_from_pos</td><td>Boolean - Does the recovery of a unique gene symbol from the ORF coordinates has to be logged? True by default. NB: We advise to set this parameter to False only if the data source does not provide any information about the gene ID nor about the transcript ID (see documentation of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ac199b748e002f7651a7ad708197c0008">symbol_from_orf_loc()</a> method for more information). Note that if several genes are overlapping with the position or if there are no genes but lncRNAs overlapping with the position, then the information will always be logged. Setting this parameter to False allows to decrease the number of logs, thus decreasing the computation time.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>chr_name: String (or None) - The chromosome name (set to None if considered as missing). </dd>
<dd>
symbol: String - The symbol or gene ID (either as provided in the source or found from the ORF coordinates or "unknown" gene ID generated). </dd>
<dd>
transcript_id: String - The transcript ID (either as provided in the source or "fake" ID generated). </dd></dl>
<div class="fragment"><div class="line"><a name="l00303"></a><span class="lineno">  303</span>&#160;    <span class="keyword">def </span>initialize_ids( self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, \</div><div class="line"><a name="l00304"></a><span class="lineno">  304</span>&#160;                        orf_stop, orf_strand, index, log_symb_from_tr=True, log_single_symb_from_pos=True ):</div><div class="line"><a name="l00305"></a><span class="lineno">  305</span>&#160;        </div><div class="line"><a name="l00306"></a><span class="lineno">  306</span>&#160;        <span class="comment"># Process the chromosome name</span></div><div class="line"><a name="l00307"></a><span class="lineno">  307</span>&#160;        chr_name = self.initialize_chromosome_name( chr_name = chr_name )</div><div class="line"><a name="l00308"></a><span class="lineno">  308</span>&#160;        </div><div class="line"><a name="l00309"></a><span class="lineno">  309</span>&#160;        <span class="comment"># Process the transcript ID</span></div><div class="line"><a name="l00310"></a><span class="lineno">  310</span>&#160;        transcript_id = self.initialize_transcript_id( transcript_id = transcript_id, </div><div class="line"><a name="l00311"></a><span class="lineno">  311</span>&#160;                                                       orf_id = orf_id,</div><div class="line"><a name="l00312"></a><span class="lineno">  312</span>&#160;                                                       index = index )</div><div class="line"><a name="l00313"></a><span class="lineno">  313</span>&#160;        </div><div class="line"><a name="l00314"></a><span class="lineno">  314</span>&#160;        <span class="comment"># Pocess the symbol </span></div><div class="line"><a name="l00315"></a><span class="lineno">  315</span>&#160;        symbol = self.initialize_gene_symbol( ensembl_db = ensembl_db, </div><div class="line"><a name="l00316"></a><span class="lineno">  316</span>&#160;                                              orf_id = orf_id, </div><div class="line"><a name="l00317"></a><span class="lineno">  317</span>&#160;                                              chr_name = chr_name, </div><div class="line"><a name="l00318"></a><span class="lineno">  318</span>&#160;                                              symbol = symbol, </div><div class="line"><a name="l00319"></a><span class="lineno">  319</span>&#160;                                              transcript_id = transcript_id, </div><div class="line"><a name="l00320"></a><span class="lineno">  320</span>&#160;                                              orf_start = orf_start, </div><div class="line"><a name="l00321"></a><span class="lineno">  321</span>&#160;                                              orf_stop = orf_stop,</div><div class="line"><a name="l00322"></a><span class="lineno">  322</span>&#160;                                              orf_strand = orf_strand,</div><div class="line"><a name="l00323"></a><span class="lineno">  323</span>&#160;                                              log_symb_from_tr = log_symb_from_tr,</div><div class="line"><a name="l00324"></a><span class="lineno">  324</span>&#160;                                              log_single_symb_from_pos = log_single_symb_from_pos )</div><div class="line"><a name="l00325"></a><span class="lineno">  325</span>&#160;        </div><div class="line"><a name="l00326"></a><span class="lineno">  326</span>&#160;        <span class="keywordflow">return</span> ( chr_name, symbol, transcript_id )</div><div class="line"><a name="l00327"></a><span class="lineno">  327</span>&#160;    </div><div class="line"><a name="l00328"></a><span class="lineno">  328</span>&#160;</div><div class="line"><a name="l00329"></a><span class="lineno">  329</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="aa01883a5db8dba8aabdb8af61d3b1f94"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa01883a5db8dba8aabdb8af61d3b1f94">&#9670;&nbsp;</a></span>initialize_transcript_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.initialize_transcript_id </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transcript_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>initialize_transcript_id </h2>
<p>This method allows to generate a "fake" transcript ID if it is not provided by the data source.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">transcript_id</td><td>String - The transcript ID (as provided in the source). </td></tr>
    <tr><td class="paramname">orf_id</td><td>String - The ORF ID (as provided in the source). </td></tr>
    <tr><td class="paramname">index</td><td>Integer - The index of the row being parsed.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>transcript_id: String - The transcript ID (either as provided in the source or as a "fake ID" generated by the method). </dd></dl>
<div class="fragment"><div class="line"><a name="l00382"></a><span class="lineno">  382</span>&#160;    <span class="keyword">def </span>initialize_transcript_id( self, transcript_id, orf_id, index ):</div><div class="line"><a name="l00383"></a><span class="lineno">  383</span>&#160;</div><div class="line"><a name="l00384"></a><span class="lineno">  384</span>&#160;        <span class="comment"># If the transcript ID is missing, generate a &quot;fake&quot; ID,</span></div><div class="line"><a name="l00385"></a><span class="lineno">  385</span>&#160;        <span class="comment"># by concatenating the name of the data source and the entry index</span></div><div class="line"><a name="l00386"></a><span class="lineno">  386</span>&#160;        <span class="keywordflow">if</span> ParserStrategy.is_empty( transcript_id ):</div><div class="line"><a name="l00387"></a><span class="lineno">  387</span>&#160;            transcript_id = Constants.PREFIX_FAKE_TRANSCRIPT + self.data_source + <span class="stringliteral">&#39;_&#39;</span> + str( index )</div><div class="line"><a name="l00388"></a><span class="lineno">  388</span>&#160;            <span class="comment"># NB: The following warning has been disabled as the number of &quot;fake&quot; transcripts</span></div><div class="line"><a name="l00389"></a><span class="lineno">  389</span>&#160;            <span class="comment">#     created for a particular source may be easily recover using the following raw SQL query: </span></div><div class="line"><a name="l00390"></a><span class="lineno">  390</span>&#160;            <span class="comment">#             SELECT COUNT( * ) </span></div><div class="line"><a name="l00391"></a><span class="lineno">  391</span>&#160;            <span class="comment">#             FROM `DSTranscript` </span></div><div class="line"><a name="l00392"></a><span class="lineno">  392</span>&#160;            <span class="comment">#             WHERE ( ( `transcript_id` LIKE &#39;UNKNOWN_%&#39; ) </span></div><div class="line"><a name="l00393"></a><span class="lineno">  393</span>&#160;            <span class="comment">#                     AND ( `data_source` = &#39;DatasourceName&#39; ) )</span></div><div class="line"><a name="l00394"></a><span class="lineno">  394</span>&#160;            <span class="comment">#Logger.get_instance().warning( &#39;There was no transcript associated with the ORF with ID: &quot;&#39; + </span></div><div class="line"><a name="l00395"></a><span class="lineno">  395</span>&#160;            <span class="comment">#                               orf_id + &#39;&quot; in &#39; + self.data_source +</span></div><div class="line"><a name="l00396"></a><span class="lineno">  396</span>&#160;            <span class="comment">#                               &#39;. Hence, this ORF will be associated with the &quot;unknown&quot; transcript &#39; + </span></div><div class="line"><a name="l00397"></a><span class="lineno">  397</span>&#160;            <span class="comment">#                               transcript_id + &#39;.&#39; +</span></div><div class="line"><a name="l00398"></a><span class="lineno">  398</span>&#160;            <span class="comment">#                               &#39; Warning code: &#39; + LogCodes.WARN_ORFASSO_TR + &#39;.&#39; )</span></div><div class="line"><a name="l00399"></a><span class="lineno">  399</span>&#160;            </div><div class="line"><a name="l00400"></a><span class="lineno">  400</span>&#160;        <span class="comment"># Otherwise, make sure this is a string</span></div><div class="line"><a name="l00401"></a><span class="lineno">  401</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00402"></a><span class="lineno">  402</span>&#160;            transcript_id = str( transcript_id )</div><div class="line"><a name="l00403"></a><span class="lineno">  403</span>&#160;                                </div><div class="line"><a name="l00404"></a><span class="lineno">  404</span>&#160;        <span class="keywordflow">return</span> transcript_id</div><div class="line"><a name="l00405"></a><span class="lineno">  405</span>&#160;</div><div class="line"><a name="l00406"></a><span class="lineno">  406</span>&#160;</div><div class="line"><a name="l00407"></a><span class="lineno">  407</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="a602be6456b764340546d4ff9fd07d19f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a602be6456b764340546d4ff9fd07d19f">&#9670;&nbsp;</a></span>is_empty()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.is_empty </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>val</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>empty_val</em> = <code>Constants.EMPTY_VALUES</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>is_empty </h2>
<p>This is a static method that allows to check if a value extracted from a Pandas data frame is empty or missing. Please see the documentation of the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1util_1_1general_1_1GeneralUtil_1_1GeneralUtil.html#a4e75317b265cc8fe9d8e5b9ef8596a70">GeneralUtil.is_empty()</a> method.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">val</td><td>String/Float (from data frame) - Value to test. </td></tr>
    <tr><td class="paramname">empty_val</td><td>List - List of values considered as empty / missing.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Boolean - True if the value is missing. </dd></dl>
<div class="fragment"><div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;    <span class="keyword">def </span>is_empty( val, empty_val=Constants.EMPTY_VALUES ):</div><div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;        </div><div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;        <span class="keywordflow">return</span> GeneralUtil.is_empty( val, empty_val )</div><div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;</div><div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;</div><div class="line"><a name="l00120"></a><span class="lineno">  120</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="ac199b748e002f7651a7ad708197c0008"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac199b748e002f7651a7ad708197c0008">&#9670;&nbsp;</a></span>symbol_from_orf_loc()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.symbol_from_orf_loc </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ensembl_db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_stop</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>log_single_symb_from_pos</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>symbol_from_orf_loc </h2>
<p>This method allows to try to get the symbol associated with an ORF using its coordinates.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ensembl_db</td><td>EnsemblRelease - The EnsemblRelease object related to the appropriate genome annotation. </td></tr>
    <tr><td class="paramname">orf_id</td><td>String - The ORF ID (as provided in the source). </td></tr>
    <tr><td class="paramname">chr_name</td><td>String - The chromosome name (get using the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a19c8d87c05bf4071bb6c7b541106a12b">initialize_chromosome_name()</a> method). </td></tr>
    <tr><td class="paramname">orf_start</td><td>String (or Integer) - The ORF start genomic coordinates. </td></tr>
    <tr><td class="paramname">orf_stop</td><td>String (or Integer) - The ORF stop genomic coordinates. </td></tr>
    <tr><td class="paramname">orf_strand</td><td>String - The ORF strand. </td></tr>
    <tr><td class="paramname">log_single_symb_from_pos</td><td>Boolean - Does the recovery of a unique gene symbol from the ORF coordinates has to be logged? True by default. NB: We advise to set this parameter to False only if the data source does not provide any information about the gene ID nor about the transcript ID (see documentation of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ac199b748e002f7651a7ad708197c0008">symbol_from_orf_loc()</a> method for more information). Note that if several genes are overlapping with the position or if there are no genes but lncRNAs overlapping with the position, then the information will always be logged. Setting this parameter to False allows to decrease the number of logs, thus decreasing the computation time.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>symbol: String - The symbol or gene ID (either an "actual" symbol found in the Ensembl database using the ORF coordinates, or as a "fake" ID generated if the actual gene ID may not be recover). </dd></dl>
<div class="fragment"><div class="line"><a name="l00562"></a><span class="lineno">  562</span>&#160;    <span class="keyword">def </span>symbol_from_orf_loc( self, ensembl_db, orf_id, chr_name, orf_start, orf_stop, orf_strand, \</div><div class="line"><a name="l00563"></a><span class="lineno">  563</span>&#160;                             log_single_symb_from_pos=True ):     </div><div class="line"><a name="l00564"></a><span class="lineno">  564</span>&#160;                    </div><div class="line"><a name="l00565"></a><span class="lineno">  565</span>&#160;        <span class="comment"># If the chromosome name is missing, set the value of the symbol to an &quot;unknown&quot; gene </span></div><div class="line"><a name="l00566"></a><span class="lineno">  566</span>&#160;        <span class="comment"># (on an &quot;unknown&quot; chromosome)</span></div><div class="line"><a name="l00567"></a><span class="lineno">  567</span>&#160;        <span class="keywordflow">if</span> ParserStrategy.is_empty( chr_name ):</div><div class="line"><a name="l00568"></a><span class="lineno">  568</span>&#160;            symbol = Constants.PREFIX_UNKNOWN_GENE + <span class="stringliteral">&#39;chr_UNKNOWN&#39;</span></div><div class="line"><a name="l00569"></a><span class="lineno">  569</span>&#160;            Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00570"></a><span class="lineno">  570</span>&#160;                                           orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00571"></a><span class="lineno">  571</span>&#160;                                           <span class="stringliteral">&#39; and the chromosome name is not known.&#39;</span> +</div><div class="line"><a name="l00572"></a><span class="lineno">  572</span>&#160;                                           <span class="stringliteral">&#39; Hence, this ORF will be associated with the &quot;unknown&quot; gene &quot;&#39;</span> + </div><div class="line"><a name="l00573"></a><span class="lineno">  573</span>&#160;                                           symbol + <span class="stringliteral">&#39;&quot;.&#39;</span> +</div><div class="line"><a name="l00574"></a><span class="lineno">  574</span>&#160;                                           <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_UNKN_MISCHR + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00575"></a><span class="lineno">  575</span>&#160;        </div><div class="line"><a name="l00576"></a><span class="lineno">  576</span>&#160;        <span class="comment"># Otherwise, try to get the symbol from the ORF coordinates</span></div><div class="line"><a name="l00577"></a><span class="lineno">  577</span>&#160;        <span class="keywordflow">else</span>:            </div><div class="line"><a name="l00578"></a><span class="lineno">  578</span>&#160;            <span class="comment"># If the ORF strand is missing, create an &quot;unknown gene&quot; (on the provided chromosome)</span></div><div class="line"><a name="l00579"></a><span class="lineno">  579</span>&#160;            <span class="keywordflow">if</span> ParserStrategy.is_empty( val = orf_strand, empty_val = Constants.EMPTY_VALUES_WO_DASH ):</div><div class="line"><a name="l00580"></a><span class="lineno">  580</span>&#160;                symbol = Constants.PREFIX_UNKNOWN_GENE + <span class="stringliteral">&#39;chr_&#39;</span> + chr_name</div><div class="line"><a name="l00581"></a><span class="lineno">  581</span>&#160;                Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00582"></a><span class="lineno">  582</span>&#160;                                               orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00583"></a><span class="lineno">  583</span>&#160;                                               <span class="stringliteral">&#39; and the strand of this ORF was missing, so overlapping&#39;</span> +</div><div class="line"><a name="l00584"></a><span class="lineno">  584</span>&#160;                                               <span class="stringliteral">&#39; genes or lncRNAs cannot be found in the Ensembl database&#39;</span> +</div><div class="line"><a name="l00585"></a><span class="lineno">  585</span>&#160;                                               <span class="stringliteral">&#39; using the ORF coordinates.&#39;</span> +</div><div class="line"><a name="l00586"></a><span class="lineno">  586</span>&#160;                                               <span class="stringliteral">&#39; Hence, this ORF will be associated with the &quot;unknown&quot;&#39;</span> +</div><div class="line"><a name="l00587"></a><span class="lineno">  587</span>&#160;                                               <span class="stringliteral">&#39; gene &quot;&#39;</span> + symbol + <span class="stringliteral">&#39;&quot;.&#39;</span> +</div><div class="line"><a name="l00588"></a><span class="lineno">  588</span>&#160;                                               <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_UNKN_MISSTRAND + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00589"></a><span class="lineno">  589</span>&#160;                </div><div class="line"><a name="l00590"></a><span class="lineno">  590</span>&#160;            <span class="comment"># Otherwise, try to get the gene(s) or lncRNA(s) overlapping with the ORF coordinates</span></div><div class="line"><a name="l00591"></a><span class="lineno">  591</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00592"></a><span class="lineno">  592</span>&#160;                <span class="comment"># Get the list of Ensembl objects (gene, lncRNA, miRNA...) overlapping with this location</span></div><div class="line"><a name="l00593"></a><span class="lineno">  593</span>&#160;                genes_at_location = ensembl_db.genes_at_locus( contig = chr_name,</div><div class="line"><a name="l00594"></a><span class="lineno">  594</span>&#160;                                                               position = int( orf_start ),</div><div class="line"><a name="l00595"></a><span class="lineno">  595</span>&#160;                                                               end = int( orf_stop ),</div><div class="line"><a name="l00596"></a><span class="lineno">  596</span>&#160;                                                               strand = orf_strand )</div><div class="line"><a name="l00597"></a><span class="lineno">  597</span>&#160;                </div><div class="line"><a name="l00598"></a><span class="lineno">  598</span>&#160;                <span class="comment"># Get the list of genes at these coordinates</span></div><div class="line"><a name="l00599"></a><span class="lineno">  599</span>&#160;                genes_at_loc = [ g.gene_name <span class="keywordflow">for</span> g <span class="keywordflow">in</span> genes_at_location <span class="keywordflow">if</span> ( g.biotype == <span class="stringliteral">&#39;protein_coding&#39;</span> ) ]</div><div class="line"><a name="l00600"></a><span class="lineno">  600</span>&#160;                </div><div class="line"><a name="l00601"></a><span class="lineno">  601</span>&#160;                <span class="comment"># If there is one single gene overlapping with these coordinates, get it</span></div><div class="line"><a name="l00602"></a><span class="lineno">  602</span>&#160;                <span class="keywordflow">if</span> ( len( genes_at_loc ) == 1 ):</div><div class="line"><a name="l00603"></a><span class="lineno">  603</span>&#160;                    symbol = genes_at_loc[ 0 ]</div><div class="line"><a name="l00604"></a><span class="lineno">  604</span>&#160;                    <span class="comment"># Only log this operation if required</span></div><div class="line"><a name="l00605"></a><span class="lineno">  605</span>&#160;                    <span class="keywordflow">if</span> log_single_symb_from_pos:</div><div class="line"><a name="l00606"></a><span class="lineno">  606</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with&#39;</span> +</div><div class="line"><a name="l00607"></a><span class="lineno">  607</span>&#160;                                                       <span class="stringliteral">&#39; original ID &quot;&#39;</span> + orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00608"></a><span class="lineno">  608</span>&#160;                                                       <span class="stringliteral">&#39;, but the ORF coordinates have been found&#39;</span> +</div><div class="line"><a name="l00609"></a><span class="lineno">  609</span>&#160;                                                       <span class="stringliteral">&#39; overlapping the gene &quot;&#39;</span> + symbol + </div><div class="line"><a name="l00610"></a><span class="lineno">  610</span>&#160;                                                       <span class="stringliteral">&#39;&quot; in the Ensembl database.&#39;</span> +</div><div class="line"><a name="l00611"></a><span class="lineno">  611</span>&#160;                                                       <span class="stringliteral">&#39; Hence, the ORF will be associated with this gene.&#39;</span> +</div><div class="line"><a name="l00612"></a><span class="lineno">  612</span>&#160;                                                       <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_SING + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00613"></a><span class="lineno">  613</span>&#160;                </div><div class="line"><a name="l00614"></a><span class="lineno">  614</span>&#160;                <span class="comment"># If there are several genes overlapping with these coordinates, concatenate them to create a gene id</span></div><div class="line"><a name="l00615"></a><span class="lineno">  615</span>&#160;                <span class="keywordflow">elif</span> ( len( genes_at_loc) &gt; 1 ):</div><div class="line"><a name="l00616"></a><span class="lineno">  616</span>&#160;                    </div><div class="line"><a name="l00617"></a><span class="lineno">  617</span>&#160;                    <span class="comment"># If the ORF is overlapping exactly two genes, concatenate them to create the gene id</span></div><div class="line"><a name="l00618"></a><span class="lineno">  618</span>&#160;                    <span class="keywordflow">if</span> ( len( genes_at_loc ) == 2 ):</div><div class="line"><a name="l00619"></a><span class="lineno">  619</span>&#160;                        symbol = ( Constants.PREFIX_OVERLAPPING_GENES + </div><div class="line"><a name="l00620"></a><span class="lineno">  620</span>&#160;                                   <span class="stringliteral">&#39;_&#39;</span>.join( sorted( genes_at_loc ) ) + </div><div class="line"><a name="l00621"></a><span class="lineno">  621</span>&#160;                                   <span class="stringliteral">&#39;_chr&#39;</span> + chr_name )</div><div class="line"><a name="l00622"></a><span class="lineno">  622</span>&#160;                        </div><div class="line"><a name="l00623"></a><span class="lineno">  623</span>&#160;                    <span class="comment"># Otherwise, only use the first and last genes to create the ID</span></div><div class="line"><a name="l00624"></a><span class="lineno">  624</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00625"></a><span class="lineno">  625</span>&#160;                        symbol = ( Constants.PREFIX_OVERLAPPING_GENES + </div><div class="line"><a name="l00626"></a><span class="lineno">  626</span>&#160;                                   genes_at_loc[ 0 ] + <span class="stringliteral">&#39;_to_&#39;</span> + genes_at_loc[ -1 ] + </div><div class="line"><a name="l00627"></a><span class="lineno">  627</span>&#160;                                   <span class="stringliteral">&#39;_chr&#39;</span> + chr_name )</div><div class="line"><a name="l00628"></a><span class="lineno">  628</span>&#160;                        </div><div class="line"><a name="l00629"></a><span class="lineno">  629</span>&#160;                    Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with original ID &quot;&#39;</span> +</div><div class="line"><a name="l00630"></a><span class="lineno">  630</span>&#160;                                                   orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00631"></a><span class="lineno">  631</span>&#160;                                                   <span class="stringliteral">&#39; but the ORF coordinates have been found overlapping&#39;</span> +</div><div class="line"><a name="l00632"></a><span class="lineno">  632</span>&#160;                                                   <span class="stringliteral">&#39; with the following genes &quot;&#39;</span> + </div><div class="line"><a name="l00633"></a><span class="lineno">  633</span>&#160;                                                   <span class="stringliteral">&#39;, &#39;</span>.join( genes_at_loc ) + <span class="stringliteral">&#39;&quot; in the Ensembl database.&#39;</span> +</div><div class="line"><a name="l00634"></a><span class="lineno">  634</span>&#160;                                                   <span class="stringliteral">&#39; Hence, this ORF will be associated with the &quot;overlapping&quot; gene &quot;&#39;</span> + </div><div class="line"><a name="l00635"></a><span class="lineno">  635</span>&#160;                                                   symbol + <span class="stringliteral">&#39;&quot;.&#39;</span> +</div><div class="line"><a name="l00636"></a><span class="lineno">  636</span>&#160;                                                   <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_SEV + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00637"></a><span class="lineno">  637</span>&#160;                </div><div class="line"><a name="l00638"></a><span class="lineno">  638</span>&#160;                <span class="comment"># Otherwise, if there is no gene overlapping with these coordinates,</span></div><div class="line"><a name="l00639"></a><span class="lineno">  639</span>&#160;                <span class="comment"># try check the lncRNAs that may be overlapping with these coordinates. </span></div><div class="line"><a name="l00640"></a><span class="lineno">  640</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00641"></a><span class="lineno">  641</span>&#160;                    <span class="comment"># Get the list of lncRNAs at these coordinates</span></div><div class="line"><a name="l00642"></a><span class="lineno">  642</span>&#160;                    lncRNAs_at_loc = [ g.gene_name <span class="keywordflow">for</span> g <span class="keywordflow">in</span> genes_at_location <span class="keywordflow">if</span> ( g.biotype == <span class="stringliteral">&#39;lincRNA&#39;</span> ) ]</div><div class="line"><a name="l00643"></a><span class="lineno">  643</span>&#160;                </div><div class="line"><a name="l00644"></a><span class="lineno">  644</span>&#160;                     <span class="comment"># If there is one single lncRNA overlapping with these coordinates, get it</span></div><div class="line"><a name="l00645"></a><span class="lineno">  645</span>&#160;                    <span class="keywordflow">if</span> ( len( lncRNAs_at_loc ) == 1 ):</div><div class="line"><a name="l00646"></a><span class="lineno">  646</span>&#160;                        symbol = lncRNAs_at_loc[ 0 ]</div><div class="line"><a name="l00647"></a><span class="lineno">  647</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00648"></a><span class="lineno">  648</span>&#160;                                                       orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00649"></a><span class="lineno">  649</span>&#160;                                                       <span class="stringliteral">&#39; nor overlapping with these coordinates,&#39;</span> +</div><div class="line"><a name="l00650"></a><span class="lineno">  650</span>&#160;                                                       <span class="stringliteral">&#39; but the ORF coordinates have been found overlapping&#39;</span> +</div><div class="line"><a name="l00651"></a><span class="lineno">  651</span>&#160;                                                       <span class="stringliteral">&#39; with the lncRNA &quot;&#39;</span> + symbol + <span class="stringliteral">&#39;&quot; in the Ensembl database.&#39;</span> +</div><div class="line"><a name="l00652"></a><span class="lineno">  652</span>&#160;                                                       <span class="stringliteral">&#39; Hence, the ORF will be associated with this lncRNA&#39;</span> +</div><div class="line"><a name="l00653"></a><span class="lineno">  653</span>&#160;                                                       <span class="stringliteral">&#39; (registered as an entry in the Gene table).&#39;</span> + </div><div class="line"><a name="l00654"></a><span class="lineno">  654</span>&#160;                                                       <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_LNCR_SING + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00655"></a><span class="lineno">  655</span>&#160;                        </div><div class="line"><a name="l00656"></a><span class="lineno">  656</span>&#160;                    <span class="comment"># If there are several lncRNAs overlapping with these coordinates, </span></div><div class="line"><a name="l00657"></a><span class="lineno">  657</span>&#160;                    <span class="comment"># concatenate them to create a gene id</span></div><div class="line"><a name="l00658"></a><span class="lineno">  658</span>&#160;                    <span class="keywordflow">elif</span> ( len( lncRNAs_at_loc ) &gt; 1 ):</div><div class="line"><a name="l00659"></a><span class="lineno">  659</span>&#160;                        </div><div class="line"><a name="l00660"></a><span class="lineno">  660</span>&#160;                        <span class="comment"># If the ORF is overlapping exactly two lncRNAs, </span></div><div class="line"><a name="l00661"></a><span class="lineno">  661</span>&#160;                        <span class="comment"># concatenate them to create the gene id</span></div><div class="line"><a name="l00662"></a><span class="lineno">  662</span>&#160;                        <span class="keywordflow">if</span> ( len( lncRNAs_at_loc ) == 2 ):</div><div class="line"><a name="l00663"></a><span class="lineno">  663</span>&#160;                            symbol = ( Constants.PREFIX_OVERLAPPING_LNCRNAS + </div><div class="line"><a name="l00664"></a><span class="lineno">  664</span>&#160;                                       <span class="stringliteral">&#39;_&#39;</span>.join( sorted( lncRNAs_at_loc ) ) + </div><div class="line"><a name="l00665"></a><span class="lineno">  665</span>&#160;                                       <span class="stringliteral">&#39;_chr&#39;</span> + chr_name )</div><div class="line"><a name="l00666"></a><span class="lineno">  666</span>&#160;                        </div><div class="line"><a name="l00667"></a><span class="lineno">  667</span>&#160;                        <span class="comment"># Otherwise, only use the first and last lncRNAs to create the ID</span></div><div class="line"><a name="l00668"></a><span class="lineno">  668</span>&#160;                        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00669"></a><span class="lineno">  669</span>&#160;                            symbol = ( Constants.PREFIX_OVERLAPPING_LNCRNAS + </div><div class="line"><a name="l00670"></a><span class="lineno">  670</span>&#160;                                       lncRNAs_at_loc[ 0 ] + <span class="stringliteral">&#39;_to_&#39;</span> + lncRNAs_at_loc[ -1 ] + </div><div class="line"><a name="l00671"></a><span class="lineno">  671</span>&#160;                                       <span class="stringliteral">&#39;_chr&#39;</span> + chr_name )</div><div class="line"><a name="l00672"></a><span class="lineno">  672</span>&#160;                        </div><div class="line"><a name="l00673"></a><span class="lineno">  673</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00674"></a><span class="lineno">  674</span>&#160;                                                       orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00675"></a><span class="lineno">  675</span>&#160;                                                       <span class="stringliteral">&#39; nor overlapping with these coordinates,&#39;</span> +</div><div class="line"><a name="l00676"></a><span class="lineno">  676</span>&#160;                                                       <span class="stringliteral">&#39; but the ORF coordinates have been found overlapping&#39;</span> +</div><div class="line"><a name="l00677"></a><span class="lineno">  677</span>&#160;                                                       <span class="stringliteral">&#39; with the following lncRNAs &quot;&#39;</span> +</div><div class="line"><a name="l00678"></a><span class="lineno">  678</span>&#160;                                                       <span class="stringliteral">&#39;, &#39;</span>.join( lncRNAs_at_loc ) + <span class="stringliteral">&#39;&quot; in the Ensembl database.&#39;</span> +</div><div class="line"><a name="l00679"></a><span class="lineno">  679</span>&#160;                                                       <span class="stringliteral">&#39; Hence, this ORF will be associated with &quot;&#39;</span> + symbol + </div><div class="line"><a name="l00680"></a><span class="lineno">  680</span>&#160;                                                       <span class="stringliteral">&#39;&quot; (registered as an entry in the Gene table).&#39;</span> +</div><div class="line"><a name="l00681"></a><span class="lineno">  681</span>&#160;                                                       <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_LNCR_SEV + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00682"></a><span class="lineno">  682</span>&#160;                </div><div class="line"><a name="l00683"></a><span class="lineno">  683</span>&#160;                    <span class="comment"># Otherwise, if there is no gene nor lncRNAs overlapping with these coordinates, create an &quot;intergenic&quot; gene. </span></div><div class="line"><a name="l00684"></a><span class="lineno">  684</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00685"></a><span class="lineno">  685</span>&#160;                        symbol = Constants.PREFIX_INTERGENIC_GENE + <span class="stringliteral">&#39;chr&#39;</span> + chr_name</div><div class="line"><a name="l00686"></a><span class="lineno">  686</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00687"></a><span class="lineno">  687</span>&#160;                                                       orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00688"></a><span class="lineno">  688</span>&#160;                                                       <span class="stringliteral">&#39; and no gene nor lncRNA has been found overlapping&#39;</span> +</div><div class="line"><a name="l00689"></a><span class="lineno">  689</span>&#160;                                                       <span class="stringliteral">&#39; the ORF coordinates in the Ensembl database.&#39;</span> +</div><div class="line"><a name="l00690"></a><span class="lineno">  690</span>&#160;                                                       <span class="stringliteral">&#39; Hence, this ORF will be associated with the &quot;intergenic&quot; gene &quot;&#39;</span> + </div><div class="line"><a name="l00691"></a><span class="lineno">  691</span>&#160;                                                       symbol + <span class="stringliteral">&#39;&quot;.&#39;</span> +</div><div class="line"><a name="l00692"></a><span class="lineno">  692</span>&#160;                                                       <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_INTERG + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00693"></a><span class="lineno">  693</span>&#160;        </div><div class="line"><a name="l00694"></a><span class="lineno">  694</span>&#160;        <span class="keywordflow">return</span> symbol</div><div class="line"><a name="l00695"></a><span class="lineno">  695</span>&#160;</div><div class="line"><a name="l00696"></a><span class="lineno">  696</span>&#160;    </div><div class="line"><a name="l00697"></a><span class="lineno">  697</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a4d5bdcc0943ba6c1c71bb3c245bc81b3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4d5bdcc0943ba6c1c71bb3c245bc81b3">&#9670;&nbsp;</a></span>symbol_from_transcript_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy.symbol_from_transcript_id </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ensembl_db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>orf_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transcript_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>log_warning</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>symbol_from_transcript_id </h2>
<p>This method allows to try to get the symbol corresponding to an Ensembl transcript ID.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ensembl_db</td><td>EnsemblRelease - The EnsemblRelease object related to the appropriate genome annotation. </td></tr>
    <tr><td class="paramname">orf_id</td><td>String - The ORF ID (as provided in the source). </td></tr>
    <tr><td class="paramname">transcript_id</td><td>String - The transcript ID (get using the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#aa01883a5db8dba8aabdb8af61d3b1f94">initialize_transcript_id()</a> method). </td></tr>
    <tr><td class="paramname">log_warning</td><td>Boolean - Does the recovery of the gene symbol from the transcript ID has to be logged? True by default. NB: We advise to set this parameter to False only if the data source does not provide any information about the gene ID (see documentation of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4d5bdcc0943ba6c1c71bb3c245bc81b3">symbol_from_transcript_id()</a> method for more information). Setting this parameter to False allows to decrease the number of logs, thus increasing the computation time.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>symbol: String / None - The symbol corresponding to the provided transcript ID. </dd></dl>
<div class="fragment"><div class="line"><a name="l00502"></a><span class="lineno">  502</span>&#160;    <span class="keyword">def </span>symbol_from_transcript_id( self, ensembl_db, orf_id, transcript_id, log_warning=True ):</div><div class="line"><a name="l00503"></a><span class="lineno">  503</span>&#160;                </div><div class="line"><a name="l00504"></a><span class="lineno">  504</span>&#160;        <span class="comment"># If the transcript ID is an Ensembl ID, try to get the gene symbol using the transcript ID</span></div><div class="line"><a name="l00505"></a><span class="lineno">  505</span>&#160;        <span class="keywordflow">if</span> ( ( <span class="keywordflow">not</span> ParserStrategy.is_empty( transcript_id ) ) <span class="keywordflow">and</span> </div><div class="line"><a name="l00506"></a><span class="lineno">  506</span>&#160;             ( transcript_id.startswith( <span class="stringliteral">&#39;ENS&#39;</span> ) ) ):</div><div class="line"><a name="l00507"></a><span class="lineno">  507</span>&#160;            <span class="keywordflow">try</span>:</div><div class="line"><a name="l00508"></a><span class="lineno">  508</span>&#160;                symbol = ensembl_db.gene_name_of_transcript_id( transcript_id )</div><div class="line"><a name="l00509"></a><span class="lineno">  509</span>&#160;            <span class="keywordflow">except</span>:</div><div class="line"><a name="l00510"></a><span class="lineno">  510</span>&#160;                symbol = <span class="keywordtype">None</span></div><div class="line"><a name="l00511"></a><span class="lineno">  511</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00512"></a><span class="lineno">  512</span>&#160;                <span class="comment"># If requested, log the recovery of the gene symbol from the transcript ID</span></div><div class="line"><a name="l00513"></a><span class="lineno">  513</span>&#160;                <span class="keywordflow">if</span> log_warning:</div><div class="line"><a name="l00514"></a><span class="lineno">  514</span>&#160;                    Logger.get_instance().warning( <span class="stringliteral">&#39;There was no gene associated with the ORF with original ID &quot;&#39;</span> + </div><div class="line"><a name="l00515"></a><span class="lineno">  515</span>&#160;                                                   orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source + </div><div class="line"><a name="l00516"></a><span class="lineno">  516</span>&#160;                                                   <span class="stringliteral">&#39; but the transcript associated with this ORF (&quot;&#39;</span> +</div><div class="line"><a name="l00517"></a><span class="lineno">  517</span>&#160;                                                   transcript_id + </div><div class="line"><a name="l00518"></a><span class="lineno">  518</span>&#160;                                                   <span class="stringliteral">&#39;&quot;) has been found associated with the gene symbol &quot;&#39;</span> +</div><div class="line"><a name="l00519"></a><span class="lineno">  519</span>&#160;                                                   symbol + <span class="stringliteral">&#39;&quot; in the Ensembl database.&#39;</span> +</div><div class="line"><a name="l00520"></a><span class="lineno">  520</span>&#160;                                                   <span class="stringliteral">&#39; Hence, this ORF will be associated with this gene.&#39;</span> +</div><div class="line"><a name="l00521"></a><span class="lineno">  521</span>&#160;                                                   <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_ORFASSO_GENE_FROM_TR_ID + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00522"></a><span class="lineno">  522</span>&#160;        </div><div class="line"><a name="l00523"></a><span class="lineno">  523</span>&#160;        <span class="comment"># Otherwise, set the symbol to None</span></div><div class="line"><a name="l00524"></a><span class="lineno">  524</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00525"></a><span class="lineno">  525</span>&#160;            symbol = <span class="keywordtype">None</span></div><div class="line"><a name="l00526"></a><span class="lineno">  526</span>&#160;            </div><div class="line"><a name="l00527"></a><span class="lineno">  527</span>&#160;        <span class="keywordflow">return</span> symbol</div><div class="line"><a name="l00528"></a><span class="lineno">  528</span>&#160;        </div><div class="line"><a name="l00529"></a><span class="lineno">  529</span>&#160;</div><div class="line"><a name="l00530"></a><span class="lineno">  530</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/execution/insertion/ParserStrategy.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
