<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.model.PRO.processed.ORFTranscriptAsso.ORFTranscriptAsso Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>model</b></li><li class="navelem"><b>PRO</b></li><li class="navelem"><b>processed</b></li><li class="navelem"><b>ORFTranscriptAsso</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORFTranscriptAsso_1_1ORFTranscriptAsso.html">ORFTranscriptAsso</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORFTranscriptAsso_1_1ORFTranscriptAsso-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.model.PRO.processed.ORFTranscriptAsso.ORFTranscriptAsso Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Inherits BasePRO.</p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORFTranscriptAsso_1_1ORFTranscriptAsso.html">ORFTranscriptAsso</a> </h1>
<p>Each <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORFTranscriptAsso_1_1ORFTranscriptAsso.html">ORFTranscriptAsso</a> contains:</p><ul>
<li>id: Integer - A unique ORF-transcript association ID (Autoincremented).</li>
<li>orf_id: Integer - The ORF ID.</li>
<li>transcript_id: Integer - The transcript ID.</li>
<li>rel_start_pos: Integer - The ORF start position in coordinates relative to the transcript.</li>
<li>rel_stop_pos: Integer - The ORF stop position in coordinates relative to the transcript.</li>
<li>predicted: Boolean - Has the ORF already been predicted by computational biology?</li>
<li>ribo_seq: Boolean - Has the ORF already been detected or predicted from Ribo-seq data?</li>
<li>ms_info: Boolean - Has the ORF already been detected by mass spectrometry?</li>
<li>start_codon_seq: String - The nucleotidic sequence consensus of the start codon.</li>
<li>start_flanking_seq: String(20) - The sequences flanking the start codon and used to compute the Kozak context.</li>
<li>sequence_nt: Text - The nucleotidic sequence consensus of the ORF.</li>
<li>sequence_aa: Text - The amino acid sequence consensus of the ORF.</li>
<li>kozak_context: Boolean - Has a Kozak context already been reported for the start codon?</li>
<li>kozak_context_comp: String(30) - The Kozak context computed from the relative start position and the flanking sequence.</li>
<li>length_nt_min: Integer - The minimal length (in nucleotides) reported for this ORF.</li>
<li>length_nt_max: Integer - The maximal length (in nucleotides) reported for this ORF.</li>
<li>length_nt_median: Float - The median length (in nucleotides) reported for this ORF.</li>
<li>length_nt_values: Integer - The list of lengths (in nucleotides) reported for this ORF.</li>
<li>length_aa_min: Integer - The minimal length (in amino acids) reported for this ORF.</li>
<li>length_aa_max: Integer - The maximal length (in amino acids) reported for this ORF.</li>
<li>length_aa_median: Float - The median length (in amino acids) reported for this ORF.</li>
<li>length_aa_values: String - The list of lengths (in amino acids) reported for this ORF.</li>
<li>orf_score_min: Float - The minimal ORF score reported for this ORF.</li>
<li>orf_score_max: Float - The maximal ORF score reported for this ORF.</li>
<li>orf_score_median: Float - The median ORF score reported for this ORF.</li>
<li>orf_score_values: String - The list of ORF score reported for this ORF.</li>
<li>phylocsf_min: Float - The minimal PhyloCSF score reported for this ORF.</li>
<li>phylocsf_max: Float - The maximal PhyloCSF score reported for this ORF.</li>
<li>phylocsf_median: Float - The maximal PhyloCSF score reported for this ORF.</li>
<li>phylocsf_values: String - The list of PhyloCSF scores reported for this ORF.</li>
<li>phastcons_min: Float - The minimal PhastCons score reported for this ORF.</li>
<li>phastcons_max: Float - The maximal PhastCons score reported for this ORF.</li>
<li>phastcons_median: Float - The median PhastCons score reported for this ORF.</li>
<li>phastcons_values: Float - The list of PhastCons scores reported for this ORF.</li>
<li>floss_min: Float - The minimal FLOSS score reported for this ORF.</li>
<li>floss_max: Float - The maximal FLOSS score reported for this ORF.</li>
<li>floss_median: Float - The median FLOSS score reported for this ORF.</li>
<li>floss_values: String - The list of FLOSS scores reported for this ORF.</li>
<li>count_ds: Integer - The number of DSORFTranscriptAsso that have been merged to create the current entry.</li>
<li>count_computed_clusters: Integer - The number of "clusters" computed based upon the amino acid ORF length reported in the DSORFTranscript entries from which the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORFTranscriptAsso_1_1ORFTranscriptAsso.html">ORFTranscriptAsso</a> has been created.</li>
<li>count_prov_lengths: Integer - The number of amino acid lengths that have been used in order to compute the "clusters".</li>
<li>computed_clusters: Text - The description of the "clusters" computed based upon the amino acid ORF length reported in the DSORFTranscript entries from which the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORFTranscriptAsso_1_1ORFTranscriptAsso.html">ORFTranscriptAsso</a> has been created.</li>
<li>gen_len_eq_orf_len: Boolean - Does the genomic length (computed using genomic coordinates) equal the provided ORF length? </li>
</ul>
</div><hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/model/PRO/processed/ORFTranscriptAsso.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
