<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.execution.merge.MergeDSOTA.MergeDSOTA Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>execution</b></li><li class="navelem"><b>merge</b></li><li class="navelem"><b>MergeDSOTA</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA.html">MergeDSOTA</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.execution.merge.MergeDSOTA.MergeDSOTA Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Inherits object.</p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a47799987aa64b9e855c62bcaada124dd"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA.html#a47799987aa64b9e855c62bcaada124dd">merge_dsota</a> (args_for_merging)</td></tr>
<tr class="separator:a47799987aa64b9e855c62bcaada124dd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7fb83900e6efe8b36cfc9b16ba3a074a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA.html#a7fb83900e6efe8b36cfc9b16ba3a074a">check_dsorftrasso_coherence</a> (dsorftranscriptasso)</td></tr>
<tr class="separator:a7fb83900e6efe8b36cfc9b16ba3a074a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af8d442d71ea042ea41e78cd11db66266"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA.html#af8d442d71ea042ea41e78cd11db66266">find_best_dsota_subpop</a> (dsorftranscriptasso_list, max_len_diff_dsota_clust)</td></tr>
<tr class="separator:af8d442d71ea042ea41e78cd11db66266"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9ffc33a50c6477debf9e9c32ea85b871"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA.html#a9ffc33a50c6477debf9e9c32ea85b871">split_dsota_using_length</a> (dsorftranscriptasso_list, max_len_diff=Constants.DEFAULT_MAX_LEN_DIFF_FOR_DSOTA_CLUSTERS)</td></tr>
<tr class="separator:a9ffc33a50c6477debf9e9c32ea85b871"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a298a5e53b8f2f4d79996a37f04f0cab6"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA.html#a298a5e53b8f2f4d79996a37f04f0cab6">DSORFTRANSCRIPTASSO_ATTRIBUTES_TO_COMPARE</a></td></tr>
<tr class="separator:a298a5e53b8f2f4d79996a37f04f0cab6"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1merge_1_1MergeDSOTA_1_1MergeDSOTA.html">MergeDSOTA</a> </h1>
<p>This contains static methods allowing to create the appropriate objects to merge DSORFTranscriptAsso entries (DS database). NB: This class has been optimized for multi-processing. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a7fb83900e6efe8b36cfc9b16ba3a074a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7fb83900e6efe8b36cfc9b16ba3a074a">&#9670;&nbsp;</a></span>check_dsorftrasso_coherence()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.merge.MergeDSOTA.MergeDSOTA.check_dsorftrasso_coherence </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dsorftranscriptasso</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>check_dsorftrasso_coherence </h2>
<p>This is a static method that allows to check the coherence of the data contained in a DSORFTranscriptAsso object. It checks:</p><ul>
<li>The agreement between the nucleic and amino acid sequences.</li>
<li>The agreement between the provided nucleic sequence and length.</li>
<li>The agreement between the provided amino acid sequence and length.</li>
<li>The agreement between the nucleic and amino acid lengths. If there is conflicting information, a warning is logged and the concerned attributes are removed from the entry.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dsorftranscriptasso</td><td>DSORFTranscriptAsso - The entry to check.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>3-tuple - A 3-elements tuple that contains the following information:<ul>
<li>dsorftranscriptasso: DSORFTranscriptAsso - The updated entry.</li>
<li>error_messages_to_log: List - The list of messages instantiated by the process and that must be logged at the error level.</li>
<li>warning_messages_to_log: List - The list of messages instantiated by the process and that must be logged at the warning level. </li>
</ul>
</dd></dl>
<div class="fragment"><div class="line"><a name="l00714"></a><span class="lineno">  714</span>&#160;    <span class="keyword">def </span>check_dsorftrasso_coherence( dsorftranscriptasso ):</div><div class="line"><a name="l00715"></a><span class="lineno">  715</span>&#160;        </div><div class="line"><a name="l00716"></a><span class="lineno">  716</span>&#160;        error_messages_to_log = []</div><div class="line"><a name="l00717"></a><span class="lineno">  717</span>&#160;        warning_messages_to_log = []</div><div class="line"><a name="l00718"></a><span class="lineno">  718</span>&#160;        </div><div class="line"><a name="l00719"></a><span class="lineno">  719</span>&#160;        </div><div class="line"><a name="l00722"></a><span class="lineno">  722</span>&#160;        <span class="keywordflow">if</span> ( dsorftranscriptasso.raw_sequence == Constants.DENCELLORFOBJ_AMBIGUOUS_ATT ):</div><div class="line"><a name="l00723"></a><span class="lineno">  723</span>&#160;            dsorftranscriptasso.raw_sequence = <span class="keywordtype">None</span></div><div class="line"><a name="l00724"></a><span class="lineno">  724</span>&#160;        </div><div class="line"><a name="l00725"></a><span class="lineno">  725</span>&#160;        <span class="keywordflow">if</span> ( dsorftranscriptasso.raw_sequence_aa == Constants.DENCELLORFOBJ_AMBIGUOUS_ATT ):</div><div class="line"><a name="l00726"></a><span class="lineno">  726</span>&#160;            dsorftranscriptasso.raw_sequence_aa = <span class="keywordtype">None</span></div><div class="line"><a name="l00727"></a><span class="lineno">  727</span>&#160;        </div><div class="line"><a name="l00728"></a><span class="lineno">  728</span>&#160;        sequence_nt = dsorftranscriptasso.raw_sequence</div><div class="line"><a name="l00729"></a><span class="lineno">  729</span>&#160;        sequence_aa = dsorftranscriptasso.raw_sequence_aa</div><div class="line"><a name="l00730"></a><span class="lineno">  730</span>&#160;            </div><div class="line"><a name="l00731"></a><span class="lineno">  731</span>&#160;        </div><div class="line"><a name="l00732"></a><span class="lineno">  732</span>&#160;        <span class="keywordflow">if</span> ( sequence_nt <span class="keywordflow">and</span> sequence_aa ):</div><div class="line"><a name="l00733"></a><span class="lineno">  733</span>&#160;            </div><div class="line"><a name="l00734"></a><span class="lineno">  734</span>&#160;            <span class="comment"># Translate the DNA sequence</span></div><div class="line"><a name="l00735"></a><span class="lineno">  735</span>&#160;            <span class="keywordflow">try</span>:</div><div class="line"><a name="l00736"></a><span class="lineno">  736</span>&#160;                translated_sequence = GeneticsUtil.translate_dna( dna_seq = sequence_nt, </div><div class="line"><a name="l00737"></a><span class="lineno">  737</span>&#160;                                                                  include_stop = <span class="keyword">False</span>,</div><div class="line"><a name="l00738"></a><span class="lineno">  738</span>&#160;                                                                  to_stop = <span class="keyword">False</span> )</div><div class="line"><a name="l00739"></a><span class="lineno">  739</span>&#160;            <span class="keywordflow">except</span> TranslateDNAException <span class="keyword">as</span> e:</div><div class="line"><a name="l00740"></a><span class="lineno">  740</span>&#160;                translated_sequence = e.get_aa_seq()</div><div class="line"><a name="l00741"></a><span class="lineno">  741</span>&#160;                error_messages_to_log.append( <span class="stringliteral">&#39;The length of the nucleic sequence of the&#39;</span> +</div><div class="line"><a name="l00742"></a><span class="lineno">  742</span>&#160;                                              <span class="stringliteral">&#39; DSORFTranscriptAsso with ID &quot;&#39;</span> + str( dsorftranscriptasso.id ) + </div><div class="line"><a name="l00743"></a><span class="lineno">  743</span>&#160;                                              <span class="stringliteral">&#39;&quot; is not a multiple of three. Hence, the sequence has been&#39;</span> +</div><div class="line"><a name="l00744"></a><span class="lineno">  744</span>&#160;                                              <span class="stringliteral">&#39; translated ignoring the last &#39;</span> + str( e.get_remainder() ) + </div><div class="line"><a name="l00745"></a><span class="lineno">  745</span>&#160;                                              <span class="stringliteral">&#39; nucleotides.&#39;</span> +</div><div class="line"><a name="l00746"></a><span class="lineno">  746</span>&#160;                                              <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_TRANSL_SEQSIZE_NOT3 + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00747"></a><span class="lineno">  747</span>&#160;            </div><div class="line"><a name="l00748"></a><span class="lineno">  748</span>&#160;            </div><div class="line"><a name="l00749"></a><span class="lineno">  749</span>&#160;            </div><div class="line"><a name="l00750"></a><span class="lineno">  750</span>&#160;            <span class="comment"># Some of the data sources have replaced the first amino acid of the ORF </span></div><div class="line"><a name="l00751"></a><span class="lineno">  751</span>&#160;            <span class="comment"># sequence by a methionine (e.g. the TTG codon is usually expected to be </span></div><div class="line"><a name="l00752"></a><span class="lineno">  752</span>&#160;            <span class="comment"># translated as a leucine while some data source have replaced it as a </span></div><div class="line"><a name="l00753"></a><span class="lineno">  753</span>&#160;            <span class="comment"># methonine; see for instance the ORF with original ID &quot;andreev_2015:128701&quot; </span></div><div class="line"><a name="l00754"></a><span class="lineno">  754</span>&#160;            <span class="comment"># from sORFs.org, Human database). Hence, when the DNA and amino acid </span></div><div class="line"><a name="l00755"></a><span class="lineno">  755</span>&#160;            <span class="comment"># sequences agree with each other excepted about the 1st amino acid, the </span></div><div class="line"><a name="l00756"></a><span class="lineno">  756</span>&#160;            <span class="comment"># sequences are considered as agreeing with each other.</span></div><div class="line"><a name="l00757"></a><span class="lineno">  757</span>&#160;            <span class="comment"># NB: It would make sense to first look if the whole translated and amino </span></div><div class="line"><a name="l00758"></a><span class="lineno">  758</span>&#160;            <span class="comment">#     acid sequences agree with each other, and only then when they do not </span></div><div class="line"><a name="l00759"></a><span class="lineno">  759</span>&#160;            <span class="comment">#     to look if this is only the first amino acid which is different. </span></div><div class="line"><a name="l00760"></a><span class="lineno">  760</span>&#160;            <span class="comment">#     Nevertheless, in order to reduce the computation time, the algorithm</span></div><div class="line"><a name="l00761"></a><span class="lineno">  761</span>&#160;            <span class="comment">#     first compares the first amino acid of the two sequences:</span></div><div class="line"><a name="l00762"></a><span class="lineno">  762</span>&#160;            <span class="comment">#         - and if this is the same, it compares the whole sequences,</span></div><div class="line"><a name="l00763"></a><span class="lineno">  763</span>&#160;            <span class="comment">#             - and if they are the same, the two sequences are kept.</span></div><div class="line"><a name="l00764"></a><span class="lineno">  764</span>&#160;            <span class="comment">#             - otherwise if they are different the two sequences are removed </span></div><div class="line"><a name="l00765"></a><span class="lineno">  765</span>&#160;            <span class="comment">#               from the entry as they do not agree with each other.</span></div><div class="line"><a name="l00766"></a><span class="lineno">  766</span>&#160;            <span class="comment">#         - and if the first amino acid is not the same in the sequences, </span></div><div class="line"><a name="l00767"></a><span class="lineno">  767</span>&#160;            <span class="comment">#           it then compares the sequence omitting the first amino acid,</span></div><div class="line"><a name="l00768"></a><span class="lineno">  768</span>&#160;            <span class="comment">#             - and if they are the same, the two sequences are kept, considering </span></div><div class="line"><a name="l00769"></a><span class="lineno">  769</span>&#160;            <span class="comment">#               they actually agree with each other despite a difference on the </span></div><div class="line"><a name="l00770"></a><span class="lineno">  770</span>&#160;            <span class="comment">#               first amino acid.</span></div><div class="line"><a name="l00771"></a><span class="lineno">  771</span>&#160;            <span class="comment">#             - otherwise if they are different the two sequences are removed from </span></div><div class="line"><a name="l00772"></a><span class="lineno">  772</span>&#160;            <span class="comment">#               the entry as they do not agree with each other.</span></div><div class="line"><a name="l00773"></a><span class="lineno">  773</span>&#160;            remove_both_sequence = <span class="keyword">False</span></div><div class="line"><a name="l00774"></a><span class="lineno">  774</span>&#160;            </div><div class="line"><a name="l00775"></a><span class="lineno">  775</span>&#160;            <span class="keywordflow">if</span> ( translated_sequence[ 0 ] == sequence_aa[ 0 ] ):</div><div class="line"><a name="l00776"></a><span class="lineno">  776</span>&#160;                <span class="keywordflow">if</span> ( translated_sequence != sequence_aa ):</div><div class="line"><a name="l00777"></a><span class="lineno">  777</span>&#160;                    remove_both_sequence = <span class="keyword">True</span></div><div class="line"><a name="l00778"></a><span class="lineno">  778</span>&#160;                                    </div><div class="line"><a name="l00779"></a><span class="lineno">  779</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00780"></a><span class="lineno">  780</span>&#160;                <span class="keywordflow">if</span> ( translated_sequence[ 1: ] != sequence_aa[ 1: ] ):</div><div class="line"><a name="l00781"></a><span class="lineno">  781</span>&#160;                    remove_both_sequence = <span class="keyword">True</span></div><div class="line"><a name="l00782"></a><span class="lineno">  782</span>&#160;            </div><div class="line"><a name="l00783"></a><span class="lineno">  783</span>&#160;            <span class="keywordflow">if</span> remove_both_sequence:</div><div class="line"><a name="l00784"></a><span class="lineno">  784</span>&#160;                warning_messages_to_log.append( <span class="stringliteral">&#39;The nucleic and amino acid sequences of the&#39;</span> +</div><div class="line"><a name="l00785"></a><span class="lineno">  785</span>&#160;                                                <span class="stringliteral">&#39; DSORFTranscriptAsso with ID &quot;&#39;</span> + </div><div class="line"><a name="l00786"></a><span class="lineno">  786</span>&#160;                                                str( dsorftranscriptasso.id ) + </div><div class="line"><a name="l00787"></a><span class="lineno">  787</span>&#160;                                                <span class="stringliteral">&#39;&quot; do not agree with each other. Hence, these sequences&#39;</span> +</div><div class="line"><a name="l00788"></a><span class="lineno">  788</span>&#160;                                                <span class="stringliteral">&#39; will not be considered to create the ORFTranscriptAsso&#39;</span> +</div><div class="line"><a name="l00789"></a><span class="lineno">  789</span>&#160;                                                <span class="stringliteral">&#39; entries (in the PRO database).&#39;</span>  +</div><div class="line"><a name="l00790"></a><span class="lineno">  790</span>&#160;                                                <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_DCORF_DSOTA_CONFL_SEQ + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00791"></a><span class="lineno">  791</span>&#160;                </div><div class="line"><a name="l00792"></a><span class="lineno">  792</span>&#160;                dsorftranscriptasso.raw_sequence = <span class="keywordtype">None</span></div><div class="line"><a name="l00793"></a><span class="lineno">  793</span>&#160;                dsorftranscriptasso.raw_sequence_aa = <span class="keywordtype">None</span></div><div class="line"><a name="l00794"></a><span class="lineno">  794</span>&#160;</div><div class="line"><a name="l00795"></a><span class="lineno">  795</span>&#160;                </div><div class="line"><a name="l00796"></a><span class="lineno">  796</span>&#160;        <span class="comment"># Else, if there is only the nucleic sequence which is provided, </span></div><div class="line"><a name="l00797"></a><span class="lineno">  797</span>&#160;        <span class="comment"># then translate it to get the amino acid sequence</span></div><div class="line"><a name="l00798"></a><span class="lineno">  798</span>&#160;        <span class="keywordflow">elif</span> sequence_nt:</div><div class="line"><a name="l00799"></a><span class="lineno">  799</span>&#160;            </div><div class="line"><a name="l00800"></a><span class="lineno">  800</span>&#160;            <span class="comment"># Translate the DNA sequence</span></div><div class="line"><a name="l00801"></a><span class="lineno">  801</span>&#160;            <span class="keywordflow">try</span>:</div><div class="line"><a name="l00802"></a><span class="lineno">  802</span>&#160;                dsorftranscriptasso.raw_sequence_aa = GeneticsUtil.translate_dna( dna_seq = sequence_nt, </div><div class="line"><a name="l00803"></a><span class="lineno">  803</span>&#160;                                                                                  include_stop = <span class="keyword">False</span>,</div><div class="line"><a name="l00804"></a><span class="lineno">  804</span>&#160;                                                                                  to_stop = <span class="keyword">False</span> )</div><div class="line"><a name="l00805"></a><span class="lineno">  805</span>&#160;            <span class="keywordflow">except</span> TranslateDNAException <span class="keyword">as</span> e:</div><div class="line"><a name="l00806"></a><span class="lineno">  806</span>&#160;                dsorftranscriptasso.raw_sequence_aa = e.get_aa_seq()</div><div class="line"><a name="l00807"></a><span class="lineno">  807</span>&#160;                warning_messages_to_log.append( <span class="stringliteral">&#39;The length of the nucleic sequence of the&#39;</span> +</div><div class="line"><a name="l00808"></a><span class="lineno">  808</span>&#160;                                                <span class="stringliteral">&#39; DSORFTranscriptAsso with ID &quot;&#39;</span> + str( dsorftranscriptasso.id ) + </div><div class="line"><a name="l00809"></a><span class="lineno">  809</span>&#160;                                                <span class="stringliteral">&#39;&quot; is not a multiple of three. Hence, the sequence has been&#39;</span> +</div><div class="line"><a name="l00810"></a><span class="lineno">  810</span>&#160;                                                <span class="stringliteral">&#39; translated ignoring the last &#39;</span> + str( e.get_remainder() ) + </div><div class="line"><a name="l00811"></a><span class="lineno">  811</span>&#160;                                                <span class="stringliteral">&#39; nucleotides.&#39;</span> +</div><div class="line"><a name="l00812"></a><span class="lineno">  812</span>&#160;                                                <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_TRANSL_SEQSIZE_NOT3 + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00813"></a><span class="lineno">  813</span>&#160;            </div><div class="line"><a name="l00814"></a><span class="lineno">  814</span>&#160;        </div><div class="line"><a name="l00815"></a><span class="lineno">  815</span>&#160;        </div><div class="line"><a name="l00816"></a><span class="lineno">  816</span>&#160;        sequence_nt = dsorftranscriptasso.raw_sequence</div><div class="line"><a name="l00817"></a><span class="lineno">  817</span>&#160;        length_nt = dsorftranscriptasso.orf_length_nt</div><div class="line"><a name="l00818"></a><span class="lineno">  818</span>&#160;        </div><div class="line"><a name="l00819"></a><span class="lineno">  819</span>&#160;        <span class="keywordflow">if</span> ( sequence_nt <span class="keywordflow">and</span> ( length_nt != <span class="keywordtype">None</span> ) ):</div><div class="line"><a name="l00820"></a><span class="lineno">  820</span>&#160;            length_sequence_nt = len( sequence_nt )</div><div class="line"><a name="l00821"></a><span class="lineno">  821</span>&#160;            length_nt = int( length_nt )</div><div class="line"><a name="l00822"></a><span class="lineno">  822</span>&#160;            </div><div class="line"><a name="l00823"></a><span class="lineno">  823</span>&#160;            <span class="keywordflow">if</span> ( length_sequence_nt != length_nt ):</div><div class="line"><a name="l00824"></a><span class="lineno">  824</span>&#160;                warning_messages_to_log.append( <span class="stringliteral">&#39;The provided nucleic length (&#39;</span> + str( length_nt ) + </div><div class="line"><a name="l00825"></a><span class="lineno">  825</span>&#160;                                                <span class="stringliteral">&#39;) and the length of the nucleic sequence (&#39;</span> + </div><div class="line"><a name="l00826"></a><span class="lineno">  826</span>&#160;                                                str( length_sequence_nt ) + </div><div class="line"><a name="l00827"></a><span class="lineno">  827</span>&#160;                                                <span class="stringliteral">&#39;) of the DSORFTranscriptAsso with ID &quot;&#39;</span> + </div><div class="line"><a name="l00828"></a><span class="lineno">  828</span>&#160;                                                str( dsorftranscriptasso.id ) + <span class="stringliteral">&#39;&quot; are not the same.&#39;</span> +</div><div class="line"><a name="l00829"></a><span class="lineno">  829</span>&#160;                                                <span class="stringliteral">&#39; Hence, this length and this sequence will not be considered&#39;</span> +</div><div class="line"><a name="l00830"></a><span class="lineno">  830</span>&#160;                                                <span class="stringliteral">&#39; to create the ORFTranscriptAsso entries (in the PRO database).&#39;</span>  +</div><div class="line"><a name="l00831"></a><span class="lineno">  831</span>&#160;                                                <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_DCORF_DSOTA_CONFL_NTLEN + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00832"></a><span class="lineno">  832</span>&#160;                </div><div class="line"><a name="l00833"></a><span class="lineno">  833</span>&#160;                dsorftranscriptasso.raw_sequence = <span class="keywordtype">None</span></div><div class="line"><a name="l00834"></a><span class="lineno">  834</span>&#160;                dsorftranscriptasso.orf_length_nt = <span class="keywordtype">None</span></div><div class="line"><a name="l00835"></a><span class="lineno">  835</span>&#160;                </div><div class="line"><a name="l00836"></a><span class="lineno">  836</span>&#160;        <span class="comment"># Else, if the nucleic sequence is provided but not the length, </span></div><div class="line"><a name="l00837"></a><span class="lineno">  837</span>&#160;        <span class="comment"># then compute the length from it</span></div><div class="line"><a name="l00838"></a><span class="lineno">  838</span>&#160;        <span class="keywordflow">elif</span> sequence_nt:</div><div class="line"><a name="l00839"></a><span class="lineno">  839</span>&#160;            dsorftranscriptasso.orf_length_nt = len( sequence_nt )</div><div class="line"><a name="l00840"></a><span class="lineno">  840</span>&#160;            </div><div class="line"><a name="l00841"></a><span class="lineno">  841</span>&#160;        </div><div class="line"><a name="l00842"></a><span class="lineno">  842</span>&#160;        </div><div class="line"><a name="l00843"></a><span class="lineno">  843</span>&#160;        sequence_aa = dsorftranscriptasso.raw_sequence_aa</div><div class="line"><a name="l00844"></a><span class="lineno">  844</span>&#160;        length_aa = dsorftranscriptasso.orf_length</div><div class="line"><a name="l00845"></a><span class="lineno">  845</span>&#160;        </div><div class="line"><a name="l00846"></a><span class="lineno">  846</span>&#160;        <span class="keywordflow">if</span> ( sequence_aa <span class="keywordflow">and</span> ( length_aa != <span class="keywordtype">None</span> ) ):</div><div class="line"><a name="l00847"></a><span class="lineno">  847</span>&#160;            length_sequence_aa = len( sequence_aa )</div><div class="line"><a name="l00848"></a><span class="lineno">  848</span>&#160;            length_aa = int( length_aa )</div><div class="line"><a name="l00849"></a><span class="lineno">  849</span>&#160;            </div><div class="line"><a name="l00850"></a><span class="lineno">  850</span>&#160;            <span class="keywordflow">if</span> ( length_sequence_aa != length_aa ):</div><div class="line"><a name="l00851"></a><span class="lineno">  851</span>&#160;                warning_messages_to_log.append( <span class="stringliteral">&#39;The provided amino acid length (&#39;</span> + str( length_aa ) + </div><div class="line"><a name="l00852"></a><span class="lineno">  852</span>&#160;                                                <span class="stringliteral">&#39;) and the length of the amino acid sequence (&#39;</span> + </div><div class="line"><a name="l00853"></a><span class="lineno">  853</span>&#160;                                                str( length_sequence_aa ) + </div><div class="line"><a name="l00854"></a><span class="lineno">  854</span>&#160;                                                <span class="stringliteral">&#39;) of the DSORFTranscriptAsso with ID &quot;&#39;</span> + </div><div class="line"><a name="l00855"></a><span class="lineno">  855</span>&#160;                                                str( dsorftranscriptasso.id ) + <span class="stringliteral">&#39;&quot; are not the same.&#39;</span> +</div><div class="line"><a name="l00856"></a><span class="lineno">  856</span>&#160;                                                <span class="stringliteral">&#39; Hence, this length and this sequence will not be considered&#39;</span> +</div><div class="line"><a name="l00857"></a><span class="lineno">  857</span>&#160;                                                <span class="stringliteral">&#39; to create the ORFTranscriptAsso entries (in the PRO database).&#39;</span> +</div><div class="line"><a name="l00858"></a><span class="lineno">  858</span>&#160;                                                <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_DCORF_DSOTA_CONFL_AALEN + <span class="stringliteral">&#39;.&#39;</span> ) </div><div class="line"><a name="l00859"></a><span class="lineno">  859</span>&#160;                </div><div class="line"><a name="l00860"></a><span class="lineno">  860</span>&#160;                dsorftranscriptasso.raw_sequence_aa = <span class="keywordtype">None</span></div><div class="line"><a name="l00861"></a><span class="lineno">  861</span>&#160;                dsorftranscriptasso.orf_length = <span class="keywordtype">None</span></div><div class="line"><a name="l00862"></a><span class="lineno">  862</span>&#160;                </div><div class="line"><a name="l00863"></a><span class="lineno">  863</span>&#160;        <span class="comment"># Else, if the nucleic sequence is provided but not the length, </span></div><div class="line"><a name="l00864"></a><span class="lineno">  864</span>&#160;        <span class="comment"># then compute the length from it</span></div><div class="line"><a name="l00865"></a><span class="lineno">  865</span>&#160;        <span class="keywordflow">elif</span> sequence_aa:</div><div class="line"><a name="l00866"></a><span class="lineno">  866</span>&#160;            dsorftranscriptasso.orf_length = len( sequence_aa )</div><div class="line"><a name="l00867"></a><span class="lineno">  867</span>&#160;            </div><div class="line"><a name="l00868"></a><span class="lineno">  868</span>&#160;        </div><div class="line"><a name="l00869"></a><span class="lineno">  869</span>&#160;        </div><div class="line"><a name="l00870"></a><span class="lineno">  870</span>&#160;        length_nt = dsorftranscriptasso.orf_length_nt</div><div class="line"><a name="l00871"></a><span class="lineno">  871</span>&#160;        length_aa = dsorftranscriptasso.orf_length</div><div class="line"><a name="l00872"></a><span class="lineno">  872</span>&#160;        </div><div class="line"><a name="l00873"></a><span class="lineno">  873</span>&#160;        <span class="keywordflow">if</span> ( ( length_nt != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( length_aa != <span class="keywordtype">None</span> ) ):</div><div class="line"><a name="l00874"></a><span class="lineno">  874</span>&#160;            length_nt = int( length_nt )</div><div class="line"><a name="l00875"></a><span class="lineno">  875</span>&#160;            length_aa = int( length_aa )</div><div class="line"><a name="l00876"></a><span class="lineno">  876</span>&#160;            length_aa_in_nt = int( ( length_aa + 1 ) * 3 )</div><div class="line"><a name="l00877"></a><span class="lineno">  877</span>&#160;            </div><div class="line"><a name="l00878"></a><span class="lineno">  878</span>&#160;            <span class="keywordflow">if</span> ( length_nt != length_aa_in_nt ):</div><div class="line"><a name="l00879"></a><span class="lineno">  879</span>&#160;                warning_messages_to_log.append( <span class="stringliteral">&#39;The provided amino acid length (&#39;</span> + str( length_aa ) + </div><div class="line"><a name="l00880"></a><span class="lineno">  880</span>&#160;                                                <span class="stringliteral">&#39;) and nucleic length (&#39;</span> + str( length_nt ) +</div><div class="line"><a name="l00881"></a><span class="lineno">  881</span>&#160;                                                + <span class="stringliteral">&#39;) of the DSORFTranscriptAsso with ID &quot;&#39;</span> + </div><div class="line"><a name="l00882"></a><span class="lineno">  882</span>&#160;                                                str( dsorftranscriptasso.id ) + <span class="stringliteral">&#39;&quot; do not agree with each&#39;</span> +</div><div class="line"><a name="l00883"></a><span class="lineno">  883</span>&#160;                                                <span class="stringliteral">&#39; other. Hence, these lengths will not be considered&#39;</span> +</div><div class="line"><a name="l00884"></a><span class="lineno">  884</span>&#160;                                                <span class="stringliteral">&#39; to create the ORFTranscriptAsso entries (in the PRO database).&#39;</span> +</div><div class="line"><a name="l00885"></a><span class="lineno">  885</span>&#160;                                                <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_DCORF_DSOTA_CONFL_AANTLEN + <span class="stringliteral">&#39;.&#39;</span> ) </div><div class="line"><a name="l00886"></a><span class="lineno">  886</span>&#160;                </div><div class="line"><a name="l00887"></a><span class="lineno">  887</span>&#160;                dsorftranscriptasso.orf_length_nt = <span class="keywordtype">None</span></div><div class="line"><a name="l00888"></a><span class="lineno">  888</span>&#160;                dsorftranscriptasso.orf_length = <span class="keywordtype">None</span></div><div class="line"><a name="l00889"></a><span class="lineno">  889</span>&#160;                </div><div class="line"><a name="l00890"></a><span class="lineno">  890</span>&#160;        <span class="comment"># Else, if the nucleic length is provided but not the amino acid length, </span></div><div class="line"><a name="l00891"></a><span class="lineno">  891</span>&#160;        <span class="comment"># then compute the amino acid length from it</span></div><div class="line"><a name="l00892"></a><span class="lineno">  892</span>&#160;        <span class="keywordflow">elif</span> ( ( length_nt != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( length_aa == <span class="keywordtype">None</span> ) ):</div><div class="line"><a name="l00893"></a><span class="lineno">  893</span>&#160;            length_nt = int( length_nt )</div><div class="line"><a name="l00894"></a><span class="lineno">  894</span>&#160;            dsorftranscriptasso.orf_length = ( ( length_nt / 3 ) - 1 )</div><div class="line"><a name="l00895"></a><span class="lineno">  895</span>&#160;            </div><div class="line"><a name="l00896"></a><span class="lineno">  896</span>&#160;        <span class="comment"># Else, if the amino acid length is provided but not the nucleic length, </span></div><div class="line"><a name="l00897"></a><span class="lineno">  897</span>&#160;        <span class="comment"># then compute the nucleic length from it</span></div><div class="line"><a name="l00898"></a><span class="lineno">  898</span>&#160;        <span class="keywordflow">elif</span> ( ( length_nt == <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( length_aa != <span class="keywordtype">None</span> ) ): </div><div class="line"><a name="l00899"></a><span class="lineno">  899</span>&#160;            length_aa = int( length_aa )</div><div class="line"><a name="l00900"></a><span class="lineno">  900</span>&#160;            dsorftranscriptasso.orf_length_nt = ( ( length_aa + 1 ) * 3 )</div><div class="line"><a name="l00901"></a><span class="lineno">  901</span>&#160;        </div><div class="line"><a name="l00902"></a><span class="lineno">  902</span>&#160;        <span class="keywordflow">return</span> ( dsorftranscriptasso, warning_messages_to_log, error_messages_to_log )</div><div class="line"><a name="l00903"></a><span class="lineno">  903</span>&#160;        </div><div class="line"><a name="l00904"></a><span class="lineno">  904</span>&#160;    </div><div class="line"><a name="l00905"></a><span class="lineno">  905</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="af8d442d71ea042ea41e78cd11db66266"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af8d442d71ea042ea41e78cd11db66266">&#9670;&nbsp;</a></span>find_best_dsota_subpop()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.merge.MergeDSOTA.MergeDSOTA.find_best_dsota_subpop </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dsorftranscriptasso_list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>max_len_diff_dsota_clust</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>find_best_dsota_subpop </h2>
<p>This method allows to compute "clusters" of DSORFTranscriptAsso entries according to their similarity in amino acid ORF length.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dsorftranscriptasso_list</td><td>List - The list of DSORFTranscriptAsso objects from which the clusters have to be computed. </td></tr>
    <tr><td class="paramname">max_len_diff_dsota_clust</td><td>Integer - The maximal length difference allowed between two DSOTA entries that belong to the same cluster.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>all_sublists: List of lists - The list of DSORFTranscriptAsso regrouped by lists of the entries defining the "cluster". </dd></dl>
<div class="fragment"><div class="line"><a name="l00921"></a><span class="lineno">  921</span>&#160;    <span class="keyword">def </span>find_best_dsota_subpop( dsorftranscriptasso_list, max_len_diff_dsota_clust ):</div><div class="line"><a name="l00922"></a><span class="lineno">  922</span>&#160;                </div><div class="line"><a name="l00923"></a><span class="lineno">  923</span>&#160;        <span class="comment"># Compute the best &quot;clusters&quot; of DSORFTranscriptAsso </span></div><div class="line"><a name="l00924"></a><span class="lineno">  924</span>&#160;        <span class="comment"># based upon the ORF amino acid length</span></div><div class="line"><a name="l00925"></a><span class="lineno">  925</span>&#160;        <span class="comment"># NB: See the documentation of the split_dsota_using_length()</span></div><div class="line"><a name="l00926"></a><span class="lineno">  926</span>&#160;        <span class="comment">#     method for more information.</span></div><div class="line"><a name="l00927"></a><span class="lineno">  927</span>&#160;        prev_all_sublists = dsorftranscriptasso_list</div><div class="line"><a name="l00928"></a><span class="lineno">  928</span>&#160;        all_sublists = MergeDSOTA.split_dsota_using_length( prev_all_sublists )</div><div class="line"><a name="l00929"></a><span class="lineno">  929</span>&#160;        </div><div class="line"><a name="l00930"></a><span class="lineno">  930</span>&#160;        <span class="comment"># While this is still possible to identify several clusters </span></div><div class="line"><a name="l00931"></a><span class="lineno">  931</span>&#160;        <span class="comment"># according to the method used, run the process</span></div><div class="line"><a name="l00932"></a><span class="lineno">  932</span>&#160;        <span class="keywordflow">while</span> ( all_sublists != prev_all_sublists ):</div><div class="line"><a name="l00933"></a><span class="lineno">  933</span>&#160;            </div><div class="line"><a name="l00934"></a><span class="lineno">  934</span>&#160;            prev_all_sublists = all_sublists</div><div class="line"><a name="l00935"></a><span class="lineno">  935</span>&#160;            all_sublists = []</div><div class="line"><a name="l00936"></a><span class="lineno">  936</span>&#160;            </div><div class="line"><a name="l00937"></a><span class="lineno">  937</span>&#160;            <span class="keywordflow">for</span> sublist <span class="keywordflow">in</span> prev_all_sublists:</div><div class="line"><a name="l00938"></a><span class="lineno">  938</span>&#160;                all_sublists += MergeDSOTA.split_dsota_using_length( dsorftranscriptasso_list = sublist, </div><div class="line"><a name="l00939"></a><span class="lineno">  939</span>&#160;                                                                        max_len_diff = max_len_diff_dsota_clust )</div><div class="line"><a name="l00940"></a><span class="lineno">  940</span>&#160;                    </div><div class="line"><a name="l00941"></a><span class="lineno">  941</span>&#160;        <span class="keywordflow">return</span> all_sublists</div><div class="line"><a name="l00942"></a><span class="lineno">  942</span>&#160;        </div><div class="line"><a name="l00943"></a><span class="lineno">  943</span>&#160;    </div><div class="line"><a name="l00944"></a><span class="lineno">  944</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a47799987aa64b9e855c62bcaada124dd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a47799987aa64b9e855c62bcaada124dd">&#9670;&nbsp;</a></span>merge_dsota()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.merge.MergeDSOTA.MergeDSOTA.merge_dsota </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>args_for_merging</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>merge_dsota </h2>
<p>This static method allows to merge a group of entries from the DSORFTranscriptAsso table of the DS database.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">args_for_merging</td><td>7-tuple - A 7-elements tuple that contains the following information:<ul>
<li>ota_id: Integer (&gt;0) - An unique integer to use as ID for the new ORFTranscriptAsso entry.</li>
<li>orf_tr_asso: Tuple - A tuple containing the IDs of the ORF and the Transcript entries related to the new ORFTranscriptAsso entry ( orf_id, transcript_id ).</li>
<li>dsorftranscriptasso_list: List of DSORFTranscriptAsso - The list of DSORFTranscriptAsso entries to merge.</li>
<li>check_dsota_coherence: Boolean - Does the coherence between the attributes of a DSORFTranscriptAsso needs to be checked?</li>
<li>compute_consensus: Boolean - Does the consensus needs to be computed when several sequences are provided?</li>
<li>sqce_consensus_ambig_threshold: Float [0,1] - The value of the threshold to use when computing the consensus sequence if this option has been selected.</li>
<li>max_len_diff_dsota_clust: Integer (&gt;0) - The maximal difference between the max. and min. lengths of DSORFTranscriptAsso entries to belong to the same "cluster".</li>
</ul>
</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>6-tuple - A 6-elements tuple that contains the following information:<ul>
<li>objects_to_insert: List - The list of new objects to insert and that CAN NOT be instantiated by any concurrent process.</li>
<li>cell_ctxt_catalog_to_insert: List - The list CellContextCatalog entries instantiated by the process (and that could eventually have been instantiated by concurrent processes)</li>
<li>provided_cat_catalog_to_insert: List - The list ProvidedCategoryCatalog entries instantiated by the process (and that could eventually have been instantiated by concurrent processes)</li>
<li>floss_class_catelog_to_insert: List - The list FlossClassCatalog entries instantiated by the process (and that could eventually have been instantiated by concurrent processes)</li>
<li>error_messages_to_log: List - The list of messages instantiated by the process and that must be logged at the error level.</li>
<li>warning_messages_to_log: List - The list of messages instantiated by the process and that must be logged at the warning level. </li>
</ul>
</dd></dl>
<div class="fragment"><div class="line"><a name="l00111"></a><span class="lineno">  111</span>&#160;    <span class="keyword">def </span>merge_dsota( args_for_merging ):</div><div class="line"><a name="l00112"></a><span class="lineno">  112</span>&#160;        </div><div class="line"><a name="l00113"></a><span class="lineno">  113</span>&#160;        <span class="comment"># Parse the arguments</span></div><div class="line"><a name="l00114"></a><span class="lineno">  114</span>&#160;        ( ota_id,</div><div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;          orf_tr_asso,</div><div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;          dsorftranscriptasso_list,</div><div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;          check_dsota_coherence,</div><div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;          compute_consensus,</div><div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;          sqce_consensus_ambig_threshold,</div><div class="line"><a name="l00120"></a><span class="lineno">  120</span>&#160;          max_len_diff_dsota_clust ) = args_for_merging</div><div class="line"><a name="l00121"></a><span class="lineno">  121</span>&#160;        </div><div class="line"><a name="l00122"></a><span class="lineno">  122</span>&#160;        <span class="comment"># Instantiate a list of new entries</span></div><div class="line"><a name="l00123"></a><span class="lineno">  123</span>&#160;        <span class="comment"># NB: The objects added to this list MUST be unique entries of the PRO database, </span></div><div class="line"><a name="l00124"></a><span class="lineno">  124</span>&#160;        <span class="comment">#     i.e. entries that could NOT be instantiated by a parallel process for which </span></div><div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;        <span class="comment">#     the provided arguments would be different. This exclude the &quot;catalog&quot; entries.</span></div><div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;        objects_to_insert = []</div><div class="line"><a name="l00127"></a><span class="lineno">  127</span>&#160;        </div><div class="line"><a name="l00128"></a><span class="lineno">  128</span>&#160;        <span class="comment"># The new CellContextCatalog, ProvidedCategoryCatalog and FlossClassCatalog </span></div><div class="line"><a name="l00129"></a><span class="lineno">  129</span>&#160;        <span class="comment"># entries are stored in separate dictionaries.</span></div><div class="line"><a name="l00130"></a><span class="lineno">  130</span>&#160;        cell_ctxt_catalog_to_insert = {}</div><div class="line"><a name="l00131"></a><span class="lineno">  131</span>&#160;        provided_cat_catalog_to_insert = {}</div><div class="line"><a name="l00132"></a><span class="lineno">  132</span>&#160;        floss_class_catelog_to_insert = {}</div><div class="line"><a name="l00133"></a><span class="lineno">  133</span>&#160;        </div><div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;        <span class="comment"># Error and warning messages expected to be logged </span></div><div class="line"><a name="l00135"></a><span class="lineno">  135</span>&#160;        <span class="comment"># are stored in separate lists and returned</span></div><div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;        error_messages_to_log = []</div><div class="line"><a name="l00137"></a><span class="lineno">  137</span>&#160;        warning_messages_to_log = []</div><div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;        </div><div class="line"><a name="l00139"></a><span class="lineno">  139</span>&#160;        <span class="comment"># Get the IDs of ORF and Transcript</span></div><div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;        ( orf_id, transcript_id ) = orf_tr_asso</div><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;        </div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;        <span class="comment"># NB: </span></div><div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;        <span class="comment"># - The original ORF IDs (from data source or generated during </span></div><div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160;        <span class="comment">#   the parsing / insertion) are not registered in the PRO database.</span></div><div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;        <span class="comment"># - The information about the equality between the genomic length</span></div><div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;        <span class="comment">#   of the ORF and the actual length of the ORF has not been yet </span></div><div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;        <span class="comment">#   computed at this stage and will be computed later.</span></div><div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;                        </div><div class="line"><a name="l00149"></a><span class="lineno">  149</span>&#160;        <span class="comment"># When the checkDSOTA option has been selected,</span></div><div class="line"><a name="l00150"></a><span class="lineno">  150</span>&#160;        <span class="comment"># for each DSORFTranscriptAsso check the raw nucleic and </span></div><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;        <span class="comment"># amino acid sequences agree with each other. If there is some </span></div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;        <span class="comment"># incoherence between them, the information will be logged and </span></div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;        <span class="comment"># the non-according attributes of the object will be set to None.</span></div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;        <span class="comment"># NB: See the documentation of the check_dsorftrasso_coherence() method</span></div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;        <span class="comment">#     for more information.</span></div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;        <span class="keywordflow">if</span> check_dsota_coherence:</div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;            dsorftranscriptasso_updated_list = []</div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;            <span class="keywordflow">for</span> dsorftranscriptasso <span class="keywordflow">in</span> dsorftranscriptasso_list:</div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;                ( checked_dsota, checked_error_to_log, checked_warning_to_log ) = MergeDSOTA.check_dsorftrasso_coherence( dsorftranscriptasso )</div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;                dsorftranscriptasso_updated_list.append( checked_dsota )</div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;                error_messages_to_log += checked_error_to_log</div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;                warning_messages_to_log += checked_warning_to_log</div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;            dsorftranscriptasso_updated_list = dsorftranscriptasso_list</div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;            </div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;        <span class="comment"># If there is one single DSORFTranscriptAsso in the list, then create </span></div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;        <span class="comment"># a new ORFTranscriptAsso using the information of its attributes</span></div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;        <span class="keywordflow">if</span> ( len( dsorftranscriptasso_updated_list ) == 1 ):</div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;            </div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;            dsorftranscriptasso = dsorftranscriptasso_updated_list[ 0 ]</div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;            </div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;            <span class="comment"># If the value of an attribute is ambiguous,</span></div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;            <span class="comment"># set it to None for the merging</span></div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;            <span class="keywordflow">for</span> att <span class="keywordflow">in</span> MergeDSOTA.DSORFTRANSCRIPTASSO_ATTRIBUTES_TO_COMPARE:</div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;                                    </div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;                att_value = getattr( dsorftranscriptasso, att ) </div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;                                   </div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;                <span class="keywordflow">if</span> ( att_value == Constants.DENCELLORFOBJ_AMBIGUOUS_ATT ):</div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;                    setattr( dsorftranscriptasso, att, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;            </div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;            </div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;            predicted = dsorftranscriptasso.predicted</div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;            <span class="keywordflow">if</span> ( predicted == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;                predicted = <span class="keyword">False</span></div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;                </div><div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;            ribo_seq = dsorftranscriptasso.ribo_seq</div><div class="line"><a name="l00187"></a><span class="lineno">  187</span>&#160;            <span class="keywordflow">if</span> ( ribo_seq == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00188"></a><span class="lineno">  188</span>&#160;                ribo_seq = <span class="keyword">False</span></div><div class="line"><a name="l00189"></a><span class="lineno">  189</span>&#160;                </div><div class="line"><a name="l00190"></a><span class="lineno">  190</span>&#160;            ms_info = dsorftranscriptasso.ms_info</div><div class="line"><a name="l00191"></a><span class="lineno">  191</span>&#160;            <span class="keywordflow">if</span> ( ms_info == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00192"></a><span class="lineno">  192</span>&#160;                ms_info = <span class="keyword">False</span></div><div class="line"><a name="l00193"></a><span class="lineno">  193</span>&#160;            </div><div class="line"><a name="l00194"></a><span class="lineno">  194</span>&#160;            <span class="comment"># Get the start codon sequence</span></div><div class="line"><a name="l00195"></a><span class="lineno">  195</span>&#160;            <span class="keywordflow">if</span> ( dsorftranscriptasso.start_codon_seq != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;                start_codon_seq = dsorftranscriptasso.start_codon_seq</div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;                <span class="keywordflow">if</span> ( dsorftranscriptasso.raw_sequence != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;                    start_codon_seq = dsorftranscriptasso.raw_sequence[ :3 ]</div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;                    start_codon_seq = <span class="keywordtype">None</span></div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;            </div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;            <span class="comment"># Get information about the Kozak context</span></div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;            <span class="keywordflow">if</span> ( str( dsorftranscriptasso.kozak_context ) <span class="keywordflow">in</span> MergeDSOTA.KOZAK_CTXT_TRUE ):</div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;                kozak_context = <span class="keyword">True</span></div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;            <span class="keywordflow">elif</span> ( str( dsorftranscriptasso.kozak_context ) <span class="keywordflow">in</span> MergeDSOTA.KOZAK_CTXT_FALSE ):</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;                kozak_context = <span class="keyword">False</span></div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;                kozak_context = <span class="keywordtype">None</span></div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;                </div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;            <span class="comment"># Define the count_prov_length attribute</span></div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;            <span class="keywordflow">if</span> ( dsorftranscriptasso.orf_length != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;                count_prov_lengths = 1</div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;                count_prov_lengths = 0</div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;            </div><div class="line"><a name="l00217"></a><span class="lineno">  217</span>&#160;            <span class="comment"># Instantiate the new object</span></div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;            orftranscriptasso = ORFTranscriptAsso( id = ota_id,</div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;                                                   orf_id = orf_id,</div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;                                                   transcript_id = transcript_id,</div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;                                                   predicted = predicted,</div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;                                                   ribo_seq = ribo_seq,</div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;                                                   ms_info = ms_info,</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;                                                   sequence_nt = dsorftranscriptasso.raw_sequence,</div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;                                                   sequence_aa = dsorftranscriptasso.raw_sequence_aa,</div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;                                                   start_codon_seq = start_codon_seq,</div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;                                                   kozak_context = kozak_context,</div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;                                                   length_nt_min = dsorftranscriptasso.orf_length_nt,</div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;                                                   length_nt_max = dsorftranscriptasso.orf_length_nt,</div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;                                                   length_nt_median = dsorftranscriptasso.orf_length_nt,</div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;                                                   length_nt_values = dsorftranscriptasso.orf_length_nt,                                                   </div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;                                                   length_aa_min = dsorftranscriptasso.orf_length,</div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                                                   length_aa_max = dsorftranscriptasso.orf_length,</div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                                                   length_aa_median = dsorftranscriptasso.orf_length,</div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;                                                   length_aa_values = dsorftranscriptasso.orf_length,</div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;                                                   orf_score_min = dsorftranscriptasso.orf_score,</div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;                                                   orf_score_max = dsorftranscriptasso.orf_score,</div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;                                                   orf_score_median = dsorftranscriptasso.orf_score,</div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;                                                   orf_score_values = dsorftranscriptasso.orf_score,                                                   </div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;                                                   phylocsf_min = dsorftranscriptasso.phylocsf,</div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;                                                   phylocsf_max = dsorftranscriptasso.phylocsf,</div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;                                                   phylocsf_median = dsorftranscriptasso.phylocsf,</div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;                                                   phylocsf_values = dsorftranscriptasso.phylocsf,</div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;                                                   phastcons_min = dsorftranscriptasso.phastcons,</div><div class="line"><a name="l00245"></a><span class="lineno">  245</span>&#160;                                                   phastcons_max = dsorftranscriptasso.phastcons,</div><div class="line"><a name="l00246"></a><span class="lineno">  246</span>&#160;                                                   phastcons_median = dsorftranscriptasso.phastcons,</div><div class="line"><a name="l00247"></a><span class="lineno">  247</span>&#160;                                                   phastcons_values = dsorftranscriptasso.phastcons,                                                   </div><div class="line"><a name="l00248"></a><span class="lineno">  248</span>&#160;                                                   floss_min = dsorftranscriptasso.floss,</div><div class="line"><a name="l00249"></a><span class="lineno">  249</span>&#160;                                                   floss_max = dsorftranscriptasso.floss,</div><div class="line"><a name="l00250"></a><span class="lineno">  250</span>&#160;                                                   floss_median = dsorftranscriptasso.floss,</div><div class="line"><a name="l00251"></a><span class="lineno">  251</span>&#160;                                                   floss_values = dsorftranscriptasso.floss,</div><div class="line"><a name="l00252"></a><span class="lineno">  252</span>&#160;                                                   count_ds = 1,</div><div class="line"><a name="l00253"></a><span class="lineno">  253</span>&#160;                                                   count_computed_clusters = 1,</div><div class="line"><a name="l00254"></a><span class="lineno">  254</span>&#160;                                                   count_prov_lengths = count_prov_lengths,</div><div class="line"><a name="l00255"></a><span class="lineno">  255</span>&#160;                                                   computed_clusters = str( [[ dsorftranscriptasso.id ]] ),</div><div class="line"><a name="l00256"></a><span class="lineno">  256</span>&#160;                                                   gen_len_eq_orf_len = <span class="keywordtype">None</span> )</div><div class="line"><a name="l00257"></a><span class="lineno">  257</span>&#160;            objects_to_insert.append( orftranscriptasso )</div><div class="line"><a name="l00258"></a><span class="lineno">  258</span>&#160;            </div><div class="line"><a name="l00259"></a><span class="lineno">  259</span>&#160;            <span class="comment"># Create new entries in the CellContext table to register the</span></div><div class="line"><a name="l00260"></a><span class="lineno">  260</span>&#160;            <span class="comment"># cellular contexts related to the ORFTranscriptAsso entry</span></div><div class="line"><a name="l00261"></a><span class="lineno">  261</span>&#160;            <span class="keywordflow">if</span> dsorftranscriptasso.cell_context:</div><div class="line"><a name="l00262"></a><span class="lineno">  262</span>&#160;                </div><div class="line"><a name="l00263"></a><span class="lineno">  263</span>&#160;                <span class="comment"># Split the cellular contexts that can eventually have been</span></div><div class="line"><a name="l00264"></a><span class="lineno">  264</span>&#160;                <span class="comment"># concatenated by the source or during the data insertion</span></div><div class="line"><a name="l00265"></a><span class="lineno">  265</span>&#160;                list_cell_context =  dsorftranscriptasso.cell_context.split( <span class="stringliteral">&#39;,&#39;</span> )</div><div class="line"><a name="l00266"></a><span class="lineno">  266</span>&#160;                    </div><div class="line"><a name="l00267"></a><span class="lineno">  267</span>&#160;                <span class="comment"># Get the set of unique values for the cellular context</span></div><div class="line"><a name="l00268"></a><span class="lineno">  268</span>&#160;                list_uniq_cell_context = set( list_cell_context )</div><div class="line"><a name="l00269"></a><span class="lineno">  269</span>&#160;                </div><div class="line"><a name="l00270"></a><span class="lineno">  270</span>&#160;                <span class="keywordflow">for</span> cell_ctxt <span class="keywordflow">in</span> list_uniq_cell_context:</div><div class="line"><a name="l00271"></a><span class="lineno">  271</span>&#160;                    </div><div class="line"><a name="l00272"></a><span class="lineno">  272</span>&#160;                    <span class="comment"># If the CellContextCatalog entry has not yet been instantiated, </span></div><div class="line"><a name="l00273"></a><span class="lineno">  273</span>&#160;                    <span class="comment"># then create a new entry in this table.</span></div><div class="line"><a name="l00274"></a><span class="lineno">  274</span>&#160;                    cell_ctxt_catalog = CellContextCatalog( context = cell_ctxt )</div><div class="line"><a name="l00275"></a><span class="lineno">  275</span>&#160;                    </div><div class="line"><a name="l00276"></a><span class="lineno">  276</span>&#160;                    existing_cell_ctxt_catalog = cell_ctxt_catalog_to_insert.get( cell_ctxt_catalog )                   </div><div class="line"><a name="l00277"></a><span class="lineno">  277</span>&#160;                    <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> existing_cell_ctxt_catalog ):</div><div class="line"><a name="l00278"></a><span class="lineno">  278</span>&#160;                        cell_ctxt_catalog_to_insert[ cell_ctxt_catalog ] = cell_ctxt_catalog</div><div class="line"><a name="l00279"></a><span class="lineno">  279</span>&#160;                    </div><div class="line"><a name="l00280"></a><span class="lineno">  280</span>&#160;                    <span class="comment"># Create a new entry in the CellContext table</span></div><div class="line"><a name="l00281"></a><span class="lineno">  281</span>&#160;                    cellcontext = CellContext( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00282"></a><span class="lineno">  282</span>&#160;                                               cell_context = cell_ctxt )</div><div class="line"><a name="l00283"></a><span class="lineno">  283</span>&#160;                    objects_to_insert.append( cellcontext )</div><div class="line"><a name="l00284"></a><span class="lineno">  284</span>&#160;            </div><div class="line"><a name="l00285"></a><span class="lineno">  285</span>&#160;            </div><div class="line"><a name="l00286"></a><span class="lineno">  286</span>&#160;            <span class="comment"># Create new entries in the ProvidedCategory table to register the </span></div><div class="line"><a name="l00287"></a><span class="lineno">  287</span>&#160;            <span class="comment"># ORF categories provided by the data source for this </span></div><div class="line"><a name="l00288"></a><span class="lineno">  288</span>&#160;            <span class="comment"># ORF - transcript association                </span></div><div class="line"><a name="l00289"></a><span class="lineno">  289</span>&#160;            <span class="keywordflow">if</span> dsorftranscriptasso.provided_category:</div><div class="line"><a name="l00290"></a><span class="lineno">  290</span>&#160;                </div><div class="line"><a name="l00291"></a><span class="lineno">  291</span>&#160;                <span class="comment"># Split the categories that can eventually have been </span></div><div class="line"><a name="l00292"></a><span class="lineno">  292</span>&#160;                <span class="comment"># concatenated by the source or during data insertion</span></div><div class="line"><a name="l00293"></a><span class="lineno">  293</span>&#160;                <span class="comment"># Replace the blanks by underscores in the provided categories</span></div><div class="line"><a name="l00294"></a><span class="lineno">  294</span>&#160;                list_provided_category = dsorftranscriptasso.provided_category.replace( <span class="stringliteral">&#39;/&#39;</span>, <span class="stringliteral">&#39;,&#39;</span> ).replace( <span class="stringliteral">&#39; &#39;</span>, <span class="stringliteral">&#39;_&#39;</span> ).split( <span class="stringliteral">&#39;,&#39;</span> )</div><div class="line"><a name="l00295"></a><span class="lineno">  295</span>&#160;                                    </div><div class="line"><a name="l00296"></a><span class="lineno">  296</span>&#160;                <span class="comment"># Get the set of unique values for the cellular context</span></div><div class="line"><a name="l00297"></a><span class="lineno">  297</span>&#160;                list_uniq_provided_category = set( list_provided_category )</div><div class="line"><a name="l00298"></a><span class="lineno">  298</span>&#160;                </div><div class="line"><a name="l00299"></a><span class="lineno">  299</span>&#160;                <span class="keywordflow">for</span> prov_cat <span class="keywordflow">in</span> list_uniq_provided_category:</div><div class="line"><a name="l00300"></a><span class="lineno">  300</span>&#160;                    </div><div class="line"><a name="l00301"></a><span class="lineno">  301</span>&#160;                    <span class="keywordflow">if</span> <span class="keywordflow">not</span> GeneralUtil.is_empty( prov_cat ):</div><div class="line"><a name="l00302"></a><span class="lineno">  302</span>&#160;                    </div><div class="line"><a name="l00303"></a><span class="lineno">  303</span>&#160;                        <span class="comment"># If the ORFCatagoryCatalog entry has not yet been instantiated, </span></div><div class="line"><a name="l00304"></a><span class="lineno">  304</span>&#160;                        <span class="comment"># then create a new entry in this table.</span></div><div class="line"><a name="l00305"></a><span class="lineno">  305</span>&#160;                        provided_cat_catalog = ProvidedCategoryCatalog( category = prov_cat )</div><div class="line"><a name="l00306"></a><span class="lineno">  306</span>&#160;                                               </div><div class="line"><a name="l00307"></a><span class="lineno">  307</span>&#160;                        existing_provided_cat_catalog = provided_cat_catalog_to_insert.get( provided_cat_catalog )                   </div><div class="line"><a name="l00308"></a><span class="lineno">  308</span>&#160;                        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> existing_provided_cat_catalog ):</div><div class="line"><a name="l00309"></a><span class="lineno">  309</span>&#160;                            provided_cat_catalog_to_insert[ provided_cat_catalog ] = provided_cat_catalog</div><div class="line"><a name="l00310"></a><span class="lineno">  310</span>&#160;                        </div><div class="line"><a name="l00311"></a><span class="lineno">  311</span>&#160;                        <span class="comment"># Create a new entry in the ProvidedCategory table</span></div><div class="line"><a name="l00312"></a><span class="lineno">  312</span>&#160;                        providedcategory = ProvidedCategory( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00313"></a><span class="lineno">  313</span>&#160;                                                             provided_category = prov_cat )</div><div class="line"><a name="l00314"></a><span class="lineno">  314</span>&#160;                        objects_to_insert.append( providedcategory )</div><div class="line"><a name="l00315"></a><span class="lineno">  315</span>&#160;                        </div><div class="line"><a name="l00316"></a><span class="lineno">  316</span>&#160;                        </div><div class="line"><a name="l00317"></a><span class="lineno">  317</span>&#160;            <span class="comment"># Create new entries in the FLOSSClass table to register</span></div><div class="line"><a name="l00318"></a><span class="lineno">  318</span>&#160;            <span class="comment"># the FLOSS categories provided by the data source</span></div><div class="line"><a name="l00319"></a><span class="lineno">  319</span>&#160;            <span class="keywordflow">if</span> dsorftranscriptasso.floss_class:                    </div><div class="line"><a name="l00320"></a><span class="lineno">  320</span>&#160;                    </div><div class="line"><a name="l00321"></a><span class="lineno">  321</span>&#160;                <span class="comment"># If the FLOSSClassCatalog entry has not yet been instantiated, </span></div><div class="line"><a name="l00322"></a><span class="lineno">  322</span>&#160;                <span class="comment"># then create a new entry in this table.</span></div><div class="line"><a name="l00323"></a><span class="lineno">  323</span>&#160;                floss_class_cat = FLOSSClassCatalog( floss_class = dsorftranscriptasso.floss_class )</div><div class="line"><a name="l00324"></a><span class="lineno">  324</span>&#160;                                               </div><div class="line"><a name="l00325"></a><span class="lineno">  325</span>&#160;                existing_floss_class_cat = floss_class_catelog_to_insert.get( floss_class_cat )                   </div><div class="line"><a name="l00326"></a><span class="lineno">  326</span>&#160;                <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> existing_floss_class_cat ):</div><div class="line"><a name="l00327"></a><span class="lineno">  327</span>&#160;                    floss_class_catelog_to_insert[ floss_class_cat ] = floss_class_cat</div><div class="line"><a name="l00328"></a><span class="lineno">  328</span>&#160;                </div><div class="line"><a name="l00329"></a><span class="lineno">  329</span>&#160;                <span class="comment"># Create a new entry in the FLOSSClass table</span></div><div class="line"><a name="l00330"></a><span class="lineno">  330</span>&#160;                flossclass = FLOSSClass( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00331"></a><span class="lineno">  331</span>&#160;                                         floss_class = dsorftranscriptasso.floss_class )</div><div class="line"><a name="l00332"></a><span class="lineno">  332</span>&#160;                objects_to_insert.append( flossclass )</div><div class="line"><a name="l00333"></a><span class="lineno">  333</span>&#160;                </div><div class="line"><a name="l00334"></a><span class="lineno">  334</span>&#160;                </div><div class="line"><a name="l00335"></a><span class="lineno">  335</span>&#160;            <span class="comment"># Create the new ORFTranscriptAssoDSAsso</span></div><div class="line"><a name="l00336"></a><span class="lineno">  336</span>&#160;            orftranscriptassodsasso = ORFTranscriptAssoDSAsso( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00337"></a><span class="lineno">  337</span>&#160;                                                               dsorftranscriptasso_id = dsorftranscriptasso.id,</div><div class="line"><a name="l00338"></a><span class="lineno">  338</span>&#160;                                                               data_source = dsorftranscriptasso.data_source )</div><div class="line"><a name="l00339"></a><span class="lineno">  339</span>&#160;            objects_to_insert.append( orftranscriptassodsasso )</div><div class="line"><a name="l00340"></a><span class="lineno">  340</span>&#160;        </div><div class="line"><a name="l00341"></a><span class="lineno">  341</span>&#160;            </div><div class="line"><a name="l00342"></a><span class="lineno">  342</span>&#160;        <span class="comment"># Otherwise, if there are several DSORFTranscriptAsso to merge in the list, </span></div><div class="line"><a name="l00343"></a><span class="lineno">  343</span>&#160;        <span class="comment"># then create a new ORFTranscriptAsso merging the information of their </span></div><div class="line"><a name="l00344"></a><span class="lineno">  344</span>&#160;        <span class="comment"># attributes</span></div><div class="line"><a name="l00345"></a><span class="lineno">  345</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00346"></a><span class="lineno">  346</span>&#160;            </div><div class="line"><a name="l00348"></a><span class="lineno">  348</span>&#160;            dsorftranscriptasso_dict = {}</div><div class="line"><a name="l00349"></a><span class="lineno">  349</span>&#160;            </div><div class="line"><a name="l00350"></a><span class="lineno">  350</span>&#160;            <span class="keywordflow">for</span> att <span class="keywordflow">in</span> MergeDSOTA.DSORFTRANSCRIPTASSO_ATTRIBUTES_TO_COMPARE:</div><div class="line"><a name="l00351"></a><span class="lineno">  351</span>&#160;                </div><div class="line"><a name="l00352"></a><span class="lineno">  352</span>&#160;                dsorftranscriptasso_dict[ att ] = []</div><div class="line"><a name="l00353"></a><span class="lineno">  353</span>&#160;                </div><div class="line"><a name="l00354"></a><span class="lineno">  354</span>&#160;                <span class="keywordflow">for</span> dsorftranscriptasso <span class="keywordflow">in</span> dsorftranscriptasso_updated_list:</div><div class="line"><a name="l00355"></a><span class="lineno">  355</span>&#160;                    att_value = getattr( dsorftranscriptasso, att )</div><div class="line"><a name="l00356"></a><span class="lineno">  356</span>&#160;                    </div><div class="line"><a name="l00357"></a><span class="lineno">  357</span>&#160;                    <span class="comment"># Store the value in the list only if it is not None and not ambiguous</span></div><div class="line"><a name="l00358"></a><span class="lineno">  358</span>&#160;                    <span class="keywordflow">if</span> ( ( att_value != <span class="keywordtype">None</span> )</div><div class="line"><a name="l00359"></a><span class="lineno">  359</span>&#160;                         <span class="keywordflow">and</span> ( att_value != Constants.DENCELLORFOBJ_AMBIGUOUS_ATT ) ):</div><div class="line"><a name="l00360"></a><span class="lineno">  360</span>&#160;                        dsorftranscriptasso_dict.get( att ).append( att_value )</div><div class="line"><a name="l00361"></a><span class="lineno">  361</span>&#160;                        </div><div class="line"><a name="l00362"></a><span class="lineno">  362</span>&#160;                <span class="comment"># Remove the empty lists from the dictionary</span></div><div class="line"><a name="l00363"></a><span class="lineno">  363</span>&#160;                <span class="keywordflow">if</span> ( dsorftranscriptasso_dict.get( att ) == [] ):</div><div class="line"><a name="l00364"></a><span class="lineno">  364</span>&#160;                    del dsorftranscriptasso_dict[ att ]</div><div class="line"><a name="l00365"></a><span class="lineno">  365</span>&#160;                    </div><div class="line"><a name="l00366"></a><span class="lineno">  366</span>&#160;            </div><div class="line"><a name="l00367"></a><span class="lineno">  367</span>&#160;            </div><div class="line"><a name="l00368"></a><span class="lineno">  368</span>&#160;            </div><div class="line"><a name="l00369"></a><span class="lineno">  369</span>&#160;            <span class="comment"># If the ORF - Transcript association has been at least predicted once,</span></div><div class="line"><a name="l00370"></a><span class="lineno">  370</span>&#160;            <span class="comment"># then set the attribute to True</span></div><div class="line"><a name="l00371"></a><span class="lineno">  371</span>&#160;            predicted = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;predicted&#39;</span>, <span class="keyword">False</span> )</div><div class="line"><a name="l00372"></a><span class="lineno">  372</span>&#160;            <span class="keywordflow">if</span> predicted:</div><div class="line"><a name="l00373"></a><span class="lineno">  373</span>&#160;                predicted = any( predicted )</div><div class="line"><a name="l00374"></a><span class="lineno">  374</span>&#160;            </div><div class="line"><a name="l00375"></a><span class="lineno">  375</span>&#160;            <span class="comment"># If the ORF - Transcript association has been at least detected once by </span></div><div class="line"><a name="l00376"></a><span class="lineno">  376</span>&#160;            <span class="comment"># Ribo-seq, then set the attribute to True</span></div><div class="line"><a name="l00377"></a><span class="lineno">  377</span>&#160;            ribo_seq = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;ribo_seq&#39;</span>, <span class="keyword">False</span> )</div><div class="line"><a name="l00378"></a><span class="lineno">  378</span>&#160;            <span class="keywordflow">if</span> ribo_seq:</div><div class="line"><a name="l00379"></a><span class="lineno">  379</span>&#160;                ribo_seq = any( ribo_seq )</div><div class="line"><a name="l00380"></a><span class="lineno">  380</span>&#160;            </div><div class="line"><a name="l00381"></a><span class="lineno">  381</span>&#160;            <span class="comment"># If the ORF - Transcript association has been at least detected once by </span></div><div class="line"><a name="l00382"></a><span class="lineno">  382</span>&#160;            <span class="comment"># mass spectrometry, then set the attribute to True</span></div><div class="line"><a name="l00383"></a><span class="lineno">  383</span>&#160;            ms_info = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;ms_info&#39;</span>, <span class="keyword">False</span> )</div><div class="line"><a name="l00384"></a><span class="lineno">  384</span>&#160;            <span class="keywordflow">if</span> ms_info:</div><div class="line"><a name="l00385"></a><span class="lineno">  385</span>&#160;                ms_info = any( ms_info )</div><div class="line"><a name="l00386"></a><span class="lineno">  386</span>&#160;            </div><div class="line"><a name="l00387"></a><span class="lineno">  387</span>&#160;            <span class="comment"># Find a consensus on the nucleic sequence</span></div><div class="line"><a name="l00388"></a><span class="lineno">  388</span>&#160;            sqce_nt_consensus = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;raw_sequence&#39;</span>, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00389"></a><span class="lineno">  389</span>&#160;            <span class="keywordflow">if</span> sqce_nt_consensus:</div><div class="line"><a name="l00390"></a><span class="lineno">  390</span>&#160;                <span class="comment"># If the computeConsensus option has been selected, find a consensus. </span></div><div class="line"><a name="l00391"></a><span class="lineno">  391</span>&#160;                <span class="keywordflow">if</span> compute_consensus:</div><div class="line"><a name="l00392"></a><span class="lineno">  392</span>&#160;                    sqce_nt_consensus = GeneticsUtil.find_sqce_consensus( list_of_sequences = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;raw_sequence&#39;</span> ), </div><div class="line"><a name="l00393"></a><span class="lineno">  393</span>&#160;                                                                          sqce_type = Constants.SEQUENCE_TYPE_DNA, </div><div class="line"><a name="l00394"></a><span class="lineno">  394</span>&#160;                                                                          threshold = sqce_consensus_ambig_threshold,</div><div class="line"><a name="l00395"></a><span class="lineno">  395</span>&#160;                                                                          fasta_end_name = str( ota_id ) + <span class="stringliteral">&#39;_nt&#39;</span> )</div><div class="line"><a name="l00396"></a><span class="lineno">  396</span>&#160;                <span class="comment"># Otherwise, instead of the sequence, store the number of sequences </span></div><div class="line"><a name="l00397"></a><span class="lineno">  397</span>&#160;                <span class="comment"># that would have been used to compute the consensus</span></div><div class="line"><a name="l00398"></a><span class="lineno">  398</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00399"></a><span class="lineno">  399</span>&#160;                    sqce_nt_consensus = len( sqce_nt_consensus )</div><div class="line"><a name="l00400"></a><span class="lineno">  400</span>&#160;            </div><div class="line"><a name="l00401"></a><span class="lineno">  401</span>&#160;            <span class="comment"># Find a consensus on the amino acid sequence</span></div><div class="line"><a name="l00402"></a><span class="lineno">  402</span>&#160;            sqce_aa_consensus = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;raw_sequence_aa&#39;</span>, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00403"></a><span class="lineno">  403</span>&#160;            <span class="keywordflow">if</span> sqce_aa_consensus:</div><div class="line"><a name="l00404"></a><span class="lineno">  404</span>&#160;                <span class="comment"># If the computeConsensus option has been selected, find a consensus. </span></div><div class="line"><a name="l00405"></a><span class="lineno">  405</span>&#160;                <span class="keywordflow">if</span> compute_consensus:</div><div class="line"><a name="l00406"></a><span class="lineno">  406</span>&#160;                    sqce_aa_consensus = GeneticsUtil.find_sqce_consensus( list_of_sequences = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;raw_sequence_aa&#39;</span> ),</div><div class="line"><a name="l00407"></a><span class="lineno">  407</span>&#160;                                                                          sqce_type = Constants.SEQUENCE_TYPE_PROT,</div><div class="line"><a name="l00408"></a><span class="lineno">  408</span>&#160;                                                                          threshold = sqce_consensus_ambig_threshold,</div><div class="line"><a name="l00409"></a><span class="lineno">  409</span>&#160;                                                                          fasta_end_name = str( ota_id ) + <span class="stringliteral">&#39;_aa&#39;</span> )</div><div class="line"><a name="l00410"></a><span class="lineno">  410</span>&#160;                <span class="comment"># Otherwise, instead of the sequence, store the number of sequences </span></div><div class="line"><a name="l00411"></a><span class="lineno">  411</span>&#160;                <span class="comment"># that would have been used to compute the consensus</span></div><div class="line"><a name="l00412"></a><span class="lineno">  412</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00413"></a><span class="lineno">  413</span>&#160;                    sqce_aa_consensus = len( sqce_aa_consensus )</div><div class="line"><a name="l00414"></a><span class="lineno">  414</span>&#160;            </div><div class="line"><a name="l00415"></a><span class="lineno">  415</span>&#160;            <span class="comment"># If a consensus has been computed for the nucleic sequence, store the three first </span></div><div class="line"><a name="l00416"></a><span class="lineno">  416</span>&#160;            <span class="comment"># letters of the nucleic sequence consensus as the start codon sequence.</span></div><div class="line"><a name="l00417"></a><span class="lineno">  417</span>&#160;            <span class="comment"># Otherwise, try to compute it from the start_codon_seq attributes if they are provided.</span></div><div class="line"><a name="l00418"></a><span class="lineno">  418</span>&#160;            <span class="keywordflow">if</span> ( sqce_nt_consensus</div><div class="line"><a name="l00419"></a><span class="lineno">  419</span>&#160;                 <span class="keywordflow">and</span> ( <span class="keywordflow">not</span> isinstance( sqce_nt_consensus, int ) )</div><div class="line"><a name="l00420"></a><span class="lineno">  420</span>&#160;                 <span class="keywordflow">and</span> ( len( sqce_nt_consensus ) &gt;= 3 ) ):</div><div class="line"><a name="l00421"></a><span class="lineno">  421</span>&#160;                sqce_start_codon_consensus = sqce_nt_consensus[ :3 ]</div><div class="line"><a name="l00422"></a><span class="lineno">  422</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00423"></a><span class="lineno">  423</span>&#160;                sqce_start_codon_consensus = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;start_codon_seq&#39;</span>, <span class="keywordtype">None</span> )                </div><div class="line"><a name="l00424"></a><span class="lineno">  424</span>&#160;                <span class="keywordflow">if</span> sqce_start_codon_consensus:</div><div class="line"><a name="l00425"></a><span class="lineno">  425</span>&#160;                    sqce_start_codon_consensus = GeneticsUtil.find_codon_consensus( sequences = sqce_start_codon_consensus, </div><div class="line"><a name="l00426"></a><span class="lineno">  426</span>&#160;                                                                                    threshold = sqce_consensus_ambig_threshold )</div><div class="line"><a name="l00427"></a><span class="lineno">  427</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00428"></a><span class="lineno">  428</span>&#160;                    sqce_start_codon_consensus = <span class="keywordtype">None</span></div><div class="line"><a name="l00429"></a><span class="lineno">  429</span>&#160;                </div><div class="line"><a name="l00430"></a><span class="lineno">  430</span>&#160;            </div><div class="line"><a name="l00431"></a><span class="lineno">  431</span>&#160;            <span class="comment"># Get the Kozak context for the start of the ORF - Transcript association.</span></div><div class="line"><a name="l00432"></a><span class="lineno">  432</span>&#160;            <span class="comment"># - By default, the Kozak context is set equal to None,</span></div><div class="line"><a name="l00433"></a><span class="lineno">  433</span>&#160;            <span class="comment"># - If the Kozak context has been at least reported once, </span></div><div class="line"><a name="l00434"></a><span class="lineno">  434</span>&#160;            <span class="comment">#   then the attribute value is set to True,</span></div><div class="line"><a name="l00435"></a><span class="lineno">  435</span>&#160;            <span class="comment"># - Otherwise, if it has already been reported that there is no Kozak context</span></div><div class="line"><a name="l00436"></a><span class="lineno">  436</span>&#160;            <span class="comment">#   for the start of this ORF - Transcript association ( AND if a Kozak context </span></div><div class="line"><a name="l00437"></a><span class="lineno">  437</span>&#160;            <span class="comment">#   has never been reported for this association), then it is set to False.</span></div><div class="line"><a name="l00438"></a><span class="lineno">  438</span>&#160;            kozak_context_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;kozak_context&#39;</span>, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00439"></a><span class="lineno">  439</span>&#160;            </div><div class="line"><a name="l00440"></a><span class="lineno">  440</span>&#160;            <span class="keywordflow">if</span> kozak_context_list:</div><div class="line"><a name="l00441"></a><span class="lineno">  441</span>&#160;                key_iterator = iter( kozak_context_list )</div><div class="line"><a name="l00442"></a><span class="lineno">  442</span>&#160;                k_ctxt = next( key_iterator, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00443"></a><span class="lineno">  443</span>&#160;                </div><div class="line"><a name="l00444"></a><span class="lineno">  444</span>&#160;                kozak_context = <span class="keywordtype">None</span></div><div class="line"><a name="l00445"></a><span class="lineno">  445</span>&#160;                <span class="keywordflow">while</span> ( k_ctxt != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( kozak_context != <span class="keyword">True</span> ):</div><div class="line"><a name="l00446"></a><span class="lineno">  446</span>&#160;                </div><div class="line"><a name="l00447"></a><span class="lineno">  447</span>&#160;                    <span class="keywordflow">if</span> str( k_ctxt ) <span class="keywordflow">in</span> MergeDSOTA.KOZAK_CTXT_TRUE:</div><div class="line"><a name="l00448"></a><span class="lineno">  448</span>&#160;                        kozak_context = <span class="keyword">True</span></div><div class="line"><a name="l00449"></a><span class="lineno">  449</span>&#160;                        </div><div class="line"><a name="l00450"></a><span class="lineno">  450</span>&#160;                    <span class="keywordflow">elif</span> str( k_ctxt ) <span class="keywordflow">in</span> MergeDSOTA.KOZAK_CTXT_FALSE:</div><div class="line"><a name="l00451"></a><span class="lineno">  451</span>&#160;                        kozak_context = <span class="keyword">False</span></div><div class="line"><a name="l00452"></a><span class="lineno">  452</span>&#160;                        </div><div class="line"><a name="l00453"></a><span class="lineno">  453</span>&#160;                    k_ctxt = next( key_iterator, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00454"></a><span class="lineno">  454</span>&#160;                    </div><div class="line"><a name="l00455"></a><span class="lineno">  455</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00456"></a><span class="lineno">  456</span>&#160;                kozak_context = <span class="keywordtype">None</span></div><div class="line"><a name="l00457"></a><span class="lineno">  457</span>&#160;            </div><div class="line"><a name="l00458"></a><span class="lineno">  458</span>&#160;            </div><div class="line"><a name="l00459"></a><span class="lineno">  459</span>&#160;            <span class="comment"># Compute the values of the amino acid ORF lengths</span></div><div class="line"><a name="l00460"></a><span class="lineno">  460</span>&#160;            orf_length_values_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;orf_length&#39;</span> )</div><div class="line"><a name="l00461"></a><span class="lineno">  461</span>&#160;            ( length_aa_min, length_aa_max, length_aa_median ) = GeneralUtil.min_max_median_of_list( list_of_values = orf_length_values_list, fct = <span class="stringliteral">&#39;int&#39;</span> )</div><div class="line"><a name="l00462"></a><span class="lineno">  462</span>&#160;            length_aa_values = GeneralUtil.list_to_string( list_to_convert = orf_length_values_list,</div><div class="line"><a name="l00463"></a><span class="lineno">  463</span>&#160;                                                           sep = Constants.OTA_LIST_VALUES_SEPARATOR,</div><div class="line"><a name="l00464"></a><span class="lineno">  464</span>&#160;                                                           not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00465"></a><span class="lineno">  465</span>&#160;                                    </div><div class="line"><a name="l00466"></a><span class="lineno">  466</span>&#160;            <span class="comment"># Compute the values of the nucleic ORF lengths</span></div><div class="line"><a name="l00467"></a><span class="lineno">  467</span>&#160;            orf_length_nt_values_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;orf_length_nt&#39;</span> )</div><div class="line"><a name="l00468"></a><span class="lineno">  468</span>&#160;            ( length_nt_min, length_nt_max, length_nt_median ) = GeneralUtil.min_max_median_of_list( list_of_values = orf_length_nt_values_list, fct = <span class="stringliteral">&#39;int&#39;</span> )</div><div class="line"><a name="l00469"></a><span class="lineno">  469</span>&#160;            length_nt_values = GeneralUtil.list_to_string( list_to_convert = orf_length_nt_values_list,</div><div class="line"><a name="l00470"></a><span class="lineno">  470</span>&#160;                                                           sep = Constants.OTA_LIST_VALUES_SEPARATOR,</div><div class="line"><a name="l00471"></a><span class="lineno">  471</span>&#160;                                                           not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00472"></a><span class="lineno">  472</span>&#160;            </div><div class="line"><a name="l00473"></a><span class="lineno">  473</span>&#160;            <span class="comment"># Compute the values for the ORF scores</span></div><div class="line"><a name="l00474"></a><span class="lineno">  474</span>&#160;            orfscore_values_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;orf_score&#39;</span> )</div><div class="line"><a name="l00475"></a><span class="lineno">  475</span>&#160;            ( orfscore_min, orfscore_max, orfscore_median ) = GeneralUtil.min_max_median_of_list( list_of_values = orfscore_values_list, fct = <span class="stringliteral">&#39;float&#39;</span> )</div><div class="line"><a name="l00476"></a><span class="lineno">  476</span>&#160;            orfscore_values = GeneralUtil.list_to_string( list_to_convert = orfscore_values_list,</div><div class="line"><a name="l00477"></a><span class="lineno">  477</span>&#160;                                                          sep = Constants.OTA_LIST_VALUES_SEPARATOR,</div><div class="line"><a name="l00478"></a><span class="lineno">  478</span>&#160;                                                          not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00479"></a><span class="lineno">  479</span>&#160;                      </div><div class="line"><a name="l00480"></a><span class="lineno">  480</span>&#160;            <span class="comment"># Compute the values for the PhyloCSF scores</span></div><div class="line"><a name="l00481"></a><span class="lineno">  481</span>&#160;            phylocsf_values_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;phylocsf&#39;</span> )</div><div class="line"><a name="l00482"></a><span class="lineno">  482</span>&#160;            ( phylocsf_min, phylocsf_max, phylocsf_median ) = GeneralUtil.min_max_median_of_list( list_of_values = phylocsf_values_list, fct = <span class="stringliteral">&#39;float&#39;</span> )</div><div class="line"><a name="l00483"></a><span class="lineno">  483</span>&#160;            phylocsf_values = GeneralUtil.list_to_string( list_to_convert = phylocsf_values_list,</div><div class="line"><a name="l00484"></a><span class="lineno">  484</span>&#160;                                                          sep = Constants.OTA_LIST_VALUES_SEPARATOR,</div><div class="line"><a name="l00485"></a><span class="lineno">  485</span>&#160;                                                          not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00486"></a><span class="lineno">  486</span>&#160;            </div><div class="line"><a name="l00487"></a><span class="lineno">  487</span>&#160;            <span class="comment"># Compute the values for the PhastCons scores</span></div><div class="line"><a name="l00488"></a><span class="lineno">  488</span>&#160;            phastcons_values_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;phastcons&#39;</span> )</div><div class="line"><a name="l00489"></a><span class="lineno">  489</span>&#160;            ( phastcons_min, phastcons_max, phastcons_median ) = GeneralUtil.min_max_median_of_list( list_of_values = phastcons_values_list, fct = <span class="stringliteral">&#39;float&#39;</span> )</div><div class="line"><a name="l00490"></a><span class="lineno">  490</span>&#160;            phastcons_values = GeneralUtil.list_to_string( list_to_convert = phastcons_values_list,</div><div class="line"><a name="l00491"></a><span class="lineno">  491</span>&#160;                                                           sep = Constants.OTA_LIST_VALUES_SEPARATOR,</div><div class="line"><a name="l00492"></a><span class="lineno">  492</span>&#160;                                                           not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00493"></a><span class="lineno">  493</span>&#160;            </div><div class="line"><a name="l00494"></a><span class="lineno">  494</span>&#160;            <span class="comment"># Compute the values for the FLOSS scores</span></div><div class="line"><a name="l00495"></a><span class="lineno">  495</span>&#160;            floss_values_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;floss&#39;</span> )</div><div class="line"><a name="l00496"></a><span class="lineno">  496</span>&#160;            ( floss_min, floss_max, floss_median ) = GeneralUtil.min_max_median_of_list( list_of_values = floss_values_list, fct = <span class="stringliteral">&#39;float&#39;</span> )</div><div class="line"><a name="l00497"></a><span class="lineno">  497</span>&#160;            floss_values = GeneralUtil.list_to_string( list_to_convert = floss_values_list,</div><div class="line"><a name="l00498"></a><span class="lineno">  498</span>&#160;                                                       sep = Constants.OTA_LIST_VALUES_SEPARATOR,</div><div class="line"><a name="l00499"></a><span class="lineno">  499</span>&#160;                                                       not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00500"></a><span class="lineno">  500</span>&#160;                        </div><div class="line"><a name="l00501"></a><span class="lineno">  501</span>&#160;            </div><div class="line"><a name="l00502"></a><span class="lineno">  502</span>&#160;            </div><div class="line"><a name="l00503"></a><span class="lineno">  503</span>&#160;            </div><div class="line"><a name="l00512"></a><span class="lineno">  512</span>&#160;            orf_length_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;orf_length&#39;</span> )</div><div class="line"><a name="l00513"></a><span class="lineno">  513</span>&#160;            </div><div class="line"><a name="l00514"></a><span class="lineno">  514</span>&#160;            <span class="keywordflow">if</span> orf_length_list:</div><div class="line"><a name="l00515"></a><span class="lineno">  515</span>&#160;                list_of_clusters = MergeDSOTA.find_best_dsota_subpop( dsorftranscriptasso_updated_list,</div><div class="line"><a name="l00516"></a><span class="lineno">  516</span>&#160;                                                                      max_len_diff_dsota_clust )</div><div class="line"><a name="l00517"></a><span class="lineno">  517</span>&#160;                </div><div class="line"><a name="l00518"></a><span class="lineno">  518</span>&#160;                <span class="comment"># Get the number of ORF amino acid lengths that have actually been </span></div><div class="line"><a name="l00519"></a><span class="lineno">  519</span>&#160;                <span class="comment"># used to compute the &quot;clusters&quot;</span></div><div class="line"><a name="l00520"></a><span class="lineno">  520</span>&#160;                count_prov_lengths = len( orf_length_list )</div><div class="line"><a name="l00521"></a><span class="lineno">  521</span>&#160;            </div><div class="line"><a name="l00522"></a><span class="lineno">  522</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00523"></a><span class="lineno">  523</span>&#160;                list_of_clusters = [ dsorftranscriptasso_updated_list ]</div><div class="line"><a name="l00524"></a><span class="lineno">  524</span>&#160;                count_prov_lengths = 0</div><div class="line"><a name="l00525"></a><span class="lineno">  525</span>&#160;            </div><div class="line"><a name="l00526"></a><span class="lineno">  526</span>&#160;            <span class="comment"># Get the number of computed &quot;clusters&quot;</span></div><div class="line"><a name="l00527"></a><span class="lineno">  527</span>&#160;            count_clusters = len( list_of_clusters )</div><div class="line"><a name="l00528"></a><span class="lineno">  528</span>&#160;            </div><div class="line"><a name="l00529"></a><span class="lineno">  529</span>&#160;            <span class="comment"># Record the computed &quot;clusters&quot; using the IDs</span></div><div class="line"><a name="l00530"></a><span class="lineno">  530</span>&#160;            list_of_clusters_with_ids = []</div><div class="line"><a name="l00531"></a><span class="lineno">  531</span>&#160;            </div><div class="line"><a name="l00532"></a><span class="lineno">  532</span>&#160;            <span class="keywordflow">for</span> cluster <span class="keywordflow">in</span> list_of_clusters:</div><div class="line"><a name="l00533"></a><span class="lineno">  533</span>&#160;                <span class="comment"># Get the IDs of DSORFTranscriptAsso that belong to the </span></div><div class="line"><a name="l00534"></a><span class="lineno">  534</span>&#160;                <span class="comment"># same &quot;cluster&quot; and add a new cluster to the list</span></div><div class="line"><a name="l00535"></a><span class="lineno">  535</span>&#160;                list_of_clusters_with_ids.append( [ str( dsota.id ) <span class="keywordflow">for</span> dsota <span class="keywordflow">in</span> cluster ] )</div><div class="line"><a name="l00536"></a><span class="lineno">  536</span>&#160;                </div><div class="line"><a name="l00537"></a><span class="lineno">  537</span>&#160;            </div><div class="line"><a name="l00538"></a><span class="lineno">  538</span>&#160;            </div><div class="line"><a name="l00539"></a><span class="lineno">  539</span>&#160;            orftranscriptasso = ORFTranscriptAsso( id = ota_id,</div><div class="line"><a name="l00540"></a><span class="lineno">  540</span>&#160;                                                   orf_id = orf_id,</div><div class="line"><a name="l00541"></a><span class="lineno">  541</span>&#160;                                                   transcript_id = transcript_id,</div><div class="line"><a name="l00542"></a><span class="lineno">  542</span>&#160;                                                   predicted = predicted,</div><div class="line"><a name="l00543"></a><span class="lineno">  543</span>&#160;                                                   ribo_seq = ribo_seq,</div><div class="line"><a name="l00544"></a><span class="lineno">  544</span>&#160;                                                   ms_info = ms_info,</div><div class="line"><a name="l00545"></a><span class="lineno">  545</span>&#160;                                                   sequence_nt = sqce_nt_consensus,</div><div class="line"><a name="l00546"></a><span class="lineno">  546</span>&#160;                                                   sequence_aa = sqce_aa_consensus,</div><div class="line"><a name="l00547"></a><span class="lineno">  547</span>&#160;                                                   start_codon_seq = sqce_start_codon_consensus,</div><div class="line"><a name="l00548"></a><span class="lineno">  548</span>&#160;                                                   kozak_context = kozak_context,</div><div class="line"><a name="l00549"></a><span class="lineno">  549</span>&#160;                                                   length_nt_min = length_nt_min,</div><div class="line"><a name="l00550"></a><span class="lineno">  550</span>&#160;                                                   length_nt_max = length_nt_max,</div><div class="line"><a name="l00551"></a><span class="lineno">  551</span>&#160;                                                   length_nt_median = length_nt_median,</div><div class="line"><a name="l00552"></a><span class="lineno">  552</span>&#160;                                                   length_nt_values = length_nt_values,</div><div class="line"><a name="l00553"></a><span class="lineno">  553</span>&#160;                                                   length_aa_min = length_aa_min,</div><div class="line"><a name="l00554"></a><span class="lineno">  554</span>&#160;                                                   length_aa_max = length_aa_max,</div><div class="line"><a name="l00555"></a><span class="lineno">  555</span>&#160;                                                   length_aa_median = length_aa_median,</div><div class="line"><a name="l00556"></a><span class="lineno">  556</span>&#160;                                                   length_aa_values = length_aa_values,                                                   </div><div class="line"><a name="l00557"></a><span class="lineno">  557</span>&#160;                                                   orf_score_min = orfscore_min,</div><div class="line"><a name="l00558"></a><span class="lineno">  558</span>&#160;                                                   orf_score_max = orfscore_max,</div><div class="line"><a name="l00559"></a><span class="lineno">  559</span>&#160;                                                   orf_score_median = orfscore_median,</div><div class="line"><a name="l00560"></a><span class="lineno">  560</span>&#160;                                                   orf_score_values = orfscore_values,                                                   </div><div class="line"><a name="l00561"></a><span class="lineno">  561</span>&#160;                                                   phylocsf_min = phylocsf_min,</div><div class="line"><a name="l00562"></a><span class="lineno">  562</span>&#160;                                                   phylocsf_max = phylocsf_max,</div><div class="line"><a name="l00563"></a><span class="lineno">  563</span>&#160;                                                   phylocsf_median = phylocsf_median,</div><div class="line"><a name="l00564"></a><span class="lineno">  564</span>&#160;                                                   phylocsf_values = phylocsf_values,                                                   </div><div class="line"><a name="l00565"></a><span class="lineno">  565</span>&#160;                                                   phastcons_min = phastcons_min,</div><div class="line"><a name="l00566"></a><span class="lineno">  566</span>&#160;                                                   phastcons_max = phastcons_max,</div><div class="line"><a name="l00567"></a><span class="lineno">  567</span>&#160;                                                   phastcons_median = phastcons_median,</div><div class="line"><a name="l00568"></a><span class="lineno">  568</span>&#160;                                                   phastcons_values = phastcons_values,                                                   </div><div class="line"><a name="l00569"></a><span class="lineno">  569</span>&#160;                                                   floss_min = floss_min,</div><div class="line"><a name="l00570"></a><span class="lineno">  570</span>&#160;                                                   floss_max = floss_max,</div><div class="line"><a name="l00571"></a><span class="lineno">  571</span>&#160;                                                   floss_median = floss_median,</div><div class="line"><a name="l00572"></a><span class="lineno">  572</span>&#160;                                                   floss_values = floss_values,</div><div class="line"><a name="l00573"></a><span class="lineno">  573</span>&#160;                                                   count_ds = len( dsorftranscriptasso_list ),</div><div class="line"><a name="l00574"></a><span class="lineno">  574</span>&#160;                                                   count_computed_clusters = count_clusters,</div><div class="line"><a name="l00575"></a><span class="lineno">  575</span>&#160;                                                   count_prov_lengths = count_prov_lengths,</div><div class="line"><a name="l00576"></a><span class="lineno">  576</span>&#160;                                                   computed_clusters = str( list_of_clusters_with_ids ),</div><div class="line"><a name="l00577"></a><span class="lineno">  577</span>&#160;                                                   gen_len_eq_orf_len = <span class="keywordtype">None</span> )</div><div class="line"><a name="l00578"></a><span class="lineno">  578</span>&#160;            objects_to_insert.append( orftranscriptasso )</div><div class="line"><a name="l00579"></a><span class="lineno">  579</span>&#160;            </div><div class="line"><a name="l00580"></a><span class="lineno">  580</span>&#160;            </div><div class="line"><a name="l00581"></a><span class="lineno">  581</span>&#160;            <span class="comment"># Create new entries in the CellContext table to register the</span></div><div class="line"><a name="l00582"></a><span class="lineno">  582</span>&#160;            <span class="comment"># cellular contexts related to the ORFTranscriptAsso entry</span></div><div class="line"><a name="l00583"></a><span class="lineno">  583</span>&#160;            cell_context_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;cell_context&#39;</span> )</div><div class="line"><a name="l00584"></a><span class="lineno">  584</span>&#160;            </div><div class="line"><a name="l00585"></a><span class="lineno">  585</span>&#160;            <span class="keywordflow">if</span> cell_context_list:</div><div class="line"><a name="l00586"></a><span class="lineno">  586</span>&#160;                </div><div class="line"><a name="l00587"></a><span class="lineno">  587</span>&#160;                <span class="comment"># Split the cellular contexts that can eventually have been</span></div><div class="line"><a name="l00588"></a><span class="lineno">  588</span>&#160;                <span class="comment"># concatenated by the source or during the data insertion</span></div><div class="line"><a name="l00589"></a><span class="lineno">  589</span>&#160;                list_cell_context = []</div><div class="line"><a name="l00590"></a><span class="lineno">  590</span>&#160;                <span class="keywordflow">for</span> cell_ctxt <span class="keywordflow">in</span> cell_context_list:</div><div class="line"><a name="l00591"></a><span class="lineno">  591</span>&#160;                    list_cell_context += cell_ctxt.split( <span class="stringliteral">&#39;,&#39;</span> )</div><div class="line"><a name="l00592"></a><span class="lineno">  592</span>&#160;                    </div><div class="line"><a name="l00593"></a><span class="lineno">  593</span>&#160;                <span class="comment"># Get the set of unique values for the cellular context</span></div><div class="line"><a name="l00594"></a><span class="lineno">  594</span>&#160;                list_uniq_cell_context = set( list_cell_context )</div><div class="line"><a name="l00595"></a><span class="lineno">  595</span>&#160;                </div><div class="line"><a name="l00596"></a><span class="lineno">  596</span>&#160;                <span class="keywordflow">for</span> cell_ctxt <span class="keywordflow">in</span> list_uniq_cell_context:</div><div class="line"><a name="l00597"></a><span class="lineno">  597</span>&#160;                    </div><div class="line"><a name="l00598"></a><span class="lineno">  598</span>&#160;                    <span class="comment"># If the CellContextCatalog entry has not yet been instantiated, </span></div><div class="line"><a name="l00599"></a><span class="lineno">  599</span>&#160;                    <span class="comment"># then create a new entry in this table.</span></div><div class="line"><a name="l00600"></a><span class="lineno">  600</span>&#160;                    cell_ctxt_catalog = CellContextCatalog( context = cell_ctxt )</div><div class="line"><a name="l00601"></a><span class="lineno">  601</span>&#160;                    </div><div class="line"><a name="l00602"></a><span class="lineno">  602</span>&#160;                    existing_cell_ctxt_catalog = cell_ctxt_catalog_to_insert.get( cell_ctxt_catalog )                   </div><div class="line"><a name="l00603"></a><span class="lineno">  603</span>&#160;                    <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> existing_cell_ctxt_catalog ):</div><div class="line"><a name="l00604"></a><span class="lineno">  604</span>&#160;                        cell_ctxt_catalog_to_insert[ cell_ctxt_catalog ] = cell_ctxt_catalog</div><div class="line"><a name="l00605"></a><span class="lineno">  605</span>&#160;                    </div><div class="line"><a name="l00606"></a><span class="lineno">  606</span>&#160;                    <span class="comment"># Create a new entry in the CellContext table</span></div><div class="line"><a name="l00607"></a><span class="lineno">  607</span>&#160;                    cellcontext = CellContext( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00608"></a><span class="lineno">  608</span>&#160;                                               cell_context = cell_ctxt )</div><div class="line"><a name="l00609"></a><span class="lineno">  609</span>&#160;                    objects_to_insert.append( cellcontext )</div><div class="line"><a name="l00610"></a><span class="lineno">  610</span>&#160;                </div><div class="line"><a name="l00611"></a><span class="lineno">  611</span>&#160;                </div><div class="line"><a name="l00612"></a><span class="lineno">  612</span>&#160;            <span class="comment"># Create new entries in the ProvidedCategory table to register the </span></div><div class="line"><a name="l00613"></a><span class="lineno">  613</span>&#160;            <span class="comment"># ORF categories provided by the data sources for this </span></div><div class="line"><a name="l00614"></a><span class="lineno">  614</span>&#160;            <span class="comment"># ORF - transcript association</span></div><div class="line"><a name="l00615"></a><span class="lineno">  615</span>&#160;            provided_category_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;provided_category&#39;</span> )</div><div class="line"><a name="l00616"></a><span class="lineno">  616</span>&#160;            </div><div class="line"><a name="l00617"></a><span class="lineno">  617</span>&#160;            <span class="keywordflow">if</span> provided_category_list:</div><div class="line"><a name="l00618"></a><span class="lineno">  618</span>&#160;                </div><div class="line"><a name="l00619"></a><span class="lineno">  619</span>&#160;                <span class="comment"># Split the categories that can eventually have been </span></div><div class="line"><a name="l00620"></a><span class="lineno">  620</span>&#160;                <span class="comment"># concatenated by the source or during data insertion</span></div><div class="line"><a name="l00621"></a><span class="lineno">  621</span>&#160;                list_provided_category = []</div><div class="line"><a name="l00622"></a><span class="lineno">  622</span>&#160;                <span class="keywordflow">for</span> prov_cat <span class="keywordflow">in</span> provided_category_list:</div><div class="line"><a name="l00623"></a><span class="lineno">  623</span>&#160;                    <span class="comment"># Replace the blanks by underscores in the provided categories</span></div><div class="line"><a name="l00624"></a><span class="lineno">  624</span>&#160;                    list_provided_category += prov_cat.replace( <span class="stringliteral">&#39;/&#39;</span>, <span class="stringliteral">&#39;,&#39;</span> ).replace( <span class="stringliteral">&#39; &#39;</span>, <span class="stringliteral">&#39;_&#39;</span> ).split( <span class="stringliteral">&#39;,&#39;</span> )</div><div class="line"><a name="l00625"></a><span class="lineno">  625</span>&#160;                </div><div class="line"><a name="l00626"></a><span class="lineno">  626</span>&#160;                </div><div class="line"><a name="l00627"></a><span class="lineno">  627</span>&#160;                <span class="comment"># Get the set of unique values for the cellular context</span></div><div class="line"><a name="l00628"></a><span class="lineno">  628</span>&#160;                list_uniq_provided_category = set( list_provided_category )</div><div class="line"><a name="l00629"></a><span class="lineno">  629</span>&#160;                </div><div class="line"><a name="l00630"></a><span class="lineno">  630</span>&#160;                <span class="keywordflow">for</span> prov_cat <span class="keywordflow">in</span> list_uniq_provided_category:</div><div class="line"><a name="l00631"></a><span class="lineno">  631</span>&#160;                    </div><div class="line"><a name="l00632"></a><span class="lineno">  632</span>&#160;                    <span class="keywordflow">if</span> <span class="keywordflow">not</span> GeneralUtil.is_empty( prov_cat ):</div><div class="line"><a name="l00633"></a><span class="lineno">  633</span>&#160;                    </div><div class="line"><a name="l00634"></a><span class="lineno">  634</span>&#160;                        <span class="comment"># If the ORFCatagoryCatalog entry has not yet been instantiated, </span></div><div class="line"><a name="l00635"></a><span class="lineno">  635</span>&#160;                        <span class="comment"># then create a new entry in this table.</span></div><div class="line"><a name="l00636"></a><span class="lineno">  636</span>&#160;                        provided_cat_catalog = ProvidedCategoryCatalog( category = prov_cat )</div><div class="line"><a name="l00637"></a><span class="lineno">  637</span>&#160;                        </div><div class="line"><a name="l00638"></a><span class="lineno">  638</span>&#160;                        existing_provided_cat_catalog = provided_cat_catalog_to_insert.get( provided_cat_catalog )                   </div><div class="line"><a name="l00639"></a><span class="lineno">  639</span>&#160;                        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> existing_provided_cat_catalog ):</div><div class="line"><a name="l00640"></a><span class="lineno">  640</span>&#160;                            provided_cat_catalog_to_insert[ provided_cat_catalog ] = provided_cat_catalog</div><div class="line"><a name="l00641"></a><span class="lineno">  641</span>&#160;                        </div><div class="line"><a name="l00642"></a><span class="lineno">  642</span>&#160;                        <span class="comment"># Create a new entry in the ProvidedCategory table</span></div><div class="line"><a name="l00643"></a><span class="lineno">  643</span>&#160;                        providedcategory = ProvidedCategory( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00644"></a><span class="lineno">  644</span>&#160;                                                             provided_category = prov_cat )</div><div class="line"><a name="l00645"></a><span class="lineno">  645</span>&#160;                        objects_to_insert.append( providedcategory )</div><div class="line"><a name="l00646"></a><span class="lineno">  646</span>&#160;                </div><div class="line"><a name="l00647"></a><span class="lineno">  647</span>&#160;                </div><div class="line"><a name="l00648"></a><span class="lineno">  648</span>&#160;            <span class="comment"># Create new entries in the FLOSSClass table to register</span></div><div class="line"><a name="l00649"></a><span class="lineno">  649</span>&#160;            <span class="comment"># the FLOSS categories provided by the data sources</span></div><div class="line"><a name="l00650"></a><span class="lineno">  650</span>&#160;            flossclass_list = dsorftranscriptasso_dict.get( <span class="stringliteral">&#39;floss_class&#39;</span> )</div><div class="line"><a name="l00651"></a><span class="lineno">  651</span>&#160;            <span class="keywordflow">if</span> flossclass_list:</div><div class="line"><a name="l00652"></a><span class="lineno">  652</span>&#160;                </div><div class="line"><a name="l00653"></a><span class="lineno">  653</span>&#160;                <span class="comment"># Get the set of unique values for the FLOSS classes</span></div><div class="line"><a name="l00654"></a><span class="lineno">  654</span>&#160;                list_uniq_flossclass = set( flossclass_list )</div><div class="line"><a name="l00655"></a><span class="lineno">  655</span>&#160;                </div><div class="line"><a name="l00656"></a><span class="lineno">  656</span>&#160;                <span class="keywordflow">for</span> floss_cls <span class="keywordflow">in</span> list_uniq_flossclass:</div><div class="line"><a name="l00657"></a><span class="lineno">  657</span>&#160;                    </div><div class="line"><a name="l00658"></a><span class="lineno">  658</span>&#160;                    </div><div class="line"><a name="l00659"></a><span class="lineno">  659</span>&#160;                    <span class="comment"># If the FLOSSClassCatalog entry has not yet been instantiated, </span></div><div class="line"><a name="l00660"></a><span class="lineno">  660</span>&#160;                    <span class="comment"># then create a new entry in this table.</span></div><div class="line"><a name="l00661"></a><span class="lineno">  661</span>&#160;                    floss_class_cat = FLOSSClassCatalog( floss_class = floss_cls )</div><div class="line"><a name="l00662"></a><span class="lineno">  662</span>&#160;                                                   </div><div class="line"><a name="l00663"></a><span class="lineno">  663</span>&#160;                    existing_floss_class_cat = floss_class_catelog_to_insert.get( floss_class_cat )                   </div><div class="line"><a name="l00664"></a><span class="lineno">  664</span>&#160;                    <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> existing_floss_class_cat ):</div><div class="line"><a name="l00665"></a><span class="lineno">  665</span>&#160;                        floss_class_catelog_to_insert[ floss_class_cat ] = floss_class_cat                              </div><div class="line"><a name="l00666"></a><span class="lineno">  666</span>&#160;                    </div><div class="line"><a name="l00667"></a><span class="lineno">  667</span>&#160;                    <span class="comment"># Create a new entry in the FLOSSClass table</span></div><div class="line"><a name="l00668"></a><span class="lineno">  668</span>&#160;                    flossclass = FLOSSClass( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00669"></a><span class="lineno">  669</span>&#160;                                             floss_class = floss_cls )</div><div class="line"><a name="l00670"></a><span class="lineno">  670</span>&#160;                    objects_to_insert.append( flossclass )</div><div class="line"><a name="l00671"></a><span class="lineno">  671</span>&#160;                </div><div class="line"><a name="l00672"></a><span class="lineno">  672</span>&#160;                </div><div class="line"><a name="l00673"></a><span class="lineno">  673</span>&#160;            <span class="comment"># Create the new ORFTranscriptAssoDSAsso</span></div><div class="line"><a name="l00674"></a><span class="lineno">  674</span>&#160;            <span class="keywordflow">for</span> k <span class="keywordflow">in</span> range( len( dsorftranscriptasso_list ) ):</div><div class="line"><a name="l00675"></a><span class="lineno">  675</span>&#160;                orftranscriptassodsasso = ORFTranscriptAssoDSAsso( orftranscriptasso_id = ota_id,</div><div class="line"><a name="l00676"></a><span class="lineno">  676</span>&#160;                                                                   dsorftranscriptasso_id = dsorftranscriptasso_list[ k ].id,</div><div class="line"><a name="l00677"></a><span class="lineno">  677</span>&#160;                                                                   data_source = dsorftranscriptasso_list[ k ].data_source )</div><div class="line"><a name="l00678"></a><span class="lineno">  678</span>&#160;                objects_to_insert.append( orftranscriptassodsasso )</div><div class="line"><a name="l00679"></a><span class="lineno">  679</span>&#160;        </div><div class="line"><a name="l00680"></a><span class="lineno">  680</span>&#160;        <span class="comment"># Turn dictionary of instantiated CellContextCatalog, ProvidedCategoryCatalog </span></div><div class="line"><a name="l00681"></a><span class="lineno">  681</span>&#160;        <span class="comment"># and FlossClassCatalog objects into lists</span></div><div class="line"><a name="l00682"></a><span class="lineno">  682</span>&#160;        cell_ctxt_catalog_to_insert = cell_ctxt_catalog_to_insert.keys()</div><div class="line"><a name="l00683"></a><span class="lineno">  683</span>&#160;        provided_cat_catalog_to_insert = provided_cat_catalog_to_insert.keys()</div><div class="line"><a name="l00684"></a><span class="lineno">  684</span>&#160;        floss_class_catelog_to_insert = floss_class_catelog_to_insert.keys()</div><div class="line"><a name="l00685"></a><span class="lineno">  685</span>&#160;        </div><div class="line"><a name="l00686"></a><span class="lineno">  686</span>&#160;        <span class="keywordflow">return</span> ( objects_to_insert, cell_ctxt_catalog_to_insert, provided_cat_catalog_to_insert, </div><div class="line"><a name="l00687"></a><span class="lineno">  687</span>&#160;                 floss_class_catelog_to_insert, error_messages_to_log, warning_messages_to_log )</div><div class="line"><a name="l00688"></a><span class="lineno">  688</span>&#160;        </div><div class="line"><a name="l00689"></a><span class="lineno">  689</span>&#160;    </div><div class="line"><a name="l00690"></a><span class="lineno">  690</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a9ffc33a50c6477debf9e9c32ea85b871"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9ffc33a50c6477debf9e9c32ea85b871">&#9670;&nbsp;</a></span>split_dsota_using_length()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.merge.MergeDSOTA.MergeDSOTA.split_dsota_using_length </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dsorftranscriptasso_list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>max_len_diff</em> = <code>Constants.DEFAULT_MAX_LEN_DIFF_FOR_DSOTA_CLUSTERS</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>split_dsota_using_length </h2>
<p>This is a static method that allows to split a list of DSORFTranscriptAsso entries into several sublists if the differences between amino acid ORF length exceed a threshold. If there are DSORFTranscriptAsso missing their ORF lengths, then they are not considered to build in the "clusters".</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dsorftranscriptasso_list</td><td>List of DSORFTranscriptAsso - The list of DSORFTranscriptAsso instances to split into clusters. </td></tr>
    <tr><td class="paramname">max_len_diff</td><td>Integer - The maximal difference between the maximal and minimal lengths of DSORFTranscriptAsso so they belong to the same group (value included). e.g. If max_len_diff = 3, Two DSORFTranscriptAsso will be considered as being part of two different "clusters" if the difference of their length exceed 4.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>sublists_dsota: List of list of DSORFTranscriptAsso - The list of DSORFTranscriptAsso objects grouped by similar lengths. </dd></dl>
<div class="fragment"><div class="line"><a name="l00966"></a><span class="lineno">  966</span>&#160;    <span class="keyword">def </span>split_dsota_using_length( dsorftranscriptasso_list, max_len_diff=Constants.DEFAULT_MAX_LEN_DIFF_FOR_DSOTA_CLUSTERS ):</div><div class="line"><a name="l00967"></a><span class="lineno">  967</span>&#160;        </div><div class="line"><a name="l00968"></a><span class="lineno">  968</span>&#160;        <span class="comment"># Build a dictionary that associate to each DSORFTranscriptAsso </span></div><div class="line"><a name="l00969"></a><span class="lineno">  969</span>&#160;        <span class="comment"># its length when it is provided</span></div><div class="line"><a name="l00970"></a><span class="lineno">  970</span>&#160;        length_dict = {}</div><div class="line"><a name="l00971"></a><span class="lineno">  971</span>&#160;        <span class="keywordflow">for</span> dsorftranscriptasso <span class="keywordflow">in</span> dsorftranscriptasso_list:</div><div class="line"><a name="l00972"></a><span class="lineno">  972</span>&#160;            <span class="comment"># Fill in the dictionary if the DSORFTranscriptAsso has a length</span></div><div class="line"><a name="l00973"></a><span class="lineno">  973</span>&#160;            <span class="keywordflow">if</span> ( dsorftranscriptasso.orf_length != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00974"></a><span class="lineno">  974</span>&#160;                length_dict[ dsorftranscriptasso ] = int( dsorftranscriptasso.orf_length )</div><div class="line"><a name="l00975"></a><span class="lineno">  975</span>&#160;        </div><div class="line"><a name="l00976"></a><span class="lineno">  976</span>&#160;        <span class="comment"># Get the sorted list of lengths and get the maximal and minimal values for these lengths</span></div><div class="line"><a name="l00977"></a><span class="lineno">  977</span>&#160;        length_list_sorted = sorted( length_dict.values() )</div><div class="line"><a name="l00978"></a><span class="lineno">  978</span>&#160;        max_len = max( length_list_sorted )</div><div class="line"><a name="l00979"></a><span class="lineno">  979</span>&#160;        min_len = min( length_list_sorted )</div><div class="line"><a name="l00980"></a><span class="lineno">  980</span>&#160;        </div><div class="line"><a name="l00981"></a><span class="lineno">  981</span>&#160;        <span class="comment"># If the absolute difference between the maximal and minimal length is</span></div><div class="line"><a name="l00982"></a><span class="lineno">  982</span>&#160;        <span class="comment"># higher than the threshold, then one may consider the ORF should not </span></div><div class="line"><a name="l00983"></a><span class="lineno">  983</span>&#160;        <span class="comment"># have been merged. Then, regroup the DSORFTranscriptAsso that belong </span></div><div class="line"><a name="l00984"></a><span class="lineno">  984</span>&#160;        <span class="comment"># to the &quot;same group&quot;.</span></div><div class="line"><a name="l00985"></a><span class="lineno">  985</span>&#160;        <span class="keywordflow">if</span> ( abs( max_len - min_len ) &gt; max_len_diff ):</div><div class="line"><a name="l00986"></a><span class="lineno">  986</span>&#160;            </div><div class="line"><a name="l00987"></a><span class="lineno">  987</span>&#160;            <span class="comment"># Compute the difference between each consecutive value</span></div><div class="line"><a name="l00988"></a><span class="lineno">  988</span>&#160;            list_of_length_diff = []</div><div class="line"><a name="l00989"></a><span class="lineno">  989</span>&#160;            </div><div class="line"><a name="l00990"></a><span class="lineno">  990</span>&#160;            <span class="keywordflow">for</span> k <span class="keywordflow">in</span> range( len( length_list_sorted ) - 1 ):</div><div class="line"><a name="l00991"></a><span class="lineno">  991</span>&#160;                list_of_length_diff.append( length_list_sorted[ k + 1 ] - length_list_sorted[ k ] )</div><div class="line"><a name="l00992"></a><span class="lineno">  992</span>&#160;                </div><div class="line"><a name="l00993"></a><span class="lineno">  993</span>&#160;            <span class="comment"># Find the highest difference(s) in the list</span></div><div class="line"><a name="l00994"></a><span class="lineno">  994</span>&#160;            max_diff = max( list_of_length_diff )</div><div class="line"><a name="l00995"></a><span class="lineno">  995</span>&#160;            </div><div class="line"><a name="l00996"></a><span class="lineno">  996</span>&#160;            <span class="comment"># Split the list of lengths into several lists, using the maximal difference</span></div><div class="line"><a name="l00997"></a><span class="lineno">  997</span>&#160;            <span class="comment"># as the point of separation</span></div><div class="line"><a name="l00998"></a><span class="lineno">  998</span>&#160;            sublists_of_lengths = [ [] ]</div><div class="line"><a name="l00999"></a><span class="lineno">  999</span>&#160;            list_nb = 0</div><div class="line"><a name="l01000"></a><span class="lineno"> 1000</span>&#160;            </div><div class="line"><a name="l01001"></a><span class="lineno"> 1001</span>&#160;            <span class="keywordflow">for</span> k <span class="keywordflow">in</span> range( len( length_list_sorted ) ):</div><div class="line"><a name="l01002"></a><span class="lineno"> 1002</span>&#160;                </div><div class="line"><a name="l01003"></a><span class="lineno"> 1003</span>&#160;                sublists_of_lengths[ list_nb ].append( length_list_sorted[ k ] )</div><div class="line"><a name="l01004"></a><span class="lineno"> 1004</span>&#160;                </div><div class="line"><a name="l01005"></a><span class="lineno"> 1005</span>&#160;                <span class="keywordflow">if</span> ( ( k &lt; len( list_of_length_diff )</div><div class="line"><a name="l01006"></a><span class="lineno"> 1006</span>&#160;                       <span class="keywordflow">and</span> ( list_of_length_diff[ k ] == max_diff ) ) ):</div><div class="line"><a name="l01007"></a><span class="lineno"> 1007</span>&#160;                    list_nb += 1</div><div class="line"><a name="l01008"></a><span class="lineno"> 1008</span>&#160;                    sublists_of_lengths.append( [] )</div><div class="line"><a name="l01009"></a><span class="lineno"> 1009</span>&#160;        </div><div class="line"><a name="l01010"></a><span class="lineno"> 1010</span>&#160;            <span class="comment"># Recover the DSORFTranscriptAsso corresponding to these lengths</span></div><div class="line"><a name="l01011"></a><span class="lineno"> 1011</span>&#160;            <span class="comment"># Instantiate an empty list that aims to receive the lists of &quot;clusters&quot;</span></div><div class="line"><a name="l01012"></a><span class="lineno"> 1012</span>&#160;            sublists_dsota = []</div><div class="line"><a name="l01013"></a><span class="lineno"> 1013</span>&#160;            </div><div class="line"><a name="l01014"></a><span class="lineno"> 1014</span>&#160;            <span class="keywordflow">for</span> sublist <span class="keywordflow">in</span> sublists_of_lengths:</div><div class="line"><a name="l01015"></a><span class="lineno"> 1015</span>&#160;                </div><div class="line"><a name="l01016"></a><span class="lineno"> 1016</span>&#160;                <span class="comment"># Get a list of unique values</span></div><div class="line"><a name="l01017"></a><span class="lineno"> 1017</span>&#160;                subls = list( set( sublist ) )</div><div class="line"><a name="l01018"></a><span class="lineno"> 1018</span>&#160;                </div><div class="line"><a name="l01019"></a><span class="lineno"> 1019</span>&#160;                all_corresp_dsota = []</div><div class="line"><a name="l01020"></a><span class="lineno"> 1020</span>&#160;                </div><div class="line"><a name="l01021"></a><span class="lineno"> 1021</span>&#160;                <span class="keywordflow">for</span> dsota_len_subls <span class="keywordflow">in</span> subls:</div><div class="line"><a name="l01022"></a><span class="lineno"> 1022</span>&#160;                    <span class="comment"># Get all DSORFTranscriptAsso belonging to this sublist</span></div><div class="line"><a name="l01023"></a><span class="lineno"> 1023</span>&#160;                    corresponding_dsota = [ dsota <span class="keywordflow">for</span> ( dsota, dsota_len ) <span class="keywordflow">in</span> length_dict.items() <span class="keywordflow">if</span> ( dsota_len == dsota_len_subls ) ]</div><div class="line"><a name="l01024"></a><span class="lineno"> 1024</span>&#160;                    all_corresp_dsota += corresponding_dsota</div><div class="line"><a name="l01025"></a><span class="lineno"> 1025</span>&#160;                </div><div class="line"><a name="l01026"></a><span class="lineno"> 1026</span>&#160;                sublists_dsota.append( all_corresp_dsota )</div><div class="line"><a name="l01027"></a><span class="lineno"> 1027</span>&#160;        </div><div class="line"><a name="l01028"></a><span class="lineno"> 1028</span>&#160;        <span class="comment"># Otherwise, all the DSORFTranscriptAsso belong to the same unique &quot;cluster&quot;</span></div><div class="line"><a name="l01029"></a><span class="lineno"> 1029</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l01030"></a><span class="lineno"> 1030</span>&#160;            sublists_dsota = [ dsorftranscriptasso_list ]</div><div class="line"><a name="l01031"></a><span class="lineno"> 1031</span>&#160;        </div><div class="line"><a name="l01032"></a><span class="lineno"> 1032</span>&#160;        <span class="keywordflow">return</span> sublists_dsota</div><div class="line"><a name="l01033"></a><span class="lineno"> 1033</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a298a5e53b8f2f4d79996a37f04f0cab6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a298a5e53b8f2f4d79996a37f04f0cab6">&#9670;&nbsp;</a></span>DSORFTRANSCRIPTASSO_ATTRIBUTES_TO_COMPARE</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">list fr.tagc.uorf.core.execution.merge.MergeDSOTA.MergeDSOTA.DSORFTRANSCRIPTASSO_ATTRIBUTES_TO_COMPARE</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">=  [ <span class="stringliteral">&#39;id&#39;</span>, <span class="stringliteral">&#39;data_source&#39;</span>, <span class="stringliteral">&#39;predicted&#39;</span>, <span class="stringliteral">&#39;ribo_seq&#39;</span>, </div><div class="line">                                                  <span class="stringliteral">&#39;cell_context&#39;</span>, <span class="stringliteral">&#39;raw_sequence&#39;</span>, <span class="stringliteral">&#39;raw_sequence_aa&#39;</span>,</div><div class="line">                                                  <span class="stringliteral">&#39;start_codon_seq&#39;</span>, <span class="stringliteral">&#39;kozak_context&#39;</span>, <span class="stringliteral">&#39;orf_length_nt&#39;</span>,</div><div class="line">                                                  <span class="stringliteral">&#39;orf_length&#39;</span>, <span class="stringliteral">&#39;provided_category&#39;</span>, <span class="stringliteral">&#39;ms_info&#39;</span>,</div><div class="line">                                                  <span class="stringliteral">&#39;orf_score&#39;</span>, <span class="stringliteral">&#39;phylocsf&#39;</span>, <span class="stringliteral">&#39;phastcons&#39;</span>, <span class="stringliteral">&#39;floss&#39;</span>, </div><div class="line">                                                  <span class="stringliteral">&#39;floss_class&#39;</span> ]</div></div><!-- fragment --><h2>Class variables </h2>
<p>List of attributes that needs to be compared to merge several DSORFTranscriptAsso entries together </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/execution/merge/MergeDSOTA.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
