<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017 Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>execution</b></li><li class="navelem"><b>insertion</b></li><li class="navelem"><b>parserstrategy</b></li><li class="navelem"><b>common</b></li><li class="navelem"><b>Samandi2017</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html">Samandi2017</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017 Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Source: Samandi et al., 2017.  
 <a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.png" usemap="#fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017_map" alt=""/>
  <map id="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017_map" name="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017_map">
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html" alt="fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy" shape="rect" coords="0,0,512,24"/>
  </map>
</div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ab35b35829138cef8d8e8ea3a671bd2b0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html#ab35b35829138cef8d8e8ea3a671bd2b0">__init__</a> (self, data_path)</td></tr>
<tr class="separator:ab35b35829138cef8d8e8ea3a671bd2b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a174b65f01d550105ef67642886958703"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html#a174b65f01d550105ef67642886958703">execute</a> (self)</td></tr>
<tr class="separator:a174b65f01d550105ef67642886958703"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td colspan="2" onclick="javascript:toggleInherit('pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy')"><img src="closed.png" alt="-"/>&#160;Public Member Functions inherited from <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html">fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy</a></td></tr>
<tr class="memitem:a8cdab4b401f878c517c4b737b9ecf778 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a8cdab4b401f878c517c4b737b9ecf778">execute</a> ()</td></tr>
<tr class="separator:a8cdab4b401f878c517c4b737b9ecf778 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11654f0533e3a163aa528c8f996c3456 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a11654f0533e3a163aa528c8f996c3456">import_file_content</a> (self, file_sep=',', dtype_val='str', encoding_val='utf-8')</td></tr>
<tr class="separator:a11654f0533e3a163aa528c8f996c3456 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb455bdb96149592f7251e5f224c7fa3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#abb455bdb96149592f7251e5f224c7fa3">check_headers</a> (self)</td></tr>
<tr class="separator:abb455bdb96149592f7251e5f224c7fa3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0b04e0792f45b98d4111aa27c51ac93 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ab0b04e0792f45b98d4111aa27c51ac93">initialize_ids</a> (self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, orf_stop, orf_strand, index, log_symb_from_tr=True, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:ab0b04e0792f45b98d4111aa27c51ac93 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa01883a5db8dba8aabdb8af61d3b1f94 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#aa01883a5db8dba8aabdb8af61d3b1f94">initialize_transcript_id</a> (self, transcript_id, orf_id, index)</td></tr>
<tr class="separator:aa01883a5db8dba8aabdb8af61d3b1f94 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a07cdd23100d43df77f4001cd81fba324 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a07cdd23100d43df77f4001cd81fba324">initialize_gene_symbol</a> (self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, orf_stop, orf_strand, log_symb_from_tr=True, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:a07cdd23100d43df77f4001cd81fba324 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d5bdcc0943ba6c1c71bb3c245bc81b3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4d5bdcc0943ba6c1c71bb3c245bc81b3">symbol_from_transcript_id</a> (self, ensembl_db, orf_id, transcript_id, log_warning=True)</td></tr>
<tr class="separator:a4d5bdcc0943ba6c1c71bb3c245bc81b3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac199b748e002f7651a7ad708197c0008 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ac199b748e002f7651a7ad708197c0008">symbol_from_orf_loc</a> (self, ensembl_db, orf_id, chr_name, orf_start, orf_stop, orf_strand, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:ac199b748e002f7651a7ad708197c0008 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4251bb5eb9243ff185efba305843d86b inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4251bb5eb9243ff185efba305843d86b">get_gene_from_symbol</a> (self, symbol, chr_name, orf_id)</td></tr>
<tr class="separator:a4251bb5eb9243ff185efba305843d86b inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a10fea64161666d9bb47ee4b8553b52e5 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a10fea64161666d9bb47ee4b8553b52e5">find_gene_from_symbol</a> (self, symbol, chr_name)</td></tr>
<tr class="separator:a10fea64161666d9bb47ee4b8553b52e5 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9d3195a3bb1fd975359b771ebee910ef inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a9d3195a3bb1fd975359b771ebee910ef">get_gene_from_conflict_id</a> (self, conflict_symbol, chr_name)</td></tr>
<tr class="separator:a9d3195a3bb1fd975359b771ebee910ef inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:abefd2f3acca36e54700b22027c0c3d5d"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html#abefd2f3acca36e54700b22027c0c3d5d">AUTHORIZED_SPECIES</a></td></tr>
<tr class="separator:abefd2f3acca36e54700b22027c0c3d5d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="inherited"></a>
Additional Inherited Members</h2></td></tr>
<tr class="inherit_header pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td colspan="2" onclick="javascript:toggleInherit('pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy')"><img src="closed.png" alt="-"/>&#160;Static Public Member Functions inherited from <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html">fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy</a></td></tr>
<tr class="memitem:a602be6456b764340546d4ff9fd07d19f inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a602be6456b764340546d4ff9fd07d19f">is_empty</a> (val, empty_val=Constants.EMPTY_VALUES)</td></tr>
<tr class="separator:a602be6456b764340546d4ff9fd07d19f inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a846efdd59ec331984a46aa80a46afcf2 inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a846efdd59ec331984a46aa80a46afcf2">get_ensembl_db</a> (annotation_version)</td></tr>
<tr class="separator:a846efdd59ec331984a46aa80a46afcf2 inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a19c8d87c05bf4071bb6c7b541106a12b inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a19c8d87c05bf4071bb6c7b541106a12b">initialize_chromosome_name</a> (chr_name)</td></tr>
<tr class="separator:a19c8d87c05bf4071bb6c7b541106a12b inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a37ad31eeda67bd00e0aba6bfe491eecf inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a37ad31eeda67bd00e0aba6bfe491eecf">get_uniq_gene_on_chromosome</a> (chr, list_of_genes)</td></tr>
<tr class="separator:a37ad31eeda67bd00e0aba6bfe491eecf inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Source: Samandi et al., 2017. </p>
<p>NB: This file contains a header. This file is expected to be in a tab separated values (TSV) format.</p>
<p>Description of the columns of the raw data file:</p><ul>
<li>[0]: Protein accession: ORF ID - String</li>
<li>[1]: Frame: Reading frame - String</li>
<li>[2]: Kozak: Kozak context - Integer</li>
<li>[3]: Gene: Gene symbol - String</li>
<li>[4]: Chromosome: Chromosome name - String</li>
<li>[5]: Start genomic coordinate - Genomic coordinates of the start codon (coordinates of the 1st nucleotide) - Integer</li>
<li>[6]: End genomic coordinate - Genomic coordinates of the stop codon (coordinates of the last nucleotide) - Integer</li>
<li><p class="startli">[7]: Protein sequence: Protein sequence - String</p>
<p class="startli">Supplementary information</p><ul>
<li>The chromosome name is provided without 'chr' prefix.</li>
<li>The genomic coordinates are reported on GRCh38 or GRCm38 annotation. (NCBI GCF_000001405.26, correspond to Ensembl v76 for H. sapiens, NCBI GCF_000001635.22, correspond to Ensembl v75 for M. musculus).</li>
<li>For '+'-strand ORFs,<ul>
<li>the first nucleotide of the start codon is located at the start position (Start genomic coordinate).</li>
<li>the last nucleotide of the stop codon is located at the stop position (End genomic coordinate).</li>
</ul>
</li>
<li>For '-'-strand ORFs,<ul>
<li>the first nucleotide of the start codon is located at the stop position (End genomic coordinate).</li>
<li>the last nucleotide of the stop codon is located at the start position (Start genomic coordinate).</li>
</ul>
</li>
<li>The nucleic sequence is not provided.</li>
<li>NB: There is no information about the strand provided in the file.</li>
<li>The amino acid sequence includes the start codon and excludes the stop codon.</li>
<li>The ORF length is not provided. <h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html" title="Source: Samandi et al., 2017.">Samandi2017</a> </h1>
</li>
</ul>
<p class="startli">This class inherits from ParserStrategy and allows to parse data from Samandi et al., 2017. </p>
</li>
</ul>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="ab35b35829138cef8d8e8ea3a671bd2b0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab35b35829138cef8d8e8ea3a671bd2b0">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>data_path</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>Constructor of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1Samandi2017_1_1Samandi2017.html" title="Source: Samandi et al., 2017.">Samandi2017</a> </h2>

<p>Reimplemented from <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#af3a41c9649531a71f9ab74bf6c5f709a">fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy</a>.</p>
<div class="fragment"><div class="line"><a name="l00123"></a><span class="lineno">  123</span>&#160;    <span class="keyword">def </span>__init__( self, data_path ):</div><div class="line"><a name="l00124"></a><span class="lineno">  124</span>&#160;        </div><div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;        ParserStrategy.__init__( self, data_path )</div><div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;</div><div class="line"><a name="l00127"></a><span class="lineno">  127</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a174b65f01d550105ef67642886958703"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a174b65f01d550105ef67642886958703">&#9670;&nbsp;</a></span>execute()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017.execute </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>execute </h2>
<p>Execute the parsing strategy. See the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a8cdab4b401f878c517c4b737b9ecf778">execute()</a> method documentation of the ParserStrategy metaclass for more information about the main steps of the algorithm run in this method.</p>
<dl class="section return"><dt>Returns</dt><dd>objects_to_insert: List - List of objects to insert in the database.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">DenCellORFException</td><td>When this parser cannot be used for the species of the database. </td></tr>
    <tr><td class="paramname">DenCellORFException</td><td>When an exception has been raised checking the headers of the file. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;    <span class="keyword">def </span>execute( self ):</div><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;        </div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;        <span class="comment"># Check this parser may be used for the species in the database</span></div><div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;        sp = DataManager.get_instance().get_data( Constants.SPECIES_SHORT )</div><div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160;        <span class="keywordflow">if</span> ( sp <span class="keywordflow">not</span> <span class="keywordflow">in</span> self.AUTHORIZED_SPECIES ):</div><div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;            <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;The parser &#39;</span> + self.data_source + </div><div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;                                       <span class="stringliteral">&#39; has been build to parse data from &#39;</span> + </div><div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;                                       <span class="stringliteral">&#39;, &#39;</span>.join( self.AUTHORIZED_SPECIES ) + </div><div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;                                       <span class="stringliteral">&#39; whilst data store in this database is related to &#39;</span> + sp )</div><div class="line"><a name="l00149"></a><span class="lineno">  149</span>&#160;        </div><div class="line"><a name="l00150"></a><span class="lineno">  150</span>&#160;        <span class="comment"># Get the appropriate annotation version and description</span></div><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;            annotation_version = self.ANNOTATION_VERSION[ sp ]</div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;            annotation_description = self.ANNOTATION_DESCRIPTION[ sp ]</div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;            ensembl_release = self.ENSEMBL_RELEASE[ sp ]</div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;        </div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;        <span class="comment"># Get the content of the file</span></div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;        self.import_file_content( file_sep = <span class="stringliteral">&#39;\t&#39;</span> )</div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;        </div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;        <span class="comment"># Check the headers of the file</span></div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;            self.check_headers()</div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;        <span class="keywordflow">except</span> DuplicateHeaderException <span class="keyword">as</span> e:</div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;            Logger.get_instance().warning( e.get_message() +</div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;                                           <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_HEADER_DUPL + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;                </div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;        <span class="comment"># Assign new variables to access some objects of the DataManager data dictionary</span></div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;        all_dsorfs = DataManager.get_instance().get_data( Constants.DM_ALL_DSORFS_FOR_SOURCE )</div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;        all_dstranscripts = DataManager.get_instance().get_data( Constants.DM_ALL_DSTRANSCRIPTS_FOR_SOURCE )</div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;        all_dsota = DataManager.get_instance().get_data( Constants.DM_ALL_DSORFTRANSCRIPTASSO_FOR_SOURCE )</div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;        </div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;        autoincrement = DataManager.get_instance().get_data( Constants.DM_AUTOINCREMENT )</div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;        </div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;        <span class="comment"># Get the appropriate Ensembl database to recover missing gene IDs</span></div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;        ensembl_db = ParserStrategy.get_ensembl_db( ensembl_release )</div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;        </div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;        <span class="comment"># Get the number of rows expected to be treated and reset the ProgressionBar instance</span></div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;        ProgressionBar.get_instance().reset_instance( total = self.file_content.shape[ 0 ] )</div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;        </div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;        <span class="comment"># Initialize the list of objects to insert</span></div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;        objects_to_insert = []</div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;        </div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;        <span class="comment"># All ORFs associated with multiple chromosomes are discarded from insertion. Keep the count of these ORFs.</span></div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;        discarded_orfs = 0 </div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;        </div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;</div><div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;        </div><div class="line"><a name="l00188"></a><span class="lineno">  188</span>&#160;        datasource = DataSource( name = Samandi2017.DATA_SOURCE,</div><div class="line"><a name="l00189"></a><span class="lineno">  189</span>&#160;                                 doi = Samandi2017.DOI,</div><div class="line"><a name="l00190"></a><span class="lineno">  190</span>&#160;                                 description = Samandi2017.SOURCE_DESCRIPTION[ sp ],</div><div class="line"><a name="l00191"></a><span class="lineno">  191</span>&#160;                                 url = Samandi2017.URL, </div><div class="line"><a name="l00192"></a><span class="lineno">  192</span>&#160;                                 annotation_version = annotation_version,</div><div class="line"><a name="l00193"></a><span class="lineno">  193</span>&#160;                                 ensembl_release = ensembl_release,</div><div class="line"><a name="l00194"></a><span class="lineno">  194</span>&#160;                                 annotation_description = annotation_description )</div><div class="line"><a name="l00195"></a><span class="lineno">  195</span>&#160;        objects_to_insert.append( datasource )</div><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;        </div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;        </div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;        <span class="comment"># Parse the content of the file and save it in appropriate objects</span></div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;        start_time = datetime.datetime.now()</div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;        <span class="keywordflow">for</span> ( index, row ) <span class="keywordflow">in</span> self.file_content.iterrows():</div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;            </div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;            <span class="comment"># Update and display the progression bar on the console</span></div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;            ProgressionBar.get_instance().increase_and_display()</div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;            </div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;            <span class="comment"># Assign useful information to new variables</span></div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;            chr_name = row[ Samandi2017.HEADER_CHROMOSOME ]</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;            </div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;            <span class="comment"># Skip the ORFs reported as being on multiple chromosomes</span></div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;            <span class="keywordflow">if</span> ( len( chr_name.split( <span class="stringliteral">&#39;,&#39;</span> ) ) &gt; 1 ):</div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;                discarded_orfs += 1</div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;            </div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;            <span class="keywordflow">elif</span> ( len( chr_name.split( <span class="stringliteral">&#39;.&#39;</span> ) ) &gt; 1 ):</div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;                discarded_orfs += 1</div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;            </div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;                <span class="comment"># Assign other useful information to new variables</span></div><div class="line"><a name="l00217"></a><span class="lineno">  217</span>&#160;                orf_id = row[ Samandi2017.HEADER_ORF_ID ]</div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;                symbol = row[ Samandi2017.HEADER_GENE_SYMBOL ]</div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;                <span class="comment"># If the gene is associated with several gene symbols, concatenate them</span></div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;                symbol = <span class="stringliteral">&#39;_&#39;</span>.join( sorted( str( symbol ).split(<span class="stringliteral">&#39; &#39;</span>) ) )</div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;                <span class="comment"># If the symbol get by concatenation is too long, use the unknown gene prefix</span></div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;                <span class="keywordflow">if</span> ( len( symbol ) &gt; Constants.MAX_LEN_STRING ):</div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;                    symbol = Constants.PREFIX_UNKNOWN_GENE + chr_name</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;                </div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;                transcript_id = <span class="keywordtype">None</span></div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;    </div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;                orf_start = row[ Samandi2017.HEADER_START_POS ]</div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;                orf_stop = row[ Samandi2017.HEADER_STOP_POS ]</div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;                orf_strand = <span class="keywordtype">None</span></div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;                </div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;                <span class="comment"># Reset the chr_name, symbol and transcript_id if necessary</span></div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;                ( chr_name, symbol, transcript_id ) = self.initialize_ids( ensembl_db = ensembl_db,</div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                                                                           orf_id = orf_id,</div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                                                                           chr_name = chr_name, </div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;                                                                           symbol = symbol, </div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;                                                                           transcript_id = <span class="keywordtype">None</span>, </div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;                                                                           orf_start = orf_start, </div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;                                                                           orf_stop = orf_stop, </div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;                                                                           orf_strand = orf_strand, </div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;                                                                           index = index )</div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;                </div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;                </div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;                dsorf = DSORF( id = autoincrement,</div><div class="line"><a name="l00245"></a><span class="lineno">  245</span>&#160;                               data_source = datasource.name,</div><div class="line"><a name="l00246"></a><span class="lineno">  246</span>&#160;                               chromosome = chr_name,</div><div class="line"><a name="l00247"></a><span class="lineno">  247</span>&#160;                               raw_strand = orf_strand,</div><div class="line"><a name="l00248"></a><span class="lineno">  248</span>&#160;                               raw_start_pos = orf_start,</div><div class="line"><a name="l00249"></a><span class="lineno">  249</span>&#160;                               raw_stop_pos = orf_stop,</div><div class="line"><a name="l00250"></a><span class="lineno">  250</span>&#160;                               spliced = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00251"></a><span class="lineno">  251</span>&#160;                               raw_splice_starts = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00252"></a><span class="lineno">  252</span>&#160;                               raw_splice_ends = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00253"></a><span class="lineno">  253</span>&#160;                               spliced_parts_count = <span class="keywordtype">None</span> )</div><div class="line"><a name="l00254"></a><span class="lineno">  254</span>&#160;                </div><div class="line"><a name="l00255"></a><span class="lineno">  255</span>&#160;                <span class="comment"># If this ORF already exists, get the existing one</span></div><div class="line"><a name="l00256"></a><span class="lineno">  256</span>&#160;                existing_dsorf = all_dsorfs.get( dsorf )</div><div class="line"><a name="l00257"></a><span class="lineno">  257</span>&#160;                <span class="keywordflow">if</span> existing_dsorf:</div><div class="line"><a name="l00258"></a><span class="lineno">  258</span>&#160;                    dsorf = existing_dsorf</div><div class="line"><a name="l00259"></a><span class="lineno">  259</span>&#160;                    </div><div class="line"><a name="l00260"></a><span class="lineno">  260</span>&#160;                <span class="comment"># Otherwise, add the newly created ORF to the list of objects to insert and keep track of it</span></div><div class="line"><a name="l00261"></a><span class="lineno">  261</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00262"></a><span class="lineno">  262</span>&#160;                    all_dsorfs[ dsorf ] = dsorf</div><div class="line"><a name="l00263"></a><span class="lineno">  263</span>&#160;                    objects_to_insert.append( dsorf )</div><div class="line"><a name="l00264"></a><span class="lineno">  264</span>&#160;</div><div class="line"><a name="l00265"></a><span class="lineno">  265</span>&#160;                    <span class="comment"># Increase the autoincrement value for ID of 1</span></div><div class="line"><a name="l00266"></a><span class="lineno">  266</span>&#160;                    autoincrement += 1</div><div class="line"><a name="l00267"></a><span class="lineno">  267</span>&#160;                    </div><div class="line"><a name="l00268"></a><span class="lineno">  268</span>&#160;                </div><div class="line"><a name="l00269"></a><span class="lineno">  269</span>&#160;                </div><div class="line"><a name="l00272"></a><span class="lineno">  272</span>&#160;                ( gene, new_obj_to_insert ) = self.get_gene_from_symbol( orf_id = orf_id,</div><div class="line"><a name="l00273"></a><span class="lineno">  273</span>&#160;                                                                         symbol = symbol,</div><div class="line"><a name="l00274"></a><span class="lineno">  274</span>&#160;                                                                         chr_name = chr_name )</div><div class="line"><a name="l00275"></a><span class="lineno">  275</span>&#160;                objects_to_insert += new_obj_to_insert</div><div class="line"><a name="l00276"></a><span class="lineno">  276</span>&#160;                </div><div class="line"><a name="l00277"></a><span class="lineno">  277</span>&#160;                </div><div class="line"><a name="l00278"></a><span class="lineno">  278</span>&#160;                </div><div class="line"><a name="l00280"></a><span class="lineno">  280</span>&#160;                dstranscript = DSTranscript( id = autoincrement,</div><div class="line"><a name="l00281"></a><span class="lineno">  281</span>&#160;                                             transcript_id = transcript_id,</div><div class="line"><a name="l00282"></a><span class="lineno">  282</span>&#160;                                             data_source = datasource.name,</div><div class="line"><a name="l00283"></a><span class="lineno">  283</span>&#160;                                             gene_id = gene.gene_id,</div><div class="line"><a name="l00284"></a><span class="lineno">  284</span>&#160;                                             strand = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00285"></a><span class="lineno">  285</span>&#160;                                             raw_start_pos = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00286"></a><span class="lineno">  286</span>&#160;                                             raw_end_pos = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00287"></a><span class="lineno">  287</span>&#160;                                             raw_cds_start_pos = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00288"></a><span class="lineno">  288</span>&#160;                                             raw_cds_stop_pos = <span class="keywordtype">None</span> )</div><div class="line"><a name="l00289"></a><span class="lineno">  289</span>&#160;                </div><div class="line"><a name="l00290"></a><span class="lineno">  290</span>&#160;                <span class="comment"># Add the newly created transcript to the list of objects to insert and keep track of it</span></div><div class="line"><a name="l00291"></a><span class="lineno">  291</span>&#160;                all_dstranscripts[ dstranscript ] = dstranscript</div><div class="line"><a name="l00292"></a><span class="lineno">  292</span>&#160;                objects_to_insert.append( dstranscript )</div><div class="line"><a name="l00293"></a><span class="lineno">  293</span>&#160;                </div><div class="line"><a name="l00294"></a><span class="lineno">  294</span>&#160;                <span class="comment"># Increase the autoincrement value for ID of 1</span></div><div class="line"><a name="l00295"></a><span class="lineno">  295</span>&#160;                autoincrement += 1</div><div class="line"><a name="l00296"></a><span class="lineno">  296</span>&#160;                </div><div class="line"><a name="l00297"></a><span class="lineno">  297</span>&#160;                </div><div class="line"><a name="l00298"></a><span class="lineno">  298</span>&#160;                </div><div class="line"><a name="l00300"></a><span class="lineno">  300</span>&#160;                orf_seq_aa = row[ Samandi2017.HEADER_ORF_AA_SEQUENCE ]</div><div class="line"><a name="l00301"></a><span class="lineno">  301</span>&#160;                <span class="keywordflow">if</span> ParserStrategy.is_empty( orf_seq_aa ):</div><div class="line"><a name="l00302"></a><span class="lineno">  302</span>&#160;                    orf_len_aa = <span class="keywordtype">None</span></div><div class="line"><a name="l00303"></a><span class="lineno">  303</span>&#160;                    orf_len_nt = <span class="keywordtype">None</span></div><div class="line"><a name="l00304"></a><span class="lineno">  304</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00305"></a><span class="lineno">  305</span>&#160;                    orf_len_aa = len( orf_seq_aa )</div><div class="line"><a name="l00306"></a><span class="lineno">  306</span>&#160;                    orf_len_nt = ( orf_len_aa + 1 ) * 3</div><div class="line"><a name="l00307"></a><span class="lineno">  307</span>&#160;                </div><div class="line"><a name="l00308"></a><span class="lineno">  308</span>&#160;                </div><div class="line"><a name="l00309"></a><span class="lineno">  309</span>&#160;                <span class="keywordflow">if</span> ( row[ Samandi2017.HEADER_FRAME ] != <span class="stringliteral">&#39;1&#39;</span> ):</div><div class="line"><a name="l00310"></a><span class="lineno">  310</span>&#160;                    orf_annot = <span class="stringliteral">&#39;alternative frame&#39;</span> </div><div class="line"><a name="l00311"></a><span class="lineno">  311</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00312"></a><span class="lineno">  312</span>&#160;                    orf_annot = <span class="keywordtype">None</span></div><div class="line"><a name="l00313"></a><span class="lineno">  313</span>&#160;                </div><div class="line"><a name="l00314"></a><span class="lineno">  314</span>&#160;                dsorftranscriptasso = DSORFTranscriptAsso( id = autoincrement,</div><div class="line"><a name="l00315"></a><span class="lineno">  315</span>&#160;                                                           data_source = datasource.name,</div><div class="line"><a name="l00316"></a><span class="lineno">  316</span>&#160;                                                           transcript_id = dstranscript.id,</div><div class="line"><a name="l00317"></a><span class="lineno">  317</span>&#160;                                                           uniq_orf_id = dsorf.id,</div><div class="line"><a name="l00318"></a><span class="lineno">  318</span>&#160;                                                           predicted = Samandi2017.PREDICTED,</div><div class="line"><a name="l00319"></a><span class="lineno">  319</span>&#160;                                                           ribo_seq = Samandi2017.RIBO_SEQ,</div><div class="line"><a name="l00320"></a><span class="lineno">  320</span>&#160;                                                           cell_context = Samandi2017.CELL_CONTEXT,</div><div class="line"><a name="l00321"></a><span class="lineno">  321</span>&#160;                                                           orf_id = orf_id,</div><div class="line"><a name="l00322"></a><span class="lineno">  322</span>&#160;                                                           raw_sequence = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00323"></a><span class="lineno">  323</span>&#160;                                                           raw_sequence_aa = orf_seq_aa,</div><div class="line"><a name="l00324"></a><span class="lineno">  324</span>&#160;                                                           start_codon_seq = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00325"></a><span class="lineno">  325</span>&#160;                                                           kozak_context = row[ Samandi2017.HEADER_KOZAK_CONTEXT ],</div><div class="line"><a name="l00326"></a><span class="lineno">  326</span>&#160;                                                           orf_length_nt = orf_len_nt,</div><div class="line"><a name="l00327"></a><span class="lineno">  327</span>&#160;                                                           orf_length = orf_len_aa,</div><div class="line"><a name="l00328"></a><span class="lineno">  328</span>&#160;                                                           provided_category = orf_annot,</div><div class="line"><a name="l00329"></a><span class="lineno">  329</span>&#160;                                                           ms_info = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00330"></a><span class="lineno">  330</span>&#160;                                                           orf_score = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00331"></a><span class="lineno">  331</span>&#160;                                                           phylocsf = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00332"></a><span class="lineno">  332</span>&#160;                                                           phastcons = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00333"></a><span class="lineno">  333</span>&#160;                                                           floss = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00334"></a><span class="lineno">  334</span>&#160;                                                           floss_class = <span class="keywordtype">None</span> )</div><div class="line"><a name="l00335"></a><span class="lineno">  335</span>&#160;            </div><div class="line"><a name="l00336"></a><span class="lineno">  336</span>&#160;                <span class="comment"># Otherwise, add the newly created DSORFTranscriptAsso to the list of objects to insert and keep track of it</span></div><div class="line"><a name="l00337"></a><span class="lineno">  337</span>&#160;                all_dsota[ dsorftranscriptasso ] = dsorftranscriptasso</div><div class="line"><a name="l00338"></a><span class="lineno">  338</span>&#160;    </div><div class="line"><a name="l00339"></a><span class="lineno">  339</span>&#160;                <span class="comment"># Add it to the list of objects to insert in the database</span></div><div class="line"><a name="l00340"></a><span class="lineno">  340</span>&#160;                objects_to_insert.append( dsorftranscriptasso )</div><div class="line"><a name="l00341"></a><span class="lineno">  341</span>&#160;                    </div><div class="line"><a name="l00342"></a><span class="lineno">  342</span>&#160;                <span class="comment"># Increase the autoincrement value for ID of 1</span></div><div class="line"><a name="l00343"></a><span class="lineno">  343</span>&#160;                autoincrement += 1</div><div class="line"><a name="l00344"></a><span class="lineno">  344</span>&#160;                </div><div class="line"><a name="l00345"></a><span class="lineno">  345</span>&#160;        <span class="comment"># Log the duration of the process</span></div><div class="line"><a name="l00346"></a><span class="lineno">  346</span>&#160;        end_time = datetime.datetime.now()</div><div class="line"><a name="l00347"></a><span class="lineno">  347</span>&#160;        Logger.get_instance().debug( <span class="stringliteral">&#39;Samandi2017.execute(): Duration of parsing: &#39;</span> + </div><div class="line"><a name="l00348"></a><span class="lineno">  348</span>&#160;                                     str( end_time - start_time ) + <span class="stringliteral">&#39; (&#39;</span> +</div><div class="line"><a name="l00349"></a><span class="lineno">  349</span>&#160;                                     str( ProgressionBar.get_instance().total_row_count ) + </div><div class="line"><a name="l00350"></a><span class="lineno">  350</span>&#160;                                     <span class="stringliteral">&#39; rows processed).&#39;</span> )</div><div class="line"><a name="l00351"></a><span class="lineno">  351</span>&#160;                    </div><div class="line"><a name="l00352"></a><span class="lineno">  352</span>&#160;        <span class="comment"># Store the current value of the autoincrement in the DataManager</span></div><div class="line"><a name="l00353"></a><span class="lineno">  353</span>&#160;        DataManager.get_instance().store_data( Constants.DM_AUTOINCREMENT, autoincrement )</div><div class="line"><a name="l00354"></a><span class="lineno">  354</span>&#160;                    </div><div class="line"><a name="l00355"></a><span class="lineno">  355</span>&#160;</div><div class="line"><a name="l00356"></a><span class="lineno">  356</span>&#160;        <span class="keywordflow">if</span> discarded_orfs &gt; 0:              </div><div class="line"><a name="l00357"></a><span class="lineno">  357</span>&#160;            Logger.get_instance().warning( str( discarded_orfs ) + <span class="stringliteral">&#39; ORFs from &#39;</span> + self.data_source + </div><div class="line"><a name="l00358"></a><span class="lineno">  358</span>&#160;                                           <span class="stringliteral">&#39; have been discarded as they were reported as being&#39;</span> +</div><div class="line"><a name="l00359"></a><span class="lineno">  359</span>&#160;                                           <span class="stringliteral">&#39; located on several chromosomes.&#39;</span> +</div><div class="line"><a name="l00360"></a><span class="lineno">  360</span>&#160;                                           <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_DISCARD_ORF + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00361"></a><span class="lineno">  361</span>&#160;                    </div><div class="line"><a name="l00362"></a><span class="lineno">  362</span>&#160;        <span class="keywordflow">return</span> objects_to_insert</div><div class="line"><a name="l00363"></a><span class="lineno">  363</span>&#160;            </div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="abefd2f3acca36e54700b22027c0c3d5d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abefd2f3acca36e54700b22027c0c3d5d">&#9670;&nbsp;</a></span>AUTHORIZED_SPECIES</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">list fr.tagc.uorf.core.execution.insertion.parserstrategy.common.Samandi2017.Samandi2017.AUTHORIZED_SPECIES</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">=  [ Constants.SPECIES_CATALOG[ Constants.HSAPIENS ],</div><div class="line">                          Constants.SPECIES_CATALOG[ Constants.MMUSCULUS ] ]</div></div><!-- fragment --><h2>Class variables </h2>
<p>General constants </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/execution/insertion/parserstrategy/common/Samandi2017.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
