<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>DenCellORF: fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">DenCellORF
   &#160;<span id="projectnumber">1.1.4</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>execution</b></li><li class="navelem"><b>ProcessingStrategy</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html">ProcessingStrategy</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Inherits object.</p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a09d62efc5a4a58d3e4f9e494870a60be"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a09d62efc5a4a58d3e4f9e494870a60be">__init__</a> (self)</td></tr>
<tr class="separator:a09d62efc5a4a58d3e4f9e494870a60be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa2a8167a66e8bf24e600fae71281674a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#aa2a8167a66e8bf24e600fae71281674a">execute</a> (self)</td></tr>
<tr class="separator:aa2a8167a66e8bf24e600fae71281674a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a159a4828f8679cc87e4f3b0308ab2249"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a159a4828f8679cc87e4f3b0308ab2249">convert_genomic_coordinates</a> (self)</td></tr>
<tr class="separator:a159a4828f8679cc87e4f3b0308ab2249"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad647bf1734add91a06570a87960a6977"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#ad647bf1734add91a06570a87960a6977">check_dsorftrasso_sequences_agreement</a> (self)</td></tr>
<tr class="separator:ad647bf1734add91a06570a87960a6977"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad323da493e31a3ce75f4767fc4d743c7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#ad323da493e31a3ce75f4767fc4d743c7">download_missing_sequences</a> (self)</td></tr>
<tr class="separator:ad323da493e31a3ce75f4767fc4d743c7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5e2942fee457da5fbde63f852639813"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#ab5e2942fee457da5fbde63f852639813">download_seq_from_ensembl</a> (self, chr, strand, start, stop, genome_version)</td></tr>
<tr class="separator:ab5e2942fee457da5fbde63f852639813"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a52b8101e3148256640cfd8d4820e0083"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a52b8101e3148256640cfd8d4820e0083">download_dsorf_seq</a> (self, chr, strand, start, stop, splicing_status, splice_starts, splice_ends, genome_version)</td></tr>
<tr class="separator:a52b8101e3148256640cfd8d4820e0083"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6a01ae4d1054a569067b4cddfdeb4f5c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a6a01ae4d1054a569067b4cddfdeb4f5c">batch_insert_to_db</a> (self, objects_to_insert, source)</td></tr>
<tr class="separator:a6a01ae4d1054a569067b4cddfdeb4f5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:ac064a9a263aed718d6070ad138a228f7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#ac064a9a263aed718d6070ad138a228f7">convert_coord</a> (chr, strand, position, lo)</td></tr>
<tr class="separator:ac064a9a263aed718d6070ad138a228f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a77d93c2c39c9c9797f1cfc620f11ddce"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a77d93c2c39c9c9797f1cfc620f11ddce">convert_dsorf_coordinates</a> (dsorf, lo)</td></tr>
<tr class="separator:a77d93c2c39c9c9797f1cfc620f11ddce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acac09e2f267b2008382df8b031ceb252"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#acac09e2f267b2008382df8b031ceb252">duplicate_dsorf_coordinates</a> (dsorf)</td></tr>
<tr class="separator:acac09e2f267b2008382df8b031ceb252"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a27a742369bcaed60067ea219599ba1e4"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a27a742369bcaed60067ea219599ba1e4">convert_dstranscript_coordinates</a> (dstranscript, lo)</td></tr>
<tr class="separator:a27a742369bcaed60067ea219599ba1e4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a250e5000352003d9f3a3862548fa8862"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a250e5000352003d9f3a3862548fa8862">duplicate_dstranscript_coordinates</a> (dstranscript)</td></tr>
<tr class="separator:a250e5000352003d9f3a3862548fa8862"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a619dba489edb08f8f896abd8bd68a712"><td class="memItemLeft" align="right" valign="top">dictionary&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html#a619dba489edb08f8f896abd8bd68a712">CHAIN_FILES_URL</a></td></tr>
<tr class="separator:a619dba489edb08f8f896abd8bd68a712"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html">ProcessingStrategy</a> </h1>
<p>This class is a strategy aiming to perform the processing of data. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a09d62efc5a4a58d3e4f9e494870a60be"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a09d62efc5a4a58d3e4f9e494870a60be">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>Constructor of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1ProcessingStrategy_1_1ProcessingStrategy.html">ProcessingStrategy</a> </h2>
<p>Instance variables:</p><ul>
<li>configfile: String - The path to the config file.</li>
<li>species: String - The name of the species in the database. </li>
</ul>
<div class="fragment"><div class="line"><a name="l00060"></a><span class="lineno">   60</span>&#160;    <span class="keyword">def </span>__init__( self ):</div><div class="line"><a name="l00061"></a><span class="lineno">   61</span>&#160;</div><div class="line"><a name="l00062"></a><span class="lineno">   62</span>&#160;        configfile = OptionManager.get_instance().get_option( OptionConstants.OPTION_CONFIG_FILE_PATH, not_none=<span class="keyword">True</span> )</div><div class="line"><a name="l00063"></a><span class="lineno">   63</span>&#160;</div><div class="line"><a name="l00064"></a><span class="lineno">   64</span>&#160;        <span class="keywordflow">if</span> configfile != <span class="keywordtype">None</span>:</div><div class="line"><a name="l00065"></a><span class="lineno">   65</span>&#160;            self.configfile = configfile</div><div class="line"><a name="l00066"></a><span class="lineno">   66</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00067"></a><span class="lineno">   67</span>&#160;            <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;ProcessingStrategy: A Config file has to be provided.&#39;</span> )</div><div class="line"><a name="l00068"></a><span class="lineno">   68</span>&#160;        </div><div class="line"><a name="l00069"></a><span class="lineno">   69</span>&#160;        self.species = <span class="keywordtype">None</span></div><div class="line"><a name="l00070"></a><span class="lineno">   70</span>&#160;    </div><div class="line"><a name="l00071"></a><span class="lineno">   71</span>&#160;        </div><div class="line"><a name="l00072"></a><span class="lineno">   72</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a6a01ae4d1054a569067b4cddfdeb4f5c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6a01ae4d1054a569067b4cddfdeb4f5c">&#9670;&nbsp;</a></span>batch_insert_to_db()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.batch_insert_to_db </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>objects_to_insert</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>source</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>batch_insert_to_db </h2>
<p>This method allows to insert a list of objects in the database. Large sets of objects are split into several batches which are inserted one at a time. Raise an exception if the insertion cannot be performed properly.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">objects_to_insert</td><td>List - The list of objects to insert in the database. </td></tr>
    <tr><td class="paramname">source</td><td>String - The name of the source from which these objects have been created. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00888"></a><span class="lineno">  888</span>&#160;    <span class="keyword">def </span>batch_insert_to_db( self, objects_to_insert, source ):</div><div class="line"><a name="l00889"></a><span class="lineno">  889</span>&#160;                </div><div class="line"><a name="l00890"></a><span class="lineno">  890</span>&#160;        <span class="comment"># Save into a temporary file the data that should be inserted.</span></div><div class="line"><a name="l00891"></a><span class="lineno">  891</span>&#160;        <span class="comment"># This allows to recover the data later if the insertion it not working, saving the processing time.</span></div><div class="line"><a name="l00892"></a><span class="lineno">  892</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00893"></a><span class="lineno">  893</span>&#160;            FileHandlerUtil.save_obj_to_file( output_folder = Constants.PROCESSING_INFORMATION_DATA_FOLDER,</div><div class="line"><a name="l00894"></a><span class="lineno">  894</span>&#160;                                                filename = source,</div><div class="line"><a name="l00895"></a><span class="lineno">  895</span>&#160;                                                objects_to_save = objects_to_insert )</div><div class="line"><a name="l00896"></a><span class="lineno">  896</span>&#160;        <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00897"></a><span class="lineno">  897</span>&#160;            Logger.get_instance().error( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): An exception was raised trying to&#39;</span> +</div><div class="line"><a name="l00898"></a><span class="lineno">  898</span>&#160;                                         <span class="stringliteral">&#39; save data &#39;</span> + source + <span class="stringliteral">&#39; Exception: \n&#39;</span> + str( e ) )</div><div class="line"><a name="l00899"></a><span class="lineno">  899</span>&#160;            </div><div class="line"><a name="l00900"></a><span class="lineno">  900</span>&#160;        <span class="comment"># Insert objects into the database</span></div><div class="line"><a name="l00901"></a><span class="lineno">  901</span>&#160;        <span class="keywordflow">if</span> ( objects_to_insert != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( len( objects_to_insert ) &gt; 0 ):</div><div class="line"><a name="l00902"></a><span class="lineno">  902</span>&#160;            <span class="comment"># Log in debug mode the size of objects expected to be inserted</span></div><div class="line"><a name="l00903"></a><span class="lineno">  903</span>&#160;            Logger.get_instance().debug( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): &#39;</span> + str( len( objects_to_insert ) ) +</div><div class="line"><a name="l00904"></a><span class="lineno">  904</span>&#160;                                         <span class="stringliteral">&#39; objects (from &#39;</span> + source + <span class="stringliteral">&#39;) are expected to be added to the database.&#39;</span> )</div><div class="line"><a name="l00905"></a><span class="lineno">  905</span>&#160;            </div><div class="line"><a name="l00906"></a><span class="lineno">  906</span>&#160;            <span class="keywordflow">if</span> len( objects_to_insert ) &lt; Constants.MAX_COUNT_TO_INSERT:</div><div class="line"><a name="l00907"></a><span class="lineno">  907</span>&#160;                </div><div class="line"><a name="l00908"></a><span class="lineno">  908</span>&#160;                <span class="comment"># Small sets of new objects are directly added to the database</span></div><div class="line"><a name="l00909"></a><span class="lineno">  909</span>&#160;                <span class="comment"># Insert objects into the database</span></div><div class="line"><a name="l00910"></a><span class="lineno">  910</span>&#160;                <span class="keywordflow">try</span>:</div><div class="line"><a name="l00911"></a><span class="lineno">  911</span>&#160;                    SQLManager.get_instance().get_session().add_all( objects_to_insert )</div><div class="line"><a name="l00912"></a><span class="lineno">  912</span>&#160;                <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00913"></a><span class="lineno">  913</span>&#160;                    <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): Error occurred trying to add following objects (from &#39;</span> +</div><div class="line"><a name="l00914"></a><span class="lineno">  914</span>&#160;                                               source  + <span class="stringliteral">&#39;) to the session: \n&#39;</span> + str( objects_to_insert ), e )</div><div class="line"><a name="l00915"></a><span class="lineno">  915</span>&#160;                    </div><div class="line"><a name="l00916"></a><span class="lineno">  916</span>&#160;                <span class="comment"># Commit changes</span></div><div class="line"><a name="l00917"></a><span class="lineno">  917</span>&#160;                <span class="keywordflow">try</span>:</div><div class="line"><a name="l00918"></a><span class="lineno">  918</span>&#160;                    SQLManager.get_instance().get_session().commit()</div><div class="line"><a name="l00919"></a><span class="lineno">  919</span>&#160;                <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00920"></a><span class="lineno">  920</span>&#160;                    <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): Error occurred trying to commit changes after addition of the following objects (from &#39;</span> +</div><div class="line"><a name="l00921"></a><span class="lineno">  921</span>&#160;                                               source  + <span class="stringliteral">&#39;): \n&#39;</span> + str( objects_to_insert ), e )</div><div class="line"><a name="l00922"></a><span class="lineno">  922</span>&#160;                    </div><div class="line"><a name="l00923"></a><span class="lineno">  923</span>&#160;                <span class="comment"># Log in debug mode the number of objects successfully inserted</span></div><div class="line"><a name="l00924"></a><span class="lineno">  924</span>&#160;                Logger.get_instance().debug( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): &#39;</span> + str( len( objects_to_insert ) ) + </div><div class="line"><a name="l00925"></a><span class="lineno">  925</span>&#160;                                             <span class="stringliteral">&#39; objects (from &#39;</span> + source + <span class="stringliteral">&#39;) have been successfully added to the database.&#39;</span> )</div><div class="line"><a name="l00926"></a><span class="lineno">  926</span>&#160;                </div><div class="line"><a name="l00927"></a><span class="lineno">  927</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00928"></a><span class="lineno">  928</span>&#160;                <span class="comment"># For large sets of objects, split the list of objects to insert into several batches </span></div><div class="line"><a name="l00929"></a><span class="lineno">  929</span>&#160;                <span class="comment"># and insert these batches one at the time</span></div><div class="line"><a name="l00930"></a><span class="lineno">  930</span>&#160;                min_bound = 0</div><div class="line"><a name="l00931"></a><span class="lineno">  931</span>&#160;                <span class="keywordflow">while</span> min_bound &lt; len( objects_to_insert ):</div><div class="line"><a name="l00932"></a><span class="lineno">  932</span>&#160;                    </div><div class="line"><a name="l00933"></a><span class="lineno">  933</span>&#160;                    <span class="keywordflow">if</span> ( min_bound + Constants.MAX_COUNT_TO_INSERT ) &lt;= len( objects_to_insert ):</div><div class="line"><a name="l00934"></a><span class="lineno">  934</span>&#160;                        max_bound = min_bound + Constants.MAX_COUNT_TO_INSERT</div><div class="line"><a name="l00935"></a><span class="lineno">  935</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00936"></a><span class="lineno">  936</span>&#160;                        max_bound = len( objects_to_insert )</div><div class="line"><a name="l00937"></a><span class="lineno">  937</span>&#160;                        </div><div class="line"><a name="l00938"></a><span class="lineno">  938</span>&#160;                    <span class="comment"># Insert objects into the database</span></div><div class="line"><a name="l00939"></a><span class="lineno">  939</span>&#160;                    <span class="keywordflow">try</span>:</div><div class="line"><a name="l00940"></a><span class="lineno">  940</span>&#160;                        SQLManager.get_instance().get_session().add_all( objects_to_insert[ min_bound : max_bound ] )</div><div class="line"><a name="l00941"></a><span class="lineno">  941</span>&#160;                    <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00942"></a><span class="lineno">  942</span>&#160;                        <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): Error occurred trying to add following objects (from &#39;</span> +</div><div class="line"><a name="l00943"></a><span class="lineno">  943</span>&#160;                                                   source  + <span class="stringliteral">&#39;) to the session: \n&#39;</span> + str( objects_to_insert[ min_bound : max_bound ] ), e )</div><div class="line"><a name="l00944"></a><span class="lineno">  944</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00945"></a><span class="lineno">  945</span>&#160;                        <span class="comment"># Commit changes</span></div><div class="line"><a name="l00946"></a><span class="lineno">  946</span>&#160;                        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00947"></a><span class="lineno">  947</span>&#160;                            SQLManager.get_instance().get_session().commit()</div><div class="line"><a name="l00948"></a><span class="lineno">  948</span>&#160;                        <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00949"></a><span class="lineno">  949</span>&#160;                            <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): Error occurred trying to commit changes after addition of the following objects (from &#39;</span> +</div><div class="line"><a name="l00950"></a><span class="lineno">  950</span>&#160;                                                       source  + <span class="stringliteral">&#39;): \n&#39;</span> + str( objects_to_insert[ min_bound : max_bound ] ) + <span class="stringliteral">&#39;.&#39;</span>, e )</div><div class="line"><a name="l00951"></a><span class="lineno">  951</span>&#160;                        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00952"></a><span class="lineno">  952</span>&#160;                            <span class="comment"># Log in debug mode the number of objects successfully inserted</span></div><div class="line"><a name="l00953"></a><span class="lineno">  953</span>&#160;                            Logger.get_instance().debug( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): &#39;</span> + str( len( objects_to_insert[ min_bound : max_bound ] ) ) + </div><div class="line"><a name="l00954"></a><span class="lineno">  954</span>&#160;                                                         <span class="stringliteral">&#39; objects (from &#39;</span> + source + <span class="stringliteral">&#39;) have been successfully added to the database.&#39;</span> )</div><div class="line"><a name="l00955"></a><span class="lineno">  955</span>&#160;                    </div><div class="line"><a name="l00956"></a><span class="lineno">  956</span>&#160;                    <span class="comment"># Redefine the minimum bound of the interval</span></div><div class="line"><a name="l00957"></a><span class="lineno">  957</span>&#160;                    min_bound = max_bound</div><div class="line"><a name="l00958"></a><span class="lineno">  958</span>&#160;</div><div class="line"><a name="l00959"></a><span class="lineno">  959</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00960"></a><span class="lineno">  960</span>&#160;            Logger.get_instance().debug( <span class="stringliteral">&#39;ProcessingStrategy.batch_insert_to_db(): There is no data to insert for &#39;</span> + source + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00961"></a><span class="lineno">  961</span>&#160;                </div></div><!-- fragment -->
</div>
</div>
<a id="ad647bf1734add91a06570a87960a6977"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad647bf1734add91a06570a87960a6977">&#9670;&nbsp;</a></span>check_dsorftrasso_sequences_agreement()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.check_dsorftrasso_sequences_agreement </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>check_dsorftrasso_sequences_agreement </h2>
<p>This method allows to check that the DNA and amino acids sequences registered in the DSORFTranscriptAsso table (i.e. the sequences provided by the sources) agree together. When this is not the case, a warning is log and a new entry is created in the UTInfoType table with a record of the DSORFTranscriptAsso ID. </p>
<div class="fragment"><div class="line"><a name="l00528"></a><span class="lineno">  528</span>&#160;    <span class="keyword">def </span>check_dsorftrasso_sequences_agreement( self ):</div><div class="line"><a name="l00529"></a><span class="lineno">  529</span>&#160;            </div><div class="line"><a name="l00530"></a><span class="lineno">  530</span>&#160;        Logger.get_instance().info( <span class="stringliteral">&#39;Checking sequence agreement in entries of the DSORFTranscriptAsso table.&#39;</span> )</div><div class="line"><a name="l00531"></a><span class="lineno">  531</span>&#160;        </div><div class="line"><a name="l00532"></a><span class="lineno">  532</span>&#160;        <span class="comment"># Variable to store eventual newly created objects</span></div><div class="line"><a name="l00533"></a><span class="lineno">  533</span>&#160;        objects_to_insert = []</div><div class="line"><a name="l00534"></a><span class="lineno">  534</span>&#160;        </div><div class="line"><a name="l00535"></a><span class="lineno">  535</span>&#160;        <span class="comment"># Get all the entries of the DSORFTranscriptAsso table</span></div><div class="line"><a name="l00536"></a><span class="lineno">  536</span>&#160;        all_dsorftranscriptasso = SQLManager.get_instance().get_session().query( DSORFTranscriptAsso ).all()</div><div class="line"><a name="l00537"></a><span class="lineno">  537</span>&#160;        </div><div class="line"><a name="l00538"></a><span class="lineno">  538</span>&#160;        <span class="comment"># Get the number of entries expected to be treated and set up a counter to follow the progression</span></div><div class="line"><a name="l00539"></a><span class="lineno">  539</span>&#160;        total_entries_count = len( all_dsorftranscriptasso )</div><div class="line"><a name="l00540"></a><span class="lineno">  540</span>&#160;        progression_count = 1</div><div class="line"><a name="l00541"></a><span class="lineno">  541</span>&#160;        </div><div class="line"><a name="l00542"></a><span class="lineno">  542</span>&#160;        <span class="comment"># For all the entires, check the DNA and amino acid sequences agree with each other</span></div><div class="line"><a name="l00543"></a><span class="lineno">  543</span>&#160;        <span class="keywordflow">for</span> dsorftranscriptasso <span class="keywordflow">in</span> all_dsorftranscriptasso:</div><div class="line"><a name="l00544"></a><span class="lineno">  544</span>&#160;            </div><div class="line"><a name="l00545"></a><span class="lineno">  545</span>&#160;            <span class="comment"># Display and update a progression bar on the console</span></div><div class="line"><a name="l00546"></a><span class="lineno">  546</span>&#160;            GraphicUtil.progression_bar( total_entries_count, progression_count )</div><div class="line"><a name="l00547"></a><span class="lineno">  547</span>&#160;            progression_count += 1</div><div class="line"><a name="l00548"></a><span class="lineno">  548</span>&#160;            </div><div class="line"><a name="l00549"></a><span class="lineno">  549</span>&#160;            <span class="keywordflow">if</span> ( dsorftranscriptasso.raw_sequence != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( dsorftranscriptasso.raw_sequence_aa != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00550"></a><span class="lineno">  550</span>&#160;                </div><div class="line"><a name="l00551"></a><span class="lineno">  551</span>&#160;                <span class="comment"># Translate the DNA sequence</span></div><div class="line"><a name="l00552"></a><span class="lineno">  552</span>&#160;                aa_sequence = GeneticsUtil.translate_dna( dsorftranscriptasso.raw_sequence )</div><div class="line"><a name="l00553"></a><span class="lineno">  553</span>&#160;                </div><div class="line"><a name="l00554"></a><span class="lineno">  554</span>&#160;                <span class="comment"># If the amino acid sequence is different from the one provided by the source</span></div><div class="line"><a name="l00555"></a><span class="lineno">  555</span>&#160;                <span class="comment"># log a warning and register this difference in the UTInfoType table</span></div><div class="line"><a name="l00556"></a><span class="lineno">  556</span>&#160;                <span class="keywordflow">if</span> aa_sequence != dsorftranscriptasso.raw_sequence_aa:</div><div class="line"><a name="l00557"></a><span class="lineno">  557</span>&#160;                    Logger.get_instance().warning( <span class="stringliteral">&#39;The DNA and amino acid sequences of the DSORFTranscriptAsso with ID &#39;</span> +</div><div class="line"><a name="l00558"></a><span class="lineno">  558</span>&#160;                                                   str( dsorftranscriptasso.id ) + <span class="stringliteral">&#39; do not agree together. An info type &#39;</span> + </div><div class="line"><a name="l00559"></a><span class="lineno">  559</span>&#160;                                                   Constants.INFO_TYPE_SEQUENCE_AGREEMENT + <span class="stringliteral">&#39;: &#39;</span> + Constants.INFO_CONTENT_DISCORDANT_SEQUENCES +</div><div class="line"><a name="l00560"></a><span class="lineno">  560</span>&#160;                                                   <span class="stringliteral">&#39; has been created in the UTInfoType table for this entry.&#39;</span>)</div><div class="line"><a name="l00561"></a><span class="lineno">  561</span>&#160;                    utinfotype = UTInfoType( table_name = <span class="stringliteral">&#39;DSORFTranscriptAsso&#39;</span>,</div><div class="line"><a name="l00562"></a><span class="lineno">  562</span>&#160;                                             object_id = dsorftranscriptasso.id, </div><div class="line"><a name="l00563"></a><span class="lineno">  563</span>&#160;                                             info_type = Constants.INFO_TYPE_SEQUENCE_AGREEMENT,</div><div class="line"><a name="l00564"></a><span class="lineno">  564</span>&#160;                                             info = Constants.INFO_CONTENT_DISCORDANT_SEQUENCES )</div><div class="line"><a name="l00565"></a><span class="lineno">  565</span>&#160;                    objects_to_insert.append( utinfotype )</div><div class="line"><a name="l00566"></a><span class="lineno">  566</span>&#160;                    </div><div class="line"><a name="l00567"></a><span class="lineno">  567</span>&#160;        <span class="comment"># Insert the new objects in the database</span></div><div class="line"><a name="l00568"></a><span class="lineno">  568</span>&#160;        <span class="keywordflow">if</span> len( objects_to_insert ) != 0:</div><div class="line"><a name="l00569"></a><span class="lineno">  569</span>&#160;            self.batch_insert_to_db( objects_to_insert, <span class="stringliteral">&#39;DSORFTrAsso_Seq_Agree&#39;</span> )</div><div class="line"><a name="l00570"></a><span class="lineno">  570</span>&#160;    </div><div class="line"><a name="l00571"></a><span class="lineno">  571</span>&#160;    </div><div class="line"><a name="l00572"></a><span class="lineno">  572</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="ac064a9a263aed718d6070ad138a228f7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac064a9a263aed718d6070ad138a228f7">&#9670;&nbsp;</a></span>convert_coord()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.convert_coord </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>lo</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>convert_coord </h2>
<p>TODO </p>
<div class="fragment"><div class="line"><a name="l00251"></a><span class="lineno">  251</span>&#160;    <span class="keyword">def </span>convert_coord( chr, strand, position, lo ):</div><div class="line"><a name="l00252"></a><span class="lineno">  252</span>&#160;        </div><div class="line"><a name="l00253"></a><span class="lineno">  253</span>&#160;        <span class="comment"># Check all the necessary information are provided</span></div><div class="line"><a name="l00254"></a><span class="lineno">  254</span>&#160;        <span class="keywordflow">if</span> ( chr == <span class="keywordtype">None</span> ) <span class="keywordflow">or</span> ( strand == <span class="keywordtype">None</span> ) <span class="keywordflow">or</span> ( position == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00255"></a><span class="lineno">  255</span>&#160;            <span class="keywordflow">raise</span> TODO Exception( <span class="stringliteral">&#39;TODO&#39;</span> )</div><div class="line"><a name="l00256"></a><span class="lineno">  256</span>&#160;        </div><div class="line"><a name="l00257"></a><span class="lineno">  257</span>&#160;        <span class="comment"># Make sure the chromosome name contains the &#39;chr&#39; prefix</span></div><div class="line"><a name="l00258"></a><span class="lineno">  258</span>&#160;        <span class="keywordflow">if</span> chr[:3] != <span class="stringliteral">&#39;chr&#39;</span>:</div><div class="line"><a name="l00259"></a><span class="lineno">  259</span>&#160;            chr = <span class="stringliteral">&#39;chr&#39;</span> + chr</div><div class="line"><a name="l00260"></a><span class="lineno">  260</span>&#160;        </div><div class="line"><a name="l00261"></a><span class="lineno">  261</span>&#160;        <span class="comment"># Try to convert the provided coordinates</span></div><div class="line"><a name="l00262"></a><span class="lineno">  262</span>&#160;        new_coord = lo.convert_coordinate( chr,</div><div class="line"><a name="l00263"></a><span class="lineno">  263</span>&#160;                                           int( position ),</div><div class="line"><a name="l00264"></a><span class="lineno">  264</span>&#160;                                           strand )</div><div class="line"><a name="l00265"></a><span class="lineno">  265</span>&#160;        </div><div class="line"><a name="l00266"></a><span class="lineno">  266</span>&#160;        <span class="keywordflow">if</span> len( new_coord ) == 1:</div><div class="line"><a name="l00267"></a><span class="lineno">  267</span>&#160;            new_chr = new_coord[0][0]</div><div class="line"><a name="l00268"></a><span class="lineno">  268</span>&#160;            new_strand = new_coord[0][1]</div><div class="line"><a name="l00269"></a><span class="lineno">  269</span>&#160;            new_pos = new_coord[0][2]</div><div class="line"><a name="l00270"></a><span class="lineno">  270</span>&#160;            <span class="keywordflow">return</span> ( new_chr, new_strand, new_pos )</div><div class="line"><a name="l00271"></a><span class="lineno">  271</span>&#160;        </div><div class="line"><a name="l00272"></a><span class="lineno">  272</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00273"></a><span class="lineno">  273</span>&#160;            <span class="keywordflow">raise</span> TODO Exception( <span class="stringliteral">&#39;ProcessingStrategy.convert_coord(): The convertion of the genomic coordinates (&#39;</span> +</div><div class="line"><a name="l00274"></a><span class="lineno">  274</span>&#160;                                  <span class="stringliteral">&#39;chr&#39;</span> + chr + strand + <span class="stringliteral">&#39;:&#39;</span> + str( position ) + <span class="stringliteral">&#39;) failed.&#39;</span> )</div><div class="line"><a name="l00275"></a><span class="lineno">  275</span>&#160;        </div><div class="line"><a name="l00276"></a><span class="lineno">  276</span>&#160;        </div><div class="line"><a name="l00277"></a><span class="lineno">  277</span>&#160;                </div></div><!-- fragment -->
</div>
</div>
<a id="a77d93c2c39c9c9797f1cfc620f11ddce"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a77d93c2c39c9c9797f1cfc620f11ddce">&#9670;&nbsp;</a></span>convert_dsorf_coordinates()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.convert_dsorf_coordinates </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dsorf</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>lo</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>convert_dsorf_coordinates </h2>
<p>This method allows to convert all the genomic coordinates of a DSORF entry. It try to convert:</p><ul>
<li>raw_start_pos (output registered in start_pos)</li>
<li>raw_stop_pos (output registered in stop_pos)</li>
<li>raw_splice_starts (output registered in splice_starts)</li>
<li>raw_splice_stops (output registered in splice_stops) If the conversion cannot be performed a warning is logged and the value of the attribute is set to None.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dsorf</td><td>DSORF - The DSORF entry for which the conversion should be performed. </td></tr>
    <tr><td class="paramname">lo</td><td>LiftOver - The LiftOver object (pyliftover package) allowing the conversion of the genomic coordinates.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>dsorf: DSORF - The updated DSORF object. </dd></dl>
<div class="fragment"><div class="line"><a name="l00297"></a><span class="lineno">  297</span>&#160;    <span class="keyword">def </span>convert_dsorf_coordinates( dsorf, lo ):</div><div class="line"><a name="l00298"></a><span class="lineno">  298</span>&#160;        </div><div class="line"><a name="l00299"></a><span class="lineno">  299</span>&#160;        <span class="comment"># Get the chromosome and the strand of the ORF </span></div><div class="line"><a name="l00300"></a><span class="lineno">  300</span>&#160;        orf_chromosome = <span class="stringliteral">&#39;chr&#39;</span> + str( dsorf.chromosome )</div><div class="line"><a name="l00301"></a><span class="lineno">  301</span>&#160;        orf_strand = str( dsorf.raw_strand )</div><div class="line"><a name="l00302"></a><span class="lineno">  302</span>&#160;        </div><div class="line"><a name="l00303"></a><span class="lineno">  303</span>&#160;        <span class="comment"># Perfom the conversion for the start genomic coordinate</span></div><div class="line"><a name="l00304"></a><span class="lineno">  304</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00305"></a><span class="lineno">  305</span>&#160;            ( start_chr, start_strand, start_pos ) = self.convert_coord( chr = orf_chromosome,</div><div class="line"><a name="l00306"></a><span class="lineno">  306</span>&#160;                                                                         strand = orf_strand,</div><div class="line"><a name="l00307"></a><span class="lineno">  307</span>&#160;                                                                         position = dsorf.raw_start_pos,</div><div class="line"><a name="l00308"></a><span class="lineno">  308</span>&#160;                                                                         lo = lo )</div><div class="line"><a name="l00309"></a><span class="lineno">  309</span>&#160;        <span class="keywordflow">except</span> TODO Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00310"></a><span class="lineno">  310</span>&#160;            Logger.get_instance().warning( e.get_message() )</div><div class="line"><a name="l00311"></a><span class="lineno">  311</span>&#160;            dsorf.start_pos = <span class="keywordtype">None</span></div><div class="line"><a name="l00312"></a><span class="lineno">  312</span>&#160;        </div><div class="line"><a name="l00313"></a><span class="lineno">  313</span>&#160;        <span class="comment"># Perform the conversion for the stop genomic coordinate</span></div><div class="line"><a name="l00314"></a><span class="lineno">  314</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00315"></a><span class="lineno">  315</span>&#160;            ( stop_chr, stop_strand, stop_pos ) = self.convert_coord( orf_chromosome,</div><div class="line"><a name="l00316"></a><span class="lineno">  316</span>&#160;                                                                      orf_strand,</div><div class="line"><a name="l00317"></a><span class="lineno">  317</span>&#160;                                                                      dsorf.raw_stop_pos )</div><div class="line"><a name="l00318"></a><span class="lineno">  318</span>&#160;        <span class="keywordflow">except</span> TODO Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00319"></a><span class="lineno">  319</span>&#160;            Logger.get_instance().warning( e.get_message() )</div><div class="line"><a name="l00320"></a><span class="lineno">  320</span>&#160;            dsorf.stop_pos = <span class="keywordtype">None</span></div><div class="line"><a name="l00321"></a><span class="lineno">  321</span>&#160;            </div><div class="line"><a name="l00322"></a><span class="lineno">  322</span>&#160;        <span class="comment"># Check the chromosome did not changed during coordinates conversion</span></div><div class="line"><a name="l00323"></a><span class="lineno">  323</span>&#160;        <span class="keywordflow">if</span> ( orf_chromosome != start_chr ) <span class="keywordflow">or</span> ( orf_chromosome != stop_chr ):</div><div class="line"><a name="l00324"></a><span class="lineno">  324</span>&#160;            Logger.get_instance().critical( <span class="stringliteral">&#39;TODO&#39;</span> )</div><div class="line"><a name="l00325"></a><span class="lineno">  325</span>&#160;        </div><div class="line"><a name="l00326"></a><span class="lineno">  326</span>&#160;        <span class="comment"># Check the strand after conversion</span></div><div class="line"><a name="l00327"></a><span class="lineno">  327</span>&#160;        <span class="keywordflow">if</span> ( start_strand != stop_strand ):</div><div class="line"><a name="l00328"></a><span class="lineno">  328</span>&#160;            Logger.get_instance().critical( <span class="stringliteral">&#39;TODO&#39;</span> )</div><div class="line"><a name="l00329"></a><span class="lineno">  329</span>&#160;            </div><div class="line"><a name="l00330"></a><span class="lineno">  330</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00331"></a><span class="lineno">  331</span>&#160;            <span class="comment"># If the ORF strand changed during the conversion, then reverse the start and stop coordinates</span></div><div class="line"><a name="l00332"></a><span class="lineno">  332</span>&#160;            <span class="keywordflow">if</span> ( start_strand != orf_strand ):</div><div class="line"><a name="l00333"></a><span class="lineno">  333</span>&#160;                ( start_pos, stop_pos ) = ( stop_pos, start_pos )</div><div class="line"><a name="l00334"></a><span class="lineno">  334</span>&#160;                </div><div class="line"><a name="l00335"></a><span class="lineno">  335</span>&#160;            dsorf.strand = start_strand</div><div class="line"><a name="l00336"></a><span class="lineno">  336</span>&#160;            dsorf.start_pos = start_pos</div><div class="line"><a name="l00337"></a><span class="lineno">  337</span>&#160;            dsorf.stop_pos = stop_pos</div><div class="line"><a name="l00338"></a><span class="lineno">  338</span>&#160;               </div><div class="line"><a name="l00339"></a><span class="lineno">  339</span>&#160;        <span class="comment"># Perform the conversion for the splicing coordinates</span></div><div class="line"><a name="l00340"></a><span class="lineno">  340</span>&#160;        <span class="keywordflow">for</span> att <span class="keywordflow">in</span> ProcessingStrategy.SPLICED_ATT_DSORF:</div><div class="line"><a name="l00341"></a><span class="lineno">  341</span>&#160;            </div><div class="line"><a name="l00342"></a><span class="lineno">  342</span>&#160;            raw_att_value = getattr( dsorf, <span class="stringliteral">&#39;raw_&#39;</span> + att )</div><div class="line"><a name="l00343"></a><span class="lineno">  343</span>&#160;            </div><div class="line"><a name="l00344"></a><span class="lineno">  344</span>&#160;            <span class="comment"># Try to convert the coordinates only if they are provided</span></div><div class="line"><a name="l00345"></a><span class="lineno">  345</span>&#160;            <span class="keywordflow">if</span> raw_att_value != <span class="keywordtype">None</span>:</div><div class="line"><a name="l00346"></a><span class="lineno">  346</span>&#160;                </div><div class="line"><a name="l00347"></a><span class="lineno">  347</span>&#160;                <span class="comment"># If the splicing value provided in the source exceeded the longest value that can be stored,</span></div><div class="line"><a name="l00348"></a><span class="lineno">  348</span>&#160;                <span class="comment"># use the same &quot;TOO_LONG&quot; string</span></div><div class="line"><a name="l00349"></a><span class="lineno">  349</span>&#160;                <span class="keywordflow">if</span> raw_att_value == Constants.REPLACE_TOO_LONG_STRINGS:</div><div class="line"><a name="l00350"></a><span class="lineno">  350</span>&#160;                    setattr( dsorf, att, raw_att_value )</div><div class="line"><a name="l00351"></a><span class="lineno">  351</span>&#160;                </div><div class="line"><a name="l00352"></a><span class="lineno">  352</span>&#160;                <span class="comment"># If splicing coordinates are provided, first parse the string </span></div><div class="line"><a name="l00353"></a><span class="lineno">  353</span>&#160;                <span class="comment"># into a list, convert each element of the list and then rebuild a string</span></div><div class="line"><a name="l00354"></a><span class="lineno">  354</span>&#160;                <span class="comment"># with the new coordinates</span></div><div class="line"><a name="l00355"></a><span class="lineno">  355</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00356"></a><span class="lineno">  356</span>&#160;                    raw_att_value = raw_att_value.split( Constants.ORF_SPLICING_COORD_SEPARATOR )</div><div class="line"><a name="l00357"></a><span class="lineno">  357</span>&#160;                    </div><div class="line"><a name="l00358"></a><span class="lineno">  358</span>&#160;                    <span class="comment"># Try to convert each element of the list,</span></div><div class="line"><a name="l00359"></a><span class="lineno">  359</span>&#160;                    <span class="comment"># if one of them cannot be converted or if one of them is not on the same strand as </span></div><div class="line"><a name="l00360"></a><span class="lineno">  360</span>&#160;                    <span class="comment"># the start and stop positions, then the whole attribute value is set to None</span></div><div class="line"><a name="l00361"></a><span class="lineno">  361</span>&#160;                    impossible_conversion = <span class="keyword">False</span></div><div class="line"><a name="l00362"></a><span class="lineno">  362</span>&#160;                    raw_att_value_index = 0</div><div class="line"><a name="l00363"></a><span class="lineno">  363</span>&#160;                    </div><div class="line"><a name="l00364"></a><span class="lineno">  364</span>&#160;                    new_att_value = []</div><div class="line"><a name="l00365"></a><span class="lineno">  365</span>&#160;                    </div><div class="line"><a name="l00366"></a><span class="lineno">  366</span>&#160;                    <span class="keywordflow">while</span> ( <span class="keywordflow">not</span> impossible_conversion ) <span class="keywordflow">and</span> ( raw_att_value_index &lt; len( raw_att_value) ):</div><div class="line"><a name="l00367"></a><span class="lineno">  367</span>&#160;                        </div><div class="line"><a name="l00368"></a><span class="lineno">  368</span>&#160;                        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00369"></a><span class="lineno">  369</span>&#160;                            ( exon_chr, exon_strand, exon_pos ) = self.convert_coord( chr = orf_chromosome,</div><div class="line"><a name="l00370"></a><span class="lineno">  370</span>&#160;                                                                                      strand = dsorf.strand,</div><div class="line"><a name="l00371"></a><span class="lineno">  371</span>&#160;                                                                                      position = raw_att_value[ raw_att_value_index ],</div><div class="line"><a name="l00372"></a><span class="lineno">  372</span>&#160;                                                                                      lo = lo )</div><div class="line"><a name="l00373"></a><span class="lineno">  373</span>&#160;                        </div><div class="line"><a name="l00374"></a><span class="lineno">  374</span>&#160;                        <span class="keywordflow">except</span> TODO Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00375"></a><span class="lineno">  375</span>&#160;                            Logger.get_instance().warning( e.get_message() + <span class="stringliteral">&#39; for the attribute &#39;</span> + att +</div><div class="line"><a name="l00376"></a><span class="lineno">  376</span>&#160;                                                           <span class="stringliteral">&#39; of the DSORF with ID &#39;</span> + str( dsorf.id ) + <span class="stringliteral">&#39; failed. Hence the value of this attribute has been set to None.&#39;</span>  )</div><div class="line"><a name="l00377"></a><span class="lineno">  377</span>&#160;                            impossible_conversion = <span class="keyword">True</span></div><div class="line"><a name="l00378"></a><span class="lineno">  378</span>&#160;                            new_att_value = <span class="keywordtype">None</span></div><div class="line"><a name="l00379"></a><span class="lineno">  379</span>&#160;                        </div><div class="line"><a name="l00380"></a><span class="lineno">  380</span>&#160;                        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00381"></a><span class="lineno">  381</span>&#160;                            raw_att_value_index += 1</div><div class="line"><a name="l00382"></a><span class="lineno">  382</span>&#160;                            <span class="keywordflow">if</span>:</div><div class="line"><a name="l00383"></a><span class="lineno">  383</span>&#160;                                new_att_value.append( )</div><div class="line"><a name="l00384"></a><span class="lineno">  384</span>&#160;                                </div><div class="line"><a name="l00385"></a><span class="lineno">  385</span>&#160;                    <span class="comment"># Rebuild the string</span></div><div class="line"><a name="l00386"></a><span class="lineno">  386</span>&#160;                    <span class="keywordflow">if</span> new_att_value != <span class="keywordtype">None</span>:</div><div class="line"><a name="l00387"></a><span class="lineno">  387</span>&#160;                        new_att_value = Constants.ORF_SPLICING_COORD_SEPARATOR.join( new_att_value )</div><div class="line"><a name="l00388"></a><span class="lineno">  388</span>&#160;                        </div><div class="line"><a name="l00389"></a><span class="lineno">  389</span>&#160;            <span class="comment"># Otherwise, set also the &quot;converted value&quot; to None</span></div><div class="line"><a name="l00390"></a><span class="lineno">  390</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00391"></a><span class="lineno">  391</span>&#160;                new_att_value = <span class="keywordtype">None</span></div><div class="line"><a name="l00392"></a><span class="lineno">  392</span>&#160;                            </div><div class="line"><a name="l00393"></a><span class="lineno">  393</span>&#160;            <span class="comment"># Register the value get after conversion</span></div><div class="line"><a name="l00394"></a><span class="lineno">  394</span>&#160;            setattr( dsorf, att, new_att_value )</div><div class="line"><a name="l00395"></a><span class="lineno">  395</span>&#160;            </div><div class="line"><a name="l00396"></a><span class="lineno">  396</span>&#160;        <span class="keywordflow">return</span> dsorf</div><div class="line"><a name="l00397"></a><span class="lineno">  397</span>&#160;        </div><div class="line"><a name="l00398"></a><span class="lineno">  398</span>&#160;        </div><div class="line"><a name="l00399"></a><span class="lineno">  399</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="a27a742369bcaed60067ea219599ba1e4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a27a742369bcaed60067ea219599ba1e4">&#9670;&nbsp;</a></span>convert_dstranscript_coordinates()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.convert_dstranscript_coordinates </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dstranscript</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>lo</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>convert_dstranscript_coordinates </h2>
<p>This method allows to convert all the genomic coordinates of a DSTranscript entry. It try to convert:</p><ul>
<li>raw_start_pos (output registered in start_pos)</li>
<li>raw_end_pos (output registered in end_pos)</li>
<li>raw_cds_start_pos (output registered in cds_start_pos)</li>
<li>raw_cds_stop_pos (output registered in cds_stop_pos)</li>
</ul>
<p>If the conversion cannot be performed a warning is logged and the value of the attribute is set to None.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dstranscript</td><td>DSTranscript - The DSTranscript entry for which the conversion should be performed. </td></tr>
    <tr><td class="paramname">lo</td><td>LiftOver - The LiftOver object (pyliftover package) allowing the conversion of the genomic coordinates.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>dstranscript: DSTranscript - The updated DSTranscript object. </dd></dl>
<div class="fragment"><div class="line"><a name="l00446"></a><span class="lineno">  446</span>&#160;    <span class="keyword">def </span>convert_dstranscript_coordinates( dstranscript, lo ):</div><div class="line"><a name="l00447"></a><span class="lineno">  447</span>&#160;        </div><div class="line"><a name="l00448"></a><span class="lineno">  448</span>&#160;        <span class="comment"># Get the chromosome of the gene related to this transcript</span></div><div class="line"><a name="l00449"></a><span class="lineno">  449</span>&#160;        transcript_gene = SQLManager.get_instance().get_session().query( Gene ).filter( Gene.gene_id == dstranscript.gene_id ).one()</div><div class="line"><a name="l00450"></a><span class="lineno">  450</span>&#160;        <span class="keywordflow">if</span> transcript_gene.chromosome == <span class="keywordtype">None</span>:</div><div class="line"><a name="l00451"></a><span class="lineno">  451</span>&#160;            transcript_chromosome = <span class="keywordtype">None</span></div><div class="line"><a name="l00452"></a><span class="lineno">  452</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00453"></a><span class="lineno">  453</span>&#160;            transcript_chromosome = <span class="stringliteral">&#39;chr&#39;</span> + transcript_gene.chromosome</div><div class="line"><a name="l00454"></a><span class="lineno">  454</span>&#160;        </div><div class="line"><a name="l00455"></a><span class="lineno">  455</span>&#160;        <span class="comment"># If the chromosome name is missing, set all the attributes values to None</span></div><div class="line"><a name="l00456"></a><span class="lineno">  456</span>&#160;        <span class="keywordflow">if</span> transcript_chromosome == <span class="keywordtype">None</span>:</div><div class="line"><a name="l00457"></a><span class="lineno">  457</span>&#160;            <span class="keywordflow">for</span> att <span class="keywordflow">in</span> ProcessingStrategy.ATT_TO_CONVERT_DSTRANSCRIPT:</div><div class="line"><a name="l00458"></a><span class="lineno">  458</span>&#160;                setattr( dstranscript, att, <span class="keywordtype">None</span> )</div><div class="line"><a name="l00459"></a><span class="lineno">  459</span>&#160;        </div><div class="line"><a name="l00460"></a><span class="lineno">  460</span>&#160;        <span class="comment"># Otherwise, try to convert the coordinates       </span></div><div class="line"><a name="l00461"></a><span class="lineno">  461</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00462"></a><span class="lineno">  462</span>&#160;            <span class="keywordflow">for</span> att <span class="keywordflow">in</span> ProcessingStrategy.ATT_TO_CONVERT_DSTRANSCRIPT:</div><div class="line"><a name="l00463"></a><span class="lineno">  463</span>&#160;    </div><div class="line"><a name="l00464"></a><span class="lineno">  464</span>&#160;                raw_att_value = getattr( dstranscript, <span class="stringliteral">&#39;raw_&#39;</span> + att )</div><div class="line"><a name="l00465"></a><span class="lineno">  465</span>&#160;                </div><div class="line"><a name="l00466"></a><span class="lineno">  466</span>&#160;                <span class="comment"># Try to convert the coordinates only if they are provided</span></div><div class="line"><a name="l00467"></a><span class="lineno">  467</span>&#160;                <span class="keywordflow">if</span> raw_att_value != <span class="keywordtype">None</span>:</div><div class="line"><a name="l00468"></a><span class="lineno">  468</span>&#160;                    new_coord = lo.convert_coordinate( transcript_chromosome,</div><div class="line"><a name="l00469"></a><span class="lineno">  469</span>&#160;                                                       int( raw_att_value ),</div><div class="line"><a name="l00470"></a><span class="lineno">  470</span>&#160;                                                       dstranscript.strand )</div><div class="line"><a name="l00471"></a><span class="lineno">  471</span>&#160;                    </div><div class="line"><a name="l00472"></a><span class="lineno">  472</span>&#160;                    <span class="keywordflow">if</span> len( new_coord ) == 1:</div><div class="line"><a name="l00473"></a><span class="lineno">  473</span>&#160;                        new_att_value = new_coord[0][1]</div><div class="line"><a name="l00474"></a><span class="lineno">  474</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00475"></a><span class="lineno">  475</span>&#160;                        new_att_value = <span class="keywordtype">None</span></div><div class="line"><a name="l00476"></a><span class="lineno">  476</span>&#160;                        Logger.get_instance().error( <span class="stringliteral">&#39;ProcessingStrategy.convert_dstranscript_coordinates(): The conversion of the genomic coordinates (&#39;</span> + </div><div class="line"><a name="l00477"></a><span class="lineno">  477</span>&#160;                                                     transcript_chromosome + <span class="stringliteral">&#39;:&#39;</span> + str( raw_att_value ) + <span class="stringliteral">&#39;) for the attribute &#39;</span> + att + </div><div class="line"><a name="l00478"></a><span class="lineno">  478</span>&#160;                                                     <span class="stringliteral">&#39; of the DSTranscript with ID &#39;</span> + str( dstranscript.id ) + </div><div class="line"><a name="l00479"></a><span class="lineno">  479</span>&#160;                                                     <span class="stringliteral">&#39; failed. Hence the value of this attribute has been set to None.&#39;</span>,</div><div class="line"><a name="l00480"></a><span class="lineno">  480</span>&#160;                                                     ex = <span class="keyword">False</span> )</div><div class="line"><a name="l00481"></a><span class="lineno">  481</span>&#160;            </div><div class="line"><a name="l00482"></a><span class="lineno">  482</span>&#160;                <span class="comment"># Otherwise, set also the &quot;converted value&quot; to None</span></div><div class="line"><a name="l00483"></a><span class="lineno">  483</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00484"></a><span class="lineno">  484</span>&#160;                    new_att_value = <span class="keywordtype">None</span></div><div class="line"><a name="l00485"></a><span class="lineno">  485</span>&#160;                    </div><div class="line"><a name="l00486"></a><span class="lineno">  486</span>&#160;                <span class="comment"># Register the value get after conversion</span></div><div class="line"><a name="l00487"></a><span class="lineno">  487</span>&#160;                setattr( dstranscript, att, new_att_value )</div><div class="line"><a name="l00488"></a><span class="lineno">  488</span>&#160;                </div><div class="line"><a name="l00489"></a><span class="lineno">  489</span>&#160;        <span class="keywordflow">return</span> dstranscript</div><div class="line"><a name="l00490"></a><span class="lineno">  490</span>&#160;        </div><div class="line"><a name="l00491"></a><span class="lineno">  491</span>&#160;        </div><div class="line"><a name="l00492"></a><span class="lineno">  492</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="a159a4828f8679cc87e4f3b0308ab2249"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a159a4828f8679cc87e4f3b0308ab2249">&#9670;&nbsp;</a></span>convert_genomic_coordinates()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.convert_genomic_coordinates </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>convert_genomic_coordinates </h2>
<p>This method allows to convert the genomic coordinates contained in the tables of the database to the current version. The entries of the following tables are updated during the execution of this method:</p><ul>
<li>DSORF table: start_pos, stop_pos, splice_starts, splice_stops attributes are filled.</li>
<li>DSTranscript table: start_pos, end_pos, cds_start_pos, cds_stop_pos are filled. When the coordinates provided by the source are in the right annotation version (attributes with a 'raw_' prefix), they are simply copied in the corresponding field. Otherwise, the coordinates are converted (using pyliftover) and saved in the corresponding field.</li>
</ul>
<p>/!\ Please be aware that this version of the method uses GRCh38 (hg38) and GRCm38 (mm10) as current annotation versions, and only allow the conversion of GRCh37 (hg19) coordinates to GRCh38 (hg38), as all the file parsed provide coordinates in the GRCh38 (hg38), GRCh37 (hg19) or GRCm38 (mm10) annotation version. </p>
<div class="fragment"><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;    <span class="keyword">def </span>convert_genomic_coordinates( self ):</div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;                </div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;        <span class="comment"># Get all the data sources contained in the database</span></div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;        datasources = DataManager.get_instance().get_data( <span class="stringliteral">&#39;datasources&#39;</span> )</div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;        </div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;        <span class="comment"># Get the dictionary which associates each data source to its annotation version</span></div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;        datasource_annot = DataManager.get_instance().get_data( <span class="stringliteral">&#39;datasource_annot&#39;</span> )</div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;        </div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;        <span class="comment"># Instantiate an empty list to receive all the objects that have been updated</span></div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;        updated_objects = []</div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;        </div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;        <span class="comment"># Get the total number of entries expected to be treated and set up a counter to follow the progression</span></div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;        total_entries_count = ( SQLManager.get_instance().get_session().query( DSORF ).count() + </div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;                                SQLManager.get_instance().get_session().query( DSTranscript ).count() )</div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;        progression_count = 1</div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;        </div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;        <span class="comment"># Get the list of the current annotation version for the species</span></div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;        current_annotation = [ Constants.CURRENT_NCBI_ANNOTATION[ self.species ], </div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;                               Constants.CURRENT_UCSC_ANNOTATION[ self.species ] ]</div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;        </div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;        <span class="comment"># For each data source, perform the conversion of coordinates on each DSORF and DSTranscript related to it</span></div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;        <span class="keywordflow">for</span> ( ds, annot ) <span class="keywordflow">in</span> datasource_annot.items():</div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;            </div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;            Logger.get_instance().debug( <span class="stringliteral">&#39;ProcessingStrategy.convert_genomic_coordinates(): Starting the conversion of DSORF and DSTranscript genomic coordinates related to &#39;</span> + </div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;                                         ds + <span class="stringliteral">&#39; (annotation version: &#39;</span> + annot + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;            </div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;            <span class="comment"># Get all the DSORF entries and all the DSTranscript entries related to this source</span></div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;            all_dsorfs_for_source = SQLManager.get_instance().get_session().query( DSORF ).filter( DSORF.data_source == ds ).all()</div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;            all_dstranscripts_for_source = SQLManager.get_instance().get_session().query( DSTranscript ).filter( DSTranscript.data_source == ds ).all()</div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;            </div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;            <span class="comment"># If the annotation of the data source is the current one, duplicate the provided genomic coordinates</span></div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;            <span class="keywordflow">if</span> annot <span class="keywordflow">in</span> current_annotation:</div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;                </div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;                <span class="comment"># Treat the DSORF entries related to the source</span></div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;                <span class="keywordflow">for</span> dsorf <span class="keywordflow">in</span> all_dsorfs_for_source:</div><div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;                    <span class="comment"># Display and update the progression bar on the console</span></div><div class="line"><a name="l00187"></a><span class="lineno">  187</span>&#160;                    GraphicUtil.progression_bar( total_entries_count, progression_count )</div><div class="line"><a name="l00188"></a><span class="lineno">  188</span>&#160;                    progression_count += 1</div><div class="line"><a name="l00189"></a><span class="lineno">  189</span>&#160;                    </div><div class="line"><a name="l00190"></a><span class="lineno">  190</span>&#160;                    <span class="comment"># Duplicate the genomic coordinates</span></div><div class="line"><a name="l00191"></a><span class="lineno">  191</span>&#160;                    updated_objects.append( self.duplicate_dsorf_coordinates( dsorf ) )</div><div class="line"><a name="l00192"></a><span class="lineno">  192</span>&#160;                </div><div class="line"><a name="l00193"></a><span class="lineno">  193</span>&#160;                <span class="comment"># Treat the DSTranscript entries related to the source</span></div><div class="line"><a name="l00194"></a><span class="lineno">  194</span>&#160;                <span class="keywordflow">for</span> dstranscript <span class="keywordflow">in</span> all_dstranscripts_for_source:</div><div class="line"><a name="l00195"></a><span class="lineno">  195</span>&#160;                    <span class="comment"># Display and update a progression bar on the console</span></div><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;                    GraphicUtil.progression_bar( total_entries_count, progression_count )</div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;                    progression_count += 1</div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;                    </div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;                    <span class="comment"># Duplicate the genomic coordinates</span></div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;                    updated_objects.append( self.duplicate_dstranscript_coordinates( dstranscript ) )</div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;                    </div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;            <span class="comment"># Check the annotation of the data source is one known</span></div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;            <span class="keywordflow">elif</span> annot <span class="keywordflow">in</span> Constants.ALL_SPECIES_ANNOTATIONS:</div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;                </div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;                <span class="comment"># Download and import the appropriate chain file (from the UCSC) allowing the conversion of annotation version</span></div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;                Logger.get_instance().debug( <span class="stringliteral">&#39;ProcessingStrategy.convert_genomic_coordinates(): Downloading the appropriate chain file from the UCSC to perfom the conversion of genomic coordinates.&#39;</span> )</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;                chain_file_url = ProcessingStrategy.CHAIN_FILES_URL[ annot ]</div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;                chain_file_path = os.path.join( Constants.TEMPORARY_FOLDER, ProcessingStrategy.CHAIN_FILENAMES[ annot ] )</div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;                wget.download( chain_file_url, out = chain_file_path, bar = <span class="keywordtype">None</span> )</div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;                </div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;                lo = pylo.LiftOver( chain_file_path )</div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;                </div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;                <span class="comment"># Treat the DSORF entries related to the source</span></div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;                <span class="keywordflow">for</span> dsorf <span class="keywordflow">in</span> all_dsorfs_for_source:</div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;                    <span class="comment"># Display and update the progression bar on the console</span></div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;                    GraphicUtil.progression_bar( total_entries_count, progression_count )</div><div class="line"><a name="l00217"></a><span class="lineno">  217</span>&#160;                    progression_count += 1</div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;                    </div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;                    <span class="comment"># Convert the genomic coordinates</span></div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;                    updated_objects.append( self.convert_dsorf_coordinates( dsorf, lo ) )</div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;                    </div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;                <span class="comment"># Treat the DSTranscript entries related to the source</span></div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;                <span class="keywordflow">for</span> dstranscript <span class="keywordflow">in</span> all_dstranscripts_for_source:</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;                    <span class="comment"># Display and update the progression bar on the console</span></div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;                    GraphicUtil.progression_bar( total_entries_count, progression_count )</div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;                    progression_count += 1</div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;                    </div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;                    <span class="comment"># Convert the genomic coordinates</span></div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;                    updated_objects.append( self.convert_dstranscript_coordinates( dstranscript, lo ) )</div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;                </div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;            <span class="comment"># Otherwise, log an error</span></div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                Logger.get_instance.error( <span class="stringliteral">&#39;ProcessingStrategy.convert_genomic_coordinates(): The annotation version provided for the data source &#39;</span> +</div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                                           ds + <span class="stringliteral">&#39;(&#39;</span> + annot + <span class="stringliteral">&#39;) is unexpected. Hence, the genomic coordinates contained in the entries related to&#39;</span> +</div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;                                           <span class="stringliteral">&#39; this data source will not be converted. Please make sure to use one of the following annotation version: &#39;</span> + </div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;                                           <span class="stringliteral">&#39;, &#39;</span>.join( Constants.ALL_SPECIES_ANNOTATIONS ) + <span class="stringliteral">&#39;.&#39;</span>,</div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;                                           ex = <span class="keyword">False</span> )</div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;            </div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;        <span class="comment"># Update the objects in the database and commit</span></div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;        <span class="keywordflow">if</span> len( updated_objects ) != 0:</div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;            self.batch_insert_to_db( updated_objects, <span class="stringliteral">&#39;objects_with_updated_coordinates&#39;</span> )</div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;            </div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;    </div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="a52b8101e3148256640cfd8d4820e0083"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a52b8101e3148256640cfd8d4820e0083">&#9670;&nbsp;</a></span>download_dsorf_seq()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.download_dsorf_seq </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>stop</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>splicing_status</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>splice_starts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>splice_ends</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>genome_version</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>download_dsorf_seq </h2>
<p>This methods allow to download the nucleic sequence of an ORF, taking into account the splicing when splicing information is provided.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chr</td><td>String - The chromosome name of the ORF. </td></tr>
    <tr><td class="paramname">strand</td><td>String - The ORF strand. </td></tr>
    <tr><td class="paramname">start</td><td>Integer - The start genomic coordinates of the ORF. </td></tr>
    <tr><td class="paramname">stop</td><td>Integer - The stop genomic coordinates of the ORF. </td></tr>
    <tr><td class="paramname">splicing_status</td><td>Boolean - Does the ORF is spliced? </td></tr>
    <tr><td class="paramname">splice_starts</td><td>String - The positions of the starts of the several "exonic" parts of the ORF. </td></tr>
    <tr><td class="paramname">splice_ends</td><td>String - The positions of the ends of the several "exonic" parts of the ORF. </td></tr>
    <tr><td class="paramname"></td><td>genome_version: String - The NCBI genome annotation version.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>: The nucleic sequence of this ORF (or of its "exonic parts"). </dd></dl>
<div class="fragment"><div class="line"><a name="l00822"></a><span class="lineno">  822</span>&#160;    <span class="keyword">def </span>download_dsorf_seq( self, chr, strand, start, stop, splicing_status, splice_starts, splice_ends, genome_version ):</div><div class="line"><a name="l00823"></a><span class="lineno">  823</span>&#160;        </div><div class="line"><a name="l00824"></a><span class="lineno">  824</span>&#160;        <span class="comment"># If the ORF is not spliced, download the sequence between the start and the stop</span></div><div class="line"><a name="l00825"></a><span class="lineno">  825</span>&#160;        <span class="keywordflow">if</span> ( splicing_status == <span class="keyword">False</span> ) <span class="keywordflow">or</span> ( splicing_status == <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00826"></a><span class="lineno">  826</span>&#160;            <span class="keywordflow">return</span> self.download_seq_from_ensembl( chr = chr,</div><div class="line"><a name="l00827"></a><span class="lineno">  827</span>&#160;                                                   strand = strand, </div><div class="line"><a name="l00828"></a><span class="lineno">  828</span>&#160;                                                   start = start, </div><div class="line"><a name="l00829"></a><span class="lineno">  829</span>&#160;                                                   stop = stop, </div><div class="line"><a name="l00830"></a><span class="lineno">  830</span>&#160;                                                   genome_version = genome_version )</div><div class="line"><a name="l00831"></a><span class="lineno">  831</span>&#160;        </div><div class="line"><a name="l00832"></a><span class="lineno">  832</span>&#160;        <span class="comment"># Otherwise, try to get the sequence of each &quot;exon&quot; and concatenate them</span></div><div class="line"><a name="l00833"></a><span class="lineno">  833</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00834"></a><span class="lineno">  834</span>&#160;            <span class="keywordflow">if</span> ( splice_starts != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( splice_ends != <span class="keywordtype">None</span> ):</div><div class="line"><a name="l00835"></a><span class="lineno">  835</span>&#160;                </div><div class="line"><a name="l00836"></a><span class="lineno">  836</span>&#160;                <span class="comment"># Get the list of starts and ends coordinates</span></div><div class="line"><a name="l00837"></a><span class="lineno">  837</span>&#160;                starts_pos = splice_starts.split( Constants.ORF_SPLICING_COORD_SEPARATOR )</div><div class="line"><a name="l00838"></a><span class="lineno">  838</span>&#160;                end_pos = splice_ends.split( Constants.ORF_SPLICING_COORD_SEPARATOR )</div><div class="line"><a name="l00839"></a><span class="lineno">  839</span>&#160;                </div><div class="line"><a name="l00840"></a><span class="lineno">  840</span>&#160;                <span class="comment"># Download the sequence of each &quot;exon&quot;</span></div><div class="line"><a name="l00841"></a><span class="lineno">  841</span>&#160;                sequence = []</div><div class="line"><a name="l00842"></a><span class="lineno">  842</span>&#160;                impossible_download = <span class="keyword">False</span></div><div class="line"><a name="l00843"></a><span class="lineno">  843</span>&#160;                idx = 0</div><div class="line"><a name="l00844"></a><span class="lineno">  844</span>&#160;                </div><div class="line"><a name="l00845"></a><span class="lineno">  845</span>&#160;                <span class="comment"># If the ORF is on the &#39;-&#39; strand, permute the start and stop coordinates </span></div><div class="line"><a name="l00846"></a><span class="lineno">  846</span>&#160;                <span class="comment"># in order to make sure the start position is lower than the stop one</span></div><div class="line"><a name="l00847"></a><span class="lineno">  847</span>&#160;                <span class="keywordflow">if</span> strand == <span class="stringliteral">&#39;-&#39;</span>:</div><div class="line"><a name="l00848"></a><span class="lineno">  848</span>&#160;                    starts_pos = starts_pos[ ::-1 ]</div><div class="line"><a name="l00849"></a><span class="lineno">  849</span>&#160;                    end_pos = end_pos[ ::-1 ]</div><div class="line"><a name="l00850"></a><span class="lineno">  850</span>&#160;                </div><div class="line"><a name="l00851"></a><span class="lineno">  851</span>&#160;                <span class="keywordflow">while</span> ( <span class="keywordflow">not</span> impossible_download ) <span class="keywordflow">and</span> ( idx &lt; len( starts_pos) ):</div><div class="line"><a name="l00852"></a><span class="lineno">  852</span>&#160;                    </div><div class="line"><a name="l00853"></a><span class="lineno">  853</span>&#160;                    exon_seq = self.download_seq_from_ensembl( chr = chr,</div><div class="line"><a name="l00854"></a><span class="lineno">  854</span>&#160;                                                               strand = strand, </div><div class="line"><a name="l00855"></a><span class="lineno">  855</span>&#160;                                                               start = starts_pos[ idx ], </div><div class="line"><a name="l00856"></a><span class="lineno">  856</span>&#160;                                                               stop = end_pos[ idx ], </div><div class="line"><a name="l00857"></a><span class="lineno">  857</span>&#160;                                                               genome_version = genome_version )</div><div class="line"><a name="l00858"></a><span class="lineno">  858</span>&#160;                    </div><div class="line"><a name="l00859"></a><span class="lineno">  859</span>&#160;                    <span class="keywordflow">if</span> exon_seq == <span class="keywordtype">None</span>:</div><div class="line"><a name="l00860"></a><span class="lineno">  860</span>&#160;                        impossible_download = <span class="keyword">True</span></div><div class="line"><a name="l00861"></a><span class="lineno">  861</span>&#160;                        sequence = <span class="keywordtype">None</span></div><div class="line"><a name="l00862"></a><span class="lineno">  862</span>&#160;                    </div><div class="line"><a name="l00863"></a><span class="lineno">  863</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00864"></a><span class="lineno">  864</span>&#160;                        idx += 1</div><div class="line"><a name="l00865"></a><span class="lineno">  865</span>&#160;                        sequence.append( exon_seq )</div><div class="line"><a name="l00866"></a><span class="lineno">  866</span>&#160;                        </div><div class="line"><a name="l00867"></a><span class="lineno">  867</span>&#160;                <span class="comment"># Concatenate the sequences and return it</span></div><div class="line"><a name="l00868"></a><span class="lineno">  868</span>&#160;                <span class="keywordflow">if</span> sequence != <span class="keywordtype">None</span>:</div><div class="line"><a name="l00869"></a><span class="lineno">  869</span>&#160;                    <span class="keywordflow">return</span> <span class="stringliteral">&#39;&#39;</span>.join( sequence )</div><div class="line"><a name="l00870"></a><span class="lineno">  870</span>&#160;                </div><div class="line"><a name="l00871"></a><span class="lineno">  871</span>&#160;                </div><div class="line"><a name="l00872"></a><span class="lineno">  872</span>&#160;            <span class="comment"># If some of the splicing coordinates are missing, set the sequence equals to None</span></div><div class="line"><a name="l00873"></a><span class="lineno">  873</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00874"></a><span class="lineno">  874</span>&#160;                <span class="keywordflow">return</span> <span class="keywordtype">None</span></div><div class="line"><a name="l00875"></a><span class="lineno">  875</span>&#160;                </div><div class="line"><a name="l00876"></a><span class="lineno">  876</span>&#160;            </div><div class="line"><a name="l00877"></a><span class="lineno">  877</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="ad323da493e31a3ce75f4767fc4d743c7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad323da493e31a3ce75f4767fc4d743c7">&#9670;&nbsp;</a></span>download_missing_sequences()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.download_missing_sequences </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>download_missing_sequences </h2>
<p>This method allow to download the nucleic sequence for all the entries of the DSORF and DSTranscript table. </p>
<div class="fragment"><div class="line"><a name="l00579"></a><span class="lineno">  579</span>&#160;    <span class="keyword">def </span>download_missing_sequences( self ):</div><div class="line"><a name="l00580"></a><span class="lineno">  580</span>&#160;        </div><div class="line"><a name="l00581"></a><span class="lineno">  581</span>&#160;        <span class="comment"># Get all the data sources contained in the database</span></div><div class="line"><a name="l00582"></a><span class="lineno">  582</span>&#160;        datasources = DataManager.get_instance().get_data( <span class="stringliteral">&#39;datasources&#39;</span> )</div><div class="line"><a name="l00583"></a><span class="lineno">  583</span>&#160;        </div><div class="line"><a name="l00584"></a><span class="lineno">  584</span>&#160;        <span class="comment"># Get the dictionary which associates each data source to its annotation version</span></div><div class="line"><a name="l00585"></a><span class="lineno">  585</span>&#160;        datasource_annot = DataManager.get_instance().get_data( <span class="stringliteral">&#39;datasource_annot&#39;</span> )</div><div class="line"><a name="l00586"></a><span class="lineno">  586</span>&#160;        </div><div class="line"><a name="l00587"></a><span class="lineno">  587</span>&#160;        <span class="comment"># Instantiate an empty list to receive all the objects that have been updated</span></div><div class="line"><a name="l00588"></a><span class="lineno">  588</span>&#160;        updated_objects = []</div><div class="line"><a name="l00589"></a><span class="lineno">  589</span>&#160;        </div><div class="line"><a name="l00590"></a><span class="lineno">  590</span>&#160;        <span class="comment"># Get the total number of entries expected to be treated and set up a counter to follow the progression</span></div><div class="line"><a name="l00591"></a><span class="lineno">  591</span>&#160;        total_entries_count = ( SQLManager.get_instance().get_session().query( DSORF ).count() + </div><div class="line"><a name="l00592"></a><span class="lineno">  592</span>&#160;                                SQLManager.get_instance().get_session().query( DSTranscript ).count() )</div><div class="line"><a name="l00593"></a><span class="lineno">  593</span>&#160;        progression_count = 1</div><div class="line"><a name="l00594"></a><span class="lineno">  594</span>&#160;        </div><div class="line"><a name="l00595"></a><span class="lineno">  595</span>&#160;        <span class="comment"># Get the NCBI current annotation version for the species</span></div><div class="line"><a name="l00596"></a><span class="lineno">  596</span>&#160;        current_ncbi_annot = Constants.CURRENT_NCBI_ANNOTATION[ self.species ]</div><div class="line"><a name="l00597"></a><span class="lineno">  597</span>&#160;        </div><div class="line"><a name="l00598"></a><span class="lineno">  598</span>&#160;        <span class="comment"># For each data source, download the nucleic sequences for each DSORF and DSTranscript related to it</span></div><div class="line"><a name="l00599"></a><span class="lineno">  599</span>&#160;        <span class="keywordflow">for</span> ( ds, annot ) <span class="keywordflow">in</span> datasource_annot.items():</div><div class="line"><a name="l00600"></a><span class="lineno">  600</span>&#160;            </div><div class="line"><a name="l00601"></a><span class="lineno">  601</span>&#160;            Logger.get_instance().debug( <span class="stringliteral">&#39;ProcessingStrategy.download_missing_info(): Starting the download of sequences for DSORF and DSTranscript related to &#39;</span> + </div><div class="line"><a name="l00602"></a><span class="lineno">  602</span>&#160;                                         ds + <span class="stringliteral">&#39; (original annotation version: &#39;</span> + annot + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l00603"></a><span class="lineno">  603</span>&#160;            </div><div class="line"><a name="l00604"></a><span class="lineno">  604</span>&#160;            <span class="comment"># Get the NCBI annotation version corresponding to the one provided</span></div><div class="line"><a name="l00605"></a><span class="lineno">  605</span>&#160;            <span class="keywordflow">if</span> annot <span class="keywordflow">in</span> Constants.ALL_NCBI_ANNOTATIONS:</div><div class="line"><a name="l00606"></a><span class="lineno">  606</span>&#160;                original_ncbi_annot = annot</div><div class="line"><a name="l00607"></a><span class="lineno">  607</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00608"></a><span class="lineno">  608</span>&#160;                original_ncbi_annot = Constants.CORRESPONDING_NCBI_FROM_UCSC[ annot ]</div><div class="line"><a name="l00609"></a><span class="lineno">  609</span>&#160;            </div><div class="line"><a name="l00610"></a><span class="lineno">  610</span>&#160;            <span class="comment"># Get all the DSORF entries and all the DSTranscript entries related to this source</span></div><div class="line"><a name="l00611"></a><span class="lineno">  611</span>&#160;            all_dsorfs_for_source = SQLManager.get_instance().get_session().query( DSORF ).filter( DSORF.data_source == ds ).all()</div><div class="line"><a name="l00612"></a><span class="lineno">  612</span>&#160;            all_dstranscripts_for_source = SQLManager.get_instance().get_session().query( DSTranscript ).filter( DSTranscript.data_source == ds ).all()</div><div class="line"><a name="l00613"></a><span class="lineno">  613</span>&#160;            </div><div class="line"><a name="l00614"></a><span class="lineno">  614</span>&#160;            <span class="comment"># Treat the DSORF entries related to the source</span></div><div class="line"><a name="l00615"></a><span class="lineno">  615</span>&#160;            <span class="keywordflow">for</span> dsorf <span class="keywordflow">in</span> all_dsorfs_for_source:</div><div class="line"><a name="l00616"></a><span class="lineno">  616</span>&#160;                <span class="comment"># Display and update the progression bar on the console</span></div><div class="line"><a name="l00617"></a><span class="lineno">  617</span>&#160;                GraphicUtil.progression_bar( total_entries_count, progression_count )</div><div class="line"><a name="l00618"></a><span class="lineno">  618</span>&#160;                progression_count += 1</div><div class="line"><a name="l00619"></a><span class="lineno">  619</span>&#160;            </div><div class="line"><a name="l00620"></a><span class="lineno">  620</span>&#160;                <span class="comment"># Download the nucleic sequence in the current annotation version</span></div><div class="line"><a name="l00621"></a><span class="lineno">  621</span>&#160;                seq_current_annot = self.download_dsorf_seq( chr = dsorf.chromosome, </div><div class="line"><a name="l00622"></a><span class="lineno">  622</span>&#160;                                                             strand = dsorf.strand,</div><div class="line"><a name="l00623"></a><span class="lineno">  623</span>&#160;                                                             start = dsorf.start_pos,</div><div class="line"><a name="l00624"></a><span class="lineno">  624</span>&#160;                                                             stop = dsorf.stop_pos, </div><div class="line"><a name="l00625"></a><span class="lineno">  625</span>&#160;                                                             splicing_status = dsorf.spliced, </div><div class="line"><a name="l00626"></a><span class="lineno">  626</span>&#160;                                                             splice_starts = dsorf.splice_starts, </div><div class="line"><a name="l00627"></a><span class="lineno">  627</span>&#160;                                                             splice_ends = dsorf.splice_ends, </div><div class="line"><a name="l00628"></a><span class="lineno">  628</span>&#160;                                                             genome_version = current_ncbi_annot )</div><div class="line"><a name="l00629"></a><span class="lineno">  629</span>&#160;                </div><div class="line"><a name="l00630"></a><span class="lineno">  630</span>&#160;                <span class="comment"># If the original version is the current one, complete the DSORF entry with the downloaded sequence</span></div><div class="line"><a name="l00631"></a><span class="lineno">  631</span>&#160;                <span class="keywordflow">if</span> original_ncbi_annot == current_ncbi_annot:</div><div class="line"><a name="l00632"></a><span class="lineno">  632</span>&#160;                    dsorf.sequence = seq_current_annot</div><div class="line"><a name="l00633"></a><span class="lineno">  633</span>&#160;                </div><div class="line"><a name="l00634"></a><span class="lineno">  634</span>&#160;                <span class="comment"># If the original version is different from the current one, download also the sequence</span></div><div class="line"><a name="l00635"></a><span class="lineno">  635</span>&#160;                <span class="comment"># in the original coordinates, and make sure it is the same as the one get in the</span></div><div class="line"><a name="l00636"></a><span class="lineno">  636</span>&#160;                <span class="comment"># coordinates get after the conversion step</span></div><div class="line"><a name="l00637"></a><span class="lineno">  637</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00638"></a><span class="lineno">  638</span>&#160;                    seq_previous_annot = self.download_dsorf_seq( chr = dsorf.chromosome, </div><div class="line"><a name="l00639"></a><span class="lineno">  639</span>&#160;                                                                  strand = dsorf.strand,</div><div class="line"><a name="l00640"></a><span class="lineno">  640</span>&#160;                                                                  start = dsorf.raw_start_pos,</div><div class="line"><a name="l00641"></a><span class="lineno">  641</span>&#160;                                                                  stop = dsorf.raw_stop_pos, </div><div class="line"><a name="l00642"></a><span class="lineno">  642</span>&#160;                                                                  splicing_status = dsorf.spliced, </div><div class="line"><a name="l00643"></a><span class="lineno">  643</span>&#160;                                                                  splice_starts = dsorf.raw_splice_starts, </div><div class="line"><a name="l00644"></a><span class="lineno">  644</span>&#160;                                                                  splice_ends = dsorf.raw_splice_ends, </div><div class="line"><a name="l00645"></a><span class="lineno">  645</span>&#160;                                                                  genome_version = original_ncbi_annot )</div><div class="line"><a name="l00646"></a><span class="lineno">  646</span>&#160;            </div><div class="line"><a name="l00647"></a><span class="lineno">  647</span>&#160;                    <span class="comment"># If the two sequences are not equal, log a warning and register this difference in the</span></div><div class="line"><a name="l00648"></a><span class="lineno">  648</span>&#160;                    <span class="comment"># UTInfoType table</span></div><div class="line"><a name="l00649"></a><span class="lineno">  649</span>&#160;                    <span class="keywordflow">if</span> seq_current_annot == seq_previous_annot:</div><div class="line"><a name="l00650"></a><span class="lineno">  650</span>&#160;                        seq_to_save = seq_current_annot</div><div class="line"><a name="l00651"></a><span class="lineno">  651</span>&#160;                        </div><div class="line"><a name="l00652"></a><span class="lineno">  652</span>&#160;                    <span class="keywordflow">elif</span> seq_current_annot == <span class="keywordtype">None</span>:</div><div class="line"><a name="l00653"></a><span class="lineno">  653</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;The sequence of the DSORF with ID &#39;</span> + str( dsorf.id ) + </div><div class="line"><a name="l00654"></a><span class="lineno">  654</span>&#160;                                                       <span class="stringliteral">&#39; cannot be found using the coordinates in the current annotation version&#39;</span> +</div><div class="line"><a name="l00655"></a><span class="lineno">  655</span>&#160;                                                       <span class="stringliteral">&#39; Hence, the sequence has been downloaded using the original annotation version (&#39;</span> +</div><div class="line"><a name="l00656"></a><span class="lineno">  656</span>&#160;                                                       original_ncbi_annot + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l00657"></a><span class="lineno">  657</span>&#160;                        seq_to_save = seq_previous_annot</div><div class="line"><a name="l00658"></a><span class="lineno">  658</span>&#160;                        </div><div class="line"><a name="l00659"></a><span class="lineno">  659</span>&#160;                    <span class="keywordflow">elif</span> seq_current_annot != seq_previous_annot:</div><div class="line"><a name="l00660"></a><span class="lineno">  660</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;The sequences of the DSORF with ID &#39;</span> + str( dsorf.id ) + </div><div class="line"><a name="l00661"></a><span class="lineno">  661</span>&#160;                                                       <span class="stringliteral">&#39; downloaded using the coordinates in the original annotation version (&#39;</span> +</div><div class="line"><a name="l00662"></a><span class="lineno">  662</span>&#160;                                                       annot + <span class="stringliteral">&#39;, &#39;</span> + seq_previous_annot + <span class="stringliteral">&#39;) and using the coordinates in the current annotation version (&#39;</span> +</div><div class="line"><a name="l00663"></a><span class="lineno">  663</span>&#160;                                                       current_ncbi_annot + <span class="stringliteral">&#39;, &#39;</span> + seq_current_annot + <span class="stringliteral">&#39;) are not the same.&#39;</span> +</div><div class="line"><a name="l00664"></a><span class="lineno">  664</span>&#160;                                                       <span class="stringliteral">&#39; The sequence in the current annotation version will be saved in the DSORF entry.&#39;</span> )</div><div class="line"><a name="l00665"></a><span class="lineno">  665</span>&#160;                        seq_to_save = seq_current_annot</div><div class="line"><a name="l00666"></a><span class="lineno">  666</span>&#160;                        utinfotype = UTInfoType( table_name = <span class="stringliteral">&#39;DSORF&#39;</span>,</div><div class="line"><a name="l00667"></a><span class="lineno">  667</span>&#160;                                                 object_id = dsorf.id,</div><div class="line"><a name="l00668"></a><span class="lineno">  668</span>&#160;                                                 info_type = Constants.INFO_TYPE_SEQUENCE_AGREEMENT_AFTER_CONVERSION,</div><div class="line"><a name="l00669"></a><span class="lineno">  669</span>&#160;                                                 info = Constants.INFO_CONTENT_DISCORDANT_SEQUENCES )</div><div class="line"><a name="l00670"></a><span class="lineno">  670</span>&#160;                    </div><div class="line"><a name="l00671"></a><span class="lineno">  671</span>&#160;                    <span class="comment"># In any case, store the original sequence in the DSORF table</span></div><div class="line"><a name="l00672"></a><span class="lineno">  672</span>&#160;                      <span class="comment"># Replace too long sequences by &quot;TOO_LONG&quot;</span></div><div class="line"><a name="l00673"></a><span class="lineno">  673</span>&#160;                    <span class="keywordflow">if</span> ( seq_to_save != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( len( seq_to_save ) &gt; Constants.MAX_LEN_TEXT ):</div><div class="line"><a name="l00674"></a><span class="lineno">  674</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;The sequence of the DSORF with ID &#39;</span> + str( dsorf.id ) + <span class="stringliteral">&#39; is too long to be stored in the database.&#39;</span> +</div><div class="line"><a name="l00675"></a><span class="lineno">  675</span>&#160;                                                       <span class="stringliteral">&#39; Hence its value will be replaced by &quot;&#39;</span> + Constants.REPLACE_TOO_LONG_TEXT + <span class="stringliteral">&#39;&quot;.&#39;</span> )</div><div class="line"><a name="l00676"></a><span class="lineno">  676</span>&#160;                        dsorf.sequence = Constants.REPLACE_TOO_LONG_TEXT</div><div class="line"><a name="l00677"></a><span class="lineno">  677</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00678"></a><span class="lineno">  678</span>&#160;                        dsorf.sequence = seq_to_save</div><div class="line"><a name="l00679"></a><span class="lineno">  679</span>&#160;                    </div><div class="line"><a name="l00680"></a><span class="lineno">  680</span>&#160;                <span class="comment"># Add this DSORF to the list of objects to update</span></div><div class="line"><a name="l00681"></a><span class="lineno">  681</span>&#160;                updated_objects.append( dsorf )</div><div class="line"><a name="l00682"></a><span class="lineno">  682</span>&#160;                </div><div class="line"><a name="l00683"></a><span class="lineno">  683</span>&#160;            <span class="comment"># Treat the DSTranscript entries related to the source</span></div><div class="line"><a name="l00684"></a><span class="lineno">  684</span>&#160;            <span class="keywordflow">for</span> dstranscript <span class="keywordflow">in</span> all_dstranscripts_for_source:</div><div class="line"><a name="l00685"></a><span class="lineno">  685</span>&#160;                <span class="comment"># Display and update the progression bar on the console</span></div><div class="line"><a name="l00686"></a><span class="lineno">  686</span>&#160;                GraphicUtil.progression_bar( total_entries_count, progression_count )</div><div class="line"><a name="l00687"></a><span class="lineno">  687</span>&#160;                progression_count += 1</div><div class="line"><a name="l00688"></a><span class="lineno">  688</span>&#160;            </div><div class="line"><a name="l00689"></a><span class="lineno">  689</span>&#160;                <span class="comment"># Get the chromosome of the gene related to this transcript</span></div><div class="line"><a name="l00690"></a><span class="lineno">  690</span>&#160;                transcript_gene = SQLManager.get_instance().get_session().query( Gene ).filter( Gene.gene_id == dstranscript.gene_id ).one()</div><div class="line"><a name="l00691"></a><span class="lineno">  691</span>&#160;                <span class="keywordflow">if</span> transcript_gene.chromosome == <span class="keywordtype">None</span>:</div><div class="line"><a name="l00692"></a><span class="lineno">  692</span>&#160;                    transcript_chromosome = <span class="keywordtype">None</span></div><div class="line"><a name="l00693"></a><span class="lineno">  693</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00694"></a><span class="lineno">  694</span>&#160;                    transcript_chromosome = <span class="stringliteral">&#39;chr&#39;</span> + transcript_gene.chromosome</div><div class="line"><a name="l00695"></a><span class="lineno">  695</span>&#160;                </div><div class="line"><a name="l00696"></a><span class="lineno">  696</span>&#160;                <span class="comment"># If the chromosome name is missing, set the sequence equal to None</span></div><div class="line"><a name="l00697"></a><span class="lineno">  697</span>&#160;                <span class="keywordflow">if</span> transcript_chromosome == <span class="keywordtype">None</span>:</div><div class="line"><a name="l00698"></a><span class="lineno">  698</span>&#160;                    dstranscript.sequence = <span class="keywordtype">None</span></div><div class="line"><a name="l00699"></a><span class="lineno">  699</span>&#160;                </div><div class="line"><a name="l00700"></a><span class="lineno">  700</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00701"></a><span class="lineno">  701</span>&#160;                    <span class="comment"># Download the nucleic sequence in the current annotation version</span></div><div class="line"><a name="l00702"></a><span class="lineno">  702</span>&#160;                    seq_current_annot = self.download_seq_from_ensembl( chr = transcript_chromosome,</div><div class="line"><a name="l00703"></a><span class="lineno">  703</span>&#160;                                                                        strand = dstranscript.strand,</div><div class="line"><a name="l00704"></a><span class="lineno">  704</span>&#160;                                                                        start = dstranscript.start_pos,</div><div class="line"><a name="l00705"></a><span class="lineno">  705</span>&#160;                                                                        stop = dstranscript.end_pos,</div><div class="line"><a name="l00706"></a><span class="lineno">  706</span>&#160;                                                                        genome_version = current_ncbi_annot )</div><div class="line"><a name="l00707"></a><span class="lineno">  707</span>&#160;                    </div><div class="line"><a name="l00708"></a><span class="lineno">  708</span>&#160;                    <span class="comment"># If the original version is the current one, complete the DSTranscript entry with the downloaded sequence</span></div><div class="line"><a name="l00709"></a><span class="lineno">  709</span>&#160;                    <span class="keywordflow">if</span> original_ncbi_annot == current_ncbi_annot:</div><div class="line"><a name="l00710"></a><span class="lineno">  710</span>&#160;                        dstranscript.sequence = seq_current_annot</div><div class="line"><a name="l00711"></a><span class="lineno">  711</span>&#160;                    </div><div class="line"><a name="l00712"></a><span class="lineno">  712</span>&#160;                    <span class="comment"># If the original version is different from the current one, download also the sequence</span></div><div class="line"><a name="l00713"></a><span class="lineno">  713</span>&#160;                    <span class="comment"># in the original coordinates, and make sure it is the same as the one get in the</span></div><div class="line"><a name="l00714"></a><span class="lineno">  714</span>&#160;                    <span class="comment"># coordinates get after the conversion step</span></div><div class="line"><a name="l00715"></a><span class="lineno">  715</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00716"></a><span class="lineno">  716</span>&#160;                        seq_previous_annot = self.download_seq_from_ensembl( chr = transcript_chromosome,</div><div class="line"><a name="l00717"></a><span class="lineno">  717</span>&#160;                                                                             strand = dstranscript.strand,</div><div class="line"><a name="l00718"></a><span class="lineno">  718</span>&#160;                                                                             start = dstranscript.raw_start_pos,</div><div class="line"><a name="l00719"></a><span class="lineno">  719</span>&#160;                                                                             stop = dstranscript.raw_end_pos,</div><div class="line"><a name="l00720"></a><span class="lineno">  720</span>&#160;                                                                             genome_version = original_ncbi_annot )</div><div class="line"><a name="l00721"></a><span class="lineno">  721</span>&#160;                </div><div class="line"><a name="l00722"></a><span class="lineno">  722</span>&#160;                        <span class="comment"># If the two sequences are not equal, log a warning and register this difference in the</span></div><div class="line"><a name="l00723"></a><span class="lineno">  723</span>&#160;                        <span class="comment"># UTInfoType table</span></div><div class="line"><a name="l00724"></a><span class="lineno">  724</span>&#160;                        <span class="keywordflow">if</span> seq_current_annot == seq_previous_annot:</div><div class="line"><a name="l00725"></a><span class="lineno">  725</span>&#160;                            seq_to_save = seq_current_annot</div><div class="line"><a name="l00726"></a><span class="lineno">  726</span>&#160;                            </div><div class="line"><a name="l00727"></a><span class="lineno">  727</span>&#160;                        <span class="keywordflow">elif</span> seq_current_annot == <span class="keywordtype">None</span>:</div><div class="line"><a name="l00728"></a><span class="lineno">  728</span>&#160;                            Logger.get_instance().warning( <span class="stringliteral">&#39;The sequence of the DSTranscript with ID &#39;</span> + str( dstranscript.id ) + </div><div class="line"><a name="l00729"></a><span class="lineno">  729</span>&#160;                                                           <span class="stringliteral">&#39; cannot be found using the coordinates in the current annotation version&#39;</span> +</div><div class="line"><a name="l00730"></a><span class="lineno">  730</span>&#160;                                                           <span class="stringliteral">&#39; Hence, the sequence has been downloaded using the original annotation version (&#39;</span> +</div><div class="line"><a name="l00731"></a><span class="lineno">  731</span>&#160;                                                           original_ncbi_annot + <span class="stringliteral">&#39;).&#39;</span> )</div><div class="line"><a name="l00732"></a><span class="lineno">  732</span>&#160;                            seq_to_save = seq_previous_annot</div><div class="line"><a name="l00733"></a><span class="lineno">  733</span>&#160;                            </div><div class="line"><a name="l00734"></a><span class="lineno">  734</span>&#160;                        <span class="keywordflow">elif</span> seq_current_annot != seq_previous_annot:</div><div class="line"><a name="l00735"></a><span class="lineno">  735</span>&#160;                            Logger.get_instance().warning( <span class="stringliteral">&#39;The sequences of the DSTranscript with ID &#39;</span> + str( dstranscript.id ) + </div><div class="line"><a name="l00736"></a><span class="lineno">  736</span>&#160;                                                           <span class="stringliteral">&#39; downloaded using the coordinates in the original annotation version (&#39;</span> +</div><div class="line"><a name="l00737"></a><span class="lineno">  737</span>&#160;                                                           annot + <span class="stringliteral">&#39;, &#39;</span> + seq_previous_annot + <span class="stringliteral">&#39;) and using the coordinates in the current annotation version (&#39;</span> +</div><div class="line"><a name="l00738"></a><span class="lineno">  738</span>&#160;                                                           current_ncbi_annot + <span class="stringliteral">&#39;, &#39;</span> + seq_current_annot + <span class="stringliteral">&#39;) are not the same.&#39;</span> +</div><div class="line"><a name="l00739"></a><span class="lineno">  739</span>&#160;                                                           <span class="stringliteral">&#39; The sequence in the current annotation version will be saved in the DSTranscript entry.&#39;</span> )</div><div class="line"><a name="l00740"></a><span class="lineno">  740</span>&#160;                            seq_to_save = seq_current_annot</div><div class="line"><a name="l00741"></a><span class="lineno">  741</span>&#160;                            utinfotype = UTInfoType( table_name = <span class="stringliteral">&#39;DSTranscript&#39;</span>,</div><div class="line"><a name="l00742"></a><span class="lineno">  742</span>&#160;                                                     object_id = dstranscript.id,</div><div class="line"><a name="l00743"></a><span class="lineno">  743</span>&#160;                                                     info_type = Constants.INFO_TYPE_SEQUENCE_AGREEMENT_AFTER_CONVERSION,</div><div class="line"><a name="l00744"></a><span class="lineno">  744</span>&#160;                                                     info = Constants.INFO_CONTENT_DISCORDANT_SEQUENCES )</div><div class="line"><a name="l00745"></a><span class="lineno">  745</span>&#160;                        </div><div class="line"><a name="l00746"></a><span class="lineno">  746</span>&#160;                        <span class="comment"># In any case, store the original sequence in the DSTranscript table</span></div><div class="line"><a name="l00747"></a><span class="lineno">  747</span>&#160;                          <span class="comment"># Replace too long sequences by &quot;TOO_LONG&quot;</span></div><div class="line"><a name="l00748"></a><span class="lineno">  748</span>&#160;                    <span class="keywordflow">if</span> ( seq_to_save != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( len( seq_to_save ) &gt; Constants.MAX_LEN_TEXT ):</div><div class="line"><a name="l00749"></a><span class="lineno">  749</span>&#160;                        Logger.get_instance().warning( <span class="stringliteral">&#39;The sequence of the DSTranscript with ID &#39;</span> + str( dstranscript.id ) + <span class="stringliteral">&#39; is too long to be stored in the database.&#39;</span> +</div><div class="line"><a name="l00750"></a><span class="lineno">  750</span>&#160;                                                       <span class="stringliteral">&#39; Hence its value will be replaced by &quot;&#39;</span> + Constants.REPLACE_TOO_LONG_TEXT + <span class="stringliteral">&#39;&quot;.&#39;</span> )</div><div class="line"><a name="l00751"></a><span class="lineno">  751</span>&#160;                        dstranscript.sequence = Constants.REPLACE_TOO_LONG_TEXT</div><div class="line"><a name="l00752"></a><span class="lineno">  752</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00753"></a><span class="lineno">  753</span>&#160;                        dstranscript.sequence = seq_to_save</div><div class="line"><a name="l00754"></a><span class="lineno">  754</span>&#160;                    </div><div class="line"><a name="l00755"></a><span class="lineno">  755</span>&#160;                <span class="comment"># Add this DSTranscript to the list of objects to update</span></div><div class="line"><a name="l00756"></a><span class="lineno">  756</span>&#160;                updated_objects.append( dstranscript ) </div><div class="line"><a name="l00757"></a><span class="lineno">  757</span>&#160;                </div><div class="line"><a name="l00758"></a><span class="lineno">  758</span>&#160;        <span class="comment"># Update the objects in the database and commit</span></div><div class="line"><a name="l00759"></a><span class="lineno">  759</span>&#160;        <span class="keywordflow">if</span> len( updated_objects ) != 0:</div><div class="line"><a name="l00760"></a><span class="lineno">  760</span>&#160;            self.batch_insert_to_db( updated_objects, <span class="stringliteral">&#39;dsorf_and_dstranscript_with_nucleic_sequences&#39;</span> )</div><div class="line"><a name="l00761"></a><span class="lineno">  761</span>&#160;    </div><div class="line"><a name="l00762"></a><span class="lineno">  762</span>&#160;    </div><div class="line"><a name="l00763"></a><span class="lineno">  763</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="ab5e2942fee457da5fbde63f852639813"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab5e2942fee457da5fbde63f852639813">&#9670;&nbsp;</a></span>download_seq_from_ensembl()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.download_seq_from_ensembl </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>stop</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>genome_version</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>download_seq_from_ensembl </h2>
<p>This method allows to download a nucleic sequence from Ensembl.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chr</td><td>String - The chromosome name (without 'chr' prefix). </td></tr>
    <tr><td class="paramname">strand</td><td>String - The DNA strand. </td></tr>
    <tr><td class="paramname">start</td><td>Integer or String - The genomic coordinates of the first position. </td></tr>
    <tr><td class="paramname">stop</td><td>Integer or String - The genomic coordinates of the last position. </td></tr>
    <tr><td class="paramname">genome_version</td><td>String - The NCBI genome version (e.g. GRCh38).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the nucleic sequence of the region </dd></dl>
<div class="fragment"><div class="line"><a name="l00777"></a><span class="lineno">  777</span>&#160;    <span class="keyword">def </span>download_seq_from_ensembl( self, chr, strand, start, stop, genome_version ):</div><div class="line"><a name="l00778"></a><span class="lineno">  778</span>&#160;        </div><div class="line"><a name="l00779"></a><span class="lineno">  779</span>&#160;        <span class="comment"># Server to query</span></div><div class="line"><a name="l00780"></a><span class="lineno">  780</span>&#160;        server = <span class="stringliteral">&#39;http://rest.ensembl.org&#39;</span></div><div class="line"><a name="l00781"></a><span class="lineno">  781</span>&#160;        </div><div class="line"><a name="l00782"></a><span class="lineno">  782</span>&#160;        <span class="comment"># If some of the feature are missing, return None</span></div><div class="line"><a name="l00783"></a><span class="lineno">  783</span>&#160;        <span class="keywordflow">if</span> ( ( chr == <span class="keywordtype">None</span> ) <span class="keywordflow">or</span> ( strand == <span class="keywordtype">None</span> ) <span class="keywordflow">or</span> ( start == <span class="keywordtype">None</span> ) <span class="keywordflow">or</span> ( stop == <span class="keywordtype">None</span> ) ):</div><div class="line"><a name="l00784"></a><span class="lineno">  784</span>&#160;            <span class="keywordflow">return</span> <span class="keywordtype">None</span></div><div class="line"><a name="l00785"></a><span class="lineno">  785</span>&#160;                </div><div class="line"><a name="l00786"></a><span class="lineno">  786</span>&#160;        <span class="comment"># URL extension to use to get the appropriate sequence</span></div><div class="line"><a name="l00787"></a><span class="lineno">  787</span>&#160;        ext = <span class="stringliteral">&#39;/sequence/region/human/{}:{}..{}:{}?coord_system_version={}&#39;</span>.format( chr, start, stop, strand, genome_version )</div><div class="line"><a name="l00788"></a><span class="lineno">  788</span>&#160;        </div><div class="line"><a name="l00789"></a><span class="lineno">  789</span>&#160;        <span class="comment"># Query the Ensembl server to get the sequence</span></div><div class="line"><a name="l00790"></a><span class="lineno">  790</span>&#160;        r = requests.get( server + ext, headers = { <span class="stringliteral">&#39;Content-Type&#39;</span> : <span class="stringliteral">&#39;text/plain&#39;</span> } )</div><div class="line"><a name="l00791"></a><span class="lineno">  791</span>&#160;        </div><div class="line"><a name="l00792"></a><span class="lineno">  792</span>&#160;        <span class="comment"># Return the sequence</span></div><div class="line"><a name="l00793"></a><span class="lineno">  793</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00794"></a><span class="lineno">  794</span>&#160;            r.ok == <span class="keyword">True</span></div><div class="line"><a name="l00795"></a><span class="lineno">  795</span>&#160;        <span class="keywordflow">except</span>:</div><div class="line"><a name="l00796"></a><span class="lineno">  796</span>&#160;            Logger.get_instance().error( <span class="stringliteral">&#39;ProcessingStrategy.download_seq_from_ensembl(): An error occured trying to get the sequence at&#39;</span> +</div><div class="line"><a name="l00797"></a><span class="lineno">  797</span>&#160;                                         <span class="stringliteral">&#39; chr&#39;</span> + str( chr ) + strand + <span class="stringliteral">&#39;:&#39;</span> + str( start ) + <span class="stringliteral">&#39;-&#39;</span> + str( stop ) + <span class="stringliteral">&#39;, on annotation version &#39;</span> + </div><div class="line"><a name="l00798"></a><span class="lineno">  798</span>&#160;                                         genome_version + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00799"></a><span class="lineno">  799</span>&#160;            <span class="keywordflow">return</span> <span class="keywordtype">None</span></div><div class="line"><a name="l00800"></a><span class="lineno">  800</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00801"></a><span class="lineno">  801</span>&#160;            <span class="keywordflow">return</span> r.text.upper()</div><div class="line"><a name="l00802"></a><span class="lineno">  802</span>&#160;        </div><div class="line"><a name="l00803"></a><span class="lineno">  803</span>&#160;        </div><div class="line"><a name="l00804"></a><span class="lineno">  804</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="acac09e2f267b2008382df8b031ceb252"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acac09e2f267b2008382df8b031ceb252">&#9670;&nbsp;</a></span>duplicate_dsorf_coordinates()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.duplicate_dsorf_coordinates </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dsorf</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>duplicate_dsorf_coordinates </h2>
<p>This method allows to copy all the genomic coordinates of a DSORF entry (performed when the DSORF has already been reported in the desired annotation version). It copies:</p><ul>
<li>raw_start_pos in start_pos</li>
<li>raw_stop_pos in stop_pos</li>
<li>raw_splice_starts in splice_starts</li>
<li>raw_splice_stops in splice_stops</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dsorf</td><td>DSORF - The DSORF entry for which the coordinates should be duplicated.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>dsorf: DSORF - The updated DSORF object. </dd></dl>
<div class="fragment"><div class="line"><a name="l00417"></a><span class="lineno">  417</span>&#160;    <span class="keyword">def </span>duplicate_dsorf_coordinates( dsorf ):</div><div class="line"><a name="l00418"></a><span class="lineno">  418</span>&#160; </div><div class="line"><a name="l00419"></a><span class="lineno">  419</span>&#160;        <span class="keywordflow">for</span> att <span class="keywordflow">in</span> ProcessingStrategy.ATT_TO_CONVERT_DSORF:</div><div class="line"><a name="l00420"></a><span class="lineno">  420</span>&#160;            setattr( dsorf, att, getattr( dsorf, <span class="stringliteral">&#39;raw_&#39;</span> + att ) )</div><div class="line"><a name="l00421"></a><span class="lineno">  421</span>&#160;            </div><div class="line"><a name="l00422"></a><span class="lineno">  422</span>&#160;        <span class="keywordflow">return</span> dsorf</div><div class="line"><a name="l00423"></a><span class="lineno">  423</span>&#160;    </div><div class="line"><a name="l00424"></a><span class="lineno">  424</span>&#160;    </div><div class="line"><a name="l00425"></a><span class="lineno">  425</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="a250e5000352003d9f3a3862548fa8862"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a250e5000352003d9f3a3862548fa8862">&#9670;&nbsp;</a></span>duplicate_dstranscript_coordinates()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.duplicate_dstranscript_coordinates </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dstranscript</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>duplicate_dstranscript_coordinates </h2>
<p>This method allows to copy all the genomic coordinates of a DSTranscript entry (performed when the transcript has already been reported in the desired annotation version). It copies:</p><ul>
<li>raw_start_pos in start_pos</li>
<li>raw_end_pos in end_pos</li>
<li>raw_cds_start_pos in cds_start_pos</li>
<li>raw_cds_stop_pos in cds_stop_pos</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dstranscript</td><td>DSTranscript - The DSTranscript entry for which the coordinates should be duplicated.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>dstranscript: DSTranscript - The updated DSTranscript object. </dd></dl>
<div class="fragment"><div class="line"><a name="l00510"></a><span class="lineno">  510</span>&#160;    <span class="keyword">def </span>duplicate_dstranscript_coordinates( dstranscript ):</div><div class="line"><a name="l00511"></a><span class="lineno">  511</span>&#160; </div><div class="line"><a name="l00512"></a><span class="lineno">  512</span>&#160;        <span class="keywordflow">for</span> att <span class="keywordflow">in</span> ProcessingStrategy.ATT_TO_CONVERT_DSTRANSCRIPT:</div><div class="line"><a name="l00513"></a><span class="lineno">  513</span>&#160;            setattr( dstranscript, att, getattr( dstranscript, <span class="stringliteral">&#39;raw_&#39;</span> + att ) )</div><div class="line"><a name="l00514"></a><span class="lineno">  514</span>&#160;            </div><div class="line"><a name="l00515"></a><span class="lineno">  515</span>&#160;        <span class="keywordflow">return</span> dstranscript</div><div class="line"><a name="l00516"></a><span class="lineno">  516</span>&#160;    </div><div class="line"><a name="l00517"></a><span class="lineno">  517</span>&#160;    </div><div class="line"><a name="l00518"></a><span class="lineno">  518</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<a id="aa2a8167a66e8bf24e600fae71281674a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa2a8167a66e8bf24e600fae71281674a">&#9670;&nbsp;</a></span>execute()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.execute </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>execute </h2>
<p>Execute the strategy to complete missing information and process data. </p>
<div class="fragment"><div class="line"><a name="l00078"></a><span class="lineno">   78</span>&#160;    <span class="keyword">def </span>execute( self ):</div><div class="line"><a name="l00079"></a><span class="lineno">   79</span>&#160;        </div><div class="line"><a name="l00080"></a><span class="lineno">   80</span>&#160;        <span class="comment"># Run DatabaseCheck in order to check the database is reachable prior to the processing of data</span></div><div class="line"><a name="l00081"></a><span class="lineno">   81</span>&#160;        <span class="comment"># If the database does not yet exists, it will be created                                                                                                # TODO: to change?!</span></div><div class="line"><a name="l00082"></a><span class="lineno">   82</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00083"></a><span class="lineno">   83</span>&#160;            Logger.get_instance().info( <span class="stringliteral">&#39;Checking the database prior to data processing...&#39;</span> )</div><div class="line"><a name="l00084"></a><span class="lineno">   84</span>&#160;            DatabaseCheckStrategy().execute()</div><div class="line"><a name="l00085"></a><span class="lineno">   85</span>&#160;        <span class="keywordflow">except</span> DenCellORFException <span class="keyword">as</span> e:</div><div class="line"><a name="l00086"></a><span class="lineno">   86</span>&#160;            Logger.get_instance().error( <span class="stringliteral">&#39;ProcessingStrategy.execute(): Error occurred whilst checking the database prior to data processing.&#39;</span> + e.get_message() )</div><div class="line"><a name="l00087"></a><span class="lineno">   87</span>&#160;        <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00088"></a><span class="lineno">   88</span>&#160;            Logger.get_instance().error( <span class="stringliteral">&#39;ProcessingStrategy.execute(): Error occurred whilst checking the database prior to data processing.&#39;</span> + str( e ) )</div><div class="line"><a name="l00089"></a><span class="lineno">   89</span>&#160;        </div><div class="line"><a name="l00090"></a><span class="lineno">   90</span>&#160;        <span class="comment"># Get the species in the database</span></div><div class="line"><a name="l00091"></a><span class="lineno">   91</span>&#160;        self.species = DataManager.get_instance().get_data( <span class="stringliteral">&#39;species_short&#39;</span> )</div><div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160;        </div><div class="line"><a name="l00093"></a><span class="lineno">   93</span>&#160;        <span class="comment"># Get all the data sources contained in the database</span></div><div class="line"><a name="l00094"></a><span class="lineno">   94</span>&#160;        DataManager.get_instance().store_query_result( <span class="stringliteral">&#39;datasources&#39;</span>, <span class="stringliteral">&#39;query( DataSource ).all()&#39;</span> )        </div><div class="line"><a name="l00095"></a><span class="lineno">   95</span>&#160;        datasources = DataManager.get_instance().get_data( <span class="stringliteral">&#39;datasources&#39;</span> )</div><div class="line"><a name="l00096"></a><span class="lineno">   96</span>&#160;        </div><div class="line"><a name="l00097"></a><span class="lineno">   97</span>&#160;        <span class="comment"># Build a dictionary which associates each data source to its annotation version</span></div><div class="line"><a name="l00098"></a><span class="lineno">   98</span>&#160;        DataManager.get_instance().store_data( <span class="stringliteral">&#39;datasource_annot&#39;</span>, {} )</div><div class="line"><a name="l00099"></a><span class="lineno">   99</span>&#160;        datasource_annot = DataManager.get_instance().get_data( <span class="stringliteral">&#39;datasource_annot&#39;</span> )</div><div class="line"><a name="l00100"></a><span class="lineno">  100</span>&#160;        <span class="keywordflow">for</span> ds <span class="keywordflow">in</span> datasources.keys():</div><div class="line"><a name="l00101"></a><span class="lineno">  101</span>&#160;            datasource_annot[ ds.name ] = ds.annotation_version</div><div class="line"><a name="l00102"></a><span class="lineno">  102</span>&#160;        </div><div class="line"><a name="l00103"></a><span class="lineno">  103</span>&#160;        <span class="comment"># Save the name of the annotation currently used in the metadata table</span></div><div class="line"><a name="l00104"></a><span class="lineno">  104</span>&#160;        metadata_current_annot = Metadata( parameter = Constants.METATABLE_CURRENT_ANNOTATION,</div><div class="line"><a name="l00105"></a><span class="lineno">  105</span>&#160;                                           value = Constants.CURRENT_NCBI_ANNOTATION[ self.species ] )</div><div class="line"><a name="l00106"></a><span class="lineno">  106</span>&#160;          <span class="comment"># Make sure the entry has not already been saved in the Metadata table</span></div><div class="line"><a name="l00107"></a><span class="lineno">  107</span>&#160;        metadata_table_content = SQLManager.get_instance().get_session().query( Metadata ).all()</div><div class="line"><a name="l00108"></a><span class="lineno">  108</span>&#160;        <span class="keywordflow">if</span> metadata_current_annot <span class="keywordflow">not</span> <span class="keywordflow">in</span> metadata_table_content:</div><div class="line"><a name="l00109"></a><span class="lineno">  109</span>&#160;            <span class="keywordflow">try</span>:</div><div class="line"><a name="l00110"></a><span class="lineno">  110</span>&#160;                SQLManager.get_instance().get_session().add( metadata_current_annot )</div><div class="line"><a name="l00111"></a><span class="lineno">  111</span>&#160;                SQLManager.get_instance().get_session().commit()</div><div class="line"><a name="l00112"></a><span class="lineno">  112</span>&#160;            <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00113"></a><span class="lineno">  113</span>&#160;                <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;ProcessingStrategy.execute(): Error occurred trying to add &#39;</span> + metadata_current_annot + </div><div class="line"><a name="l00114"></a><span class="lineno">  114</span>&#160;                                           <span class="stringliteral">&#39; and commit to the session.&#39;</span>, e )</div><div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;            </div><div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;        <span class="comment"># Convert the genomic coordinates to the current genome assembly</span></div><div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;        self.convert_genomic_coordinates()</div><div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;    </div><div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;        <span class="comment"># Check the sequences stored in the DSORFTranscriptAsso table</span></div><div class="line"><a name="l00120"></a><span class="lineno">  120</span>&#160;        <span class="comment"># agree with each other</span></div><div class="line"><a name="l00121"></a><span class="lineno">  121</span>&#160;        self.check_dsorftrasso_sequences_agreement()</div><div class="line"><a name="l00122"></a><span class="lineno">  122</span>&#160;    </div><div class="line"><a name="l00123"></a><span class="lineno">  123</span>&#160;        <span class="comment"># Download missing information</span></div><div class="line"><a name="l00124"></a><span class="lineno">  124</span>&#160;        self.download_missing_sequences()</div><div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;        </div><div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;        <span class="comment"># Compute missing information</span></div><div class="line"><a name="l00127"></a><span class="lineno">  127</span>&#160;    </div><div class="line"><a name="l00128"></a><span class="lineno">  128</span>&#160;    </div><div class="line"><a name="l00129"></a><span class="lineno">  129</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a619dba489edb08f8f896abd8bd68a712"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a619dba489edb08f8f896abd8bd68a712">&#9670;&nbsp;</a></span>CHAIN_FILES_URL</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">dictionary fr.tagc.uorf.core.execution.ProcessingStrategy.ProcessingStrategy.CHAIN_FILES_URL</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">=  { Constants.ANNOTATION_VERSION_HG19 : <span class="stringliteral">&#39;http://hgdownload.cse.ucsc.edu/goldenpath/hg19/liftOver/hg19ToHg38.over.chain.gz&#39;</span>,</div><div class="line">                        Constants.ANNOTATION_VERSION_GRCH37 : <span class="stringliteral">&#39;http://hgdownload.cse.ucsc.edu/goldenpath/hg19/liftOver/hg19ToHg38.over.chain.gz&#39;</span> }</div></div><!-- fragment --><h2>Class variables </h2>
<p>URL of chain files (necessary for the conversion of genomic positions) </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/volume1/choteau/workspace/tagc-uorf/step1/src/fr/tagc/uorf/core/execution/ProcessingStrategy.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
