<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>execution</b></li><li class="navelem"><b>insertion</b></li><li class="navelem"><b>parserstrategy</b></li><li class="navelem"><b>common</b></li><li class="navelem"><b>sORFs_org</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html">sORFs_org</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Source: sORFs.org.  
 <a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.png" usemap="#fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_5Forg.sORFs_5Forg_map" alt=""/>
  <map id="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_5Forg.sORFs_5Forg_map" name="fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_5Forg.sORFs_5Forg_map">
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html" alt="fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy" shape="rect" coords="290,0,861,24"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1sORFs__org__Human_1_1sORFs__org__Human.html" title="Source: sORFs.org." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.human.sORFs_org_Human.sORFs_org_Human" shape="rect" coords="0,112,571,136"/>
<area href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1mouse_1_1sORFs__org__Mouse_1_1sORFs__org__Mouse.html" title="Source: sORFs.org." alt="fr.tagc.uorf.core.execution.insertion.parserstrategy.mouse.sORFs_org_Mouse.sORFs_org_Mouse" shape="rect" coords="581,112,1152,136"/>
  </map>
</div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a35fbfff77c3823848928613b72dde15d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html#a35fbfff77c3823848928613b72dde15d">__init__</a> (self, data_path)</td></tr>
<tr class="separator:a35fbfff77c3823848928613b72dde15d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac52da5716a05346fe6811a3de0d6d858"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html#ac52da5716a05346fe6811a3de0d6d858">parse_file_content</a> (self)</td></tr>
<tr class="separator:ac52da5716a05346fe6811a3de0d6d858"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1591cc0613ec664ba31a39310fb34c42"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html#a1591cc0613ec664ba31a39310fb34c42">execute</a> (self)</td></tr>
<tr class="separator:a1591cc0613ec664ba31a39310fb34c42"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td colspan="2" onclick="javascript:toggleInherit('pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy')"><img src="closed.png" alt="-"/>&#160;Public Member Functions inherited from <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html">fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy</a></td></tr>
<tr class="memitem:a8cdab4b401f878c517c4b737b9ecf778 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a8cdab4b401f878c517c4b737b9ecf778">execute</a> ()</td></tr>
<tr class="separator:a8cdab4b401f878c517c4b737b9ecf778 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11654f0533e3a163aa528c8f996c3456 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a11654f0533e3a163aa528c8f996c3456">import_file_content</a> (self, file_sep=',', dtype_val='str', encoding_val='utf-8')</td></tr>
<tr class="separator:a11654f0533e3a163aa528c8f996c3456 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb455bdb96149592f7251e5f224c7fa3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#abb455bdb96149592f7251e5f224c7fa3">check_headers</a> (self)</td></tr>
<tr class="separator:abb455bdb96149592f7251e5f224c7fa3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0b04e0792f45b98d4111aa27c51ac93 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ab0b04e0792f45b98d4111aa27c51ac93">initialize_ids</a> (self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, orf_stop, orf_strand, index, log_symb_from_tr=True, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:ab0b04e0792f45b98d4111aa27c51ac93 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa01883a5db8dba8aabdb8af61d3b1f94 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#aa01883a5db8dba8aabdb8af61d3b1f94">initialize_transcript_id</a> (self, transcript_id, orf_id, index)</td></tr>
<tr class="separator:aa01883a5db8dba8aabdb8af61d3b1f94 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a07cdd23100d43df77f4001cd81fba324 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a07cdd23100d43df77f4001cd81fba324">initialize_gene_symbol</a> (self, ensembl_db, orf_id, chr_name, symbol, transcript_id, orf_start, orf_stop, orf_strand, log_symb_from_tr=True, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:a07cdd23100d43df77f4001cd81fba324 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d5bdcc0943ba6c1c71bb3c245bc81b3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4d5bdcc0943ba6c1c71bb3c245bc81b3">symbol_from_transcript_id</a> (self, ensembl_db, orf_id, transcript_id, log_warning=True)</td></tr>
<tr class="separator:a4d5bdcc0943ba6c1c71bb3c245bc81b3 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac199b748e002f7651a7ad708197c0008 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#ac199b748e002f7651a7ad708197c0008">symbol_from_orf_loc</a> (self, ensembl_db, orf_id, chr_name, orf_start, orf_stop, orf_strand, log_single_symb_from_pos=True)</td></tr>
<tr class="separator:ac199b748e002f7651a7ad708197c0008 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4251bb5eb9243ff185efba305843d86b inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a4251bb5eb9243ff185efba305843d86b">get_gene_from_symbol</a> (self, symbol, chr_name, orf_id)</td></tr>
<tr class="separator:a4251bb5eb9243ff185efba305843d86b inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a10fea64161666d9bb47ee4b8553b52e5 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a10fea64161666d9bb47ee4b8553b52e5">find_gene_from_symbol</a> (self, symbol, chr_name)</td></tr>
<tr class="separator:a10fea64161666d9bb47ee4b8553b52e5 inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9d3195a3bb1fd975359b771ebee910ef inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a9d3195a3bb1fd975359b771ebee910ef">get_gene_from_conflict_id</a> (self, conflict_symbol, chr_name)</td></tr>
<tr class="separator:a9d3195a3bb1fd975359b771ebee910ef inherit pub_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:ad35950e4eecbe3a6ac244a108cb79ee6"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html#ad35950e4eecbe3a6ac244a108cb79ee6">rm_stop_from_seq</a> (aa_sequence)</td></tr>
<tr class="separator:ad35950e4eecbe3a6ac244a108cb79ee6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td colspan="2" onclick="javascript:toggleInherit('pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy')"><img src="closed.png" alt="-"/>&#160;Static Public Member Functions inherited from <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html">fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy</a></td></tr>
<tr class="memitem:a602be6456b764340546d4ff9fd07d19f inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a602be6456b764340546d4ff9fd07d19f">is_empty</a> (val, empty_val=Constants.EMPTY_VALUES)</td></tr>
<tr class="separator:a602be6456b764340546d4ff9fd07d19f inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a846efdd59ec331984a46aa80a46afcf2 inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a846efdd59ec331984a46aa80a46afcf2">get_ensembl_db</a> (annotation_version)</td></tr>
<tr class="separator:a846efdd59ec331984a46aa80a46afcf2 inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a19c8d87c05bf4071bb6c7b541106a12b inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a19c8d87c05bf4071bb6c7b541106a12b">initialize_chromosome_name</a> (chr_name)</td></tr>
<tr class="separator:a19c8d87c05bf4071bb6c7b541106a12b inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a37ad31eeda67bd00e0aba6bfe491eecf inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a37ad31eeda67bd00e0aba6bfe491eecf">get_uniq_gene_on_chromosome</a> (chr, list_of_genes)</td></tr>
<tr class="separator:a37ad31eeda67bd00e0aba6bfe491eecf inherit pub_static_methods_classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a612dc7af609862589f6de0806d5c0b77"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html#a612dc7af609862589f6de0806d5c0b77">DOI</a> = '10.1093/nar/gkx1130'</td></tr>
<tr class="separator:a612dc7af609862589f6de0806d5c0b77"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Source: sORFs.org. </p>
<p>NB: This file contains a header. This file is expected to be in a tab-delimited format. More information about each column may be found at: <a href="http://sorfs.org/biomart_manual">http://sorfs.org/biomart_manual</a>.</p>
<p>Description of the columns of the raw data file: Extensive information about each feature may be found at: <a href="http://sorfs.org/biomart_manual">http://sorfs.org/biomart_manual</a> NB: Columns described here are the one used for the M. musculus database. Column names used in the other files are described in the child parsers.</p><ul>
<li>[0]: Sorf ID: Unique ORF ID - String ORF ID formatted as: Data source name + ":" + unique integer</li>
<li>[1]: Cell line: The name of the data source - String</li>
<li>[2]: Chromosome: The chromosome name - String</li>
<li>[3]: Strand: Strand - String</li>
<li>[4]: Sorf start: Genomic start coordinate of the ORF - Integer</li>
<li>[5]: Sorf end: Genomic end coordinate of the ORF - Integer</li>
<li>[6]: Spliced: Is the ORF derived from splice variation of a mRNA transcript? - String NO = Canonical mRNA transcripts YES = Splice variation of these mRNAs transcript NA = No mRNA transcript information available</li>
<li>[7]: Spliced start parts: Spliced start positions - String When the ORF is annotated on spliced mRNA variants, its genomic positions contain multiple starts and stops. In such case, the coordinates are separated by an underscore ("_") and the values of start and stop positions are provided in the same order.</li>
<li>[8]: Spliced stop parts: SPliced stop positions - String See comment for "Spliced start parts"</li>
<li>[9]: Sorf length: ORF length (in amino acids, include the stop codon) - Integer</li>
<li>[10]: Start codon: Start codon sequence - String</li>
<li>[11]: Downstream gene distance: Genomic distance to the closest downstream located gene (only for intergenic genes) - Integer</li>
<li>[12]: Upstream gene distance: Genomic distance to the closest upstream located gene (only for intergenic genes) - Integer</li>
<li>[13]: Transcript sequence: ORF nucleotide sequence - String</li>
<li>[14]: AA-sequence: ORF Amino acids sequence - String</li>
<li>[15]: Mass: Hypothetic mass for the resulting micropeptide translated from the ORF - Float</li>
<li>[16]: Annotation: Annotation based on the location of the ORF relative to Ensembl mRNA annotation. NB: in the list: exonic, intronic, intergenic, ncRNA, 3UTR, 5UTR, sORF (correspond to Ensembl protein coding ORF &lt;= 100 AA), NMD (Nonsense mediated decay), NSD (Non stop decay)</li>
<li>[17]: Biotype: Based on Ensembl annotation system - String NB: in the list: protein_coding, pseudogene, long_noncoding, short_noncoding</li>
<li>[18]: RPKM: RPKM, Represents the ribosomal density across the ORF - Float</li>
<li>[19]: RPF coverage: Percentage of nucleotides covered by RPFs - Float</li>
<li>[20]: RPF coverage uniformity: Expresses how uniform the ribosome footprints are distributed over the ORF sequence - Float</li>
<li>[21]: Exon overlap: Amount of overlap between the ORF and exon regions of annotated protein coding sequence - Float NB: Ranges between 0 (completetly outside exonic regions) and 1 (completely overlapping with exonic regions)</li>
<li>[22]: In frame coverage: ORF score, percentagee of nucleotides covered by in-frame situated RPF - Float NB: RPF lengths as defined in the ORFscore</li>
<li>[23]: In frame: In-frame with annotated protein coding sequences? - Boolean (String Yes/No) NB: ORFs located on annotated mRNA transcripts can be in-frame with annotated protein coding sequences on this mRNA (=Yes) or out-of-frame (=No). For ORFs located on mRNA transcripts without annotated protein coding sequence or ORFs not located on annotated mRNA transcripts the in-frame attribute cannot be computed (=NA).</li>
<li>[24]: overlap with PC exons: Does the ORF overlap exon regions of annotated protein coding sequences - Boolean (String Yes/No)</li>
<li>[25]: Ensembl transcript ID: Ensembl transcript ID - String</li>
<li>[26]: Rltm min Rchx: Difference between ribosome accumulation on TIS-candidates from LTM/HARR treated and CHX/EM treated RIBO-seq data - Float NB: Acts as a criterium in the TIS-calling algorithm.</li>
<li>[27]: FLOSS score: FLOSS score - Float NB: the FLOSS algorithm provides a score based on the comparison between the RPF-length distribution of the ORF and the RPF-length distribution of canonical protein-coding sequences.</li>
<li>[28]: Floss-Classification: Classification based on the FLOSS score - String NB: This classification represents the tendency of the ORF to be conding. In the list: Good, Extreme (extreme but still in cutoff range), Not in cutoff range.</li>
<li>[29]: Orfscore: ORF score - Float NB: The ORF score calculates the preference of RPFs to accumulate in the first frame of coding sequences.</li>
<li>[30]: Peak shift: Peak shift - String NB: Ribosome profiles are combined into peaks. When ribosome occupancy is detected on the first position of an ATG, or a near-cognate start codon, it is defined as a peak. However, if the +/- 1 position is an ATG near-cognate, this ribosome profile will also be defined as a peak and the position will change. As such, different profiles can be combined into a single peak. If this is the case, the ribosome profile hits are added up. This value represents the peak shifts. A peak shift value of 1/-1 indicates a near cognate start codon at the +/- 1 position, a peak shift value of 10/-10 means a near cognate start codon right before/after the annotated peak, a peak shift position of 0 indicates no ATG near cognate in the neighbourhood. The peak shift filter allow to select sORFs with a specific peak shift.</li>
<li>[31]: PhastCon: PhastCon score - Float NB: PhastCon examines evolutionary signatures characteristic to alignments of conserved coding region in order to determine whether a multi-species nucleotide sequence alignment is likely to represent a protein-coding region and provides a score based on this alignment.</li>
<li>[32]: PhyloP: PhyloP score - Float NB: PhyloP examines evolutionary signatures characteristic to alignments of conserved coding regions in order to determine whether a multi-species nucleotide sequence alignment is likely to represent a protein-coding region.</li>
<li><p class="startli">[33]: cutoff p-value: ??</p>
<p class="startli">Supplementary information</p><ul>
<li>The chromosome name is provided without 'chr' prefix.</li>
<li>The genomic coordinates are reported on GRCh38 (v90) or GRCm38 (v90) annotation.</li>
<li>For '+'-strand ORFs,<ul>
<li>the first nucleotide of the start codon is located at start position (Sorf start).</li>
<li>the last nucleotide of the stop codon is located at stop position (Sorf end). and, for each exon, [ e.g. when Spliced start parts = start1_start2 and Spliced stop parts = stop1_stop2 ]</li>
<li>the first nucleotide of the exon is located at the Spliced start parts position, in the same order (e.g. start1 for exon1, start2 for exon2)</li>
<li>the last nucleotide of the exon is located at the Spliced stop parts position, in the same order (e.g. stop1 for exon1, stop2 for exon2)</li>
</ul>
</li>
<li>For '-'-strand ORFs,<ul>
<li>the first nucleotide of the start codon is located at stop position (Sorf end).</li>
<li>the last nucleotide of the stop codon is located at start position (Sorf start). and, for each exon, [ e.g. when Spliced start parts = start1_start2 and Spliced stop parts = stop1_stop2 ]</li>
<li>the first nucleotide of the exon is located at the Spliced start parts position, in the same order (e.g. start1 for exon1, start2 for exon2)</li>
<li>the last nucleotide of the exon is located at the Spliced stop parts position, in the same order (e.g. stop1 for exon1, stop2 for exon2) /!\ NB: contrary to the stop and start position, the spliced positions are NOT reverted!</li>
</ul>
</li>
<li>The nucleic sequence (Transcript sequence) includes both the start and the stop codons.</li>
<li>The amino acid sequence (AA-sequence) includes both the start and the stop codons.</li>
<li>The ORF length (Sorf length) is in amino acids and includes both the start and the stop codons. <h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html" title="Source: sORFs.org.">sORFs_org</a> </h1>
</li>
</ul>
<p class="startli">This class inherits from ParserStrategy and is a metaclass aiming to parse data from the sORFs.org databases. </p>
</li>
</ul>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a35fbfff77c3823848928613b72dde15d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a35fbfff77c3823848928613b72dde15d">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>data_path</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>Constructor of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1common_1_1sORFs__org_1_1sORFs__org.html" title="Source: sORFs.org.">sORFs_org</a> </h2>

<p>Reimplemented from <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#af3a41c9649531a71f9ab74bf6c5f709a">fr.tagc.uorf.core.execution.insertion.ParserStrategy.ParserStrategy</a>.</p>

<p>Reimplemented in <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1mouse_1_1sORFs__org__Mouse_1_1sORFs__org__Mouse.html#a2641e96fe30067be4e62f2ffc6056fc0">fr.tagc.uorf.core.execution.insertion.parserstrategy.mouse.sORFs_org_Mouse.sORFs_org_Mouse</a>, and <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1sORFs__org__Human_1_1sORFs__org__Human.html#a2dfe4412ba42ed1fff25abe9cfff4555">fr.tagc.uorf.core.execution.insertion.parserstrategy.human.sORFs_org_Human.sORFs_org_Human</a>.</p>
<div class="fragment"><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;    <span class="keyword">def </span>__init__( self, data_path ):</div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;        </div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;        ParserStrategy.__init__( self, data_path )</div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;</div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;</div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a1591cc0613ec664ba31a39310fb34c42"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1591cc0613ec664ba31a39310fb34c42">&#9670;&nbsp;</a></span>execute()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org.execute </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>execute </h2>
<p>Log an error to inform the user that he/she has to used one of the child class to parse the data from sORF.org. The data will not be parsed using the current metaclass.</p>
<dl class="section return"><dt>Returns</dt><dd>None </dd></dl>

<p>Reimplemented in <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1mouse_1_1sORFs__org__Mouse_1_1sORFs__org__Mouse.html#a15dd68b5285e4a4d431ca9083fc54e74">fr.tagc.uorf.core.execution.insertion.parserstrategy.mouse.sORFs_org_Mouse.sORFs_org_Mouse</a>, and <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1parserstrategy_1_1human_1_1sORFs__org__Human_1_1sORFs__org__Human.html#a5b5bf523f0bca1d9c5ab778160c33da9">fr.tagc.uorf.core.execution.insertion.parserstrategy.human.sORFs_org_Human.sORFs_org_Human</a>.</p>
<div class="fragment"><div class="line"><a name="l00563"></a><span class="lineno">  563</span>&#160;    <span class="keyword">def </span>execute(self):</div><div class="line"><a name="l00564"></a><span class="lineno">  564</span>&#160;    </div><div class="line"><a name="l00565"></a><span class="lineno">  565</span>&#160;        Logger.get_instance().error( <span class="stringliteral">&#39;Several parsers inheriting from sORF_org are existing to parse data&#39;</span> +</div><div class="line"><a name="l00566"></a><span class="lineno">  566</span>&#160;                                     <span class="stringliteral">&#39; from this source (one for each species). Please select the&#39;</span> +</div><div class="line"><a name="l00567"></a><span class="lineno">  567</span>&#160;                                     <span class="stringliteral">&#39; appropriate one (in the config file) to insert data in the database.&#39;</span>,</div><div class="line"><a name="l00568"></a><span class="lineno">  568</span>&#160;                                     ex = <span class="keyword">False</span> )</div><div class="line"><a name="l00569"></a><span class="lineno">  569</span>&#160;        </div><div class="line"><a name="l00570"></a><span class="lineno">  570</span>&#160;        <span class="keywordflow">return</span> <span class="keywordtype">None</span></div></div><!-- fragment -->
</div>
</div>
<a id="ac52da5716a05346fe6811a3de0d6d858"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac52da5716a05346fe6811a3de0d6d858">&#9670;&nbsp;</a></span>parse_file_content()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org.parse_file_content </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>parse_file_content </h2>
<p>Parse the content of the file and return the list of object that have to be inserted in the database. See the <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1insertion_1_1ParserStrategy_1_1ParserStrategy.html#a8cdab4b401f878c517c4b737b9ecf778">execute()</a> method documentation of the ParserStrategy metaclass for more information about the main steps of the algorithm run in this method.</p>
<dl class="section return"><dt>Returns</dt><dd>objects_to_insert: List - List of objects to insert in the database.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">DenCellORFException</td><td>When this parser cannot be used for the species of the database. </td></tr>
    <tr><td class="paramname">DenCellORFException</td><td>When an exception has been raised checking the headers of the file. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;    <span class="keyword">def </span>parse_file_content( self ):</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;        </div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;        <span class="comment"># Get the content of the file</span></div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;        self.import_file_content( file_sep = <span class="stringliteral">&#39;\t&#39;</span> )</div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;        </div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;        <span class="comment"># Check the headers of the file</span></div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;            self.check_headers()</div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;        <span class="keywordflow">except</span> DuplicateHeaderException <span class="keyword">as</span> e:</div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;            Logger.get_instance().warning( e.get_message() +</div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                                           <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_HEADER_DUPL + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                </div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;        <span class="comment"># Assign new variables to access some objects of the DataManager data dictionary</span></div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;        all_dsorfs = DataManager.get_instance().get_data( Constants.DM_ALL_DSORFS_FOR_SOURCE )</div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;        all_dstranscripts = DataManager.get_instance().get_data( Constants.DM_ALL_DSTRANSCRIPTS_FOR_SOURCE )</div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;        all_dsota = DataManager.get_instance().get_data( Constants.DM_ALL_DSORFTRANSCRIPTASSO_FOR_SOURCE ) </div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;        all_utdstranscriptgeneconflict = DataManager.get_instance().get_data( Constants.DM_ALL_UTDSTRANSCRIPTGENECONFLICT )</div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;        </div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;        autoincrement = DataManager.get_instance().get_data( Constants.DM_AUTOINCREMENT )</div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;        </div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;        <span class="comment"># Get the appropriate Ensembl database to recover missing gene IDs</span></div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;        ensembl_db = ParserStrategy.get_ensembl_db( self.ENSEMBL_RELEASE )</div><div class="line"><a name="l00245"></a><span class="lineno">  245</span>&#160;        </div><div class="line"><a name="l00246"></a><span class="lineno">  246</span>&#160;        <span class="comment"># Get the number of rows expected to be treated and reset the ProgressionBar instance</span></div><div class="line"><a name="l00247"></a><span class="lineno">  247</span>&#160;        ProgressionBar.get_instance().reset_instance( total = self.file_content.shape[ 0 ] )</div><div class="line"><a name="l00248"></a><span class="lineno">  248</span>&#160;        </div><div class="line"><a name="l00249"></a><span class="lineno">  249</span>&#160;        <span class="comment"># Initialize the list of objects to insert</span></div><div class="line"><a name="l00250"></a><span class="lineno">  250</span>&#160;        objects_to_insert = []</div><div class="line"><a name="l00251"></a><span class="lineno">  251</span>&#160;</div><div class="line"><a name="l00252"></a><span class="lineno">  252</span>&#160;</div><div class="line"><a name="l00253"></a><span class="lineno">  253</span>&#160;        </div><div class="line"><a name="l00255"></a><span class="lineno">  255</span>&#160;        datasource = DataSource( name = self.DATA_SOURCE,</div><div class="line"><a name="l00256"></a><span class="lineno">  256</span>&#160;                                 doi = sORFs_org.DOI,</div><div class="line"><a name="l00257"></a><span class="lineno">  257</span>&#160;                                 description = self.SOURCE_DESCRIPTION,</div><div class="line"><a name="l00258"></a><span class="lineno">  258</span>&#160;                                 url = self.URL, </div><div class="line"><a name="l00259"></a><span class="lineno">  259</span>&#160;                                 annotation_version = self.ANNOTATION_VERSION,</div><div class="line"><a name="l00260"></a><span class="lineno">  260</span>&#160;                                 ensembl_release = self.ENSEMBL_RELEASE,</div><div class="line"><a name="l00261"></a><span class="lineno">  261</span>&#160;                                 annotation_description = self.ANNOTATION_DESCRIPTION )</div><div class="line"><a name="l00262"></a><span class="lineno">  262</span>&#160;        objects_to_insert.append( datasource )</div><div class="line"><a name="l00263"></a><span class="lineno">  263</span>&#160;        </div><div class="line"><a name="l00264"></a><span class="lineno">  264</span>&#160;        </div><div class="line"><a name="l00265"></a><span class="lineno">  265</span>&#160;        <span class="comment"># Parse the content of the file and save it in appropriate objects</span></div><div class="line"><a name="l00266"></a><span class="lineno">  266</span>&#160;        start_time = datetime.datetime.now()</div><div class="line"><a name="l00267"></a><span class="lineno">  267</span>&#160;        <span class="keywordflow">for</span> ( index, row ) <span class="keywordflow">in</span> self.file_content.iterrows():</div><div class="line"><a name="l00268"></a><span class="lineno">  268</span>&#160;            </div><div class="line"><a name="l00269"></a><span class="lineno">  269</span>&#160;            <span class="comment"># Update and display the progression bar on the console</span></div><div class="line"><a name="l00270"></a><span class="lineno">  270</span>&#160;            ProgressionBar.get_instance().increase_and_display()</div><div class="line"><a name="l00271"></a><span class="lineno">  271</span>&#160;            </div><div class="line"><a name="l00272"></a><span class="lineno">  272</span>&#160;            <span class="comment"># Assign useful information to new variables</span></div><div class="line"><a name="l00273"></a><span class="lineno">  273</span>&#160;            chr_name = row[ self.HEADER_CHROMOSOME ]</div><div class="line"><a name="l00274"></a><span class="lineno">  274</span>&#160;            orf_id = row[ self.HEADER_ORF_ID ]</div><div class="line"><a name="l00275"></a><span class="lineno">  275</span>&#160;            transcript_id = row[ self.HEADER_TRANSCRIPT_ID ]</div><div class="line"><a name="l00276"></a><span class="lineno">  276</span>&#160;            orf_annotation = row[ self.HEADER_ORF_ANNOTATION ]</div><div class="line"><a name="l00277"></a><span class="lineno">  277</span>&#160;            </div><div class="line"><a name="l00278"></a><span class="lineno">  278</span>&#160;            <span class="keywordflow">if</span> int(row[self.HEADER_STRAND]) == 1:</div><div class="line"><a name="l00279"></a><span class="lineno">  279</span>&#160;                orf_strand = <span class="stringliteral">&quot;+&quot;</span></div><div class="line"><a name="l00280"></a><span class="lineno">  280</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00281"></a><span class="lineno">  281</span>&#160;                orf_strand = <span class="stringliteral">&quot;-&quot;</span></div><div class="line"><a name="l00282"></a><span class="lineno">  282</span>&#160;                </div><div class="line"><a name="l00283"></a><span class="lineno">  283</span>&#160;            orf_start = row[ self.HEADER_START_POS ]</div><div class="line"><a name="l00284"></a><span class="lineno">  284</span>&#160;            orf_stop = row[ self.HEADER_STOP_POS ]</div><div class="line"><a name="l00285"></a><span class="lineno">  285</span>&#160;            </div><div class="line"><a name="l00286"></a><span class="lineno">  286</span>&#160;            <span class="keywordflow">if</span> ( row[ self.HEADER_SPLICED ] == <span class="stringliteral">&#39;Yes&#39;</span> ):</div><div class="line"><a name="l00287"></a><span class="lineno">  287</span>&#160;                orf_spliced = <span class="keyword">True</span></div><div class="line"><a name="l00288"></a><span class="lineno">  288</span>&#160;            <span class="keywordflow">elif</span> ( row[ self.HEADER_SPLICED ] == <span class="stringliteral">&#39;No&#39;</span> ):</div><div class="line"><a name="l00289"></a><span class="lineno">  289</span>&#160;                orf_spliced = <span class="keyword">False</span></div><div class="line"><a name="l00290"></a><span class="lineno">  290</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00291"></a><span class="lineno">  291</span>&#160;                orf_spliced = <span class="keywordtype">None</span></div><div class="line"><a name="l00292"></a><span class="lineno">  292</span>&#160;                </div><div class="line"><a name="l00293"></a><span class="lineno">  293</span>&#160;            orf_splice_starts = row[ self.HEADER_SPLICE_STARTS ]</div><div class="line"><a name="l00294"></a><span class="lineno">  294</span>&#160;            orf_splice_ends = row[ self.HEADER_SPLICE_ENDS ]</div><div class="line"><a name="l00295"></a><span class="lineno">  295</span>&#160;            <span class="comment"># Get the number of &quot;exonic&quot; parts of the ORF</span></div><div class="line"><a name="l00296"></a><span class="lineno">  296</span>&#160;            <span class="keywordflow">if</span> orf_spliced:</div><div class="line"><a name="l00297"></a><span class="lineno">  297</span>&#160;                </div><div class="line"><a name="l00298"></a><span class="lineno">  298</span>&#160;                <span class="keywordflow">if</span> ( ( orf_splice_starts != <span class="keywordtype">None</span> ) <span class="keywordflow">and</span> ( orf_splice_ends != <span class="keywordtype">None</span> ) ):</div><div class="line"><a name="l00299"></a><span class="lineno">  299</span>&#160;                    splice_starts_count = len( orf_splice_starts.split( <span class="stringliteral">&#39;_&#39;</span> ) )</div><div class="line"><a name="l00300"></a><span class="lineno">  300</span>&#160;                    splice_ends_count = len( orf_splice_ends.split( <span class="stringliteral">&#39;_&#39;</span> ) )</div><div class="line"><a name="l00301"></a><span class="lineno">  301</span>&#160;                    </div><div class="line"><a name="l00302"></a><span class="lineno">  302</span>&#160;                    <span class="comment"># If the number of starts is different from the number of ends, raise an exception</span></div><div class="line"><a name="l00303"></a><span class="lineno">  303</span>&#160;                    <span class="keywordflow">if</span> ( splice_starts_count != splice_ends_count ):</div><div class="line"><a name="l00304"></a><span class="lineno">  304</span>&#160;                        <span class="keywordflow">raise</span> DenCellORFException( self.data_source + <span class="stringliteral">&#39;.execute():&#39;</span> +</div><div class="line"><a name="l00305"></a><span class="lineno">  305</span>&#160;                                                   <span class="stringliteral">&#39; The number of starts for exonic parts (&#39;</span> + </div><div class="line"><a name="l00306"></a><span class="lineno">  306</span>&#160;                                                   str( splice_starts_count ) + </div><div class="line"><a name="l00307"></a><span class="lineno">  307</span>&#160;                                                   <span class="stringliteral">&#39;) does not equal the number of ends for exonic parts (&#39;</span> +</div><div class="line"><a name="l00308"></a><span class="lineno">  308</span>&#160;                                                   str( splice_ends_count ) + </div><div class="line"><a name="l00309"></a><span class="lineno">  309</span>&#160;                                                   <span class="stringliteral">&#39;) for the ORF with original ID &quot;&#39;</span> + orf_id + <span class="stringliteral">&#39;&quot;.&#39;</span> )</div><div class="line"><a name="l00310"></a><span class="lineno">  310</span>&#160;                        spliced_parts_count = <span class="keywordtype">None</span></div><div class="line"><a name="l00311"></a><span class="lineno">  311</span>&#160;                        </div><div class="line"><a name="l00312"></a><span class="lineno">  312</span>&#160;                    <span class="keywordflow">else</span>:</div><div class="line"><a name="l00313"></a><span class="lineno">  313</span>&#160;                        spliced_parts_count = splice_starts_count</div><div class="line"><a name="l00314"></a><span class="lineno">  314</span>&#160;                    </div><div class="line"><a name="l00315"></a><span class="lineno">  315</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00316"></a><span class="lineno">  316</span>&#160;                    Logger.get_instance().warning( <span class="stringliteral">&#39;The ORF with original ID &quot;&#39;</span> + orf_id + </div><div class="line"><a name="l00317"></a><span class="lineno">  317</span>&#160;                                                   <span class="stringliteral">&#39;&quot; is reported to be spliced in &#39;</span> + self.data_source +</div><div class="line"><a name="l00318"></a><span class="lineno">  318</span>&#160;                                                   <span class="stringliteral">&#39; but no location of start and ends of &quot;exonic&quot; parts are provided.&#39;</span> +</div><div class="line"><a name="l00319"></a><span class="lineno">  319</span>&#160;                                                   <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_SPLIC_MISS + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00320"></a><span class="lineno">  320</span>&#160;                    spliced_parts_count = <span class="keywordtype">None</span></div><div class="line"><a name="l00321"></a><span class="lineno">  321</span>&#160;                </div><div class="line"><a name="l00322"></a><span class="lineno">  322</span>&#160;                <span class="comment"># Replace too long strings of coordinates by &quot;TOO_LONG&quot;</span></div><div class="line"><a name="l00323"></a><span class="lineno">  323</span>&#160;                <span class="keywordflow">if</span> ( len( str( orf_splice_starts ) ) &gt; Constants.MAX_LEN_STRING ):</div><div class="line"><a name="l00324"></a><span class="lineno">  324</span>&#160;                    Logger.get_instance().warning( <span class="stringliteral">&#39;The splicing information for the ORF with original ID&quot;&#39;</span> + </div><div class="line"><a name="l00325"></a><span class="lineno">  325</span>&#160;                                                   orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source +</div><div class="line"><a name="l00326"></a><span class="lineno">  326</span>&#160;                                                   <span class="stringliteral">&#39; was too long to be stored in the database&#39;</span> +</div><div class="line"><a name="l00327"></a><span class="lineno">  327</span>&#160;                                                   <span class="stringliteral">&#39; (raw splice starts: &#39;</span> + orf_splice_starts + </div><div class="line"><a name="l00328"></a><span class="lineno">  328</span>&#160;                                                   <span class="stringliteral">&#39;, raw splice ends: &#39;</span> + orf_splice_ends + <span class="stringliteral">&#39;).&#39;</span></div><div class="line"><a name="l00329"></a><span class="lineno">  329</span>&#160;                                                   <span class="stringliteral">&#39; Hence these values will be changed for &quot;&#39;</span> + </div><div class="line"><a name="l00330"></a><span class="lineno">  330</span>&#160;                                                   Constants.REPLACE_TOO_LONG_STRINGS + <span class="stringliteral">&#39;&quot;.&#39;</span> +</div><div class="line"><a name="l00331"></a><span class="lineno">  331</span>&#160;                                                   <span class="stringliteral">&#39; Warning code: &#39;</span> + LogCodes.WARN_SPLIC_TOOLONG + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00332"></a><span class="lineno">  332</span>&#160;                    orf_splice_starts = Constants.REPLACE_TOO_LONG_STRINGS</div><div class="line"><a name="l00333"></a><span class="lineno">  333</span>&#160;                    orf_splice_ends = Constants.REPLACE_TOO_LONG_STRINGS</div><div class="line"><a name="l00334"></a><span class="lineno">  334</span>&#160;                    </div><div class="line"><a name="l00335"></a><span class="lineno">  335</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00336"></a><span class="lineno">  336</span>&#160;                spliced_parts_count = 1</div><div class="line"><a name="l00337"></a><span class="lineno">  337</span>&#160;            </div><div class="line"><a name="l00338"></a><span class="lineno">  338</span>&#160;            <span class="comment"># Reset the chr_name, symbol and transcript_id if necessary</span></div><div class="line"><a name="l00339"></a><span class="lineno">  339</span>&#160;            ( chr_name, symbol, transcript_id ) = self.initialize_ids( ensembl_db = ensembl_db,</div><div class="line"><a name="l00340"></a><span class="lineno">  340</span>&#160;                                                                       orf_id = orf_id,</div><div class="line"><a name="l00341"></a><span class="lineno">  341</span>&#160;                                                                       chr_name = chr_name,</div><div class="line"><a name="l00342"></a><span class="lineno">  342</span>&#160;                                                                       symbol = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00343"></a><span class="lineno">  343</span>&#160;                                                                       transcript_id = transcript_id, </div><div class="line"><a name="l00344"></a><span class="lineno">  344</span>&#160;                                                                       orf_start = orf_start,</div><div class="line"><a name="l00345"></a><span class="lineno">  345</span>&#160;                                                                       orf_stop = orf_stop,</div><div class="line"><a name="l00346"></a><span class="lineno">  346</span>&#160;                                                                       orf_strand = orf_strand,</div><div class="line"><a name="l00347"></a><span class="lineno">  347</span>&#160;                                                                       index = index,</div><div class="line"><a name="l00348"></a><span class="lineno">  348</span>&#160;                                                                       log_symb_from_tr = <span class="keyword">False</span> )</div><div class="line"><a name="l00349"></a><span class="lineno">  349</span>&#160;</div><div class="line"><a name="l00350"></a><span class="lineno">  350</span>&#160;</div><div class="line"><a name="l00351"></a><span class="lineno">  351</span>&#160;            </div><div class="line"><a name="l00353"></a><span class="lineno">  353</span>&#160;            dsorf = DSORF( id = autoincrement,</div><div class="line"><a name="l00354"></a><span class="lineno">  354</span>&#160;                           data_source = datasource.name,</div><div class="line"><a name="l00355"></a><span class="lineno">  355</span>&#160;                           chromosome = chr_name,</div><div class="line"><a name="l00356"></a><span class="lineno">  356</span>&#160;                           raw_strand = orf_strand,</div><div class="line"><a name="l00357"></a><span class="lineno">  357</span>&#160;                           raw_start_pos = orf_start,</div><div class="line"><a name="l00358"></a><span class="lineno">  358</span>&#160;                           raw_stop_pos = orf_stop,</div><div class="line"><a name="l00359"></a><span class="lineno">  359</span>&#160;                           spliced = orf_spliced,</div><div class="line"><a name="l00360"></a><span class="lineno">  360</span>&#160;                           raw_splice_starts = orf_splice_starts,</div><div class="line"><a name="l00361"></a><span class="lineno">  361</span>&#160;                           raw_splice_ends = orf_splice_ends,</div><div class="line"><a name="l00362"></a><span class="lineno">  362</span>&#160;                           spliced_parts_count = spliced_parts_count )</div><div class="line"><a name="l00363"></a><span class="lineno">  363</span>&#160;            </div><div class="line"><a name="l00364"></a><span class="lineno">  364</span>&#160;            <span class="comment"># If this ORF already exists, get the existing one</span></div><div class="line"><a name="l00365"></a><span class="lineno">  365</span>&#160;            existing_dsorf = all_dsorfs.get( dsorf )</div><div class="line"><a name="l00366"></a><span class="lineno">  366</span>&#160;            <span class="keywordflow">if</span> existing_dsorf:</div><div class="line"><a name="l00367"></a><span class="lineno">  367</span>&#160;                dsorf = existing_dsorf</div><div class="line"><a name="l00368"></a><span class="lineno">  368</span>&#160;                </div><div class="line"><a name="l00369"></a><span class="lineno">  369</span>&#160;            <span class="comment"># Otherwise, add the newly created ORF to the list of objects to insert and keep track of it</span></div><div class="line"><a name="l00370"></a><span class="lineno">  370</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00371"></a><span class="lineno">  371</span>&#160;                all_dsorfs[ dsorf ] = dsorf</div><div class="line"><a name="l00372"></a><span class="lineno">  372</span>&#160;                objects_to_insert.append( dsorf )</div><div class="line"><a name="l00373"></a><span class="lineno">  373</span>&#160;                </div><div class="line"><a name="l00374"></a><span class="lineno">  374</span>&#160;                <span class="comment"># Increase the autoincrement value for ID of 1</span></div><div class="line"><a name="l00375"></a><span class="lineno">  375</span>&#160;                autoincrement += 1</div><div class="line"><a name="l00376"></a><span class="lineno">  376</span>&#160;                </div><div class="line"><a name="l00377"></a><span class="lineno">  377</span>&#160;            </div><div class="line"><a name="l00378"></a><span class="lineno">  378</span>&#160;            </div><div class="line"><a name="l00381"></a><span class="lineno">  381</span>&#160;            ( gene, new_obj_to_insert ) = self.get_gene_from_symbol( orf_id = orf_id,</div><div class="line"><a name="l00382"></a><span class="lineno">  382</span>&#160;                                                                     symbol = symbol,</div><div class="line"><a name="l00383"></a><span class="lineno">  383</span>&#160;                                                                     chr_name = chr_name )</div><div class="line"><a name="l00384"></a><span class="lineno">  384</span>&#160;            objects_to_insert += new_obj_to_insert</div><div class="line"><a name="l00385"></a><span class="lineno">  385</span>&#160;            </div><div class="line"><a name="l00386"></a><span class="lineno">  386</span>&#160;            </div><div class="line"><a name="l00387"></a><span class="lineno">  387</span>&#160;            </div><div class="line"><a name="l00389"></a><span class="lineno">  389</span>&#160;            dstranscript = DSTranscript( id = autoincrement,</div><div class="line"><a name="l00390"></a><span class="lineno">  390</span>&#160;                                         transcript_id = transcript_id,</div><div class="line"><a name="l00391"></a><span class="lineno">  391</span>&#160;                                         data_source = datasource.name,</div><div class="line"><a name="l00392"></a><span class="lineno">  392</span>&#160;                                         gene_id = gene.gene_id,</div><div class="line"><a name="l00393"></a><span class="lineno">  393</span>&#160;                                         strand = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00394"></a><span class="lineno">  394</span>&#160;                                         raw_start_pos = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00395"></a><span class="lineno">  395</span>&#160;                                         raw_end_pos = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00396"></a><span class="lineno">  396</span>&#160;                                         raw_cds_start_pos = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00397"></a><span class="lineno">  397</span>&#160;                                         raw_cds_stop_pos = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00398"></a><span class="lineno">  398</span>&#160;                                         rna_biotype = row[ self.HEADER_ORF_BIOTYPE ] )</div><div class="line"><a name="l00399"></a><span class="lineno">  399</span>&#160;            </div><div class="line"><a name="l00400"></a><span class="lineno">  400</span>&#160;            <span class="comment"># If this transcript already exists, get the existing one</span></div><div class="line"><a name="l00401"></a><span class="lineno">  401</span>&#160;            existing_dstranscript = all_dstranscripts.get( dstranscript )</div><div class="line"><a name="l00402"></a><span class="lineno">  402</span>&#160;            <span class="keywordflow">if</span> existing_dstranscript:</div><div class="line"><a name="l00403"></a><span class="lineno">  403</span>&#160;                </div><div class="line"><a name="l00404"></a><span class="lineno">  404</span>&#160;                <span class="comment"># If a conflict about the genes associated with this transcript has already been found, </span></div><div class="line"><a name="l00405"></a><span class="lineno">  405</span>&#160;                <span class="comment"># get the UTDSTranscriptGeneConflict entry that keep record of it</span></div><div class="line"><a name="l00406"></a><span class="lineno">  406</span>&#160;                utdstranscriptgeneconflict = UTDSTranscriptGeneConflict( transcript_id = dstranscript.transcript_id,</div><div class="line"><a name="l00407"></a><span class="lineno">  407</span>&#160;                                                                         data_source = dstranscript.data_source )</div><div class="line"><a name="l00408"></a><span class="lineno">  408</span>&#160;                </div><div class="line"><a name="l00409"></a><span class="lineno">  409</span>&#160;                existing_utdstranscriptgeneconflict = all_utdstranscriptgeneconflict.get( utdstranscriptgeneconflict )</div><div class="line"><a name="l00410"></a><span class="lineno">  410</span>&#160;                <span class="keywordflow">if</span> existing_utdstranscriptgeneconflict:</div><div class="line"><a name="l00411"></a><span class="lineno">  411</span>&#160;                    <span class="comment"># Get the list of gene already reported as being associated with the transcript</span></div><div class="line"><a name="l00412"></a><span class="lineno">  412</span>&#160;                    gene_id_conflicts = existing_utdstranscriptgeneconflict.gene_ids.split( Constants.UTDSTRANSCRIPTGENECONFLICT_SEPARATOR_IDS )</div><div class="line"><a name="l00413"></a><span class="lineno">  413</span>&#160;                <span class="keywordflow">else</span>:</div><div class="line"><a name="l00414"></a><span class="lineno">  414</span>&#160;                    gene_id_conflicts = <span class="keywordtype">None</span></div><div class="line"><a name="l00415"></a><span class="lineno">  415</span>&#160;                </div><div class="line"><a name="l00416"></a><span class="lineno">  416</span>&#160;                <span class="comment"># Update the existing transcript using information from the current entry if necessary</span></div><div class="line"><a name="l00417"></a><span class="lineno">  417</span>&#160;                <span class="keywordflow">try</span>:</div><div class="line"><a name="l00418"></a><span class="lineno">  418</span>&#160;                    existing_dstranscript.update( dstranscript, gene_id_conflicts )</div><div class="line"><a name="l00419"></a><span class="lineno">  419</span>&#160;                    </div><div class="line"><a name="l00420"></a><span class="lineno">  420</span>&#160;                <span class="keywordflow">except</span> DenCellORFTrUpdateException <span class="keyword">as</span> e:</div><div class="line"><a name="l00421"></a><span class="lineno">  421</span>&#160;                    Logger.get_instance().error( <span class="stringliteral">&#39;The transcript &#39;</span> + transcript_id + <span class="stringliteral">&#39; (from &#39;</span> + self.data_source +</div><div class="line"><a name="l00422"></a><span class="lineno">  422</span>&#160;                                                 <span class="stringliteral">&#39;) has been found several time with different properties: &#39;</span> + e.get_message() +</div><div class="line"><a name="l00423"></a><span class="lineno">  423</span>&#160;                                                 <span class="stringliteral">&#39; Error code: &#39;</span> + LogCodes.ERR_DCORF_CONFL_DSTR + <span class="stringliteral">&#39;.&#39;</span>,</div><div class="line"><a name="l00424"></a><span class="lineno">  424</span>&#160;                                                 ex = <span class="keyword">False</span> )</div><div class="line"><a name="l00425"></a><span class="lineno">  425</span>&#160;                    </div><div class="line"><a name="l00426"></a><span class="lineno">  426</span>&#160;                    <span class="comment"># If a conflict has been found on the gene ID, associate this transcript to its new &quot;conflict&quot; gene ID</span></div><div class="line"><a name="l00427"></a><span class="lineno">  427</span>&#160;                    new_gene_id = e.get_gene_id()</div><div class="line"><a name="l00428"></a><span class="lineno">  428</span>&#160;                    </div><div class="line"><a name="l00429"></a><span class="lineno">  429</span>&#160;                    <span class="keywordflow">if</span> new_gene_id:</div><div class="line"><a name="l00430"></a><span class="lineno">  430</span>&#160;                        <span class="comment"># Get the lists of all the genes that have been found conflicting</span></div><div class="line"><a name="l00431"></a><span class="lineno">  431</span>&#160;                        <span class="keywordflow">if</span> e.get_conflict_list():</div><div class="line"><a name="l00432"></a><span class="lineno">  432</span>&#160;                            <span class="comment"># If an UTDSTranscriptGeneConflict entry was already existing, update the list of gene IDs,</span></div><div class="line"><a name="l00433"></a><span class="lineno">  433</span>&#160;                            <span class="comment"># Otherwise, add the new entry to the list of objects to insert</span></div><div class="line"><a name="l00434"></a><span class="lineno">  434</span>&#160;                            <span class="keywordflow">if</span> existing_utdstranscriptgeneconflict:</div><div class="line"><a name="l00435"></a><span class="lineno">  435</span>&#160;                                existing_utdstranscriptgeneconflict.gene_ids = Constants.UTDSTRANSCRIPTGENECONFLICT_SEPARATOR_IDS.join( e.get_conflict_list() )</div><div class="line"><a name="l00436"></a><span class="lineno">  436</span>&#160;                                utdstranscriptgeneconflict = existing_utdstranscriptgeneconflict</div><div class="line"><a name="l00437"></a><span class="lineno">  437</span>&#160;                            </div><div class="line"><a name="l00438"></a><span class="lineno">  438</span>&#160;                            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00439"></a><span class="lineno">  439</span>&#160;                                utdstranscriptgeneconflict.gene_ids = Constants.UTDSTRANSCRIPTGENECONFLICT_SEPARATOR_IDS.join( e.get_conflict_list() )</div><div class="line"><a name="l00440"></a><span class="lineno">  440</span>&#160;                                all_utdstranscriptgeneconflict[ utdstranscriptgeneconflict ] = utdstranscriptgeneconflict</div><div class="line"><a name="l00441"></a><span class="lineno">  441</span>&#160;                                objects_to_insert.append( utdstranscriptgeneconflict )</div><div class="line"><a name="l00442"></a><span class="lineno">  442</span>&#160;                            </div><div class="line"><a name="l00443"></a><span class="lineno">  443</span>&#160;                        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00444"></a><span class="lineno">  444</span>&#160;                            <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;sORFs_org.parse_file_content(): The gene ID associated with the DSTranscript with ID &quot;&#39;</span> +</div><div class="line"><a name="l00445"></a><span class="lineno">  445</span>&#160;                                                       existing_dstranscript.id + <span class="stringliteral">&#39;&quot; has been updated whilst the list of genes related to the&#39;</span> +</div><div class="line"><a name="l00446"></a><span class="lineno">  446</span>&#160;                                                       <span class="stringliteral">&#39; DSTranscript cannot be get. Please contact the developer if you see this message.&#39;</span> )</div><div class="line"><a name="l00447"></a><span class="lineno">  447</span>&#160;                    </div><div class="line"><a name="l00448"></a><span class="lineno">  448</span>&#160;                        <span class="comment"># Get the appropriate gene object from this new gene ID</span></div><div class="line"><a name="l00449"></a><span class="lineno">  449</span>&#160;                        ( gene, new_obj_to_insert ) = self.get_gene_from_conflict_id( conflict_symbol = new_gene_id,</div><div class="line"><a name="l00450"></a><span class="lineno">  450</span>&#160;                                                                                      chr_name = chr_name )</div><div class="line"><a name="l00451"></a><span class="lineno">  451</span>&#160;                        </div><div class="line"><a name="l00452"></a><span class="lineno">  452</span>&#160;                        <span class="comment"># Add the newly created object to the list</span></div><div class="line"><a name="l00453"></a><span class="lineno">  453</span>&#160;                        <span class="comment"># NB: The DSTranscript entry has been created before the Gene, GeneAlias and UTGeneFromAlias </span></div><div class="line"><a name="l00454"></a><span class="lineno">  454</span>&#160;                        <span class="comment">#     entries but the gene_id attribute of the DSTranscript is a foreign key from the Gene</span></div><div class="line"><a name="l00455"></a><span class="lineno">  455</span>&#160;                        <span class="comment">#     table. Hence, this is necessary to add these new objects to the session prior than the</span></div><div class="line"><a name="l00456"></a><span class="lineno">  456</span>&#160;                        <span class="comment">#     existing DSTranscript.</span></div><div class="line"><a name="l00457"></a><span class="lineno">  457</span>&#160;                        objects_to_insert = new_obj_to_insert + objects_to_insert</div><div class="line"><a name="l00458"></a><span class="lineno">  458</span>&#160;                        existing_dstranscript.gene_id = gene.gene_id</div><div class="line"><a name="l00459"></a><span class="lineno">  459</span>&#160;                        </div><div class="line"><a name="l00460"></a><span class="lineno">  460</span>&#160;                dstranscript = existing_dstranscript</div><div class="line"><a name="l00461"></a><span class="lineno">  461</span>&#160;</div><div class="line"><a name="l00462"></a><span class="lineno">  462</span>&#160;            <span class="comment"># Otherwise, add the newly created transcript to the list of objects to insert and keep track of it</span></div><div class="line"><a name="l00463"></a><span class="lineno">  463</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00464"></a><span class="lineno">  464</span>&#160;                all_dstranscripts[ dstranscript ] = dstranscript</div><div class="line"><a name="l00465"></a><span class="lineno">  465</span>&#160;                objects_to_insert.append( dstranscript )</div><div class="line"><a name="l00466"></a><span class="lineno">  466</span>&#160;                </div><div class="line"><a name="l00467"></a><span class="lineno">  467</span>&#160;                <span class="comment"># Increase the autoincrement value for ID of 1</span></div><div class="line"><a name="l00468"></a><span class="lineno">  468</span>&#160;                autoincrement += 1</div><div class="line"><a name="l00469"></a><span class="lineno">  469</span>&#160;                        </div><div class="line"><a name="l00470"></a><span class="lineno">  470</span>&#160;            </div><div class="line"><a name="l00471"></a><span class="lineno">  471</span>&#160;            </div><div class="line"><a name="l00473"></a><span class="lineno">  473</span>&#160;            </div><div class="line"><a name="l00474"></a><span class="lineno">  474</span>&#160;            <span class="comment"># Get the ORF length</span></div><div class="line"><a name="l00475"></a><span class="lineno">  475</span>&#160;            raw_sequence = row[ self.HEADER_ORF_SEQUENCE ]</div><div class="line"><a name="l00476"></a><span class="lineno">  476</span>&#160;            <span class="keywordflow">if</span> ParserStrategy.is_empty( raw_sequence ):</div><div class="line"><a name="l00477"></a><span class="lineno">  477</span>&#160;                raw_sequence = <span class="keywordtype">None</span></div><div class="line"><a name="l00478"></a><span class="lineno">  478</span>&#160;                start_codon_seq = <span class="keywordtype">None</span></div><div class="line"><a name="l00479"></a><span class="lineno">  479</span>&#160;                orf_len_nt = <span class="keywordtype">None</span></div><div class="line"><a name="l00480"></a><span class="lineno">  480</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00481"></a><span class="lineno">  481</span>&#160;                start_codon_seq = raw_sequence[:3]</div><div class="line"><a name="l00482"></a><span class="lineno">  482</span>&#160;                orf_len_nt = len( raw_sequence )</div><div class="line"><a name="l00483"></a><span class="lineno">  483</span>&#160;            </div><div class="line"><a name="l00484"></a><span class="lineno">  484</span>&#160;            <span class="keywordflow">if</span> ParserStrategy.is_empty( row[ self.HEADER_ORF_LENGTH ] ):</div><div class="line"><a name="l00485"></a><span class="lineno">  485</span>&#160;                orf_len = <span class="keywordtype">None</span></div><div class="line"><a name="l00486"></a><span class="lineno">  486</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00487"></a><span class="lineno">  487</span>&#160;                orf_len = int( row[ self.HEADER_ORF_LENGTH ] ) - 1</div><div class="line"><a name="l00488"></a><span class="lineno">  488</span>&#160;                </div><div class="line"><a name="l00489"></a><span class="lineno">  489</span>&#160;            dsorftranscriptasso = DSORFTranscriptAsso( id = autoincrement,</div><div class="line"><a name="l00490"></a><span class="lineno">  490</span>&#160;                                                       data_source = datasource.name,</div><div class="line"><a name="l00491"></a><span class="lineno">  491</span>&#160;                                                       transcript_id = dstranscript.id,</div><div class="line"><a name="l00492"></a><span class="lineno">  492</span>&#160;                                                       uniq_orf_id = dsorf.id,</div><div class="line"><a name="l00493"></a><span class="lineno">  493</span>&#160;                                                       predicted = self.PREDICTED,</div><div class="line"><a name="l00494"></a><span class="lineno">  494</span>&#160;                                                       ribo_seq = self.RIBO_SEQ,</div><div class="line"><a name="l00495"></a><span class="lineno">  495</span>&#160;                                                       cell_context = row[ self.HEADER_CELL_CONTEXT ],</div><div class="line"><a name="l00496"></a><span class="lineno">  496</span>&#160;                                                       orf_id = orf_id,</div><div class="line"><a name="l00497"></a><span class="lineno">  497</span>&#160;                                                       raw_sequence = raw_sequence,</div><div class="line"><a name="l00498"></a><span class="lineno">  498</span>&#160;                                                       raw_sequence_aa = self.rm_stop_from_seq( row[ self.HEADER_ORF_SEQUENCE_AA ] ),</div><div class="line"><a name="l00499"></a><span class="lineno">  499</span>&#160;                                                       start_codon_seq = start_codon_seq,</div><div class="line"><a name="l00500"></a><span class="lineno">  500</span>&#160;                                                       kozak_context = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00501"></a><span class="lineno">  501</span>&#160;                                                       orf_length_nt = orf_len_nt,</div><div class="line"><a name="l00502"></a><span class="lineno">  502</span>&#160;                                                       orf_length = orf_len,</div><div class="line"><a name="l00503"></a><span class="lineno">  503</span>&#160;                                                       provided_category = orf_annotation,</div><div class="line"><a name="l00504"></a><span class="lineno">  504</span>&#160;                                                       ms_info = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00505"></a><span class="lineno">  505</span>&#160;                                                       orf_score = row[ self.HEADER_ORF_SCORE ],</div><div class="line"><a name="l00506"></a><span class="lineno">  506</span>&#160;                                                       phylocsf = <span class="keywordtype">None</span>,</div><div class="line"><a name="l00507"></a><span class="lineno">  507</span>&#160;                                                       phastcons = row[ self.HEADER_PHASTCONS ],</div><div class="line"><a name="l00508"></a><span class="lineno">  508</span>&#160;                                                       floss = row[ self.HEADER_FLOSS_SCORE ],</div><div class="line"><a name="l00509"></a><span class="lineno">  509</span>&#160;                                                       floss_class = row[ self.HEADER_FLOSS_CLASS ] )</div><div class="line"><a name="l00510"></a><span class="lineno">  510</span>&#160;            </div><div class="line"><a name="l00511"></a><span class="lineno">  511</span>&#160;            <span class="comment"># If this DSORFTranscriptAsso has already been reported (in the current source), get the existing one</span></div><div class="line"><a name="l00512"></a><span class="lineno">  512</span>&#160;            existing_dsota = all_dsota.get( dsorftranscriptasso )</div><div class="line"><a name="l00513"></a><span class="lineno">  513</span>&#160;            <span class="keywordflow">if</span> existing_dsota:</div><div class="line"><a name="l00514"></a><span class="lineno">  514</span>&#160;                    </div><div class="line"><a name="l00515"></a><span class="lineno">  515</span>&#160;                <span class="comment"># Try to update the existing DSORFTranscriptAsso</span></div><div class="line"><a name="l00516"></a><span class="lineno">  516</span>&#160;                <span class="keywordflow">try</span>:</div><div class="line"><a name="l00517"></a><span class="lineno">  517</span>&#160;                    existing_dsota.update( dsorftranscriptasso )</div><div class="line"><a name="l00518"></a><span class="lineno">  518</span>&#160;                <span class="keywordflow">except</span> DenCellORFObjUpdateException <span class="keyword">as</span> e:</div><div class="line"><a name="l00519"></a><span class="lineno">  519</span>&#160;                    Logger.get_instance().error( <span class="stringliteral">&#39;The same ORF has been found associated several time&#39;</span> +</div><div class="line"><a name="l00520"></a><span class="lineno">  520</span>&#160;                                                 <span class="stringliteral">&#39; with the same transcript (&quot;&#39;</span> + dstranscript.transcript_id +</div><div class="line"><a name="l00521"></a><span class="lineno">  521</span>&#160;                                                 <span class="stringliteral">&#39;&quot;) under original ORF IDs &quot;&#39;</span> + orf_id + <span class="stringliteral">&#39;&quot; and &quot;&#39;</span> + </div><div class="line"><a name="l00522"></a><span class="lineno">  522</span>&#160;                                                 existing_dsota.orf_id + <span class="stringliteral">&#39;&quot; in &#39;</span> + self.data_source + <span class="stringliteral">&#39;.&#39;</span> +</div><div class="line"><a name="l00523"></a><span class="lineno">  523</span>&#160;                                                 <span class="stringliteral">&#39;, with different properties: &#39;</span> + e.get_message() +</div><div class="line"><a name="l00524"></a><span class="lineno">  524</span>&#160;                                                 <span class="stringliteral">&#39; Error code: &#39;</span> + LogCodes.ERR_DCORF_CONFL_DSORF + <span class="stringliteral">&#39;.&#39;</span>,</div><div class="line"><a name="l00525"></a><span class="lineno">  525</span>&#160;                                                 ex = <span class="keyword">False</span> )</div><div class="line"><a name="l00526"></a><span class="lineno">  526</span>&#160;                </div><div class="line"><a name="l00527"></a><span class="lineno">  527</span>&#160;                dsorftranscriptasso = existing_dsota</div><div class="line"><a name="l00528"></a><span class="lineno">  528</span>&#160;</div><div class="line"><a name="l00529"></a><span class="lineno">  529</span>&#160;                </div><div class="line"><a name="l00530"></a><span class="lineno">  530</span>&#160;            <span class="comment"># Otherwise, add the newly created DSORFTranscriptAsso to the list of objects to insert and keep track of it</span></div><div class="line"><a name="l00531"></a><span class="lineno">  531</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00532"></a><span class="lineno">  532</span>&#160;                all_dsota[ dsorftranscriptasso ] = dsorftranscriptasso</div><div class="line"><a name="l00533"></a><span class="lineno">  533</span>&#160;</div><div class="line"><a name="l00534"></a><span class="lineno">  534</span>&#160;                <span class="comment"># Add it to the list of objects to insert in the database</span></div><div class="line"><a name="l00535"></a><span class="lineno">  535</span>&#160;                objects_to_insert.append( dsorftranscriptasso )</div><div class="line"><a name="l00536"></a><span class="lineno">  536</span>&#160;                </div><div class="line"><a name="l00537"></a><span class="lineno">  537</span>&#160;                <span class="comment"># Increase the autoincrement value for ID of 1</span></div><div class="line"><a name="l00538"></a><span class="lineno">  538</span>&#160;                autoincrement += 1</div><div class="line"><a name="l00539"></a><span class="lineno">  539</span>&#160;            </div><div class="line"><a name="l00540"></a><span class="lineno">  540</span>&#160;        <span class="comment"># Log the duration of the process</span></div><div class="line"><a name="l00541"></a><span class="lineno">  541</span>&#160;        end_time = datetime.datetime.now()</div><div class="line"><a name="l00542"></a><span class="lineno">  542</span>&#160;        Logger.get_instance().debug( <span class="stringliteral">&#39;sORFs_org.parse_file_content(): Duration of parsing: &#39;</span> + </div><div class="line"><a name="l00543"></a><span class="lineno">  543</span>&#160;                                     str( end_time - start_time ) + <span class="stringliteral">&#39; (&#39;</span> +</div><div class="line"><a name="l00544"></a><span class="lineno">  544</span>&#160;                                     str( ProgressionBar.get_instance().total_row_count ) + </div><div class="line"><a name="l00545"></a><span class="lineno">  545</span>&#160;                                     <span class="stringliteral">&#39; rows processed).&#39;</span> )</div><div class="line"><a name="l00546"></a><span class="lineno">  546</span>&#160;                </div><div class="line"><a name="l00547"></a><span class="lineno">  547</span>&#160;        <span class="comment"># Store the current value of the autoincrement in the DataManager</span></div><div class="line"><a name="l00548"></a><span class="lineno">  548</span>&#160;        DataManager.get_instance().store_data( Constants.DM_AUTOINCREMENT, autoincrement )</div><div class="line"><a name="l00549"></a><span class="lineno">  549</span>&#160;        </div><div class="line"><a name="l00550"></a><span class="lineno">  550</span>&#160;        <span class="keywordflow">return</span> objects_to_insert</div><div class="line"><a name="l00551"></a><span class="lineno">  551</span>&#160;    </div><div class="line"><a name="l00552"></a><span class="lineno">  552</span>&#160;</div><div class="line"><a name="l00553"></a><span class="lineno">  553</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a id="ad35950e4eecbe3a6ac244a108cb79ee6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad35950e4eecbe3a6ac244a108cb79ee6">&#9670;&nbsp;</a></span>rm_stop_from_seq()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org.rm_stop_from_seq </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>aa_sequence</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>rm_stop_from_seq </h2>
<p>This is a static method that allows to return an amino acid sequence without its stop.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">aa_sequence</td><td>String - The sequence (in amino acid).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>String - The amino acid sequence without </dd></dl>
<div class="fragment"><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;    <span class="keyword">def </span>rm_stop_from_seq( aa_sequence ):</div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;        </div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;        <span class="keywordflow">if</span> ( isinstance( aa_sequence, basestring )</div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;             <span class="keywordflow">and</span> len( aa_sequence ) &gt; 0 ):</div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;            <span class="keywordflow">if</span> aa_sequence[-1] == <span class="stringliteral">&#39;*&#39;</span>:</div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;                <span class="keywordflow">return</span> aa_sequence[:-1]</div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;            <span class="keywordflow">else</span>:</div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;                <span class="keywordflow">return</span> aa_sequence</div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;            <span class="keywordflow">return</span> <span class="keywordtype">None</span></div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;</div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a612dc7af609862589f6de0806d5c0b77"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a612dc7af609862589f6de0806d5c0b77">&#9670;&nbsp;</a></span>DOI</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">string fr.tagc.uorf.core.execution.insertion.parserstrategy.common.sORFs_org.sORFs_org.DOI = '10.1093/nar/gkx1130'</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>Class variables </h2>
<p>Constants </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/execution/insertion/parserstrategy/common/sORFs_org.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
