<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: fr.tagc.uorf.core.execution.GenerateFastaFileStrategy.GenerateFastaFileStrategy Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>fr</b></li><li class="navelem"><b>tagc</b></li><li class="navelem"><b>uorf</b></li><li class="navelem"><b>core</b></li><li class="navelem"><b>execution</b></li><li class="navelem"><b>GenerateFastaFileStrategy</b></li><li class="navelem"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html">GenerateFastaFileStrategy</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">fr.tagc.uorf.core.execution.GenerateFastaFileStrategy.GenerateFastaFileStrategy Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Inherits object.</p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a2cffa3319182875b710b276a54af874b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html#a2cffa3319182875b710b276a54af874b">__init__</a> (self)</td></tr>
<tr class="separator:a2cffa3319182875b710b276a54af874b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3368641e2274153ed6d30f279453f914"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html#a3368641e2274153ed6d30f279453f914">execute</a> (self)</td></tr>
<tr class="separator:a3368641e2274153ed6d30f279453f914"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a2e308323e2accd275d470365e06c5b1a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html#a2e308323e2accd275d470365e06c5b1a">check_stop_codons_in_sqce</a> (row, seq_type, seq_attribute_name)</td></tr>
<tr class="separator:a2e308323e2accd275d470365e06c5b1a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9680d0a2df2b41b32db8de889ee46044"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html#a9680d0a2df2b41b32db8de889ee46044">generate_header</a> (row, taxon_sc_name, taxon_code, taxon_id, table, db_release, long_header)</td></tr>
<tr class="separator:a9680d0a2df2b41b32db8de889ee46044"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a936c81e088788e99b88638c5c0735c92"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html#a936c81e088788e99b88638c5c0735c92">ORF_TABLE</a> = '<a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORF_1_1ORF.html">ORF</a>'</td></tr>
<tr class="separator:a936c81e088788e99b88638c5c0735c92"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html">GenerateFastaFileStrategy</a> </h1>
<p>This class is a strategy aiming to generate a FASTA file using the information contained either in the ORF table or in the ORFTranscriptAsso table of the PRO database. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a2cffa3319182875b710b276a54af874b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2cffa3319182875b710b276a54af874b">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.GenerateFastaFileStrategy.GenerateFastaFileStrategy.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>Constructor of <a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1execution_1_1GenerateFastaFileStrategy_1_1GenerateFastaFileStrategy.html">GenerateFastaFileStrategy</a> </h2>
<p>Instance variables:</p><ul>
<li>db_settings: Dictionary - A dictionary of settings. This may include:<ul>
<li>The database name.</li>
<li>The database type (SQLite / MySQL).</li>
<li>For SQLite databases: the folder of SQLite file.</li>
<li>For MySQL databases: the MySQL user, password, host IP and port.</li>
</ul>
</li>
<li>output_folder: String - The name of the folder where to export the FASTA file.</li>
<li>filename: String - The name of the FASTA file.</li>
<li>seq_type: String - The type of sequence to use ('DNA' for nucleic sequences or 'PROT' for amino acid sequences).</li>
<li>table_type: String - The name of the table to query ('ORF' or 'OTA') to get the sequences. </li>
</ul>
<div class="fragment"><div class="line"><a name="l00129"></a><span class="lineno">  129</span>&#160;    <span class="keyword">def </span>__init__( self ):</div><div class="line"><a name="l00130"></a><span class="lineno">  130</span>&#160;        </div><div class="line"><a name="l00131"></a><span class="lineno">  131</span>&#160;        <span class="comment"># Get the options necessary to establish the connection to the database</span></div><div class="line"><a name="l00132"></a><span class="lineno">  132</span>&#160;        self.db_settings = {}</div><div class="line"><a name="l00133"></a><span class="lineno">  133</span>&#160;        self.db_settings[ Constants.DB_SETTINGS_DB_TYPE ] = OptionManager.get_instance().get_option( OptionConstants.OPTION_DB_TYPE )</div><div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;        self.db_settings[ Constants.DB_SETTINGS_DB_NAME ] = OptionManager.get_instance().get_option( OptionConstants.OPTION_DB_NAME, </div><div class="line"><a name="l00135"></a><span class="lineno">  135</span>&#160;                                                                                                     not_none = <span class="keyword">True</span> )</div><div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;        </div><div class="line"><a name="l00137"></a><span class="lineno">  137</span>&#160;        <span class="keywordflow">if</span> ( self.db_settings[ Constants.DB_SETTINGS_DB_TYPE ] == SQLConstants.DB_TYPE_MYSQL ):</div><div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;            self.db_settings[ Constants.DB_SETTINGS_MYSQL_USER ] = OptionManager.get_instance().get_option( OptionConstants.OPTION_DB_MYSQL_USER )</div><div class="line"><a name="l00139"></a><span class="lineno">  139</span>&#160;            self.db_settings[ Constants.DB_SETTINGS_MYSQL_PASSWD ] = OptionManager.get_instance().get_option( OptionConstants.OPTION_DB_MYSQL_PASSWD )</div><div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;            self.db_settings[ Constants.DB_SETTINGS_MYSQL_HOST ] = OptionManager.get_instance().get_option( OptionConstants.OPTION_DB_MYSQL_HOST_IP )</div><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;            self.db_settings[ Constants.DB_SETTINGS_MYSQL_PORT ] = OptionManager.get_instance().get_option( OptionConstants.OPTION_DB_MYSQL_PORT )</div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;            </div><div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;        <span class="keywordflow">elif</span> ( self.db_settings[ Constants.DB_SETTINGS_DB_TYPE ] == SQLConstants.DB_TYPE_SQLITE ):</div><div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160;            self.db_settings[ Constants.DB_SETTINGS_DB_FOLDER ] = OptionManager.get_instance().get_option( OptionConstants.OPTION_DB_FOLDER )</div><div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;        </div><div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;        <span class="comment"># Get the output folder</span></div><div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;        self.output_folder = OptionManager.get_instance().get_option( OptionConstants.OPTION_OUTPUT_FOLDER, </div><div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;                                                                      not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00149"></a><span class="lineno">  149</span>&#160;        <span class="comment"># By default, save the file in a FASTA folder</span></div><div class="line"><a name="l00150"></a><span class="lineno">  150</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> self.output_folder ):</div><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;            self.output_folder = Constants.FASTA_FOLDER</div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;        </div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;        <span class="comment"># Get the eventual filename</span></div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;        self.filename = OptionManager.get_instance().get_option( OptionConstants.OPTION_FASTA_FILENAME, </div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;                                                                 not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;        <span class="comment"># By default, name the file &#39;DenCellORF_ORF&#39;</span></div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> self.filename ):</div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;            self.filename = Constants.DEFAULT_FASTA_FILENAME</div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;        </div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;        <span class="comment"># Get the type of sequence to use to generate the FASTA file (DNA or PROT)</span></div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;        self.seq_type = OptionManager.get_instance().get_option( OptionConstants.OPTION_FASTA_SEQ_TYPE,</div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;                                                                 not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;        <span class="comment"># By default, generate a FASTA of amino acid sequences</span></div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> self.seq_type ):</div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;            self.seq_type = Constants.SEQUENCE_TYPE_PROT</div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;        </div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;        <span class="comment"># Get the table to query to get the sequences (ORF or ORFTranscriptAsso)</span></div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;        self.table_type = OptionManager.get_instance().get_option( OptionConstants.OPTION_FASTA_QUERY_TABLE,</div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;                                                                   not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;    </div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;        <span class="comment"># By default, generate the file using the content of the ORF table</span></div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> self.table_type ):</div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;            self.table_type = self.ORF_TABLE</div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;        </div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;        <span class="comment"># Get the name of the attribute to use to get the sequence</span></div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;        self.seq_attribute_name = self.SEQUENCE_ATTRIBUTES[ self.seq_type ][ self.table_type ]</div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;            </div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;        <span class="comment"># Get the option informing if the sequences that have stop codons have to be excluded</span></div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;        self.exclude_sqce_with_stops = OptionManager.get_instance().get_option( OptionConstants.OPTION_FASTA_EXCLUDE_SQCES_WITH_STOP,</div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;                                                                                not_none = <span class="keyword">False</span> )</div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;        </div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;        <span class="comment"># Get the option allowing to generate long headers</span></div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;        self.long_header = OptionManager.get_instance().get_option( OptionConstants.OPTION_FASTA_LONG_HEADERS,</div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;                                                                    not_none = <span class="keyword">False</span> )        </div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;    </div><div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;        </div><div class="line"><a name="l00187"></a><span class="lineno">  187</span>&#160;    </div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a2e308323e2accd275d470365e06c5b1a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2e308323e2accd275d470365e06c5b1a">&#9670;&nbsp;</a></span>check_stop_codons_in_sqce()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.GenerateFastaFileStrategy.GenerateFastaFileStrategy.check_stop_codons_in_sqce </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>row</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq_type</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq_attribute_name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>check_stop_codons_in_sqce </h2>
<p>This is a static method allowing to check if a a sequence contains stop codons.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">row</td><td>Pandas Series - A row from the ORF Pandas data frame. </td></tr>
    <tr><td class="paramname">seq_type</td><td>String - The type of sequence (DNA or PROT). </td></tr>
    <tr><td class="paramname">seq_attribute_name</td><td>String - The name of the attribute containing the sequence.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>stop_in_seq: Boolean - True if the sequence contains a stop codon. </dd></dl>
<div class="fragment"><div class="line"><a name="l00358"></a><span class="lineno">  358</span>&#160;    <span class="keyword">def </span>check_stop_codons_in_sqce( row, seq_type, seq_attribute_name ):</div><div class="line"><a name="l00359"></a><span class="lineno">  359</span>&#160;        </div><div class="line"><a name="l00360"></a><span class="lineno">  360</span>&#160;        seq = row[ seq_attribute_name ]</div><div class="line"><a name="l00361"></a><span class="lineno">  361</span>&#160;        </div><div class="line"><a name="l00362"></a><span class="lineno">  362</span>&#160;        <span class="keywordflow">if</span> seq_type == Constants.SEQUENCE_TYPE_PROT:</div><div class="line"><a name="l00363"></a><span class="lineno">  363</span>&#160;            stop_codons_list = [ GeneticsConstants.STOP_CODON ]</div><div class="line"><a name="l00364"></a><span class="lineno">  364</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00365"></a><span class="lineno">  365</span>&#160;            stop_codons_list = GeneticsConstants.STOP_CODON_SEQUENCES</div><div class="line"><a name="l00366"></a><span class="lineno">  366</span>&#160;            </div><div class="line"><a name="l00367"></a><span class="lineno">  367</span>&#160;            <span class="comment"># As the stop codons have to be found in the same frame than</span></div><div class="line"><a name="l00368"></a><span class="lineno">  368</span>&#160;            <span class="comment"># the start codon, split the sequence into a list of </span></div><div class="line"><a name="l00369"></a><span class="lineno">  369</span>&#160;            <span class="comment"># 3-characters strings</span></div><div class="line"><a name="l00370"></a><span class="lineno">  370</span>&#160;            seq = [ seq[ k : k+3 ] <span class="keywordflow">for</span> k <span class="keywordflow">in</span> range( 0, len( seq ), 3 ) ]</div><div class="line"><a name="l00371"></a><span class="lineno">  371</span>&#160;            </div><div class="line"><a name="l00372"></a><span class="lineno">  372</span>&#160;            <span class="comment"># Exclude of the list the last codon of the list, </span></div><div class="line"><a name="l00373"></a><span class="lineno">  373</span>&#160;            <span class="comment"># as it is expected to be a stop codon</span></div><div class="line"><a name="l00374"></a><span class="lineno">  374</span>&#160;            seq = seq[ :-1 ]</div><div class="line"><a name="l00375"></a><span class="lineno">  375</span>&#160;        </div><div class="line"><a name="l00376"></a><span class="lineno">  376</span>&#160;        stop_in_seq = <span class="keyword">False</span></div><div class="line"><a name="l00377"></a><span class="lineno">  377</span>&#160;        k = 0</div><div class="line"><a name="l00378"></a><span class="lineno">  378</span>&#160;        <span class="keywordflow">while</span> ( ( <span class="keywordflow">not</span> stop_in_seq ) <span class="keywordflow">and</span> ( k &lt; len( stop_codons_list ) ) ):</div><div class="line"><a name="l00379"></a><span class="lineno">  379</span>&#160;            stop = stop_codons_list[ k ]</div><div class="line"><a name="l00380"></a><span class="lineno">  380</span>&#160;            <span class="keywordflow">if</span> stop <span class="keywordflow">in</span> seq:</div><div class="line"><a name="l00381"></a><span class="lineno">  381</span>&#160;                stop_in_seq = <span class="keyword">True</span></div><div class="line"><a name="l00382"></a><span class="lineno">  382</span>&#160;            k += 1</div><div class="line"><a name="l00383"></a><span class="lineno">  383</span>&#160;            </div><div class="line"><a name="l00384"></a><span class="lineno">  384</span>&#160;        <span class="keywordflow">return</span> stop_in_seq</div><div class="line"><a name="l00385"></a><span class="lineno">  385</span>&#160;        </div><div class="line"><a name="l00386"></a><span class="lineno">  386</span>&#160;        </div><div class="line"><a name="l00387"></a><span class="lineno">  387</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="a3368641e2274153ed6d30f279453f914"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3368641e2274153ed6d30f279453f914">&#9670;&nbsp;</a></span>execute()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.GenerateFastaFileStrategy.GenerateFastaFileStrategy.execute </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<h2>execute </h2>
<p>Execute the strategy to generate the FASTA file.</p>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">DenCellORFException</td><td>When an exception has been raised creating a session to the database. </td></tr>
    <tr><td class="paramname">DenCellORFException</td><td>When there is no sequence to write in the file. </td></tr>
  </table>
  </dd>
</dl>
<div class="fragment"><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;    <span class="keyword">def </span>execute( self ):</div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;        </div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;        <span class="comment"># Create a session to the &quot;PRO-like&quot; database</span></div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;        SQLManagerPRO.get_instance().set_db_settings( self.db_settings )</div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;        </div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;        <span class="keywordflow">try</span>:</div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;            SQLManagerPRO.get_instance().get_session()</div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;        <span class="keywordflow">except</span> Exception <span class="keyword">as</span> e:</div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;            <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;GenerateFastaFileStrategy.execute(): An error occurred trying to&#39;</span> +</div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;                                       <span class="stringliteral">&#39; create a session to the database.&#39;</span> +</div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;                                        <span class="stringliteral">&#39;\n Error code: &#39;</span> + LogCodes.ERR_SQL_SESSION + <span class="stringliteral">&#39;.&#39;</span>, e)</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;        SQLManagerPRO.get_instance().close_session()</div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;        </div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;        </div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;        Logger.get_instance().info( <span class="stringliteral">&#39;Starting to build the FASTA file.&#39;</span> )</div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;        Logger.get_instance().info( <span class="stringliteral">&#39;The fasta file will be created querying the &#39;</span> + self.table_type +</div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;                                    <span class="stringliteral">&#39; table and using the &#39;</span> + self.seq_type + <span class="stringliteral">&#39; sequences.&#39;</span> )</div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;        </div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;        <span class="comment"># Create the output folder if it does not yet exist</span></div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;        <span class="comment"># (and its parent folders if necessary )</span></div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;        <span class="keywordflow">if</span> ( <span class="keywordflow">not</span> os.path.isdir( self.output_folder ) ):</div><div class="line"><a name="l00217"></a><span class="lineno">  217</span>&#160;            os.makedirs( self.output_folder )</div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;            </div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;        file_path = os.path.join( self.output_folder, self.filename ) + self.FASTA_FILE_EXTENSION</div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;        </div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;        <span class="comment"># Get the name of the species</span></div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;        sp = SQLManagerPRO.get_instance().get_session().query( PROSpeciesCatalog ).one().name</div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;        SQLManagerPRO.get_instance().close_session()</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;        </div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;        <span class="comment"># Get the informations related to the species</span></div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;        <span class="comment"># NB: These information will be used in the headers</span></div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;        taxon_sc_name = Constants.SPECIES_CATALOG_FULL_NAMES_WITH_CAPS[ sp ]</div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;        taxon_code = Constants.SPECIES_CATALOG_CODE[ sp ]</div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;        taxon_id = Constants.SPECIES_CATALOG_TAXON_ID[ sp ]</div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;        </div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;        <span class="comment"># Get the version number of the database</span></div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;        db_release = SQLManagerPRO.get_instance().get_session().query( </div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;                                                                        PROMetadata.value </div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;                                                                      ).filter( </div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;                                                                                    PROMetadata.parameter == Constants.METATABLE_DATABASE_VERSION_NUMBER</div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;                                                                                ).all()</div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;        db_release = GeneralUtil.query_result_to_list( db_release )</div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;        <span class="keywordflow">if</span> ( len( db_release ) == 1 ):</div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;            db_release = db_release[ 0 ]</div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;            db_release = <span class="stringliteral">&#39;&#39;</span></div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;            </div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;        </div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;        <span class="comment"># Create the FASTA file</span></div><div class="line"><a name="l00245"></a><span class="lineno">  245</span>&#160;        <span class="comment"># ---------------------                </span></div><div class="line"><a name="l00246"></a><span class="lineno">  246</span>&#160;        <span class="comment"># Get the information necessary to compute the file content</span></div><div class="line"><a name="l00247"></a><span class="lineno">  247</span>&#160;        <span class="keywordflow">if</span> ( self.table_type == self.ORF_TABLE ):</div><div class="line"><a name="l00248"></a><span class="lineno">  248</span>&#160;            <span class="comment"># Get the necessary information from the ORF table</span></div><div class="line"><a name="l00249"></a><span class="lineno">  249</span>&#160;            all_orfs_query = SQLManagerPRO.get_instance().get_session().query( </div><div class="line"><a name="l00250"></a><span class="lineno">  250</span>&#160;                                                                                ORF.id,</div><div class="line"><a name="l00251"></a><span class="lineno">  251</span>&#160;                                                                                ORF.chromosome,</div><div class="line"><a name="l00252"></a><span class="lineno">  252</span>&#160;                                                                                ORF.strand,</div><div class="line"><a name="l00253"></a><span class="lineno">  253</span>&#160;                                                                                ORF.start_pos,</div><div class="line"><a name="l00254"></a><span class="lineno">  254</span>&#160;                                                                                ORF.stop_pos,</div><div class="line"><a name="l00255"></a><span class="lineno">  255</span>&#160;                                                                                ORF.spliced_parts_count,</div><div class="line"><a name="l00256"></a><span class="lineno">  256</span>&#160;                                                                                eval( <span class="stringliteral">&#39;ORF.&#39;</span> + self.seq_attribute_name ) </div><div class="line"><a name="l00257"></a><span class="lineno">  257</span>&#160;                                                                               ).filter(</div><div class="line"><a name="l00258"></a><span class="lineno">  258</span>&#160;                                                                                            eval( <span class="stringliteral">&#39;ORF.&#39;</span> + self.seq_attribute_name ) != <span class="keywordtype">None</span></div><div class="line"><a name="l00259"></a><span class="lineno">  259</span>&#160;                                                                                        )</div><div class="line"><a name="l00260"></a><span class="lineno">  260</span>&#160;                </div><div class="line"><a name="l00261"></a><span class="lineno">  261</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00262"></a><span class="lineno">  262</span>&#160;            <span class="comment"># Get the necessary information from the ORFTranscriptAsso table</span></div><div class="line"><a name="l00263"></a><span class="lineno">  263</span>&#160;            all_orfs_query = SQLManagerPRO.get_instance().get_session().query( </div><div class="line"><a name="l00264"></a><span class="lineno">  264</span>&#160;                                                                                ORFTranscriptAsso.id,</div><div class="line"><a name="l00265"></a><span class="lineno">  265</span>&#160;                                                                                ORFTranscriptAsso.orf_id,</div><div class="line"><a name="l00266"></a><span class="lineno">  266</span>&#160;                                                                                ORFTranscriptAsso.transcript_id,</div><div class="line"><a name="l00267"></a><span class="lineno">  267</span>&#160;                                                                                ORFTranscriptAsso.rel_start_pos,</div><div class="line"><a name="l00268"></a><span class="lineno">  268</span>&#160;                                                                                ORFTranscriptAsso.rel_stop_pos,</div><div class="line"><a name="l00269"></a><span class="lineno">  269</span>&#160;                                                                                eval( <span class="stringliteral">&#39;ORFTranscriptAsso.&#39;</span> + self.seq_attribute_name )</div><div class="line"><a name="l00270"></a><span class="lineno">  270</span>&#160;                                                                               ).filter(</div><div class="line"><a name="l00271"></a><span class="lineno">  271</span>&#160;                                                                                            eval( <span class="stringliteral">&#39;ORFTranscriptAsso.&#39;</span> + self.seq_attribute_name ) != <span class="keywordtype">None</span></div><div class="line"><a name="l00272"></a><span class="lineno">  272</span>&#160;                                                                                        )</div><div class="line"><a name="l00273"></a><span class="lineno">  273</span>&#160;        </div><div class="line"><a name="l00274"></a><span class="lineno">  274</span>&#160;        <span class="comment"># Run the query and get the results as a Pandas data frame</span></div><div class="line"><a name="l00275"></a><span class="lineno">  275</span>&#160;        all_orfs_df = pd.read_sql( all_orfs_query.statement,</div><div class="line"><a name="l00276"></a><span class="lineno">  276</span>&#160;                                   SQLManagerPRO.get_instance().get_engine() )</div><div class="line"><a name="l00277"></a><span class="lineno">  277</span>&#160;        SQLManagerPRO.get_instance().close_session()</div><div class="line"><a name="l00278"></a><span class="lineno">  278</span>&#160;        </div><div class="line"><a name="l00279"></a><span class="lineno">  279</span>&#160;        <span class="comment"># Check the query returned a result</span></div><div class="line"><a name="l00280"></a><span class="lineno">  280</span>&#160;        total_sqce_count = all_orfs_df.shape[0]</div><div class="line"><a name="l00281"></a><span class="lineno">  281</span>&#160;        <span class="keywordflow">if</span> ( total_sqce_count == 0 ):</div><div class="line"><a name="l00282"></a><span class="lineno">  282</span>&#160;            <span class="keywordflow">raise</span> DenCellORFException( <span class="stringliteral">&#39;It seems that the database you are querying do not contain any&#39;</span> +</div><div class="line"><a name="l00283"></a><span class="lineno">  283</span>&#160;                                       <span class="stringliteral">&#39; entry with sequence (&#39;</span> + self.seq_type + <span class="stringliteral">&#39;) in its &#39;</span> + </div><div class="line"><a name="l00284"></a><span class="lineno">  284</span>&#160;                                       self.table_type + <span class="stringliteral">&#39; table. Hence, the generation of the fasta file&#39;</span> +</div><div class="line"><a name="l00285"></a><span class="lineno">  285</span>&#160;                                       <span class="stringliteral">&#39; has been stopped.&#39;</span> )</div><div class="line"><a name="l00286"></a><span class="lineno">  286</span>&#160;        </div><div class="line"><a name="l00287"></a><span class="lineno">  287</span>&#160;        all_orfs_df = all_orfs_df.astype( str )</div><div class="line"><a name="l00288"></a><span class="lineno">  288</span>&#160;        </div><div class="line"><a name="l00289"></a><span class="lineno">  289</span>&#160;        <span class="comment"># If the excludeSqcesWithStop option has been selected, </span></div><div class="line"><a name="l00290"></a><span class="lineno">  290</span>&#160;        <span class="comment"># then exclude from the data frame all the sequences </span></div><div class="line"><a name="l00291"></a><span class="lineno">  291</span>&#160;        <span class="comment"># that contains at least a stop, then remove it</span></div><div class="line"><a name="l00292"></a><span class="lineno">  292</span>&#160;        <span class="keywordflow">if</span> self.exclude_sqce_with_stops:</div><div class="line"><a name="l00293"></a><span class="lineno">  293</span>&#160;            contains_stop_codon = all_orfs_df.apply( self.check_stop_codons_in_sqce,</div><div class="line"><a name="l00294"></a><span class="lineno">  294</span>&#160;                                                     seq_type = self.seq_type,</div><div class="line"><a name="l00295"></a><span class="lineno">  295</span>&#160;                                                     seq_attribute_name = self.seq_attribute_name,</div><div class="line"><a name="l00296"></a><span class="lineno">  296</span>&#160;                                                     axis=1 ).to_frame()</div><div class="line"><a name="l00297"></a><span class="lineno">  297</span>&#160;            contains_stop_codon = contains_stop_codon.rename( columns = { 0: <span class="stringliteral">&#39;contains_stop_codon&#39;</span> } )</div><div class="line"><a name="l00298"></a><span class="lineno">  298</span>&#160;            all_orfs_df = pd.concat( [ all_orfs_df, contains_stop_codon ],</div><div class="line"><a name="l00299"></a><span class="lineno">  299</span>&#160;                                     axis = 1 )</div><div class="line"><a name="l00300"></a><span class="lineno">  300</span>&#160;            </div><div class="line"><a name="l00301"></a><span class="lineno">  301</span>&#160;            <span class="comment"># Extract from the data frame the ORF for which the sequence</span></div><div class="line"><a name="l00302"></a><span class="lineno">  302</span>&#160;            <span class="comment"># do not contains stop codons</span></div><div class="line"><a name="l00303"></a><span class="lineno">  303</span>&#160;            all_orfs_df = all_orfs_df[ all_orfs_df.contains_stop_codon == <span class="keyword">False</span> ]</div><div class="line"><a name="l00304"></a><span class="lineno">  304</span>&#160;            </div><div class="line"><a name="l00305"></a><span class="lineno">  305</span>&#160;            Logger.get_instance().info( str( total_sqce_count - all_orfs_df.shape[0] ) +</div><div class="line"><a name="l00306"></a><span class="lineno">  306</span>&#160;                                        <span class="stringliteral">&#39; sequences (/&#39;</span> + str( total_sqce_count ) + </div><div class="line"><a name="l00307"></a><span class="lineno">  307</span>&#160;                                        <span class="stringliteral">&#39;) have been removed as they were containing stop codons&#39;</span> )</div><div class="line"><a name="l00308"></a><span class="lineno">  308</span>&#160;            </div><div class="line"><a name="l00309"></a><span class="lineno">  309</span>&#160;        </div><div class="line"><a name="l00310"></a><span class="lineno">  310</span>&#160;        <span class="comment"># For each row, build a string that will be used </span></div><div class="line"><a name="l00311"></a><span class="lineno">  311</span>&#160;        <span class="comment"># as header line in the FASTA file</span></div><div class="line"><a name="l00312"></a><span class="lineno">  312</span>&#160;        header = all_orfs_df.apply( self.generate_header, </div><div class="line"><a name="l00313"></a><span class="lineno">  313</span>&#160;                                    axis=1,</div><div class="line"><a name="l00314"></a><span class="lineno">  314</span>&#160;                                    taxon_sc_name = taxon_sc_name, </div><div class="line"><a name="l00315"></a><span class="lineno">  315</span>&#160;                                    taxon_code = taxon_code,</div><div class="line"><a name="l00316"></a><span class="lineno">  316</span>&#160;                                    taxon_id = str( taxon_id ), </div><div class="line"><a name="l00317"></a><span class="lineno">  317</span>&#160;                                    table = self.table_type,</div><div class="line"><a name="l00318"></a><span class="lineno">  318</span>&#160;                                    db_release = db_release,</div><div class="line"><a name="l00319"></a><span class="lineno">  319</span>&#160;                                    long_header = self.long_header ).to_frame()</div><div class="line"><a name="l00320"></a><span class="lineno">  320</span>&#160;        header = header.rename( columns = { 0: <span class="stringliteral">&#39;header&#39;</span> } )</div><div class="line"><a name="l00321"></a><span class="lineno">  321</span>&#160;        all_orfs_df = pd.concat( [ all_orfs_df, header ],</div><div class="line"><a name="l00322"></a><span class="lineno">  322</span>&#160;                                 axis = 1 )</div><div class="line"><a name="l00323"></a><span class="lineno">  323</span>&#160;</div><div class="line"><a name="l00324"></a><span class="lineno">  324</span>&#160;        <span class="comment"># Write the FASTA file one line at a time</span></div><div class="line"><a name="l00325"></a><span class="lineno">  325</span>&#160;        <span class="keyword">with</span> open( file_path, <span class="stringliteral">&#39;w&#39;</span> ) <span class="keyword">as</span> fasta_file:</div><div class="line"><a name="l00326"></a><span class="lineno">  326</span>&#160;    </div><div class="line"><a name="l00327"></a><span class="lineno">  327</span>&#160;            <span class="keywordflow">for</span> ( index, row ) <span class="keywordflow">in</span> all_orfs_df.iterrows():</div><div class="line"><a name="l00328"></a><span class="lineno">  328</span>&#160;                                </div><div class="line"><a name="l00329"></a><span class="lineno">  329</span>&#160;                <span class="comment"># Write the header line</span></div><div class="line"><a name="l00330"></a><span class="lineno">  330</span>&#160;                fasta_file.write( <span class="stringliteral">&#39;&gt;&#39;</span> + row[ <span class="stringliteral">&#39;header&#39;</span> ] + <span class="stringliteral">&#39;\n&#39;</span> )</div><div class="line"><a name="l00331"></a><span class="lineno">  331</span>&#160;                </div><div class="line"><a name="l00332"></a><span class="lineno">  332</span>&#160;                <span class="comment"># Write the sequence line(s)</span></div><div class="line"><a name="l00333"></a><span class="lineno">  333</span>&#160;                  <span class="comment"># Split the sequence if it has to be written on several lines</span></div><div class="line"><a name="l00334"></a><span class="lineno">  334</span>&#160;                full_seq = row[ self.seq_attribute_name ]</div><div class="line"><a name="l00335"></a><span class="lineno">  335</span>&#160;                seq = <span class="stringliteral">&#39;\n&#39;</span>.join( [ full_seq[ k : k + self.MAX_SEQ_LINE_LENGTH ] <span class="keywordflow">for</span> k <span class="keywordflow">in</span> range( 0, len( full_seq ), self.MAX_SEQ_LINE_LENGTH ) ] )</div><div class="line"><a name="l00336"></a><span class="lineno">  336</span>&#160;                  <span class="comment"># Write the sequence line(s)</span></div><div class="line"><a name="l00337"></a><span class="lineno">  337</span>&#160;                fasta_file.write( seq + <span class="stringliteral">&#39;\n&#39;</span> )</div><div class="line"><a name="l00338"></a><span class="lineno">  338</span>&#160;                </div><div class="line"><a name="l00339"></a><span class="lineno">  339</span>&#160;        Logger.get_instance().info( <span class="stringliteral">&#39;The fasta file has been created at &#39;</span> + file_path + <span class="stringliteral">&#39;.&#39;</span> )</div><div class="line"><a name="l00340"></a><span class="lineno">  340</span>&#160;                </div><div class="line"><a name="l00341"></a><span class="lineno">  341</span>&#160;        </div><div class="line"><a name="l00342"></a><span class="lineno">  342</span>&#160;        </div><div class="line"><a name="l00343"></a><span class="lineno">  343</span>&#160;        </div></div><!-- fragment -->
</div>
</div>
<a id="a9680d0a2df2b41b32db8de889ee46044"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9680d0a2df2b41b32db8de889ee46044">&#9670;&nbsp;</a></span>generate_header()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def fr.tagc.uorf.core.execution.GenerateFastaFileStrategy.GenerateFastaFileStrategy.generate_header </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>row</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>taxon_sc_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>taxon_code</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>taxon_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>table</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>db_release</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>long_header</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>generate_header </h2>
<p>This is a static method allowing to compute the header of a sequence.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">row</td><td>Pandas Series - A row from the ORF Pandas data frame. </td></tr>
    <tr><td class="paramname">taxon_sc_name</td><td>String - The species scientific full name (e.g. Homo sapiens). </td></tr>
    <tr><td class="paramname">taxon_code</td><td>String - The taxonomic code (e.g. HUMAN). </td></tr>
    <tr><td class="paramname">taxon_id</td><td>String - The taxonomic ID (e.g. 9606). </td></tr>
    <tr><td class="paramname">table</td><td>String - The table used to perform queries (ORF/OTA). </td></tr>
    <tr><td class="paramname">db_release</td><td>String - The database version number. </td></tr>
    <tr><td class="paramname">long_header</td><td>Boolean - Should long headers be generated?</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>header: String - The header value. </dd></dl>
<div class="fragment"><div class="line"><a name="l00405"></a><span class="lineno">  405</span>&#160;    <span class="keyword">def </span>generate_header( row, taxon_sc_name, taxon_code, taxon_id, table, db_release, long_header ):</div><div class="line"><a name="l00406"></a><span class="lineno">  406</span>&#160;        </div><div class="line"><a name="l00407"></a><span class="lineno">  407</span>&#160;        header = [ GenerateFastaFileStrategy.FASTA_SOURCE_NAME ]</div><div class="line"><a name="l00408"></a><span class="lineno">  408</span>&#160;        </div><div class="line"><a name="l00409"></a><span class="lineno">  409</span>&#160;        <span class="comment"># Define each part of the header</span></div><div class="line"><a name="l00410"></a><span class="lineno">  410</span>&#160;        <span class="keywordflow">if</span> ( table == GenerateFastaFileStrategy.ORF_TABLE ):</div><div class="line"><a name="l00411"></a><span class="lineno">  411</span>&#160;            header += [ <span class="stringliteral">&#39;ORF&#39;</span> + row[ <span class="stringliteral">&#39;id&#39;</span> ],</div><div class="line"><a name="l00412"></a><span class="lineno">  412</span>&#160;                        <span class="stringliteral">&#39;ORF&#39;</span> + row[ <span class="stringliteral">&#39;id&#39;</span> ] + <span class="stringliteral">&#39;_&#39;</span> + taxon_code,</div><div class="line"><a name="l00413"></a><span class="lineno">  413</span>&#160;                        <span class="stringliteral">&#39;OS=&#39;</span> + taxon_sc_name,</div><div class="line"><a name="l00414"></a><span class="lineno">  414</span>&#160;                        <span class="stringliteral">&#39;OX=&#39;</span> + taxon_id,</div><div class="line"><a name="l00415"></a><span class="lineno">  415</span>&#160;                        <span class="stringliteral">&#39;chr=&#39;</span> + row[ <span class="stringliteral">&#39;chromosome&#39;</span> ],</div><div class="line"><a name="l00416"></a><span class="lineno">  416</span>&#160;                        <span class="stringliteral">&#39;strand=&#39;</span> + row[ <span class="stringliteral">&#39;strand&#39;</span> ],</div><div class="line"><a name="l00417"></a><span class="lineno">  417</span>&#160;                        <span class="stringliteral">&#39;start_pos=&#39;</span> + row[ <span class="stringliteral">&#39;start_pos&#39;</span> ],</div><div class="line"><a name="l00418"></a><span class="lineno">  418</span>&#160;                        <span class="stringliteral">&#39;stop_pos=&#39;</span> + row[ <span class="stringliteral">&#39;stop_pos&#39;</span> ],</div><div class="line"><a name="l00419"></a><span class="lineno">  419</span>&#160;                        <span class="stringliteral">&#39;exons=&#39;</span> + row[ <span class="stringliteral">&#39;spliced_parts_count&#39;</span> ] ]</div><div class="line"><a name="l00420"></a><span class="lineno">  420</span>&#160;        </div><div class="line"><a name="l00421"></a><span class="lineno">  421</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00422"></a><span class="lineno">  422</span>&#160;            header += [ <span class="stringliteral">&#39;OTA&#39;</span> + row[ <span class="stringliteral">&#39;id&#39;</span> ],</div><div class="line"><a name="l00423"></a><span class="lineno">  423</span>&#160;                        <span class="stringliteral">&#39;OTA&#39;</span> + row[ <span class="stringliteral">&#39;id&#39;</span> ] + <span class="stringliteral">&#39;_&#39;</span> + taxon_code,</div><div class="line"><a name="l00424"></a><span class="lineno">  424</span>&#160;                        <span class="stringliteral">&#39;OS=&#39;</span> + taxon_sc_name,</div><div class="line"><a name="l00425"></a><span class="lineno">  425</span>&#160;                        <span class="stringliteral">&#39;OX=&#39;</span> + taxon_id,</div><div class="line"><a name="l00426"></a><span class="lineno">  426</span>&#160;                        <span class="stringliteral">&#39;rel_start_pos=&#39;</span> + row[ <span class="stringliteral">&#39;rel_start_pos&#39;</span> ],</div><div class="line"><a name="l00427"></a><span class="lineno">  427</span>&#160;                        <span class="stringliteral">&#39;rel_stop_pos=&#39;</span> + row[ <span class="stringliteral">&#39;rel_stop_pos&#39;</span> ],</div><div class="line"><a name="l00428"></a><span class="lineno">  428</span>&#160;                        <span class="stringliteral">&#39;orf_id=&#39;</span> + row[ <span class="stringliteral">&#39;orf_id&#39;</span> ],</div><div class="line"><a name="l00429"></a><span class="lineno">  429</span>&#160;                        <span class="stringliteral">&#39;transcript_id=&#39;</span> + row[ <span class="stringliteral">&#39;transcript_id&#39;</span> ] ]</div><div class="line"><a name="l00430"></a><span class="lineno">  430</span>&#160;        </div><div class="line"><a name="l00431"></a><span class="lineno">  431</span>&#160;        <span class="comment"># Add the database url and version number if available</span></div><div class="line"><a name="l00432"></a><span class="lineno">  432</span>&#160;        header += [ <span class="stringliteral">&#39;database=&#39;</span> + Constants.PROJECT_NAME,</div><div class="line"><a name="l00433"></a><span class="lineno">  433</span>&#160;                    <span class="stringliteral">&#39;url=&#39;</span> + Constants.WEBSITE_URL ]</div><div class="line"><a name="l00434"></a><span class="lineno">  434</span>&#160;        <span class="keywordflow">if</span> ( db_release ):</div><div class="line"><a name="l00435"></a><span class="lineno">  435</span>&#160;            header.append( <span class="stringliteral">&#39;release=&#39;</span> + db_release )  </div><div class="line"><a name="l00436"></a><span class="lineno">  436</span>&#160;        </div><div class="line"><a name="l00437"></a><span class="lineno">  437</span>&#160;        <span class="comment"># Convert the header into a string</span></div><div class="line"><a name="l00438"></a><span class="lineno">  438</span>&#160;        header_string = <span class="stringliteral">&#39;|&#39;</span>.join( header[0:3] )</div><div class="line"><a name="l00439"></a><span class="lineno">  439</span>&#160;        <span class="keywordflow">if</span> long_header:</div><div class="line"><a name="l00440"></a><span class="lineno">  440</span>&#160;            header_string += <span class="stringliteral">&#39; &#39;</span> + <span class="stringliteral">&#39; &#39;</span>.join( header[3:] )</div><div class="line"><a name="l00441"></a><span class="lineno">  441</span>&#160;        </div><div class="line"><a name="l00442"></a><span class="lineno">  442</span>&#160;        <span class="keywordflow">return</span> header_string</div></div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a936c81e088788e99b88638c5c0735c92"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a936c81e088788e99b88638c5c0735c92">&#9670;&nbsp;</a></span>ORF_TABLE</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">string fr.tagc.uorf.core.execution.GenerateFastaFileStrategy.GenerateFastaFileStrategy.ORF_TABLE = '<a class="el" href="classfr_1_1tagc_1_1uorf_1_1core_1_1model_1_1PRO_1_1processed_1_1ORF_1_1ORF.html">ORF</a>'</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<h2>Class variables </h2>
<p>Tables that may be used to generate the FASTA file </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/mnt/NAS7/Workspace/choteaus/tagc-uorf/project/01_ORF_datafreeze/06_src/fr/tagc/uorf/core/execution/GenerateFastaFileStrategy.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
