<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.15"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>sORFdatafreezer: sORF datafreezer</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">sORFdatafreezer
   &#160;<span id="projectnumber">1.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.15 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">sORF datafreezer </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><h1>Description</h1>
<p>This source code has been created to perform a data freeze of <b>short Open Reading Frame (ORFs) related data</b>. It is able to perform the parsing and insertion of several data sources and perform normalization, merging and summary of redundant information and computation of several interesting features, such as ORF annotations or estimation of Kozak contexts. We strongly suggest to read the <b>user's manual</b> for detailed explanations about the available strategies, and general structure of the algorithms for each of them. The current manual is exclusively <b>dedicated to developers</b> and does not intend to explain the interest of each strategy, the expected formats of inputs, formats of outputs or the way the data are handled!</p>
<p>The present page summarizes some useful information about the most important functions and classes defined in the source code. For precise information about a particular class or method, you will be able to find a precise documentation that includes:</p>
<ul>
<li>The aim of the class or method,</li>
<li>The inputs of the method (with detailed explanation of their format).</li>
<li>The outputs of the method (with detailed explanation of their format).</li>
<li>The exceptions susceptible to be raised by the method.</li>
<li>For the class constructors, the instance attributes defined.</li>
</ul>
<p><b>Any</b> module of the source code contains <b>one single and unique class</b>.</p>
<h1>Entrypoint</h1>
<p>The <code>fr/tagc/uorf/uorf.py</code> file constitutes the main file of the program (the entrypoint). It contains a <code>__main__</code> function that is run when launching the script with a Python interpreter. The user should never try to run a strategy otherwise than using this entrypoint.</p>
<h1>Strategies</h1>
<p>The program uses strategy patterns to perform the expected tasks.</p>
<p>For each strategy defined in the user's manual, there is a class that use the same name followed by the suffix <code>Strategy</code> that is located in the <code>fr.tagc.uorf.core.execution</code> package. This section briefly describes the main modules related to each strategy.</p>
<p>NB: Utils modules (<code>fr.tagc.uorf.core.util</code>) and external packages are <b>not</b> referenced here. We advice to read the documentation of the strategy for more information regarding the modules or external packages it requires.</p>
<ul>
<li><b>DatabaseCheck</b><ul>
<li>The list of strategies using the <b>DatabaseCheck</b> strategy may be found at the end of the <code>fr.tagc.uorf.core.util.option.OptionConstants</code> module.</li>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.DatabaseCheckStrategy</code>: Strategy class.</li>
<li><code>fr.tagc.uorf.core.execution.dbcheck</code> package: defines <code>CheckDatabase</code> meta / abstract class and <code>CheckDSDatabase</code>, <code>CheckFILTDatabase</code>, <code>CheckPRODatabase</code> classes.</li>
</ul>
</li>
<li>Related database models: DS, PRO.</li>
</ul>
</li>
<li><b>AddReleaseVersion</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.AddReleaseVersion</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: DS, PRO.</li>
</ul>
</li>
<li><b>Insertion</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.InsertionStrategy</code>: Strategy class.</li>
<li><code>fr.tagc.uorf.core.execution.insertion</code> package includes:<ul>
<li><code>ParserWrapper</code>: A wrapper that allows to execute the appropriate parser and return the list of objects that needs to be inserted in the database.</li>
<li><code>ParserStrategy</code>: Meta and abstract class that includes useuful methods for the parsing of data sources.</li>
<li><code>parserstrategy</code> packages includes <code>common</code>, <code>human</code> and <code>mouse</code> sub-packages. Any parser of this module is a new class inheriting from the <code>ParserStrategy</code> class. Any parser is able to import the data from the source, parse the data and instantiate the appropriate objects that need to be inserted in the database.</li>
</ul>
</li>
</ul>
</li>
<li>Related database models: DS.</li>
</ul>
</li>
<li><b>ForceInsertion</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.ForceInsertionStrategy</code>: Strategy class.</li>
<li>See modules and packages related to the <b>Insertion</b> strategy.</li>
</ul>
</li>
<li>Related database models: DS.</li>
</ul>
</li>
<li><b>Deletion</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.DeletionStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: DS.</li>
</ul>
</li>
<li><b>LiftOver</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.LiftOverStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: DS.</li>
</ul>
</li>
<li><b>Merge</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.MergeStrategy</code>: Strategy class (and meta class).</li>
<li><code>fr.tagc.uorf.core.execution.merge</code> package includes:<ul>
<li><code>MergeDSORF</code>: A static class that help performing the merge of <em>DSORF</em> entries. This module has been adapted to multi-processing.</li>
<li><code>MergeDSOTA</code>: A static class that help performing the merge of <em>DSORFTranscriptAsso</em> entries. This module has been adapted to multi-processing.</li>
</ul>
</li>
</ul>
</li>
<li>Related database models: DS, PRO.</li>
</ul>
</li>
<li><b>ResumeMerge</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.ResumeMergeStrategy</code>: Strategy class (inherit from the <code>MergeStrategy</code> class).</li>
<li>See modules and packages related to the <b>Merge</b> strategy.</li>
</ul>
</li>
<li>Related database models: DS, PRO.</li>
</ul>
</li>
<li><b>ComputeMissingInformation</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.ComputeMissingInformationStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>ComputeRelInformation</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.ComputeRelInformationStrategy</code>: Strategy class.</li>
<li><code>fr.tagc.uorf.core.execution.relative_coord</code> package includes:<ul>
<li>R scripts that aims to be run as subprocesses.</li>
</ul>
</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>ComputeKozakContext</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.ComputeKozakContextStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>AnnotateORF</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.AnnotateORFStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>Filter</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.FilterStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>GenerateFastaFile</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.GenerateFastaFileStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>GenerateBEDContent</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.GenerateBEDContentStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>GenerateBEDFile</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.GenerateBEDContentStrategy</code>: Strategy class (inherit from the <code>GenerateBEDContentStrategy</code> class).</li>
<li><code>fr.tagc.uorf.core.execution.GenerateBEDFileStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>GenerateTrackDbFile</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.GenerateTrackDbFileFileStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>GenerateGFFFile</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.GenerateGFFFileStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: PRO.</li>
</ul>
</li>
<li><b>AssessDatabaseContent</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.AssessDatabaseContentStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: DS, PRO.</li>
</ul>
</li>
<li><b>GenerateStatFiles</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.GenerateStatFilesStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: None.</li>
</ul>
</li>
<li><b>Backup strategy</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.BackupStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: DS, PRO.</li>
</ul>
</li>
<li><b>Restore strategy</b><ul>
<li>Main modules related to this strategy:<ul>
<li><code>fr.tagc.uorf.core.execution.RestoreStrategy</code>: Strategy class.</li>
</ul>
</li>
<li>Related database models: DS, PRO.</li>
</ul>
</li>
</ul>
<h1>Database models</h1>
<p>The <code>fr.tagc.uorf.core.model</code> package contains the database models, <em>i.e</em> all the classes that are required by SQLALchemy (ORM) to create the tables of the databases and to handle their content.</p>
<p>Two models of databases (<em>DS</em> standing for DataSource and <em>PRO</em> standing for PROcessed) are available. The classes necessary to handle these model have been created in separated sub-packages. For more information about the database structures and table attributes (type, primary keys...), please see the database schemas provided with the source code and the documentation of these classes.</p>
<p>The program is able to handle several databases (with distinct structures) at the same time (see documentation of the <code>fr.tagc.uorf.core.util.sql</code> package for more information).</p>
<p>The <code>fr.tagc.uorf.core.model</code> package contains:</p>
<ul>
<li>The <code>DS</code> package containing the classes necessary to handle the <em>DS</em> database. It contains the following packages:<ul>
<li><code>meta</code> contains the modules related to the tables that aim to store meta data about the database (species, annotation version information...). It contains the <code>Metadata</code> and <code>SpeciesCatalog</code> modules.</li>
<li><code>datasource</code> contains the modules related to the tables that aim to store the content that has been parsed from the data sources. It contains the <code>DataSource</code>, <code>DSORF</code>, <code>DSTranscript</code> and <code>DSORFTranscriptAsso</code> modules.</li>
<li><code>utilrelation</code> contains the modules related to the tables that are not <em>per se</em> of interest for the user but which essentially aims to store data generated during the execution of the program and that need to persist between the execution of two strategies (<em>e.g</em>. because their computation may need a long time). It contains the <code>UTGeneFromAlias</code> and <code>UTDSTranscriptGeneConflict</code> modules.</li>
<li><code>Gene</code> and <code>GeneAlias</code> modules define the tables that contain the cross-references.</li>
</ul>
</li>
<li>The <code>PRO</code> package containing the classes necessary to handle the <em>PRO</em> database. It contains the following packages:<ul>
<li><code>meta</code> contains the modules related to the tables that aim to store meta data about the database (species, annotation version information...). It contains the <code>PROMetadata</code> and <code>PROSpeciesCatalog</code> modules.</li>
<li><code>processed</code> contains the modules related to the tables that aim to store the content merged from the <em>DS</em> database. It contains the <code>ORF</code>, <code>Transcript</code>, <code>ORFTranscriptAsso</code>, <code>CellContext</code>, <code>FLOSSClass</code>, <code>ProvidedCategory</code>, <code>ORFDSAsso</code>, <code>TranscriptDSAsso</code> and <code>ORFTranscriptAssoDSAsso</code> modules. It also contains the sub-package <code>catalog</code> with <code>CellContextCatalog</code>, <code>FLOSSClassCatalog</code> and <code>ProvidedCategory</code> modules.</li>
<li><code>computed</code>contains the modules related to the tables that store computed data. It contains the <code>ORFCategory</code>, <code>ORFCategoryCatalog</code>, <code>ORFAnnotation</code> and <code>ORFAnnotationCatalog</code> modules.</li>
<li><code>utilrelation</code> contains the modules related to the tables that are not <em>per se</em> of interest for the user but which essentially aims to store data generated during the execution of the program and that need to persist between the execution of two strategies (<em>e.g</em>. because their computation may need a long time). It contains the <code>UTRNABiotypeCatalog</code> and <code>UTBEDContent</code> modules.</li>
<li><code>PROGene</code> and <code>PROGeneAlias</code> modules define the tables that contain the cross-references.</li>
</ul>
</li>
</ul>
<h1>Utils</h1>
<p>Any class or method defined in the <code>fr.tagc.uorf.core.util</code> module are non-specific classes and methods that have been built to be used anywhere in the program.</p>
<ul>
<li><b>Loggers</b>: <code>fr.tagc.uorf.core.util.log</code> package<ul>
<li>Two loggers defined:<ul>
<li><code>Logger</code> module defines the main logger. The class defined is a singleton.</li>
<li><code>GeneRefLogger</code> module defines the logger dedicated to gene references problems. The class defined is a singleton.</li>
</ul>
</li>
<li>Classes defined in these modules allow to log message anywhere in the program.</li>
<li>External modules related to this package:<ul>
<li><code>fr.tagc.uorf.core.util.LogCodes</code>: file defining constants which are used as warning and error codes.</li>
</ul>
</li>
</ul>
</li>
<li><b>Command line arguments</b> (options): <code>fr.tagc.uorf.core.util.option</code> package<ul>
<li>Modules:<ul>
<li><code>OptionManager</code> modules define a command line arguments parser and option manager. The class defined is a singleton.</li>
<li><code>OptionConstants</code> module defines constants related to the options that may be provided in the command line and constants helping to manage them. No class defined in this module, constants file.</li>
</ul>
</li>
</ul>
</li>
<li><b>SQL manager</b>: <code>fr.tagc.uorf.core.util.sql</code> package<ul>
<li>Modules:<ul>
<li><code>SQLManager</code> module defines an abstract meta class of singletons which children aim to manage database connections.</li>
<li><code>SQLManagerDS</code>, <code>SQLManagerPRO</code>, <code>SQLManagerFILT</code> modules contains classes inheriting from the <code>SQLManager</code> class and aim to manage respectively <em>DS</em>, <em>PRO</em> (<em>PRO</em>-like) and <em>FILT</em> databases. The classes defined in these modules are singleton.</li>
<li><code>Base</code> module instantiates objects of the <code>declarative_base</code> SQLALchemy class. No class defined in this module, procedural-like code defined there.</li>
<li><code>SQLContants</code> module defines constants related to the database managment, such as connection settings (user name, URI etc.) or default value for some MySQL variables.</li>
<li><code>SQLCollationManager</code> module defines a class to manage the database collation for all the columns of the database containing text (such as <code>VARCHAR</code> or <code>TEXT</code> columns). The class defined is a singleton.</li>
</ul>
</li>
<li>Classes defined in these modules facilitate the creation of engine and seessions to the databases. They also facilitates the creation of databases as well as addition, deletion and update of entries in the database. They are able to manage the commit or roll-back of sessions. Methods of these classes are using <b>SQLAlchemy</b> (ORM) tools.</li>
</ul>
</li>
<li><b>Exceptions</b>: <code>fr.tagc.uorf.core.util.exception</code> package<ul>
<li><code>DenCellORFException</code> module defines a meta class of custom exceptions. The class defined inherit itself from the <code>Exception</code> Python class.</li>
<li>The instances of the classes defined in this package are exception that may at least contain a message and an other exception. The class defines also methods to display the message of the exception (and an eventual trace track). All class of exceptions defined by these modules obviously aim to be manipulated anywhere in the program.</li>
<li>Each package contains one or several module(s), each defining one unique exception class:<ul>
<li><code>dencellorfobject</code> exceptions: Used when an error related to the manipulation of object specific to the present source code is met.</li>
<li><code>insertexception</code> exceptions: Used when problems are met during the data insertion.</li>
<li><code>headerexception</code> exceptions: Used when problems related to the headers of pandas data frames are met.</li>
<li><code>refexception</code> exceptions: Used when problems related to the gene references and cross-references are met.</li>
<li><code>processingexception</code> exceptions: Used when problems related to the processing of data are met.</li>
<li><code>httpexception</code> exceptions: Used when exceptions / errors have been raised by an HTTP request.</li>
<li><code>geneticsexception</code> exceptions: Used when problems related to the manipulation of genetics objects are met.</li>
</ul>
</li>
</ul>
</li>
<li><b>Shared data</b> (DataManager): <code>fr.tagc.uorf.core.util.data</code> package<ul>
<li>Modules:<ul>
<li><code>DataManager</code> module helps managing objects of any class shared by different classes and/or methods of the program. The class defined is a singleton.</li>
</ul>
</li>
</ul>
</li>
<li><b>Genetics-related data and data from Ensembl</b>: <code>fr.tagc.uorf.core.util.genetics</code> and <code>fr.tagc.uorf.core.util.ensembl</code> packages<ul>
<li>Ensembl modules:<ul>
<li><code>EnsemblUtil</code> module defines a static class that contains methods helping to use Ensembl tools (and the <code>PyEnsembl</code> package).</li>
<li><code>EnsemblRestClient</code> module defines a client helping to use the Ensembl REST API. The class defined is a singleton.</li>
</ul>
</li>
<li>Genetics modules:<ul>
<li><code>GeneticsUtil</code> module defines a static class that contains methods helping to handle genetics-related data (such as performing DNA translation).</li>
<li><code>GeneticsConstants</code> module defines constants related to genetics data. No class defined in this module, constants file.</li>
</ul>
</li>
</ul>
</li>
<li><b>Graphics</b>: <code>fr.tagc.uorf.core.util.graphics</code> package<ul>
<li>Modules:<ul>
<li><code>ProgressionBar</code> module aims to help the display of progression bars on the console. The class defined is a singleton.</li>
</ul>
</li>
</ul>
</li>
<li><b>General and miscellaneous utils</b>: <code>fr.tagc.uorf.core.util.general</code> package<ul>
<li>This package contains highly generic functions that may be useful anywhere in the program.</li>
<li>Modules:<ul>
<li><code>GeneralUtil</code> module defines a static class dedicated to these general utils.</li>
<li><code>FileHandlerUtil</code> module defines a static class dedicated to the manipulation of files.</li>
</ul>
</li>
</ul>
</li>
<li><b>Constants</b>: <code>fr.tagc.uorf.core.util.Constants</code> module (see following section)</li>
<li><b>Default output and temporary folders</b>: <code>fr.tagc.uorf.core.util.DefaultOutputFolder</code> and <code>fr.tagc.uorf.core.util.DefaultTemporaryFolder</code> modules<ul>
<li>Modules:<ul>
<li><code>DefaultOutputFolder</code> defines one <b>unique</b> variable that contains the path of the folder to use as output folder (in particular for the log files). A different output folder may be defined for some strategies (see <b>user's manual</b> for more information).</li>
<li><code>DefaultTemporaryFolder</code> defines one <b>unique</b> variable that contains the path of the folder to use as temporary folder.</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1>Constants</h1>
<p>The <code>fr.tagc.uorf.core.util.Constants</code> module define numerous constants that are used in most of the modules.</p>
<p>Here is a brief description of the sections it contains:</p>
<ul>
<li><b>Project informations</b>: General information about the project (project name, alias, URL to user-friendly web interface).</li>
<li><b>Config file constants</b>: Definition of the config file sections and items.</li>
<li><b>DataManager constants</b>: Definition of the keys of the DataManager main dictionary.</li>
<li><b>General constants</b>: General constants, mainly regarding the species allowed.</li>
<li><b>Constants relative to the annotation versions</b>: Definition of the annotation version and correspondences between the versions (NCBI, UCSC).</li>
<li><b>Constants relative to the Gene / GeneAlias / util tables of the database</b>: Definition of prefix or separators to use in the gene tables.</li>
<li><b>Constants relative to the connection to the database</b>: Definition of SQLManager keys.</li>
<li><b>Constants relative to logs</b>: Default settings of the main logger.</li>
<li><b>Constants relative to gene reference logs</b>: Default settings of the gene references logger.</li>
<li><b>Constants relative to the files created by the program</b>: Definition of default path to use for output and temporary files and folders.</li>
<li><b>Constants relative to the insertion of data in the database</b>: Default settings for the insertion of new entries in the databases.</li>
<li><b>Constants relative to the built of the PRO database</b>: Default settings mainly related to the strategies manipulating the <b>PRO</b> database.</li>
<li><b>Constants relative to the DenCellORFObjects</b>: Definition of constants related to the database contents. This includes definition of chromosome names, gene and transcript prefixes etc.</li>
<li><b>Constants relative to the Metadata table</b>: Definition of keys and definitions used in the metadata table.</li>
<li><b>Constants relative to the sequences</b>: Default settings related to DNA and protein sequences (letter to use for ambiguous residue etc.).</li>
<li><b>Constants relative to the computation of relative coordinates from absolute coordinates relative</b>: Constants specifically related to the <b>ComputeRelCoord</b> strategy.</li>
<li><b>Constants relative to the merging of CellContext entries</b>: Default settings for the cell merging (see documentation of the <b>ComputeMissingInfo</b> strategy).</li>
<li><b>Constants relative to the ORF annotations</b>: Default settings for the ORF categories and annotation computation (see documentation of the <b>AnnotateORF</b> strategy).</li>
<li><b>Constants relative to the filtering</b>: Default settings for the filtering of a PRO-like database (see documentation of the <b>Filter</b> strategy).</li>
</ul>
<p><b>NB:</b> The <b>Information to developers</b> section of the <b>user's manual</b> contains complementary information regarding the constants.</p>
<h1>Authors, license and copyright</h1>
<p>Sébastien A. Choteau<sup>1,2</sup>, Lionel Spinelli<sup>1,2</sup>, Philippe Pierre<sup>2</sup>, Christine Brun<sup>1,3</sup></p>
<ol type="1">
<li>Aix-Marseille Univ, INSERM, TAGC, CENTURI, Marseille, France</li>
<li>Aix-Marseille Univ, CNRS, INSERM, CIML, CENTURI, Marseille, France</li>
<li>CNRS, Marseille, France</li>
</ol>
<p>If you have questions or comments, please write to:</p>
<ul>
<li>Sébastien A. Choteau at <a href="#" onclick="location.href='mai'+'lto:'+'seb'+'as'+'tie'+'n.'+'cho'+'te'+'au@'+'un'+'iv-'+'am'+'u.f'+'r'; return false;">sebastien.choteau@univ-amu.fr</a></li>
</ul>
<p>Last update of this page: <b>24/05/2020</b>. Please note that we ensure <b>continuous maintenance</b> of all other pages of the current documentation dedicated to developers. </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.15
</small></address>
</body>
</html>
